<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Map positions within the CDS to coordinates relative to the start of
the transcript — cdsToTranscript • ensembldb</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Map positions within the CDS to coordinates relative to the start of
the transcript — cdsToTranscript" />
<meta property="og:description" content="Converts CDS-relative coordinates to positions within the transcript, i.e.
relative to the start of the transcript and hence including its 5' UTR." />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ensembldb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.17.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/ensembldb.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/MySQL-backend.html">Using a MySQL server backend</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping-use-cases.html">Use cases for coordinate mapping with ensembldb</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping.html">Mapping between genome, transcript and protein coordinates</a>
    </li>
    <li>
      <a href="../articles/proteins.html">Querying protein features</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jorainer/ensembldb/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Map positions within the CDS to coordinates relative to the start of
the transcript</h1>
    <small class="dont-index">Source: <a href='https://github.com/jorainer/ensembldb/blob/master/R/transcriptToX.R'><code>R/transcriptToX.R</code></a></small>
    <div class="hidden name"><code>cdsToTranscript.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Converts CDS-relative coordinates to positions within the transcript, i.e.
relative to the start of the transcript and hence including its 5' UTR.</p>
    </div>

    <div class="ref-usage sourceCode"><pre class='sourceCode r'><code><span class='fu'>cdsToTranscript</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>db</span>, id <span class='op'>=</span> <span class='st'>"name"</span><span class='op'>)</span></code></pre></div>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p><code>IRanges</code> with the coordinates within the CDS. Coordinates
are expected to be relative to the transcription start (the first
nucleotide of the transcript). The Ensembl IDs of the corresponding
transcripts have to be provided either as <code>names</code> of the <code>IRanges</code>, or
in one of its metadata columns.</p></td>
    </tr>
    <tr>
      <th>db</th>
      <td><p><code>EnsDb</code> object.</p></td>
    </tr>
    <tr>
      <th>id</th>
      <td><p><code>character(1)</code> specifying where the transcript identifier can be
found. Has to be either <code>"name"</code> or one of <code>colnames(mcols(prng))</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>IRanges</code> with the same length (and order) than the input <code>IRanges</code>
<code>x</code>. Each element in <code>IRanges</code> provides the coordinates within the
transcripts CDS. The transcript-relative coordinates are provided
as metadata columns.
<code>IRanges</code> with a start coordinate of <code>-1</code> is returned for transcripts
that are not known in the database, non-coding transcripts or if the
provided start and/or end coordinates are not within the coding region.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other coordinate mapping functions: 
<code><a href='genomeToProtein.html'>genomeToProtein</a>()</code>,
<code><a href='genomeToTranscript.html'>genomeToTranscript</a>()</code>,
<code><a href='proteinToGenome.html'>proteinToGenome</a>()</code>,
<code><a href='proteinToTranscript.html'>proteinToTranscript</a>()</code>,
<code><a href='transcriptToCds.html'>transcriptToCds</a>()</code>,
<code><a href='transcriptToGenome.html'>transcriptToGenome</a>()</code>,
<code><a href='transcriptToProtein.html'>transcriptToProtein</a>()</code></p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Johannes Rainer</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <div class="ref-examples sourceCode"><pre class='sourceCode r'><code><span class='r-in'></span>
<span class='r-in'><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>EnsDb.Hsapiens.v86</span><span class='op'>)</span></span>
<span class='r-in'><span class='co'>## Defining transcript-relative coordinates for 4 transcripts of the gene</span></span>
<span class='r-in'><span class='co'>## BCL2</span></span>
<span class='r-in'><span class='va'>txcoords</span> <span class='op'>&lt;-</span> <span class='fu'>IRanges</span><span class='op'>(</span>start <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>4</span>, <span class='fl'>3</span>, <span class='fl'>143</span>, <span class='fl'>147</span><span class='op'>)</span>, width <span class='op'>=</span> <span class='fl'>1</span>,</span>
<span class='r-in'>    names <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"ENST00000398117"</span>, <span class='st'>"ENST00000333681"</span>,</span>
<span class='r-in'>    <span class='st'>"ENST00000590515"</span>, <span class='st'>"ENST00000589955"</span><span class='op'>)</span><span class='op'>)</span></span>
<span class='r-in'></span>
<span class='r-in'><span class='fu'>cdsToTranscript</span><span class='op'>(</span><span class='va'>txcoords</span>, <span class='va'>EnsDb.Hsapiens.v86</span><span class='op'>)</span></span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>1 of 4 transcript(s) are non-coding: ENST00000590515</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> IRanges object with 4 ranges and 2 metadata columns:</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                       start       end     width | cds_start   cds_end</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                   &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; | &lt;integer&gt; &lt;integer&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENST00000398117      1466      1466         1 |         4         4</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENST00000333681       902       902         1 |         3         3</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENST00000590515        -1        -1         1 |       143       143</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENST00000589955       293       293         1 |       147       147</span>
<span class='r-in'></span>
<span class='r-in'><span class='co'>## Next we map the coordinate for variants within the gene PKP2 to the</span></span>
<span class='r-in'><span class='co'>## genome. The variants is PKP2 c.1643DelG and the provided</span></span>
<span class='r-in'><span class='co'>## position is thus relative to the CDS. We have to convert the</span></span>
<span class='r-in'><span class='co'>## position first to transcript-relative coordinates.</span></span>
<span class='r-in'><span class='va'>pkp2</span> <span class='op'>&lt;-</span> <span class='fu'>IRanges</span><span class='op'>(</span>start <span class='op'>=</span> <span class='fl'>1643</span>, width <span class='op'>=</span> <span class='fl'>1</span>, name <span class='op'>=</span> <span class='st'>"ENST00000070846"</span><span class='op'>)</span></span>
<span class='r-in'></span>
<span class='r-in'><span class='co'>## Map the coordinates by first converting the CDS- to transcript-relative</span></span>
<span class='r-in'><span class='co'>## coordinates</span></span>
<span class='r-in'><span class='fu'><a href='transcriptToGenome.html'>transcriptToGenome</a></span><span class='op'>(</span><span class='fu'>cdsToTranscript</span><span class='op'>(</span><span class='va'>pkp2</span>, <span class='va'>EnsDb.Hsapiens.v86</span><span class='op'>)</span>,</span>
<span class='r-in'>    <span class='va'>EnsDb.Hsapiens.v86</span><span class='op'>)</span></span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> GRangesList object of length 1:</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> $ENST00000070846</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> GRanges object with 1 range and 5 metadata columns:</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>       seqnames    ranges strand |         exon_id           tx_id exon_rank</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;     &lt;character&gt; &lt;integer&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   [1]       12  32841073      - | ENSE00000936503 ENST00000070846         7</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>        tx_start    tx_end</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>       &lt;integer&gt; &lt;integer&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   [1]      1668      1668</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   -------</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   seqinfo: 1 sequence from GRCh38 genome</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> </span>
</code></pre></div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by Johannes Rainer.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p></p>
</div>

      </footer>
   </div>

  


  

  </body>
</html>


