<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Functionality related to DNA/RNA sequences — getGenomeFaFile • ensembldb</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Functionality related to DNA/RNA sequences — getGenomeFaFile" />
<meta property="og:description" content="Utility functions related to RNA/DNA sequences, such as extracting
  RNA/DNA sequences for features defined in Ensb." />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->




  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ensembldb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.17.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/ensembldb.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/MySQL-backend.html">Using a MySQL server backend</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping-use-cases.html">Use cases for coordinate mapping with ensembldb</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping.html">Mapping between genome, transcript and protein coordinates</a>
    </li>
    <li>
      <a href="../articles/proteins.html">Querying protein features</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jorainer/ensembldb/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Functionality related to DNA/RNA sequences</h1>
    
    <div class="hidden name"><code>EnsDb-sequences.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Utility functions related to RNA/DNA sequences, such as extracting
  RNA/DNA sequences for features defined in <code>Ensb</code>.</p>
    </div>

    <div class="ref-usage sourceCode"><pre class='sourceCode r'><code><span class='co'># S4 method for EnsDb</span>
<span class='fu'>getGenomeFaFile</span><span class='op'>(</span><span class='va'>x</span>, pattern<span class='op'>=</span><span class='st'>"dna.toplevel.fa"</span><span class='op'>)</span>

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>getGenomeTwoBitFile</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span></code></pre></div>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>pattern</th>
      <td><p>For method <code>getGenomeFaFile</code>: the pattern to be used to
    identify the fasta file representing genomic DNA sequence.</p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>An <code>EnsDb</code> instance.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="methods-and-functions"><a class="anchor" href="#methods-and-functions"></a>Methods and Functions</h2>

    <p></p><dl>

    <dt>getGenomeFaFile</dt><dd><p>Returns a <code><a href='https://rdrr.io/pkg/Rsamtools/man/FaFile-class.html'>FaFile-class</a></code> (defined in
      <code>Rsamtools</code>) with the genomic sequence of the genome build
      matching the Ensembl version of the <code>EnsDb</code> object.
      The file is retrieved using the <code>AnnotationHub</code> package,
      thus, at least for the first invocation, an internet connection is
      required to locate and download the file; subsequent calls will
      load the cached file instead.
      If no fasta file for the actual Ensembl version is available the
      function tries to identify a file matching the species and genome
      build version of the closest Ensembl release and returns that
      instead.
      See the vignette for an example to work with such files.</p></dd>

    <dt>getGenomeTwoBitFile</dt><dd><p>Returns a <code>TwoBitFile-class</code> (defined in the
      <code>rtracklayer</code> package) with the genomeic sequence of the
      genome build matching the Ensembl version of the <code>EnsDb</code>
      object. The file is retrieved from <code>AnnotationHub</code> and hence
      requires (at least for the first query) an active internet
      connection to download the respective resource. If no DNA sequence
      matching the Ensembl version of <code>x</code> is available, the
      function tries to find the genomic sequence of the best matching
      genome build (closest Ensembl release) and returns that.      See the <code>ensembldb</code> vignette for details.</p></dd>
  
</dl>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>For <code>getGenomeFaFile</code>: a <code><a href='https://rdrr.io/pkg/Rsamtools/man/FaFile-class.html'>FaFile-class</a></code>
  object with the genomic DNA sequence.</p>
<p>For <code>getGenomeTwoBitFile</code>: a <code>TwoBitFile-class</code>
  object with the genome sequence.</p>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Johannes Rainer</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='EnsDb-exonsBy.html'>transcripts</a></code>
  <code><a href='EnsDb-exonsBy.html'>exonsBy</a></code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <div class="ref-examples sourceCode"><pre class='sourceCode r'><code><span class='r-in'></span>
<span class='r-in'><span class='co'>## Loading an EnsDb for Ensembl version 86 (genome GRCh38):</span></span>
<span class='r-in'><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>EnsDb.Hsapiens.v86</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>edb</span> <span class='op'>&lt;-</span> <span class='va'>EnsDb.Hsapiens.v86</span></span>
<span class='r-in'></span>
<span class='r-in'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span></span>
<span class='r-in'>    <span class='co'>## Retrieve a TwoBitFile with the gneomic DNA sequence matching the organism,</span></span>
<span class='r-in'>    <span class='co'>## genome release version and, if possible, the Ensembl version of the</span></span>
<span class='r-in'>    <span class='co'>## EnsDb object.</span></span>
<span class='r-in'>    <span class='va'>Dna</span> <span class='op'>&lt;-</span> <span class='fu'>getGenomeTwoBitFile</span><span class='op'>(</span><span class='va'>edb</span><span class='op'>)</span></span>
<span class='r-in'>    <span class='co'>## Extract the transcript sequence for all transcripts encoded on chromosome</span></span>
<span class='r-in'>    <span class='co'>## Y.</span></span>
<span class='r-in'>    <span class='co'>##extractTranscriptSeqs(Dna, edb, filter=SeqNameFilter("Y"))</span></span>
<span class='r-in'></span>
<span class='r-in'><span class='op'>}</span></span>
<span class='r-in'></span>
</code></pre></div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by Johannes Rainer.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p></p>
</div>

      </footer>
   </div>

  


  

  </body>
</html>


