<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Retrieve annotation data from an Ensembl based package — exonsBy • ensembldb</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Retrieve annotation data from an Ensembl based package — exonsBy" />
<meta property="og:description" content="Retrieve gene/transcript/exons annotations stored in an Ensembl based
  database package generated with the makeEnsembldbPackage
  function. Parameter filter enables to define filters to
  retrieve only specific data. Alternatively, a global filter might be
  added to the EnsDb object using the addFilter
  method." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ensembldb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.11.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/ensembldb.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/MySQL-backend.html">Using a MySQL server backend</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping-use-cases.html">Use cases for coordinate mapping with ensembldb</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping.html">Mapping between genome, transcript and protein coordinates</a>
    </li>
    <li>
      <a href="../articles/proteins.html">Querying protein features</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jorainer/ensembldb/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Retrieve annotation data from an Ensembl based package</h1>
    
    <div class="hidden name"><code>EnsDb-exonsBy.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Retrieve gene/transcript/exons annotations stored in an Ensembl based
  database package generated with the <code><a href='makeEnsemblDbPackage.html'>makeEnsembldbPackage</a></code>
  function. Parameter <code>filter</code> enables to define filters to
  retrieve only specific data. Alternatively, a global filter might be
  added to the <code>EnsDb</code> object using the <code><a href='global-filters.html'>addFilter</a></code>
  method.</p>
    </div>

    <pre class="usage"><span class='co'># S4 method for EnsDb</span>
<span class='fu'>exons</span>(<span class='no'>x</span>, <span class='kw'>columns</span> <span class='kw'>=</span> <span class='fu'><a href='EnsDb-class.html'>listColumns</a></span>(<span class='no'>x</span>,<span class='st'>"exon"</span>),
        <span class='kw'>filter</span> <span class='kw'>=</span> <span class='fu'>AnnotationFilterList</span>(), <span class='no'>order.by</span>,
        <span class='kw'>order.type</span> <span class='kw'>=</span> <span class='st'>"asc"</span>, <span class='kw'>return.type</span> <span class='kw'>=</span> <span class='st'>"GRanges"</span>)

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>exonsBy</span>(<span class='no'>x</span>, <span class='kw'>by</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"tx"</span>, <span class='st'>"gene"</span>),
        <span class='kw'>columns</span> <span class='kw'>=</span> <span class='fu'><a href='EnsDb-class.html'>listColumns</a></span>(<span class='no'>x</span>, <span class='st'>"exon"</span>), <span class='kw'>filter</span> <span class='kw'>=</span>
        <span class='fu'>AnnotationFilterList</span>(), <span class='kw'>use.names</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>intronsByTranscript</span>(<span class='no'>x</span>, <span class='no'>...</span>, <span class='kw'>use.names</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>exonsByOverlaps</span>(<span class='no'>x</span>, <span class='no'>ranges</span>, <span class='kw'>maxgap</span> <span class='kw'>=</span> -<span class='fl'>1L</span>, <span class='kw'>minoverlap</span> <span class='kw'>=</span> <span class='fl'>0L</span>,
        <span class='kw'>type</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"any"</span>, <span class='st'>"start"</span>, <span class='st'>"end"</span>), <span class='kw'>columns</span> <span class='kw'>=</span> <span class='fu'><a href='EnsDb-class.html'>listColumns</a></span>(<span class='no'>x</span>, <span class='st'>"exon"</span>),
        <span class='kw'>filter</span> <span class='kw'>=</span> <span class='fu'>AnnotationFilterList</span>())

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>transcripts</span>(<span class='no'>x</span>, <span class='kw'>columns</span> <span class='kw'>=</span> <span class='fu'><a href='EnsDb-class.html'>listColumns</a></span>(<span class='no'>x</span>, <span class='st'>"tx"</span>),
        <span class='kw'>filter</span> <span class='kw'>=</span> <span class='fu'>AnnotationFilterList</span>(), <span class='no'>order.by</span>, <span class='kw'>order.type</span> <span class='kw'>=</span> <span class='st'>"asc"</span>,
        <span class='kw'>return.type</span> <span class='kw'>=</span> <span class='st'>"GRanges"</span>)

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>transcriptsBy</span>(<span class='no'>x</span>, <span class='kw'>by</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"gene"</span>, <span class='st'>"exon"</span>),
        <span class='kw'>columns</span> <span class='kw'>=</span> <span class='fu'><a href='EnsDb-class.html'>listColumns</a></span>(<span class='no'>x</span>, <span class='st'>"tx"</span>), <span class='kw'>filter</span> <span class='kw'>=</span> <span class='fu'>AnnotationFilterList</span>())

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>transcriptsByOverlaps</span>(<span class='no'>x</span>, <span class='no'>ranges</span>, <span class='kw'>maxgap</span> <span class='kw'>=</span> -<span class='fl'>1L</span>,
        <span class='kw'>minoverlap</span> <span class='kw'>=</span> <span class='fl'>0L</span>, <span class='kw'>type</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"any"</span>, <span class='st'>"start"</span>, <span class='st'>"end"</span>),
        <span class='kw'>columns</span> <span class='kw'>=</span> <span class='fu'><a href='EnsDb-class.html'>listColumns</a></span>(<span class='no'>x</span>, <span class='st'>"tx"</span>), <span class='kw'>filter</span> <span class='kw'>=</span> <span class='fu'>AnnotationFilterList</span>())

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>promoters</span>(<span class='no'>x</span>, <span class='kw'>upstream</span> <span class='kw'>=</span> <span class='fl'>2000</span>, <span class='kw'>downstream</span> <span class='kw'>=</span> <span class='fl'>200</span>,
        <span class='kw'>use.names</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='no'>...</span>)

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>genes</span>(<span class='no'>x</span>, <span class='kw'>columns</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fu'><a href='EnsDb-class.html'>listColumns</a></span>(<span class='no'>x</span>, <span class='st'>"gene"</span>), <span class='st'>"entrezid"</span>),
        <span class='kw'>filter</span> <span class='kw'>=</span> <span class='fu'>AnnotationFilterList</span>(), <span class='no'>order.by</span>, <span class='kw'>order.type</span> <span class='kw'>=</span> <span class='st'>"asc"</span>,
        <span class='kw'>return.type</span> <span class='kw'>=</span> <span class='st'>"GRanges"</span>)

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>disjointExons</span>(<span class='no'>x</span>, <span class='kw'>aggregateGenes</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
        <span class='kw'>includeTranscripts</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>filter</span> <span class='kw'>=</span> <span class='fu'>AnnotationFilterList</span>(), <span class='no'>...</span>)

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>cdsBy</span>(<span class='no'>x</span>, <span class='kw'>by</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"tx"</span>, <span class='st'>"gene"</span>), <span class='kw'>columns</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
        <span class='kw'>filter</span> <span class='kw'>=</span> <span class='fu'>AnnotationFilterList</span>(), <span class='kw'>use.names</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>fiveUTRsByTranscript</span>(<span class='no'>x</span>, <span class='kw'>columns</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
        <span class='kw'>filter</span> <span class='kw'>=</span> <span class='fu'>AnnotationFilterList</span>())

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>threeUTRsByTranscript</span>(<span class='no'>x</span>, <span class='kw'>columns</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
        <span class='kw'>filter</span> <span class='kw'>=</span> <span class='fu'>AnnotationFilterList</span>())

<span class='co'># S4 method for GRangesList</span>
<span class='fu'>toSAF</span>(<span class='no'>x</span>, <span class='no'>...</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>...</th>
      <td><p>For <code>promoters</code>: additional arguments to be passed to the
    <code>transcripts</code> method. For <code>intronsByTranscript</code>:
    additional arguments such as <code>filter</code>.</p></td>
    </tr>
    <tr>
      <th>aggregateGenes</th>
      <td><p>For <code>disjointExons</code>: When <code>FALSE</code> (default) exon fragments
    that overlap multiple genes are dropped. When <code>TRUE</code>, all
    fragments are kept and the <code>gene_id</code> metadata column includes
    all gene IDs that overlap the exon fragment.</p></td>
    </tr>
    <tr>
      <th>by</th>
      <td><p>For <code>exonsBy</code>: wheter exons sould be fetched by genes
    or by transcripts; as in the corresponding function of the
    <code>GenomicFeatures</code> package.
    For <code>transcriptsBy</code>: whether
    transcripts should be fetched by genes or by exons; fetching
    transcripts by cds as supported by the
    <code><a href='https://rdrr.io/pkg/GenomicFeatures/man/transcriptsBy.html'>transcriptsBy</a></code> method in the
    <code>GenomicFeatures</code> package is currently not implemented.
    For <code>cdsBy</code>: whether cds should be fetched by transcript of by
    gene.</p></td>
    </tr>
    <tr>
      <th>columns</th>
      <td><p>Columns to be retrieved from the database tables.</p>
<p>Default values for <code>genes</code>  are all columns from the <code>gene</code>
    database table, for <code>exons</code> and <code>exonsBy</code> the column names of
    the <code>exon</code> database table table and for <code>transcript</code> and
    <code>transcriptBy</code> the columns of the <code>tx</code> data base table
    (see details below for more information).</p>
<p>Note that any of the column names of the database tables can be
    submitted to any of the methods (use <code><a href='EnsDb-class.html'>listTables</a></code> or
    <code><a href='EnsDb-class.html'>listColumns</a></code> methods for a complete list of allowed
    column names).</p>
<p>For <code>cdsBy</code>: this argument is only supported for for
    <code>by="tx"</code>.</p></td>
    </tr>
    <tr>
      <th>downstream</th>
      <td><p>For method <code>promoters</code>: the number of nucleotides downstream of
    the transcription start site that should be included in the promoter region.</p></td>
    </tr>
    <tr>
      <th>filter</th>
      <td><p>A filter describing which results to retrieve from the database. Can
    be a single object extending
    <code><a href='https://rdrr.io/pkg/AnnotationFilter/man/AnnotationFilter.html'>AnnotationFilter</a></code>, an
    <code><a href='https://rdrr.io/pkg/AnnotationFilter/man/AnnotationFilterList.html'>AnnotationFilterList</a></code> object
    combining several such objects or a <code>formula</code> representing a
    filter expression (see examples below or
    <code><a href='https://rdrr.io/pkg/AnnotationFilter/man/AnnotationFilter.html'>AnnotationFilter</a></code> for more
    details). Use the <code>supportedFilters</code> method to get an
    overview of supported filter classes and related fields.</p></td>
    </tr>
    <tr>
      <th>includeTranscripts</th>
      <td><p>For <code>disjointExons</code>: When <code>TRUE</code> (default) a
    <code>tx_name</code> metadata column is included that lists all transcript
    IDs that overlap the exon fragment. Note: this is different to the
    <code><a href='https://rdrr.io/pkg/GenomicFeatures/man/disjointExons.html'>disjointExons</a></code> function in the
    <code>GenomicFeatures</code> package, that lists the transcript names, not
    IDs.</p></td>
    </tr>
    <tr>
      <th>maxgap</th>
      <td><p>For <code>exonsByOverlaps</code> and <code>transcriptsByOverlaps</code>: see
    <code>exonsByOverlaps</code> in <code>GenomicFeatures</code> for more
    information.</p></td>
    </tr>
    <tr>
      <th>minoverlap</th>
      <td><p>For <code>exonsByOverlaps</code> and <code>transcriptsByOverlaps</code>: see
    <code>exonsByOverlaps</code> in <code>GenomicFeatures</code> for more
    information.</p></td>
    </tr>
    <tr>
      <th>order.by</th>
      <td><p>Character vector specifying the column(s) by which the result should
    be ordered. This can be either in the form of
    <code>"gene_id, seq_name"</code> or <code><a href='https://rdrr.io/r/base/c.html'>c("gene_id", "seq_name")</a></code>.</p></td>
    </tr>
    <tr>
      <th>order.type</th>
      <td><p>If the results should be ordered ascending
    (<code>asc</code>, default) or descending (<code>desc</code>).</p></td>
    </tr>
    <tr>
      <th>ranges</th>
      <td><p>For <code>exonsByOverlaps</code> and <code>transcriptsByOverlaps</code>: a
    <code>GRanges</code> object specifying the genomic regions.</p></td>
    </tr>
    <tr>
      <th>return.type</th>
      <td><p>Type of the returned object. Can be either
    <code>"data.frame"</code>, <code>"DataFrame"</code> or <code>"GRanges"</code>. In the
    latter case the return object will be a <code>GRanges</code> object with
    the GRanges specifying the chromosomal start and end coordinates of
    the feature (gene, transcript or exon, depending whether <code>genes</code>,
    <code>transcripts</code> or <code>exons</code> was called). All additional
    columns are added as metadata columns to the GRanges object.</p></td>
    </tr>
    <tr>
      <th>type</th>
      <td><p>For <code>exonsByOverlaps</code> and <code>transcriptsByOverlaps</code>: see
    <code>exonsByOverlaps</code> in <code>GenomicFeatures</code> for more
    information.</p></td>
    </tr>
    <tr>
      <th>upstream</th>
      <td><p>For method <code>promoters</code>: the number of nucleotides upstream of
    the transcription start site that should be included in the promoter region.</p></td>
    </tr>
    <tr>
      <th>use.names</th>
      <td><p>For <code>cdsBy</code> and <code>exonsBy</code>: only for <code>by="gene"</code>: use
    the names of the genes instead of their IDs as names of the resulting
    <code>GRangesList</code>.</p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>For <code>toSAF</code> a <code>GRangesList</code> object.
    For all other methods an <code>EnsDb</code> instance.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="methods-and-functions"><a class="anchor" href="#methods-and-functions"></a>Methods and Functions</h2>

    <p></p><dl'>

    <dt>exons</dt><dd><p>Retrieve exon information from the database. Additional
      columns from transcripts or genes associated with the exons can be specified
      and are added to the respective exon annotation.</p></dd>

    <dt>exonsBy</dt><dd><p>Retrieve exons grouped by transcript or by gene. This
      function returns a <code>GRangesList</code> as does the analogous function
      in the <code>GenomicFeatures</code> package. Using the <code>columns</code>
      parameter it is possible to determine which additional values should
      be retrieved from the database. These will be included in the
      <code>GRanges</code> object for the exons as metadata columns.
      The exons in the inner <code>GRanges</code> are ordered by the exon
      index within the transcript (if <code>by="tx"</code>), or increasingly by the
      chromosomal start position of the exon or decreasingly by the chromosomal end
      position of the exon depending whether the gene is encoded on the
      + or - strand (for <code>by="gene"</code>).
      The <code>GRanges</code> in the <code>GRangesList</code> will be ordered by
      the name of the gene or transcript.</p></dd>

    <dt>intronsByTranscript</dt><dd><p>Retrieve introns by transcripts. Filters can also be passed to the
      function. For more information see the <code>intronsByTranscript</code>
      method in the <code>GenomicFeatures</code> package.</p></dd>

    <dt>exonsByOverlaps</dt><dd><p>Retrieve exons overlapping specified genomic ranges. For
      more information see the
      <code>exonsByOverlaps</code> method in the
      <code>GenomicFeatures</code> package. The functionality is to some
      extent similar and redundant to the <code>exons</code> method in
      combination with <code>GRangesFilter</code> filter.</p></dd>

    <dt>transcripts</dt><dd><p>Retrieve transcript information from the database. Additional
      columns from genes or exons associated with the transcripts can be specified
      and are added to the respective transcript annotation.</p></dd>

    <dt>transcriptsBy</dt><dd><p>Retrieve transcripts grouped by gene or exon. This
      function returns a <code>GRangesList</code> as does the analogous function
      in the <code>GenomicFeatures</code> package. Using the <code>columns</code>
      parameter it is possible to determine which additional values should
      be retrieved from the database. These will be included in the
      <code>GRanges</code> object for the transcripts as metadata columns.
      The transcripts in the inner <code>GRanges</code> are ordered  increasingly by the
      chromosomal start position of the transcript for genes encoded on
      the + strand and in a decreasing manner by the chromosomal end
      position of the transcript for genes encoded on the - strand.
      The <code>GRanges</code> in the <code>GRangesList</code> will be ordered by
      the name of the gene or exon.</p></dd>

    <dt>transcriptsByOverlaps</dt><dd><p>Retrieve transcripts overlapping specified genomic ranges. For
      more information see
      <code><a href='https://rdrr.io/pkg/GenomicFeatures/man/transcriptsByOverlaps.html'>transcriptsByOverlaps</a></code> method in the
      <code>GenomicFeatures</code> package. The functionality is to some
      extent similar and redundant to the <code>transcripts</code> method in
      combination with <code>GRangesFilter</code> filter.</p></dd>

    <dt>promoters</dt><dd><p>Retrieve promoter information from the database. Additional
      columns from genes or exons associated with the promoters can be specified
      and are added to the respective promoter annotation.</p></dd>
    <dt>genes</dt><dd><p>Retrieve gene information from the database. Additional columns
      from transcripts or exons associated with the genes can be
      specified and are added to the respective gene annotation. Note
      that column <code>"entrezid"</code> is a <code>list</code> of Entrezgene
      identifiers to accomodate the potential 1:n mapping between
      Ensembl genes and Entrezgene IDs.</p></dd>

    <dt>disjointExons</dt><dd><p>This method is identical to
      <code><a href='https://rdrr.io/pkg/GenomicFeatures/man/disjointExons.html'>disjointExons</a></code> defined in the
      <code>GenomicFeatures</code> package. It creates a <code>GRanges</code> of
      non-overlapping exon parts with metadata columns of <code>gene_id</code>
      and <code>exonic_part</code>. Exon parts that overlap more than one gene
      can be dropped with <code>aggregateGenes=FALSE</code>.</p></dd>

    <dt>cdsBy</dt><dd><p>Returns the coding region grouped either by transcript or by
      gene. Each element in the <code>GRangesList</code> represents the cds
      for one transcript or gene, with the individual ranges
      corresponding to the coding part of its exons.
      For <code>by="tx"</code> additional annotation columns can be added to
      the individual <code>GRanges</code> (in addition to the default columns
      <code>exon_id</code> and <code>exon_rank</code>).
      Note that the <code>GRangesList</code> is sorted by its names.</p></dd>

    <dt>fiveUTRsByTranscript</dt><dd><p>Returns the 5' untranslated region for protein coding
      transcripts.</p></dd>

    <dt>threeUTRsByTranscript</dt><dd><p>Returns the 3' untranslated region for protein coding
      transcripts.</p></dd>

    <dt>toSAF</dt><dd><p>Reformats a <code>GRangesList</code> object into a
      <code>data.frame</code> corresponding to a standard SAF (Simplified
      Annotation Format) file (i.e. with column names <code>"GeneID"</code>,
      <code>"Chr"</code>, <code>"Start"</code>, <code>"End"</code> and
      <code>"Strand"</code>). Note: this method makes only sense on a
      <code>GRangesList</code> that groups features (exons, transcripts) by gene.</p></dd>

  
</dl>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>A detailed description of all database tables and the associated
  attributes/column names is also given in the vignette of this package.
  An overview of the columns is given below:</p><dl'>
    <dt>gene_id</dt><dd><p>the Ensembl gene ID of the gene.</p></dd>
    <dt>gene_name</dt><dd><p>the name of the gene (in most cases its official symbol).</p></dd>
    <dt>entrezid</dt><dd><p>the NCBI Entrezgene ID of the gene. Note that this
      column contains a <code>list</code> of Entrezgene identifiers to
      accommodate the potential 1:n mapping between Ensembl genes and
      Entrezgene IDs.</p></dd>
    <dt>gene_biotype</dt><dd><p>the biotype of the gene.</p></dd>
    <dt>gene_seq_start</dt><dd><p>the start coordinate of the gene on the
      sequence (usually a chromosome).</p></dd>
    <dt>gene_seq_end</dt><dd><p>the end coordinate of the gene.</p></dd>
    <dt>seq_name</dt><dd><p>the name of the sequence the gene is encoded
      (usually a chromosome).</p></dd>
    <dt>seq_strand</dt><dd><p>the strand on which the gene is encoded</p></dd>
    <dt>seq_coord_system</dt><dd><p>the coordinate system of the sequence.</p></dd>
    <dt>tx_id</dt><dd><p>the Ensembl transcript ID.</p></dd>
    <dt>tx_biotype</dt><dd><p>the biotype of the transcript.</p></dd>
    <dt>tx_seq_start</dt><dd><p>the chromosomal start coordinate of the transcript.</p></dd>
    <dt>tx_seq_end</dt><dd><p>the chromosomal end coordinate of the transcript.</p></dd>
    <dt>tx_cds_seq_start</dt><dd><p>the start coordinate of the coding region of
      the transcript (NULL for non-coding transcripts).</p></dd>
    <dt>tx_cds_seq_end</dt><dd><p>the end coordinate of the coding region.</p></dd>
    <dt>gc_content</dt><dd><p>the G and C nucleotide content of the transcript's
    sequence expressed as a percentage (i.e. between 0 and 100).</p></dd>
    <dt>exon_id</dt><dd><p>the ID of the exon. In Ensembl, each exon specified
      by a unique chromosomal start and end position has its own
      ID. Thus, the same exon might be part of several transcripts.</p></dd>
    <dt>exon_seq_start</dt><dd><p>the chromosomal start coordinate of the exon.</p></dd>
    <dt>exon_seq_end</dt><dd><p>the chromosomal end coordinate of the exon.</p></dd>
    <dt>exon_idx</dt><dd><p>the index of the exon in the transcript model. As
      noted above, an exon can be part of several transcripts and thus
      its position inside these transcript might differ.</p></dd>
  
</dl>

<p>Many <code>EnsDb</code> databases provide also protein related
  annotations. See <code><a href='ProteinFunctionality.html'>listProteinColumns</a></code> for more information.</p>
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>Ensembl defines genes not only on standard chromosomes, but also on
  patched chromosomes and chromosome variants. Thus it might be
  advisable to restrict the queries to just those chromosomes of
  interest (e.g. by specifying a <code>SeqNameFilter(c(1:22, "X", "Y"))</code>).
  In addition, also so called LRG genes (Locus Reference Genomic) are defined in
  Ensembl. Their gene id starts with LRG instead of ENS for Ensembl
  genes, thus, a filter can be applied to specifically select those
  genes or exclude those genes (see examples below).</p>
<p>Depending on the value of the global option
  <code>"ucscChromosomeNames"</code> (use
  <code><a href='https://rdrr.io/r/base/options.html'>getOption(ucscChromosomeNames, FALSE)</a></code> to get its value or
  <code><a href='https://rdrr.io/r/base/options.html'>option(ucscChromosomeNames=TRUE)</a></code> to change its value)
  the sequence/chromosome names of the returned <code>GRanges</code> objects
  or provided in the returned <code>data.frame</code> or <code>DataFrame</code>
  correspond to Ensembl chromosome names (if value is <code>FALSE</code>) or
  UCSC chromosome names (if <code>TRUE</code>). This ensures a better
  integration with the <code>Gviz</code> package, in which this option is set
  by default to <code>TRUE</code>.</p>
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>For <code>exons</code>, <code>transcripts</code> and <code>genes</code>,
  a <code>data.frame</code>, <code>DataFrame</code>
  or a <code>GRanges</code>, depending on the value of the
  <code>return.type</code> parameter. The result is ordered as specified by
  the parameter <code>order.by</code> or, if not provided, by <code>seq_name</code>
  and chromosomal start coordinate, but NOT by any ordering of values in
  eventually submitted filter objects.</p>
<p>For <code>exonsBy</code>, <code>transcriptsBy</code>:
  a <code>GRangesList</code>, depending on the value of the
  <code>return.type</code> parameter. The results are ordered by the value of the
  <code>by</code> parameter.</p>
<p>For <code>exonsByOverlaps</code> and <code>transcriptsByOverlaps</code>: a
  <code>GRanges</code> with the exons or transcripts overlapping the specified
  regions.</p>
<p>For <code>toSAF</code>: a <code>data.frame</code> with column names
  <code>"GeneID"</code> (the group name from the <code>GRangesList</code>, i.e. the
  ID by which the <code>GRanges</code> are split), <code>"Chr"</code> (the seqnames
  from the <code>GRanges</code>), <code>"Start"</code> (the start coordinate),
  <code>"End"</code> (the end coordinate) and <code>"Strand"</code> (the strand).</p>
<p>For <code>disjointExons</code>: a <code>GRanges</code> of non-overlapping exon
  parts.</p>
<p>For <code>cdsBy</code>: a <code>GRangesList</code> with <code>GRanges</code> per either
  transcript or exon specifying the start and end coordinates of the
  coding region of the transcript or gene.</p>
<p>For <code>fiveUTRsByTranscript</code>: a <code>GRangesList</code> with
  <code>GRanges</code> for each protein coding transcript representing the
  start and end coordinates of full or partial exons that constitute the
  5' untranslated region of the transcript.</p>
<p>For <code>threeUTRsByTranscript</code>: a <code>GRangesList</code> with
  <code>GRanges</code> for each protein coding transcript representing the
  start and end coordinates of full or partial exons that constitute the
  3' untranslated region of the transcript.</p>
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>While it is possible to request values from a column <code>"tx_name"</code>
  (with the <code>columns</code> argument), no such column is present in the
  database. The returned values correspond to the ID of the transcripts.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code>supportedFilters</code> to get an overview of supported filters.
  <code><a href='makeEnsemblDbPackage.html'>makeEnsembldbPackage</a></code>,
  <code><a href='EnsDb-class.html'>listColumns</a></code>, <code><a href='EnsDb-lengths.html'>lengthOf</a></code></p>
<p><code><a href='global-filters.html'>addFilter</a></code> for globally adding filters to an <code>EnsDb</code>
  object.</p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='fu'><a href='https://rdrr.io/r/base/library.html'>library</a></span>(<span class='no'>EnsDb.Hsapiens.v86</span>)
<span class='no'>edb</span> <span class='kw'>&lt;-</span> <span class='no'>EnsDb.Hsapiens.v86</span>

<span class='co'>######   genes</span>
<span class='co'>##</span>
<span class='co'>## Get all genes encoded on chromosome Y</span>
<span class='no'>AllY</span> <span class='kw'>&lt;-</span> <span class='fu'>genes</span>(<span class='no'>edb</span>, <span class='kw'>filter</span> <span class='kw'>=</span> <span class='fu'>SeqNameFilter</span>(<span class='st'>"Y"</span>))
<span class='no'>AllY</span></div><div class='output co'>#&gt; GRanges object with 523 ranges and 6 metadata columns:
#&gt;                   seqnames            ranges strand |         gene_id
#&gt;                      &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
#&gt;   ENSG00000251841        Y   2784749-2784853      + | ENSG00000251841
#&gt;   ENSG00000184895        Y   2786855-2787699      - | ENSG00000184895
#&gt;   ENSG00000237659        Y   2789827-2790328      + | ENSG00000237659
#&gt;   ENSG00000232195        Y   2827982-2828218      + | ENSG00000232195
#&gt;   ENSG00000129824        Y   2841486-2932000      + | ENSG00000129824
#&gt;               ...      ...               ...    ... .             ...
#&gt;   ENSG00000224240        Y 26549425-26549743      + | ENSG00000224240
#&gt;   ENSG00000227629        Y 26586642-26591601      - | ENSG00000227629
#&gt;   ENSG00000237917        Y 26594851-26634652      - | ENSG00000237917
#&gt;   ENSG00000231514        Y 26626520-26627159      - | ENSG00000231514
#&gt;   ENSG00000235857        Y 56855244-56855488      + | ENSG00000235857
#&gt;                     gene_name           gene_biotype seq_coord_system
#&gt;                   &lt;character&gt;            &lt;character&gt;      &lt;character&gt;
#&gt;   ENSG00000251841  RNU6-1334P                  snRNA       chromosome
#&gt;   ENSG00000184895         SRY         protein_coding       chromosome
#&gt;   ENSG00000237659  RNASEH2CP1   processed_pseudogene       chromosome
#&gt;   ENSG00000232195    TOMM22P2   processed_pseudogene       chromosome
#&gt;   ENSG00000129824      RPS4Y1         protein_coding       chromosome
#&gt;               ...         ...                    ...              ...
#&gt;   ENSG00000224240     CYCSP49   processed_pseudogene       chromosome
#&gt;   ENSG00000227629  SLC25A15P1 unprocessed_pseudogene       chromosome
#&gt;   ENSG00000237917     PARP4P1 unprocessed_pseudogene       chromosome
#&gt;   ENSG00000231514     FAM58CP   processed_pseudogene       chromosome
#&gt;   ENSG00000235857     CTBP2P1   processed_pseudogene       chromosome
#&gt;                        symbol entrezid
#&gt;                   &lt;character&gt;   &lt;list&gt;
#&gt;   ENSG00000251841  RNU6-1334P       NA
#&gt;   ENSG00000184895         SRY     6736
#&gt;   ENSG00000237659  RNASEH2CP1       NA
#&gt;   ENSG00000232195    TOMM22P2       NA
#&gt;   ENSG00000129824      RPS4Y1     6192
#&gt;               ...         ...      ...
#&gt;   ENSG00000224240     CYCSP49       NA
#&gt;   ENSG00000227629  SLC25A15P1       NA
#&gt;   ENSG00000237917     PARP4P1       NA
#&gt;   ENSG00000231514     FAM58CP       NA
#&gt;   ENSG00000235857     CTBP2P1       NA
#&gt;   -------
#&gt;   seqinfo: 1 sequence from GRCh38 genome</div><div class='input'>
<span class='co'>## Return the result as a DataFrame; also, we use a filter expression here</span>
<span class='co'>## to define which features to extract from the database.</span>
<span class='no'>AllY.granges</span> <span class='kw'>&lt;-</span> <span class='fu'>genes</span>(<span class='no'>edb</span>,
                      <span class='kw'>filter</span> <span class='kw'>=</span> ~ <span class='no'>seq_name</span> <span class='kw'>==</span> <span class='st'>"Y"</span>,
                      <span class='kw'>return.type</span><span class='kw'>=</span><span class='st'>"DataFrame"</span>)
<span class='no'>AllY.granges</span></div><div class='output co'>#&gt; DataFrame with 523 rows and 10 columns
#&gt;             gene_id   gene_name           gene_biotype gene_seq_start
#&gt;         &lt;character&gt; &lt;character&gt;            &lt;character&gt;      &lt;integer&gt;
#&gt; 1   ENSG00000251841  RNU6-1334P                  snRNA        2784749
#&gt; 2   ENSG00000184895         SRY         protein_coding        2786855
#&gt; 3   ENSG00000237659  RNASEH2CP1   processed_pseudogene        2789827
#&gt; 4   ENSG00000232195    TOMM22P2   processed_pseudogene        2827982
#&gt; 5   ENSG00000129824      RPS4Y1         protein_coding        2841486
#&gt; ...             ...         ...                    ...            ...
#&gt; 519 ENSG00000224240     CYCSP49   processed_pseudogene       26549425
#&gt; 520 ENSG00000227629  SLC25A15P1 unprocessed_pseudogene       26586642
#&gt; 521 ENSG00000237917     PARP4P1 unprocessed_pseudogene       26594851
#&gt; 522 ENSG00000231514     FAM58CP   processed_pseudogene       26626520
#&gt; 523 ENSG00000235857     CTBP2P1   processed_pseudogene       56855244
#&gt;     gene_seq_end    seq_name seq_strand seq_coord_system      symbol entrezid
#&gt;        &lt;integer&gt; &lt;character&gt;  &lt;integer&gt;      &lt;character&gt; &lt;character&gt;   &lt;list&gt;
#&gt; 1        2784853           Y          1       chromosome  RNU6-1334P       NA
#&gt; 2        2787699           Y         -1       chromosome         SRY     6736
#&gt; 3        2790328           Y          1       chromosome  RNASEH2CP1       NA
#&gt; 4        2828218           Y          1       chromosome    TOMM22P2       NA
#&gt; 5        2932000           Y          1       chromosome      RPS4Y1     6192
#&gt; ...          ...         ...        ...              ...         ...      ...
#&gt; 519     26549743           Y          1       chromosome     CYCSP49       NA
#&gt; 520     26591601           Y         -1       chromosome  SLC25A15P1       NA
#&gt; 521     26634652           Y         -1       chromosome     PARP4P1       NA
#&gt; 522     26627159           Y         -1       chromosome     FAM58CP       NA
#&gt; 523     56855488           Y          1       chromosome     CTBP2P1       NA</div><div class='input'>
<span class='co'>## Include all transcripts of the gene and their chromosomal</span>
<span class='co'>## coordinates, sort by chrom start of transcripts and return as</span>
<span class='co'>## GRanges.</span>
<span class='no'>AllY.granges.tx</span> <span class='kw'>&lt;-</span> <span class='fu'>genes</span>(<span class='no'>edb</span>,
                         <span class='kw'>filter</span> <span class='kw'>=</span> <span class='fu'>SeqNameFilter</span>(<span class='st'>"Y"</span>),
                         <span class='kw'>columns</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"gene_id"</span>, <span class='st'>"seq_name"</span>,
                                     <span class='st'>"seq_strand"</span>, <span class='st'>"tx_id"</span>, <span class='st'>"tx_biotype"</span>,
                                     <span class='st'>"tx_seq_start"</span>, <span class='st'>"tx_seq_end"</span>),
                         <span class='kw'>order.by</span> <span class='kw'>=</span> <span class='st'>"tx_seq_start"</span>)
<span class='no'>AllY.granges.tx</span></div><div class='output co'>#&gt; GRanges object with 740 ranges and 5 metadata columns:
#&gt;                   seqnames            ranges strand |         gene_id
#&gt;                      &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
#&gt;   ENSG00000251841        Y   2784749-2784853      + | ENSG00000251841
#&gt;   ENSG00000184895        Y   2786855-2787699      - | ENSG00000184895
#&gt;   ENSG00000237659        Y   2789827-2790328      + | ENSG00000237659
#&gt;   ENSG00000232195        Y   2827982-2828218      + | ENSG00000232195
#&gt;   ENSG00000129824        Y   2841486-2932000      + | ENSG00000129824
#&gt;               ...      ...               ...    ... .             ...
#&gt;   ENSG00000224240        Y 26549425-26549743      + | ENSG00000224240
#&gt;   ENSG00000227629        Y 26586642-26591601      - | ENSG00000227629
#&gt;   ENSG00000237917        Y 26594851-26634652      - | ENSG00000237917
#&gt;   ENSG00000231514        Y 26626520-26627159      - | ENSG00000231514
#&gt;   ENSG00000235857        Y 56855244-56855488      + | ENSG00000235857
#&gt;                             tx_id             tx_biotype tx_seq_start
#&gt;                       &lt;character&gt;            &lt;character&gt;    &lt;integer&gt;
#&gt;   ENSG00000251841 ENST00000516032                  snRNA      2784749
#&gt;   ENSG00000184895 ENST00000383070         protein_coding      2786855
#&gt;   ENSG00000237659 ENST00000454281   processed_pseudogene      2789827
#&gt;   ENSG00000232195 ENST00000430735   processed_pseudogene      2827982
#&gt;   ENSG00000129824 ENST00000250784         protein_coding      2841486
#&gt;               ...             ...                    ...          ...
#&gt;   ENSG00000224240 ENST00000420810   processed_pseudogene     26549425
#&gt;   ENSG00000227629 ENST00000456738 unprocessed_pseudogene     26586642
#&gt;   ENSG00000237917 ENST00000435945 unprocessed_pseudogene     26594851
#&gt;   ENSG00000231514 ENST00000435741   processed_pseudogene     26626520
#&gt;   ENSG00000235857 ENST00000431853   processed_pseudogene     56855244
#&gt;                   tx_seq_end
#&gt;                    &lt;integer&gt;
#&gt;   ENSG00000251841    2784853
#&gt;   ENSG00000184895    2787699
#&gt;   ENSG00000237659    2790328
#&gt;   ENSG00000232195    2828218
#&gt;   ENSG00000129824    2867268
#&gt;               ...        ...
#&gt;   ENSG00000224240   26549743
#&gt;   ENSG00000227629   26591601
#&gt;   ENSG00000237917   26634652
#&gt;   ENSG00000231514   26627159
#&gt;   ENSG00000235857   56855488
#&gt;   -------
#&gt;   seqinfo: 1 sequence from GRCh38 genome</div><div class='input'>


<span class='co'>######   transcripts</span>
<span class='co'>##</span>
<span class='co'>## Get all transcripts of a gene</span>
<span class='no'>Tx</span> <span class='kw'>&lt;-</span> <span class='fu'>transcripts</span>(<span class='no'>edb</span>,
                  <span class='kw'>filter</span> <span class='kw'>=</span> <span class='fu'>GeneIdFilter</span>(<span class='st'>"ENSG00000184895"</span>),
                  <span class='kw'>order.by</span> <span class='kw'>=</span> <span class='st'>"tx_seq_start"</span>)
<span class='no'>Tx</span></div><div class='output co'>#&gt; GRanges object with 1 range and 6 metadata columns:
#&gt;                   seqnames          ranges strand |           tx_id
#&gt;                      &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
#&gt;   ENST00000383070        Y 2786855-2787699      - | ENST00000383070
#&gt;                       tx_biotype tx_cds_seq_start tx_cds_seq_end
#&gt;                      &lt;character&gt;        &lt;integer&gt;      &lt;integer&gt;
#&gt;   ENST00000383070 protein_coding          2786989        2787603
#&gt;                           gene_id         tx_name
#&gt;                       &lt;character&gt;     &lt;character&gt;
#&gt;   ENST00000383070 ENSG00000184895 ENST00000383070
#&gt;   -------
#&gt;   seqinfo: 1 sequence from GRCh38 genome</div><div class='input'>
<span class='co'>## Get all transcripts of two genes along with some information on the</span>
<span class='co'>## gene and transcript</span>
<span class='no'>Tx</span> <span class='kw'>&lt;-</span> <span class='fu'>transcripts</span>(<span class='no'>edb</span>,
                  <span class='kw'>filter</span> <span class='kw'>=</span> <span class='fu'>GeneIdFilter</span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"ENSG00000184895"</span>,
                                          <span class='st'>"ENSG00000092377"</span>)),
                  <span class='kw'>columns</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"gene_id"</span>, <span class='st'>"gene_seq_start"</span>, <span class='st'>"gene_seq_end"</span>,
                              <span class='st'>"gene_biotype"</span>, <span class='st'>"tx_biotype"</span>))
<span class='no'>Tx</span></div><div class='output co'>#&gt; GRanges object with 4 ranges and 6 metadata columns:
#&gt;                   seqnames          ranges strand |         gene_id
#&gt;                      &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
#&gt;   ENST00000383070        Y 2786855-2787699      - | ENSG00000184895
#&gt;   ENST00000383032        Y 6910686-7091683      + | ENSG00000092377
#&gt;   ENST00000355162        Y 6910686-7091683      + | ENSG00000092377
#&gt;   ENST00000346432        Y 6910686-7091683      + | ENSG00000092377
#&gt;                   gene_seq_start gene_seq_end   gene_biotype     tx_biotype
#&gt;                        &lt;integer&gt;    &lt;integer&gt;    &lt;character&gt;    &lt;character&gt;
#&gt;   ENST00000383070        2786855      2787699 protein_coding protein_coding
#&gt;   ENST00000383032        6910686      7091683 protein_coding protein_coding
#&gt;   ENST00000355162        6910686      7091683 protein_coding protein_coding
#&gt;   ENST00000346432        6910686      7091683 protein_coding protein_coding
#&gt;                             tx_id
#&gt;                       &lt;character&gt;
#&gt;   ENST00000383070 ENST00000383070
#&gt;   ENST00000383032 ENST00000383032
#&gt;   ENST00000355162 ENST00000355162
#&gt;   ENST00000346432 ENST00000346432
#&gt;   -------
#&gt;   seqinfo: 1 sequence from GRCh38 genome</div><div class='input'>
<span class='co'>######   promoters</span>
<span class='co'>##</span>
<span class='co'>## Get the bona-fide promoters (2k up- to 200nt downstream of TSS)</span>
<span class='fu'>promoters</span>(<span class='no'>edb</span>, <span class='kw'>filter</span> <span class='kw'>=</span> <span class='fu'>GeneIdFilter</span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"ENSG00000184895"</span>,
                                       <span class='st'>"ENSG00000092377"</span>)))</div><div class='output co'>#&gt; GRanges object with 4 ranges and 6 metadata columns:
#&gt;                   seqnames          ranges strand |           tx_id
#&gt;                      &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
#&gt;   ENST00000383070        Y 2787500-2789699      - | ENST00000383070
#&gt;   ENST00000383032        Y 6908686-6910885      + | ENST00000383032
#&gt;   ENST00000355162        Y 6908686-6910885      + | ENST00000355162
#&gt;   ENST00000346432        Y 6908686-6910885      + | ENST00000346432
#&gt;                       tx_biotype tx_cds_seq_start tx_cds_seq_end
#&gt;                      &lt;character&gt;        &lt;integer&gt;      &lt;integer&gt;
#&gt;   ENST00000383070 protein_coding          2786989        2787603
#&gt;   ENST00000383032 protein_coding          7025085        7091492
#&gt;   ENST00000355162 protein_coding          7025085        7091492
#&gt;   ENST00000346432 protein_coding          7025085        7091492
#&gt;                           gene_id         tx_name
#&gt;                       &lt;character&gt;     &lt;character&gt;
#&gt;   ENST00000383070 ENSG00000184895 ENST00000383070
#&gt;   ENST00000383032 ENSG00000092377 ENST00000383032
#&gt;   ENST00000355162 ENSG00000092377 ENST00000355162
#&gt;   ENST00000346432 ENSG00000092377 ENST00000346432
#&gt;   -------
#&gt;   seqinfo: 1 sequence from GRCh38 genome</div><div class='input'>
<span class='co'>######   exons</span>
<span class='co'>##</span>
<span class='co'>## Get all exons of protein coding transcript for the gene ENSG00000184895</span>
<span class='no'>Exon</span> <span class='kw'>&lt;-</span> <span class='fu'>exons</span>(<span class='no'>edb</span>,
              <span class='kw'>filter</span> <span class='kw'>=</span> ~ <span class='no'>gene_id</span> <span class='kw'>==</span> <span class='st'>"ENSG00000184895"</span> <span class='kw'>&amp;</span>
                  <span class='no'>tx_biotype</span> <span class='kw'>==</span> <span class='st'>"protein_coding"</span>,
              <span class='kw'>columns</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"gene_id"</span>, <span class='st'>"gene_seq_start"</span>, <span class='st'>"gene_seq_end"</span>,
                          <span class='st'>"tx_biotype"</span>, <span class='st'>"gene_biotype"</span>))
<span class='no'>Exon</span></div><div class='output co'>#&gt; GRanges object with 1 range and 6 metadata columns:
#&gt;                   seqnames          ranges strand |         gene_id
#&gt;                      &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
#&gt;   ENSE00001494622        Y 2786855-2787699      - | ENSG00000184895
#&gt;                   gene_seq_start gene_seq_end     tx_biotype   gene_biotype
#&gt;                        &lt;integer&gt;    &lt;integer&gt;    &lt;character&gt;    &lt;character&gt;
#&gt;   ENSE00001494622        2786855      2787699 protein_coding protein_coding
#&gt;                           exon_id
#&gt;                       &lt;character&gt;
#&gt;   ENSE00001494622 ENSE00001494622
#&gt;   -------
#&gt;   seqinfo: 1 sequence from GRCh38 genome</div><div class='input'>


<span class='co'>#####    exonsBy</span>
<span class='co'>##</span>
<span class='co'>## Get all exons for transcripts encoded on chromosomes X and Y.</span>
<span class='no'>ETx</span> <span class='kw'>&lt;-</span> <span class='fu'>exonsBy</span>(<span class='no'>edb</span>, <span class='kw'>by</span> <span class='kw'>=</span> <span class='st'>"tx"</span>,
               <span class='kw'>filter</span> <span class='kw'>=</span> <span class='fu'>SeqNameFilter</span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"X"</span>, <span class='st'>"Y"</span>)))
<span class='no'>ETx</span></div><div class='output co'>#&gt; GRangesList object of length 6829:
#&gt; $ENST00000014935
#&gt; GRanges object with 9 ranges and 2 metadata columns:
#&gt;       seqnames              ranges strand |         exon_id exon_rank
#&gt;          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt; &lt;integer&gt;
#&gt;   [1]        X 154409418-154410039      - | ENSE00001050512         1
#&gt;   [2]        X 154409112-154409196      - | ENSE00001050508         2
#&gt;   [3]        X 154405434-154405655      - | ENSE00001450968         3
#&gt;   [4]        X 154404995-154405083      - | ENSE00000678418         4
#&gt;   [5]        X 154404828-154404914      - | ENSE00003790806         5
#&gt;   [6]        X 154403522-154403622      - | ENSE00000678413         6
#&gt;   [7]        X 154403269-154403381      - | ENSE00003549633         7
#&gt;   [8]        X 154402942-154403190      - | ENSE00000678409         8
#&gt;   [9]        X 154401758-154402841      - | ENSE00001200781         9
#&gt;   -------
#&gt;   seqinfo: 2 sequences from GRCh38 genome
#&gt; 
#&gt; ...
#&gt; &lt;6828 more elements&gt;</div><div class='input'><span class='co'>## Get all exons for genes encoded on chromosome 1 to 22, X and Y and</span>
<span class='co'>## include additional annotation columns in the result</span>
<span class='no'>EGenes</span> <span class='kw'>&lt;-</span> <span class='fu'>exonsBy</span>(<span class='no'>edb</span>, <span class='kw'>by</span> <span class='kw'>=</span> <span class='st'>"gene"</span>,
                  <span class='kw'>filter</span> <span class='kw'>=</span> <span class='fu'>SeqNameFilter</span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"X"</span>, <span class='st'>"Y"</span>)),
                  <span class='kw'>columns</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"gene_biotype"</span>, <span class='st'>"gene_name"</span>))
<span class='no'>EGenes</span></div><div class='output co'>#&gt; GRangesList object of length 2922:
#&gt; $ENSG00000000003
#&gt; GRanges object with 20 ranges and 3 metadata columns:
#&gt;        seqnames              ranges strand |   gene_biotype   gene_name
#&gt;           &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;character&gt; &lt;character&gt;
#&gt;    [1]        X 100639945-100639991      - | protein_coding      TSPAN6
#&gt;    [2]        X 100636793-100637104      - | protein_coding      TSPAN6
#&gt;    [3]        X 100636608-100636806      - | protein_coding      TSPAN6
#&gt;    [4]        X 100636191-100636689      - | protein_coding      TSPAN6
#&gt;    [5]        X 100635558-100635746      - | protein_coding      TSPAN6
#&gt;    ...      ...                 ...    ... .            ...         ...
#&gt;   [16]        X 100632541-100632568      - | protein_coding      TSPAN6
#&gt;   [17]        X 100632063-100632068      - | protein_coding      TSPAN6
#&gt;   [18]        X 100630759-100630866      - | protein_coding      TSPAN6
#&gt;   [19]        X 100628670-100629986      - | protein_coding      TSPAN6
#&gt;   [20]        X 100627109-100629986      - | protein_coding      TSPAN6
#&gt;                exon_id
#&gt;            &lt;character&gt;
#&gt;    [1] ENSE00001828996
#&gt;    [2] ENSE00001863395
#&gt;    [3] ENSE00001855382
#&gt;    [4] ENSE00001886883
#&gt;    [5] ENSE00003662440
#&gt;    ...             ...
#&gt;   [16] ENSE00001849132
#&gt;   [17] ENSE00003731560
#&gt;   [18] ENSE00000868868
#&gt;   [19] ENSE00001459322
#&gt;   [20] ENSE00003730948
#&gt;   -------
#&gt;   seqinfo: 2 sequences from GRCh38 genome
#&gt; 
#&gt; ...
#&gt; &lt;2921 more elements&gt;</div><div class='input'>
<span class='co'>## Note that this might also contain "LRG" genes.</span>
<span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/grep.html'>grep</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span>(<span class='no'>EGenes</span>), <span class='kw'>pattern</span><span class='kw'>=</span><span class='st'>"LRG"</span>))</div><div class='output co'>#&gt; [1] 40</div><div class='input'>
<span class='co'>## to fetch just Ensemblgenes, use an GeneIdFilter with value</span>
<span class='co'>## "ENS%" and condition "like"</span>
<span class='no'>eg</span> <span class='kw'>&lt;-</span> <span class='fu'>exonsBy</span>(<span class='no'>edb</span>, <span class='kw'>by</span> <span class='kw'>=</span> <span class='st'>"gene"</span>,
              <span class='kw'>filter</span> <span class='kw'>=</span> <span class='fu'>AnnotationFilterList</span>(<span class='fu'>SeqNameFilter</span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"X"</span>, <span class='st'>"Y"</span>)),
                                            <span class='fu'>GeneIdFilter</span>(<span class='st'>"ENS"</span>, <span class='st'>"startsWith"</span>)),
              <span class='kw'>columns</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"gene_biotype"</span>, <span class='st'>"gene_name"</span>))
<span class='no'>eg</span></div><div class='output co'>#&gt; GRangesList object of length 2882:
#&gt; $ENSG00000000003
#&gt; GRanges object with 20 ranges and 4 metadata columns:
#&gt;        seqnames              ranges strand |   gene_biotype   gene_name
#&gt;           &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;character&gt; &lt;character&gt;
#&gt;    [1]        X 100639945-100639991      - | protein_coding      TSPAN6
#&gt;    [2]        X 100636793-100637104      - | protein_coding      TSPAN6
#&gt;    [3]        X 100636608-100636806      - | protein_coding      TSPAN6
#&gt;    [4]        X 100636191-100636689      - | protein_coding      TSPAN6
#&gt;    [5]        X 100635558-100635746      - | protein_coding      TSPAN6
#&gt;    ...      ...                 ...    ... .            ...         ...
#&gt;   [16]        X 100632541-100632568      - | protein_coding      TSPAN6
#&gt;   [17]        X 100632063-100632068      - | protein_coding      TSPAN6
#&gt;   [18]        X 100630759-100630866      - | protein_coding      TSPAN6
#&gt;   [19]        X 100628670-100629986      - | protein_coding      TSPAN6
#&gt;   [20]        X 100627109-100629986      - | protein_coding      TSPAN6
#&gt;                exon_id         gene_id
#&gt;            &lt;character&gt;     &lt;character&gt;
#&gt;    [1] ENSE00001828996 ENSG00000000003
#&gt;    [2] ENSE00001863395 ENSG00000000003
#&gt;    [3] ENSE00001855382 ENSG00000000003
#&gt;    [4] ENSE00001886883 ENSG00000000003
#&gt;    [5] ENSE00003662440 ENSG00000000003
#&gt;    ...             ...             ...
#&gt;   [16] ENSE00001849132 ENSG00000000003
#&gt;   [17] ENSE00003731560 ENSG00000000003
#&gt;   [18] ENSE00000868868 ENSG00000000003
#&gt;   [19] ENSE00001459322 ENSG00000000003
#&gt;   [20] ENSE00003730948 ENSG00000000003
#&gt;   -------
#&gt;   seqinfo: 2 sequences from GRCh38 genome
#&gt; 
#&gt; ...
#&gt; &lt;2881 more elements&gt;</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/grep.html'>grep</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span>(<span class='no'>eg</span>), <span class='kw'>pattern</span><span class='kw'>=</span><span class='st'>"LRG"</span>))</div><div class='output co'>#&gt; [1] 0</div><div class='input'>
<span class='co'>#####    transcriptsBy</span>
<span class='co'>##</span>
<span class='no'>TGenes</span> <span class='kw'>&lt;-</span> <span class='fu'>transcriptsBy</span>(<span class='no'>edb</span>, <span class='kw'>by</span> <span class='kw'>=</span> <span class='st'>"gene"</span>,
                        <span class='kw'>filter</span> <span class='kw'>=</span> <span class='fu'>SeqNameFilter</span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"X"</span>, <span class='st'>"Y"</span>)))
<span class='no'>TGenes</span></div><div class='output co'>#&gt; GRangesList object of length 2922:
#&gt; $ENSG00000000003
#&gt; GRanges object with 5 ranges and 6 metadata columns:
#&gt;       seqnames              ranges strand |           tx_id
#&gt;          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
#&gt;   [1]        X 100633442-100639991      - | ENST00000494424
#&gt;   [2]        X 100627109-100637104      - | ENST00000612152
#&gt;   [3]        X 100632063-100637104      - | ENST00000614008
#&gt;   [4]        X 100628670-100636806      - | ENST00000373020
#&gt;   [5]        X 100632541-100636689      - | ENST00000496771
#&gt;                 tx_biotype tx_cds_seq_start tx_cds_seq_end         gene_id
#&gt;                &lt;character&gt;        &lt;integer&gt;      &lt;integer&gt;     &lt;character&gt;
#&gt;   [1] processed_transcript             &lt;NA&gt;           &lt;NA&gt; ENSG00000000003
#&gt;   [2]       protein_coding        100630798      100635569 ENSG00000000003
#&gt;   [3]       protein_coding        100632063      100635569 ENSG00000000003
#&gt;   [4]       protein_coding        100630798      100636694 ENSG00000000003
#&gt;   [5] processed_transcript             &lt;NA&gt;           &lt;NA&gt; ENSG00000000003
#&gt;               tx_name
#&gt;           &lt;character&gt;
#&gt;   [1] ENST00000494424
#&gt;   [2] ENST00000612152
#&gt;   [3] ENST00000614008
#&gt;   [4] ENST00000373020
#&gt;   [5] ENST00000496771
#&gt;   -------
#&gt;   seqinfo: 2 sequences from GRCh38 genome
#&gt; 
#&gt; $ENSG00000000005
#&gt; GRanges object with 2 ranges and 6 metadata columns:
#&gt;       seqnames              ranges strand |           tx_id
#&gt;          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
#&gt;   [1]        X 100584802-100599885      + | ENST00000373031
#&gt;   [2]        X 100593624-100597531      + | ENST00000485971
#&gt;                 tx_biotype tx_cds_seq_start tx_cds_seq_end         gene_id
#&gt;                &lt;character&gt;        &lt;integer&gt;      &lt;integer&gt;     &lt;character&gt;
#&gt;   [1]       protein_coding        100585019      100599717 ENSG00000000005
#&gt;   [2] processed_transcript             &lt;NA&gt;           &lt;NA&gt; ENSG00000000005
#&gt;               tx_name
#&gt;           &lt;character&gt;
#&gt;   [1] ENST00000373031
#&gt;   [2] ENST00000485971
#&gt;   -------
#&gt;   seqinfo: 2 sequences from GRCh38 genome
#&gt; 
#&gt; $ENSG00000001497
#&gt; GRanges object with 5 ranges and 6 metadata columns:
#&gt;       seqnames            ranges strand |           tx_id
#&gt;          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
#&gt;   [1]        X 65512583-65534775      - | ENST00000484069
#&gt;   [2]        X 65512582-65534756      - | ENST00000374811
#&gt;   [3]        X 65512583-65534756      - | ENST00000374804
#&gt;   [4]        X 65512582-65534754      - | ENST00000374807
#&gt;   [5]        X 65520429-65523617      - | ENST00000469091
#&gt;                    tx_biotype tx_cds_seq_start tx_cds_seq_end         gene_id
#&gt;                   &lt;character&gt;        &lt;integer&gt;      &lt;integer&gt;     &lt;character&gt;
#&gt;   [1] nonsense_mediated_decay         65525021       65534715 ENSG00000001497
#&gt;   [2]          protein_coding         65512775       65534715 ENSG00000001497
#&gt;   [3]          protein_coding         65512775       65534715 ENSG00000001497
#&gt;   [4]          protein_coding         65512775       65534715 ENSG00000001497
#&gt;   [5]          protein_coding         65520655       65523617 ENSG00000001497
#&gt;               tx_name
#&gt;           &lt;character&gt;
#&gt;   [1] ENST00000484069
#&gt;   [2] ENST00000374811
#&gt;   [3] ENST00000374804
#&gt;   [4] ENST00000374807
#&gt;   [5] ENST00000469091
#&gt;   -------
#&gt;   seqinfo: 2 sequences from GRCh38 genome
#&gt; 
#&gt; ...
#&gt; &lt;2919 more elements&gt;</div><div class='input'>
<span class='co'>## convert this to a SAF formatted data.frame that can be used by the</span>
<span class='co'>## featureCounts function from the Rsubreader package.</span>
<span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span>(<span class='fu'>toSAF</span>(<span class='no'>TGenes</span>))</div><div class='output co'>#&gt;            GeneID Chr     Start       End Strand
#&gt; 1 ENSG00000000003   X 100633442 100639991      -
#&gt; 2 ENSG00000000003   X 100627109 100637104      -
#&gt; 3 ENSG00000000003   X 100632063 100637104      -
#&gt; 4 ENSG00000000003   X 100628670 100636806      -
#&gt; 5 ENSG00000000003   X 100632541 100636689      -
#&gt; 6 ENSG00000000005   X 100584802 100599885      +</div><div class='input'>

<span class='co'>#####   transcriptsByOverlaps</span>
<span class='co'>##</span>
<span class='no'>ir</span> <span class='kw'>&lt;-</span> <span class='fu'>IRanges</span>(<span class='kw'>start</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>2654890</span>, <span class='fl'>2709520</span>, <span class='fl'>28111770</span>),
              <span class='kw'>end</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>2654900</span>, <span class='fl'>2709550</span>, <span class='fl'>28111790</span>))
<span class='no'>gr</span> <span class='kw'>&lt;-</span> <span class='fu'>GRanges</span>(<span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span>(<span class='st'>"Y"</span>, <span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span>(<span class='no'>ir</span>)), <span class='no'>ir</span>)

<span class='co'>## Retrieve all transcripts overlapping any of the regions.</span>
<span class='no'>txs</span> <span class='kw'>&lt;-</span> <span class='fu'>transcriptsByOverlaps</span>(<span class='no'>edb</span>, <span class='no'>gr</span>)
<span class='no'>txs</span></div><div class='output co'>#&gt; GRanges object with 0 ranges and 6 metadata columns:
#&gt;    seqnames    ranges strand |       tx_id  tx_biotype tx_cds_seq_start
#&gt;       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;        &lt;integer&gt;
#&gt;    tx_cds_seq_end     gene_id     tx_name
#&gt;         &lt;integer&gt; &lt;character&gt; &lt;character&gt;
#&gt;   -------
#&gt;   seqinfo: 1 sequence from GRCh38 genome</div><div class='input'>
<span class='co'>## Alternatively, use a GRangesFilter</span>
<span class='no'>grf</span> <span class='kw'>&lt;-</span> <span class='fu'>GRangesFilter</span>(<span class='no'>gr</span>, <span class='kw'>type</span> <span class='kw'>=</span> <span class='st'>"any"</span>)
<span class='no'>txs</span> <span class='kw'>&lt;-</span> <span class='fu'>transcripts</span>(<span class='no'>edb</span>, <span class='kw'>filter</span> <span class='kw'>=</span> <span class='no'>grf</span>)
<span class='no'>txs</span></div><div class='output co'>#&gt; GRanges object with 0 ranges and 6 metadata columns:
#&gt;    seqnames    ranges strand |       tx_id  tx_biotype tx_cds_seq_start
#&gt;       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;        &lt;integer&gt;
#&gt;    tx_cds_seq_end     gene_id     tx_name
#&gt;         &lt;integer&gt; &lt;character&gt; &lt;character&gt;
#&gt;   -------
#&gt;   seqinfo: no sequences</div><div class='input'>

<span class='co'>####    cdsBy</span>
<span class='co'>## Get the coding region for all transcripts on chromosome Y.</span>
<span class='co'>## Specifying also additional annotation columns (in addition to the default</span>
<span class='co'>## exon_id and exon_rank).</span>
<span class='no'>cds</span> <span class='kw'>&lt;-</span> <span class='fu'>cdsBy</span>(<span class='no'>edb</span>, <span class='kw'>by</span> <span class='kw'>=</span> <span class='st'>"tx"</span>, <span class='kw'>filter</span> <span class='kw'>=</span> <span class='fu'>SeqNameFilter</span>(<span class='st'>"Y"</span>),
             <span class='kw'>columns</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"tx_biotype"</span>, <span class='st'>"gene_name"</span>))

<span class='co'>####    the 5' untranslated regions:</span>
<span class='no'>fUTRs</span> <span class='kw'>&lt;-</span> <span class='fu'>fiveUTRsByTranscript</span>(<span class='no'>edb</span>, <span class='kw'>filter</span> <span class='kw'>=</span> <span class='fu'>SeqNameFilter</span>(<span class='st'>"Y"</span>))

<span class='co'>####    the 3' untranslated regions with additional column gene_name.</span>
<span class='no'>tUTRs</span> <span class='kw'>&lt;-</span> <span class='fu'>threeUTRsByTranscript</span>(<span class='no'>edb</span>, <span class='kw'>filter</span> <span class='kw'>=</span> <span class='fu'>SeqNameFilter</span>(<span class='st'>"Y"</span>),
                               <span class='kw'>columns</span> <span class='kw'>=</span> <span class='st'>"gene_name"</span>)</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Johannes Rainer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


