<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Basic usage of an Ensembl based annotation database — EnsDb-class • ensembldb</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Basic usage of an Ensembl based annotation database — EnsDb-class" />
<meta property="og:description" content="The EnsDb class provides access to an Ensembl-based annotation
  package. This help page describes functions to get some basic
  informations from such an object." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ensembldb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.11.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/ensembldb.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/MySQL-backend.html">Using a MySQL server backend</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping-use-cases.html">Use cases for coordinate mapping with ensembldb</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping.html">Mapping between genome, transcript and protein coordinates</a>
    </li>
    <li>
      <a href="../articles/proteins.html">Querying protein features</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jorainer/ensembldb/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Basic usage of an Ensembl based annotation database</h1>
    
    <div class="hidden name"><code>EnsDb-class.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The <code>EnsDb</code> class provides access to an Ensembl-based annotation
  package. This help page describes functions to get some basic
  informations from such an object.</p>
    </div>

    <pre class="usage"><span class='co'># S4 method for EnsDb</span>
<span class='fu'>dbconn</span>(<span class='no'>x</span>)

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>ensemblVersion</span>(<span class='no'>x</span>)

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>listColumns</span>(<span class='no'>x</span>, <span class='no'>table</span>, <span class='kw'>skip.keys</span><span class='kw'>=</span><span class='fl'>TRUE</span>, <span class='no'>...</span>)

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>listGenebiotypes</span>(<span class='no'>x</span>, <span class='no'>...</span>)

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>listTxbiotypes</span>(<span class='no'>x</span>, <span class='no'>...</span>)

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>listTables</span>(<span class='no'>x</span>, <span class='no'>...</span>)

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>metadata</span>(<span class='no'>x</span>, <span class='no'>...</span>)

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>organism</span>(<span class='no'>object</span>)

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>returnFilterColumns</span>(<span class='no'>x</span>)

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>returnFilterColumns</span>(<span class='no'>x</span>)

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>returnFilterColumns</span>(<span class='no'>x</span>) <span class='kw'>&lt;-</span> <span class='no'>value</span>

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>seqinfo</span>(<span class='no'>x</span>)

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>seqlevels</span>(<span class='no'>x</span>)

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>updateEnsDb</span>(<span class='no'>x</span>, <span class='no'>...</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.
    Not used.</p></td>
    </tr>
    <tr>
      <th>object</th>
      <td><p>For <code>organism</code>: an <code>EnsDb</code> instance.</p></td>
    </tr>
    <tr>
      <th>skip.keys</th>
      <td><p>for <code>listColumns</code>: whether primary and foreign keys (not
    being e.g. <code>"gene_id"</code> or alike) should be returned or not. By
    default these will not be returned.</p></td>
    </tr>
    <tr>
      <th>table</th>
      <td><p>For <code>listColumns</code>: optionally specify the table name(s) for
    which the columns should be returned.</p></td>
    </tr>
    <tr>
      <th>value</th>
      <td><p>For <code>returnFilterColumns</code>: a logical of length one specifying
    whether columns that are used for eventual filters should also be
    returned.</p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>An <code>EnsDb</code> instance.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="objects-from-the-class"><a class="anchor" href="#objects-from-the-class"></a>Objects from the Class</h2>

    <p>A connection to the respective annotation database is created upon
  loading of an annotation package created with the
  <code><a href='makeEnsemblDbPackage.html'>makeEnsembldbPackage</a></code> function. In addition, the
  <code><a href='EnsDb.html'>EnsDb</a></code> constructor specifying the SQLite database file can be
  called to generate an instance of the object (see
  <code><a href='makeEnsemblDbPackage.html'>makeEnsemblSQLiteFromTables</a></code> for an example).</p>
    <h2 class="hasAnchor" id="slots"><a class="anchor" href="#slots"></a>Slots</h2>

    <p></p><dl'>
    <dt>ensdb</dt><dd><p>Object of class <code>"DBIConnection"</code>: the
      connection to the database.</p></dd>

    <dt>tables</dt><dd><p>Named list of database table columns with the names being the
      database table names. The tables are ordered by their degree,
      i.e. the number of other tables they can be joined with.</p></dd>

    <dt>.properties</dt><dd><p>Internal list storing user-defined properties. Should not be
      directly accessed.</p></dd>
  
</dl>

    <h2 class="hasAnchor" id="methods-and-functions"><a class="anchor" href="#methods-and-functions"></a>Methods and Functions</h2>

    <p></p><dl'>

    <dt>dbconn</dt><dd><p>Returns the connection to the internal SQL database.</p></dd>

    <dt>ensemblVersion</dt><dd><p>Returns the Ensembl version on which the package was built.</p></dd>

    <dt>listColumns</dt><dd><p>Lists all columns of all tables in the database, or, if
      <code>table</code> is specified, of the respective table.</p></dd>

    <dt>listGenebiotypes</dt><dd><p>Lists all gene biotypes defined in the database.</p></dd>

    <dt>listTxbiotypes</dt><dd><p>Lists all transcript biotypes defined in the database.</p></dd>

    <dt>listTables</dt><dd><p>Returns a named list of database table columns (names of the
      list being the database table names).</p></dd>

    <dt>metadata</dt><dd><p>Returns a <code>data.frame</code> with the metadata information from the
      database, i.e. informations about the Ensembl version or Genome
      build the database was build upon.</p></dd>

    <dt>organism</dt><dd><p>Returns the organism name (e.g. <code>"homo_sapiens"</code>).</p></dd>

    <dt>returnFilterColumns, returnFilterColumns&lt;-</dt><dd><p>Get or set the option which results in columns that are used for
      eventually specified filters to be added as result columns. The
      default value is <code>TRUE</code> (i.e. filter columns are returned).</p></dd>

    <dt>seqinfo</dt><dd><p>Returns the sequence/chromosome information from the database.</p></dd>

    <dt>seqlevels</dt><dd><p>Returns the chromosome/sequence names that are available in the
      database.</p></dd>

    <dt>show</dt><dd><p>Displays some informations from the database.</p></dd>

    <dt>updateEnsDb</dt><dd><p>Updates the <code>EnsDb</code> object to the most recent implementation.</p></dd>

  
</dl>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p></p><dl'>
    <dt>For <code>connection</code></dt><dd><p>The SQL connection to the RSQLite database.</p></dd>

    <dt>For <code>EnsDb</code></dt><dd><p>An <code>EnsDb</code> instance.</p></dd>

    <dt>For <code>lengthOf</code></dt><dd><p>A named integer vector with the length of the genes or transcripts.</p></dd>

    <dt>For <code>listColumns</code></dt><dd><p>A character vector with the column names.</p></dd>

    <dt>For <code>listGenebiotypes</code></dt><dd><p>A character vector with the biotypes of the genes in the database.</p></dd>

    <dt>For <code>listTxbiotypes</code></dt><dd><p>A character vector with the biotypes of the transcripts in the database.</p></dd>

    <dt>For <code>listTables</code></dt><dd><p>A list with the names corresponding to the database table names
      and the elements being the attribute (column) names of the table.</p></dd>

    <dt>For <code>metadata</code></dt><dd><p>A <code>data.frame</code>.</p></dd>

    <dt>For <code>organism</code></dt><dd><p>A character string.</p></dd>

    <dt>For <code>returnFilterColumns</code></dt><dd><p>A logical of length 1.</p></dd>

    <dt>For <code>seqinfo</code></dt><dd><p>A <code>Seqinfo</code> class.</p></dd>

    <dt>For <code>updateEnsDb</code></dt><dd><p>A <code>EnsDb</code> object.</p></dd>
  
</dl>

    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>While a column named <code>"tx_name"</code> is listed by the
  <code>listTables</code> and <code>listColumns</code> method, no such column is
  present in the database. Transcript names returned by the methods are
  actually the transcript IDs. This <em>virtual</em> column was only
  introduced to be compliant with <code>TxDb</code> objects (which provide
  transcript names).</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='EnsDb.html'>EnsDb</a></code>,
  <code><a href='makeEnsemblDbPackage.html'>makeEnsembldbPackage</a></code>,
  <code><a href='EnsDb-exonsBy.html'>exonsBy</a></code>, <code><a href='EnsDb-exonsBy.html'>genes</a></code>,
  <code><a href='EnsDb-exonsBy.html'>transcripts</a></code>,
  <code><a href='makeEnsemblDbPackage.html'>makeEnsemblSQLiteFromTables</a></code></p>  
<p><code><a href='global-filters.html'>addFilter</a></code> for globally adding filters to an <code>EnsDb</code>
  object.</p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='fu'><a href='https://rdrr.io/r/base/library.html'>library</a></span>(<span class='no'>EnsDb.Hsapiens.v86</span>)

<span class='co'>## Display some information:</span>
<span class='no'>EnsDb.Hsapiens.v86</span></div><div class='output co'>#&gt; EnsDb for Ensembl:
#&gt; |Backend: SQLite
#&gt; |Db type: EnsDb
#&gt; |Type of Gene ID: Ensembl Gene ID
#&gt; |Supporting package: ensembldb
#&gt; |Db created by: ensembldb package from Bioconductor
#&gt; |script_version: 0.3.0
#&gt; |Creation time: Thu May 18 16:32:27 2017
#&gt; |ensembl_version: 86
#&gt; |ensembl_host: localhost
#&gt; |Organism: homo_sapiens
#&gt; |taxonomy_id: 9606
#&gt; |genome_build: GRCh38
#&gt; |DBSCHEMAVERSION: 2.0
#&gt; | No. of genes: 63970.
#&gt; | No. of transcripts: 216741.
#&gt; |Protein data available.</div><div class='input'>
<span class='co'>## Show the tables along with its columns</span>
<span class='fu'>listTables</span>(<span class='no'>EnsDb.Hsapiens.v86</span>)</div><div class='output co'>#&gt; $gene
#&gt; [1] "gene_id"          "gene_name"        "gene_biotype"     "gene_seq_start"  
#&gt; [5] "gene_seq_end"     "seq_name"         "seq_strand"       "seq_coord_system"
#&gt; [9] "symbol"          
#&gt; 
#&gt; $tx
#&gt; [1] "tx_id"            "tx_biotype"       "tx_seq_start"     "tx_seq_end"      
#&gt; [5] "tx_cds_seq_start" "tx_cds_seq_end"   "gene_id"          "tx_name"         
#&gt; 
#&gt; $tx2exon
#&gt; [1] "tx_id"    "exon_id"  "exon_idx"
#&gt; 
#&gt; $exon
#&gt; [1] "exon_id"        "exon_seq_start" "exon_seq_end"  
#&gt; 
#&gt; $chromosome
#&gt; [1] "seq_name"    "seq_length"  "is_circular"
#&gt; 
#&gt; $protein
#&gt; [1] "tx_id"            "protein_id"       "protein_sequence"
#&gt; 
#&gt; $uniprot
#&gt; [1] "protein_id"           "uniprot_id"           "uniprot_db"          
#&gt; [4] "uniprot_mapping_type"
#&gt; 
#&gt; $protein_domain
#&gt; [1] "protein_id"            "protein_domain_id"     "protein_domain_source"
#&gt; [4] "interpro_accession"    "prot_dom_start"        "prot_dom_end"         
#&gt; 
#&gt; $entrezgene
#&gt; [1] "gene_id"  "entrezid"
#&gt; 
#&gt; $metadata
#&gt; [1] "name"  "value"
#&gt; </div><div class='input'>
<span class='co'>## For what species is this database?</span>
<span class='fu'>organism</span>(<span class='no'>EnsDb.Hsapiens.v86</span>)</div><div class='output co'>#&gt; [1] "Homo sapiens"</div><div class='input'>
<span class='co'>## What Ensembl version if the database based on?</span>
<span class='fu'>ensemblVersion</span>(<span class='no'>EnsDb.Hsapiens.v86</span>)</div><div class='output co'>#&gt; [1] "86"</div><div class='input'>
<span class='co'>## Get some more information from the database</span>
<span class='fu'>metadata</span>(<span class='no'>EnsDb.Hsapiens.v86</span>)</div><div class='output co'>#&gt;                  name                               value
#&gt; 1             Db type                               EnsDb
#&gt; 2     Type of Gene ID                     Ensembl Gene ID
#&gt; 3  Supporting package                           ensembldb
#&gt; 4       Db created by ensembldb package from Bioconductor
#&gt; 5      script_version                               0.3.0
#&gt; 6       Creation time            Thu May 18 16:32:27 2017
#&gt; 7     ensembl_version                                  86
#&gt; 8        ensembl_host                           localhost
#&gt; 9            Organism                        homo_sapiens
#&gt; 10        taxonomy_id                                9606
#&gt; 11       genome_build                              GRCh38
#&gt; 12    DBSCHEMAVERSION                                 2.0</div><div class='input'>
<span class='co'>## Get all the sequence names.</span>
<span class='fu'>seqlevels</span>(<span class='no'>EnsDb.Hsapiens.v86</span>)</div><div class='output co'>#&gt;   [1] "1"                                     
#&gt;   [2] "10"                                    
#&gt;   [3] "11"                                    
#&gt;   [4] "12"                                    
#&gt;   [5] "13"                                    
#&gt;   [6] "14"                                    
#&gt;   [7] "15"                                    
#&gt;   [8] "16"                                    
#&gt;   [9] "17"                                    
#&gt;  [10] "18"                                    
#&gt;  [11] "19"                                    
#&gt;  [12] "2"                                     
#&gt;  [13] "20"                                    
#&gt;  [14] "21"                                    
#&gt;  [15] "22"                                    
#&gt;  [16] "3"                                     
#&gt;  [17] "4"                                     
#&gt;  [18] "5"                                     
#&gt;  [19] "6"                                     
#&gt;  [20] "7"                                     
#&gt;  [21] "8"                                     
#&gt;  [22] "9"                                     
#&gt;  [23] "CHR_HG107_PATCH"                       
#&gt;  [24] "CHR_HG126_PATCH"                       
#&gt;  [25] "CHR_HG1311_PATCH"                      
#&gt;  [26] "CHR_HG1342_HG2282_PATCH"               
#&gt;  [27] "CHR_HG1362_PATCH"                      
#&gt;  [28] "CHR_HG142_HG150_NOVEL_TEST"            
#&gt;  [29] "CHR_HG151_NOVEL_TEST"                  
#&gt;  [30] "CHR_HG1651_PATCH"                      
#&gt;  [31] "CHR_HG1832_PATCH"                      
#&gt;  [32] "CHR_HG2021_PATCH"                      
#&gt;  [33] "CHR_HG2022_PATCH"                      
#&gt;  [34] "CHR_HG2023_PATCH"                      
#&gt;  [35] "CHR_HG2030_PATCH"                      
#&gt;  [36] "CHR_HG2058_PATCH"                      
#&gt;  [37] "CHR_HG2062_PATCH"                      
#&gt;  [38] "CHR_HG2063_PATCH"                      
#&gt;  [39] "CHR_HG2066_PATCH"                      
#&gt;  [40] "CHR_HG2072_PATCH"                      
#&gt;  [41] "CHR_HG2095_PATCH"                      
#&gt;  [42] "CHR_HG2104_PATCH"                      
#&gt;  [43] "CHR_HG2116_PATCH"                      
#&gt;  [44] "CHR_HG2128_PATCH"                      
#&gt;  [45] "CHR_HG2191_PATCH"                      
#&gt;  [46] "CHR_HG2213_PATCH"                      
#&gt;  [47] "CHR_HG2217_PATCH"                      
#&gt;  [48] "CHR_HG2232_PATCH"                      
#&gt;  [49] "CHR_HG2233_PATCH"                      
#&gt;  [50] "CHR_HG2235_PATCH"                      
#&gt;  [51] "CHR_HG2239_PATCH"                      
#&gt;  [52] "CHR_HG2247_PATCH"                      
#&gt;  [53] "CHR_HG2249_PATCH"                      
#&gt;  [54] "CHR_HG2288_HG2289_PATCH"               
#&gt;  [55] "CHR_HG2290_PATCH"                      
#&gt;  [56] "CHR_HG2291_PATCH"                      
#&gt;  [57] "CHR_HG2334_PATCH"                      
#&gt;  [58] "CHR_HG26_PATCH"                        
#&gt;  [59] "CHR_HG986_PATCH"                       
#&gt;  [60] "CHR_HSCHR10_1_CTG1"                    
#&gt;  [61] "CHR_HSCHR10_1_CTG2"                    
#&gt;  [62] "CHR_HSCHR10_1_CTG3"                    
#&gt;  [63] "CHR_HSCHR10_1_CTG4"                    
#&gt;  [64] "CHR_HSCHR10_1_CTG6"                    
#&gt;  [65] "CHR_HSCHR11_1_CTG1_2"                  
#&gt;  [66] "CHR_HSCHR11_1_CTG5"                    
#&gt;  [67] "CHR_HSCHR11_1_CTG6"                    
#&gt;  [68] "CHR_HSCHR11_1_CTG7"                    
#&gt;  [69] "CHR_HSCHR11_1_CTG8"                    
#&gt;  [70] "CHR_HSCHR11_2_CTG1"                    
#&gt;  [71] "CHR_HSCHR11_2_CTG1_1"                  
#&gt;  [72] "CHR_HSCHR11_3_CTG1"                    
#&gt;  [73] "CHR_HSCHR12_1_CTG1"                    
#&gt;  [74] "CHR_HSCHR12_1_CTG2_1"                  
#&gt;  [75] "CHR_HSCHR12_2_CTG1"                    
#&gt;  [76] "CHR_HSCHR12_2_CTG2"                    
#&gt;  [77] "CHR_HSCHR12_2_CTG2_1"                  
#&gt;  [78] "CHR_HSCHR12_3_CTG2"                    
#&gt;  [79] "CHR_HSCHR12_3_CTG2_1"                  
#&gt;  [80] "CHR_HSCHR12_4_CTG2"                    
#&gt;  [81] "CHR_HSCHR12_4_CTG2_1"                  
#&gt;  [82] "CHR_HSCHR12_5_CTG2"                    
#&gt;  [83] "CHR_HSCHR12_5_CTG2_1"                  
#&gt;  [84] "CHR_HSCHR12_6_CTG2_1"                  
#&gt;  [85] "CHR_HSCHR13_1_CTG1"                    
#&gt;  [86] "CHR_HSCHR13_1_CTG3"                    
#&gt;  [87] "CHR_HSCHR13_1_CTG5"                    
#&gt;  [88] "CHR_HSCHR13_1_CTG8"                    
#&gt;  [89] "CHR_HSCHR14_1_CTG1"                    
#&gt;  [90] "CHR_HSCHR14_2_CTG1"                    
#&gt;  [91] "CHR_HSCHR14_3_CTG1"                    
#&gt;  [92] "CHR_HSCHR14_7_CTG1"                    
#&gt;  [93] "CHR_HSCHR15_1_CTG1"                    
#&gt;  [94] "CHR_HSCHR15_1_CTG3"                    
#&gt;  [95] "CHR_HSCHR15_1_CTG8"                    
#&gt;  [96] "CHR_HSCHR15_2_CTG3"                    
#&gt;  [97] "CHR_HSCHR15_2_CTG8"                    
#&gt;  [98] "CHR_HSCHR15_3_CTG3"                    
#&gt;  [99] "CHR_HSCHR15_3_CTG8"                    
#&gt; [100] "CHR_HSCHR15_4_CTG8"                    
#&gt; [101] "CHR_HSCHR15_5_CTG8"                    
#&gt; [102] "CHR_HSCHR15_6_CTG8"                    
#&gt; [103] "CHR_HSCHR16_1_CTG1"                    
#&gt; [104] "CHR_HSCHR16_1_CTG3_1"                  
#&gt; [105] "CHR_HSCHR16_2_CTG3_1"                  
#&gt; [106] "CHR_HSCHR16_3_CTG1"                    
#&gt; [107] "CHR_HSCHR16_4_CTG1"                    
#&gt; [108] "CHR_HSCHR16_4_CTG3_1"                  
#&gt; [109] "CHR_HSCHR16_5_CTG1"                    
#&gt; [110] "CHR_HSCHR16_CTG2"                      
#&gt; [111] "CHR_HSCHR17_10_CTG4"                   
#&gt; [112] "CHR_HSCHR17_1_CTG1"                    
#&gt; [113] "CHR_HSCHR17_1_CTG2"                    
#&gt; [114] "CHR_HSCHR17_1_CTG4"                    
#&gt; [115] "CHR_HSCHR17_1_CTG5"                    
#&gt; [116] "CHR_HSCHR17_1_CTG9"                    
#&gt; [117] "CHR_HSCHR17_2_CTG1"                    
#&gt; [118] "CHR_HSCHR17_2_CTG2"                    
#&gt; [119] "CHR_HSCHR17_2_CTG4"                    
#&gt; [120] "CHR_HSCHR17_2_CTG5"                    
#&gt; [121] "CHR_HSCHR17_3_CTG2"                    
#&gt; [122] "CHR_HSCHR17_3_CTG4"                    
#&gt; [123] "CHR_HSCHR17_4_CTG4"                    
#&gt; [124] "CHR_HSCHR17_5_CTG4"                    
#&gt; [125] "CHR_HSCHR17_6_CTG4"                    
#&gt; [126] "CHR_HSCHR17_7_CTG4"                    
#&gt; [127] "CHR_HSCHR17_8_CTG4"                    
#&gt; [128] "CHR_HSCHR17_9_CTG4"                    
#&gt; [129] "CHR_HSCHR18_1_CTG1_1"                  
#&gt; [130] "CHR_HSCHR18_1_CTG2_1"                  
#&gt; [131] "CHR_HSCHR18_2_CTG1_1"                  
#&gt; [132] "CHR_HSCHR18_2_CTG2"                    
#&gt; [133] "CHR_HSCHR18_2_CTG2_1"                  
#&gt; [134] "CHR_HSCHR18_3_CTG2_1"                  
#&gt; [135] "CHR_HSCHR18_5_CTG1_1"                  
#&gt; [136] "CHR_HSCHR18_ALT21_CTG2_1"              
#&gt; [137] "CHR_HSCHR18_ALT2_CTG2_1"               
#&gt; [138] "CHR_HSCHR19KIR_ABC08_A1_HAP_CTG3_1"    
#&gt; [139] "CHR_HSCHR19KIR_ABC08_AB_HAP_C_P_CTG3_1"
#&gt; [140] "CHR_HSCHR19KIR_ABC08_AB_HAP_T_P_CTG3_1"
#&gt; [141] "CHR_HSCHR19KIR_FH05_A_HAP_CTG3_1"      
#&gt; [142] "CHR_HSCHR19KIR_FH05_B_HAP_CTG3_1"      
#&gt; [143] "CHR_HSCHR19KIR_FH06_A_HAP_CTG3_1"      
#&gt; [144] "CHR_HSCHR19KIR_FH06_BA1_HAP_CTG3_1"    
#&gt; [145] "CHR_HSCHR19KIR_FH08_A_HAP_CTG3_1"      
#&gt; [146] "CHR_HSCHR19KIR_FH08_BAX_HAP_CTG3_1"    
#&gt; [147] "CHR_HSCHR19KIR_FH13_A_HAP_CTG3_1"      
#&gt; [148] "CHR_HSCHR19KIR_FH13_BA2_HAP_CTG3_1"    
#&gt; [149] "CHR_HSCHR19KIR_FH15_A_HAP_CTG3_1"      
#&gt; [150] "CHR_HSCHR19KIR_FH15_B_HAP_CTG3_1"      
#&gt; [151] "CHR_HSCHR19KIR_G085_A_HAP_CTG3_1"      
#&gt; [152] "CHR_HSCHR19KIR_G085_BA1_HAP_CTG3_1"    
#&gt; [153] "CHR_HSCHR19KIR_G248_A_HAP_CTG3_1"      
#&gt; [154] "CHR_HSCHR19KIR_G248_BA2_HAP_CTG3_1"    
#&gt; [155] "CHR_HSCHR19KIR_GRC212_AB_HAP_CTG3_1"   
#&gt; [156] "CHR_HSCHR19KIR_GRC212_BA1_HAP_CTG3_1"  
#&gt; [157] "CHR_HSCHR19KIR_LUCE_A_HAP_CTG3_1"      
#&gt; [158] "CHR_HSCHR19KIR_LUCE_BDEL_HAP_CTG3_1"   
#&gt; [159] "CHR_HSCHR19KIR_RP5_B_HAP_CTG3_1"       
#&gt; [160] "CHR_HSCHR19KIR_RSH_A_HAP_CTG3_1"       
#&gt; [161] "CHR_HSCHR19KIR_RSH_BA2_HAP_CTG3_1"     
#&gt; [162] "CHR_HSCHR19KIR_T7526_A_HAP_CTG3_1"     
#&gt; [163] "CHR_HSCHR19KIR_T7526_BDEL_HAP_CTG3_1"  
#&gt; [164] "CHR_HSCHR19LRC_COX1_CTG3_1"            
#&gt; [165] "CHR_HSCHR19LRC_COX2_CTG3_1"            
#&gt; [166] "CHR_HSCHR19LRC_LRC_I_CTG3_1"           
#&gt; [167] "CHR_HSCHR19LRC_LRC_J_CTG3_1"           
#&gt; [168] "CHR_HSCHR19LRC_LRC_S_CTG3_1"           
#&gt; [169] "CHR_HSCHR19LRC_LRC_T_CTG3_1"           
#&gt; [170] "CHR_HSCHR19LRC_PGF1_CTG3_1"            
#&gt; [171] "CHR_HSCHR19LRC_PGF2_CTG3_1"            
#&gt; [172] "CHR_HSCHR19_1_CTG2"                    
#&gt; [173] "CHR_HSCHR19_1_CTG3_1"                  
#&gt; [174] "CHR_HSCHR19_2_CTG2"                    
#&gt; [175] "CHR_HSCHR19_2_CTG3_1"                  
#&gt; [176] "CHR_HSCHR19_3_CTG2"                    
#&gt; [177] "CHR_HSCHR19_3_CTG3_1"                  
#&gt; [178] "CHR_HSCHR19_4_CTG2"                    
#&gt; [179] "CHR_HSCHR19_4_CTG3_1"                  
#&gt; [180] "CHR_HSCHR19_5_CTG2"                    
#&gt; [181] "CHR_HSCHR1_1_CTG11"                    
#&gt; [182] "CHR_HSCHR1_1_CTG3"                     
#&gt; [183] "CHR_HSCHR1_1_CTG31"                    
#&gt; [184] "CHR_HSCHR1_1_CTG32_1"                  
#&gt; [185] "CHR_HSCHR1_2_CTG3"                     
#&gt; [186] "CHR_HSCHR1_2_CTG31"                    
#&gt; [187] "CHR_HSCHR1_2_CTG32_1"                  
#&gt; [188] "CHR_HSCHR1_3_CTG3"                     
#&gt; [189] "CHR_HSCHR1_3_CTG31"                    
#&gt; [190] "CHR_HSCHR1_3_CTG32_1"                  
#&gt; [191] "CHR_HSCHR1_4_CTG3"                     
#&gt; [192] "CHR_HSCHR1_4_CTG31"                    
#&gt; [193] "CHR_HSCHR1_5_CTG3"                     
#&gt; [194] "CHR_HSCHR1_5_CTG32_1"                  
#&gt; [195] "CHR_HSCHR1_ALT2_1_CTG32_1"             
#&gt; [196] "CHR_HSCHR20_1_CTG1"                    
#&gt; [197] "CHR_HSCHR20_1_CTG2"                    
#&gt; [198] "CHR_HSCHR20_1_CTG3"                    
#&gt; [199] "CHR_HSCHR20_1_CTG4"                    
#&gt; [200] "CHR_HSCHR21_2_CTG1_1"                  
#&gt; [201] "CHR_HSCHR21_3_CTG1_1"                  
#&gt; [202] "CHR_HSCHR21_4_CTG1_1"                  
#&gt; [203] "CHR_HSCHR21_5_CTG2"                    
#&gt; [204] "CHR_HSCHR21_6_CTG1_1"                  
#&gt; [205] "CHR_HSCHR21_8_CTG1_1"                  
#&gt; [206] "CHR_HSCHR22_1_CTG1"                    
#&gt; [207] "CHR_HSCHR22_1_CTG2"                    
#&gt; [208] "CHR_HSCHR22_1_CTG3"                    
#&gt; [209] "CHR_HSCHR22_1_CTG4"                    
#&gt; [210] "CHR_HSCHR22_1_CTG5"                    
#&gt; [211] "CHR_HSCHR22_1_CTG6"                    
#&gt; [212] "CHR_HSCHR22_1_CTG7"                    
#&gt; [213] "CHR_HSCHR22_2_CTG1"                    
#&gt; [214] "CHR_HSCHR22_3_CTG1"                    
#&gt; [215] "CHR_HSCHR22_4_CTG1"                    
#&gt; [216] "CHR_HSCHR22_5_CTG1"                    
#&gt; [217] "CHR_HSCHR22_6_CTG1"                    
#&gt; [218] "CHR_HSCHR22_7_CTG1"                    
#&gt; [219] "CHR_HSCHR22_8_CTG1"                    
#&gt; [220] "CHR_HSCHR2_1_CTG1"                     
#&gt; [221] "CHR_HSCHR2_1_CTG15"                    
#&gt; [222] "CHR_HSCHR2_1_CTG5"                     
#&gt; [223] "CHR_HSCHR2_1_CTG7"                     
#&gt; [224] "CHR_HSCHR2_1_CTG7_2"                   
#&gt; [225] "CHR_HSCHR2_2_CTG1"                     
#&gt; [226] "CHR_HSCHR2_2_CTG15"                    
#&gt; [227] "CHR_HSCHR2_2_CTG7"                     
#&gt; [228] "CHR_HSCHR2_2_CTG7_2"                   
#&gt; [229] "CHR_HSCHR2_3_CTG1"                     
#&gt; [230] "CHR_HSCHR2_3_CTG15"                    
#&gt; [231] "CHR_HSCHR2_3_CTG7_2"                   
#&gt; [232] "CHR_HSCHR2_4_CTG1"                     
#&gt; [233] "CHR_HSCHR2_6_CTG7_2"                   
#&gt; [234] "CHR_HSCHR3_1_CTG1"                     
#&gt; [235] "CHR_HSCHR3_1_CTG2_1"                   
#&gt; [236] "CHR_HSCHR3_1_CTG3"                     
#&gt; [237] "CHR_HSCHR3_2_CTG2_1"                   
#&gt; [238] "CHR_HSCHR3_2_CTG3"                     
#&gt; [239] "CHR_HSCHR3_3_CTG1"                     
#&gt; [240] "CHR_HSCHR3_3_CTG3"                     
#&gt; [241] "CHR_HSCHR3_4_CTG2_1"                   
#&gt; [242] "CHR_HSCHR3_4_CTG3"                     
#&gt; [243] "CHR_HSCHR3_5_CTG2_1"                   
#&gt; [244] "CHR_HSCHR3_5_CTG3"                     
#&gt; [245] "CHR_HSCHR3_6_CTG3"                     
#&gt; [246] "CHR_HSCHR3_7_CTG3"                     
#&gt; [247] "CHR_HSCHR3_8_CTG3"                     
#&gt; [248] "CHR_HSCHR3_9_CTG3"                     
#&gt; [249] "CHR_HSCHR4_11_CTG12"                   
#&gt; [250] "CHR_HSCHR4_1_CTG12"                    
#&gt; [251] "CHR_HSCHR4_1_CTG4"                     
#&gt; [252] "CHR_HSCHR4_1_CTG6"                     
#&gt; [253] "CHR_HSCHR4_1_CTG9"                     
#&gt; [254] "CHR_HSCHR4_2_CTG12"                    
#&gt; [255] "CHR_HSCHR4_2_CTG4"                     
#&gt; [256] "CHR_HSCHR4_3_CTG12"                    
#&gt; [257] "CHR_HSCHR4_4_CTG12"                    
#&gt; [258] "CHR_HSCHR4_5_CTG12"                    
#&gt; [259] "CHR_HSCHR4_6_CTG12"                    
#&gt; [260] "CHR_HSCHR4_7_CTG12"                    
#&gt; [261] "CHR_HSCHR4_8_CTG12"                    
#&gt; [262] "CHR_HSCHR4_9_CTG12"                    
#&gt; [263] "CHR_HSCHR5_1_CTG1"                     
#&gt; [264] "CHR_HSCHR5_1_CTG1_1"                   
#&gt; [265] "CHR_HSCHR5_1_CTG5"                     
#&gt; [266] "CHR_HSCHR5_2_CTG1"                     
#&gt; [267] "CHR_HSCHR5_2_CTG1_1"                   
#&gt; [268] "CHR_HSCHR5_2_CTG5"                     
#&gt; [269] "CHR_HSCHR5_3_CTG1"                     
#&gt; [270] "CHR_HSCHR5_3_CTG5"                     
#&gt; [271] "CHR_HSCHR5_4_CTG1"                     
#&gt; [272] "CHR_HSCHR5_4_CTG1_1"                   
#&gt; [273] "CHR_HSCHR5_5_CTG1"                     
#&gt; [274] "CHR_HSCHR5_6_CTG1"                     
#&gt; [275] "CHR_HSCHR5_7_CTG1"                     
#&gt; [276] "CHR_HSCHR6_1_CTG10"                    
#&gt; [277] "CHR_HSCHR6_1_CTG2"                     
#&gt; [278] "CHR_HSCHR6_1_CTG3"                     
#&gt; [279] "CHR_HSCHR6_1_CTG4"                     
#&gt; [280] "CHR_HSCHR6_1_CTG5"                     
#&gt; [281] "CHR_HSCHR6_1_CTG6"                     
#&gt; [282] "CHR_HSCHR6_1_CTG7"                     
#&gt; [283] "CHR_HSCHR6_1_CTG8"                     
#&gt; [284] "CHR_HSCHR6_1_CTG9"                     
#&gt; [285] "CHR_HSCHR6_8_CTG1"                     
#&gt; [286] "CHR_HSCHR6_MHC_APD_CTG1"               
#&gt; [287] "CHR_HSCHR6_MHC_COX_CTG1"               
#&gt; [288] "CHR_HSCHR6_MHC_DBB_CTG1"               
#&gt; [289] "CHR_HSCHR6_MHC_MANN_CTG1"              
#&gt; [290] "CHR_HSCHR6_MHC_MCF_CTG1"               
#&gt; [291] "CHR_HSCHR6_MHC_QBL_CTG1"               
#&gt; [292] "CHR_HSCHR6_MHC_SSTO_CTG1"              
#&gt; [293] "CHR_HSCHR7_1_CTG1"                     
#&gt; [294] "CHR_HSCHR7_1_CTG4_4"                   
#&gt; [295] "CHR_HSCHR7_1_CTG6"                     
#&gt; [296] "CHR_HSCHR7_1_CTG7"                     
#&gt; [297] "CHR_HSCHR7_2_CTG1"                     
#&gt; [298] "CHR_HSCHR7_2_CTG4_4"                   
#&gt; [299] "CHR_HSCHR7_2_CTG6"                     
#&gt; [300] "CHR_HSCHR7_2_CTG7"                     
#&gt; [301] "CHR_HSCHR7_3_CTG6"                     
#&gt; [302] "CHR_HSCHR8_1_CTG1"                     
#&gt; [303] "CHR_HSCHR8_1_CTG6"                     
#&gt; [304] "CHR_HSCHR8_1_CTG7"                     
#&gt; [305] "CHR_HSCHR8_2_CTG1"                     
#&gt; [306] "CHR_HSCHR8_2_CTG7"                     
#&gt; [307] "CHR_HSCHR8_3_CTG1"                     
#&gt; [308] "CHR_HSCHR8_3_CTG7"                     
#&gt; [309] "CHR_HSCHR8_4_CTG1"                     
#&gt; [310] "CHR_HSCHR8_4_CTG7"                     
#&gt; [311] "CHR_HSCHR8_5_CTG1"                     
#&gt; [312] "CHR_HSCHR8_5_CTG7"                     
#&gt; [313] "CHR_HSCHR8_6_CTG1"                     
#&gt; [314] "CHR_HSCHR8_7_CTG1"                     
#&gt; [315] "CHR_HSCHR8_8_CTG1"                     
#&gt; [316] "CHR_HSCHR8_9_CTG1"                     
#&gt; [317] "CHR_HSCHR9_1_CTG1"                     
#&gt; [318] "CHR_HSCHR9_1_CTG2"                     
#&gt; [319] "CHR_HSCHR9_1_CTG3"                     
#&gt; [320] "CHR_HSCHR9_1_CTG4"                     
#&gt; [321] "CHR_HSCHR9_1_CTG5"                     
#&gt; [322] "CHR_HSCHR9_1_CTG6"                     
#&gt; [323] "CHR_HSCHRX_1_CTG3"                     
#&gt; [324] "CHR_HSCHRX_2_CTG12"                    
#&gt; [325] "CHR_HSCHRX_2_CTG3"                     
#&gt; [326] "GL000009.2"                            
#&gt; [327] "GL000194.1"                            
#&gt; [328] "GL000195.1"                            
#&gt; [329] "GL000205.2"                            
#&gt; [330] "GL000213.1"                            
#&gt; [331] "GL000216.2"                            
#&gt; [332] "GL000218.1"                            
#&gt; [333] "GL000219.1"                            
#&gt; [334] "GL000220.1"                            
#&gt; [335] "GL000225.1"                            
#&gt; [336] "KI270442.1"                            
#&gt; [337] "KI270711.1"                            
#&gt; [338] "KI270713.1"                            
#&gt; [339] "KI270721.1"                            
#&gt; [340] "KI270726.1"                            
#&gt; [341] "KI270727.1"                            
#&gt; [342] "KI270728.1"                            
#&gt; [343] "KI270731.1"                            
#&gt; [344] "KI270733.1"                            
#&gt; [345] "KI270734.1"                            
#&gt; [346] "KI270744.1"                            
#&gt; [347] "KI270750.1"                            
#&gt; [348] "LRG_183"                               
#&gt; [349] "LRG_187"                               
#&gt; [350] "LRG_239"                               
#&gt; [351] "LRG_311"                               
#&gt; [352] "LRG_721"                               
#&gt; [353] "LRG_741"                               
#&gt; [354] "LRG_93"                                
#&gt; [355] "MT"                                    
#&gt; [356] "X"                                     
#&gt; [357] "Y"                                     </div><div class='input'>
<span class='co'>## List all available gene biotypes from the database:</span>
<span class='fu'>listGenebiotypes</span>(<span class='no'>EnsDb.Hsapiens.v86</span>)</div><div class='output co'>#&gt;  [1] "protein_coding"                     "unitary_pseudogene"                
#&gt;  [3] "unprocessed_pseudogene"             "processed_pseudogene"              
#&gt;  [5] "processed_transcript"               "transcribed_unprocessed_pseudogene"
#&gt;  [7] "antisense"                          "transcribed_unitary_pseudogene"    
#&gt;  [9] "polymorphic_pseudogene"             "lincRNA"                           
#&gt; [11] "sense_intronic"                     "transcribed_processed_pseudogene"  
#&gt; [13] "sense_overlapping"                  "IG_V_pseudogene"                   
#&gt; [15] "pseudogene"                         "TR_V_gene"                         
#&gt; [17] "3prime_overlapping_ncRNA"           "IG_V_gene"                         
#&gt; [19] "bidirectional_promoter_lncRNA"      "snRNA"                             
#&gt; [21] "miRNA"                              "misc_RNA"                          
#&gt; [23] "snoRNA"                             "rRNA"                              
#&gt; [25] "Mt_tRNA"                            "Mt_rRNA"                           
#&gt; [27] "IG_C_gene"                          "IG_J_gene"                         
#&gt; [29] "TR_J_gene"                          "TR_C_gene"                         
#&gt; [31] "TR_V_pseudogene"                    "TR_J_pseudogene"                   
#&gt; [33] "IG_D_gene"                          "ribozyme"                          
#&gt; [35] "IG_C_pseudogene"                    "TR_D_gene"                         
#&gt; [37] "TEC"                                "IG_J_pseudogene"                   
#&gt; [39] "scRNA"                              "scaRNA"                            
#&gt; [41] "vaultRNA"                           "sRNA"                              
#&gt; [43] "macro_lncRNA"                       "non_coding"                        
#&gt; [45] "IG_pseudogene"                      "LRG_gene"                          </div><div class='input'>
<span class='co'>## List all available transcript biotypes:</span>
<span class='fu'>listTxbiotypes</span>(<span class='no'>EnsDb.Hsapiens.v86</span>)</div><div class='output co'>#&gt;  [1] "protein_coding"                     "processed_transcript"              
#&gt;  [3] "nonsense_mediated_decay"            "retained_intron"                   
#&gt;  [5] "unitary_pseudogene"                 "TEC"                               
#&gt;  [7] "miRNA"                              "misc_RNA"                          
#&gt;  [9] "non_stop_decay"                     "unprocessed_pseudogene"            
#&gt; [11] "processed_pseudogene"               "transcribed_unprocessed_pseudogene"
#&gt; [13] "lincRNA"                            "antisense"                         
#&gt; [15] "transcribed_unitary_pseudogene"     "polymorphic_pseudogene"            
#&gt; [17] "sense_intronic"                     "transcribed_processed_pseudogene"  
#&gt; [19] "sense_overlapping"                  "IG_V_pseudogene"                   
#&gt; [21] "pseudogene"                         "TR_V_gene"                         
#&gt; [23] "3prime_overlapping_ncRNA"           "IG_V_gene"                         
#&gt; [25] "bidirectional_promoter_lncRNA"      "snRNA"                             
#&gt; [27] "snoRNA"                             "rRNA"                              
#&gt; [29] "Mt_tRNA"                            "Mt_rRNA"                           
#&gt; [31] "IG_C_gene"                          "IG_J_gene"                         
#&gt; [33] "TR_J_gene"                          "TR_C_gene"                         
#&gt; [35] "TR_V_pseudogene"                    "TR_J_pseudogene"                   
#&gt; [37] "IG_D_gene"                          "ribozyme"                          
#&gt; [39] "IG_C_pseudogene"                    "TR_D_gene"                         
#&gt; [41] "IG_J_pseudogene"                    "scRNA"                             
#&gt; [43] "scaRNA"                             "vaultRNA"                          
#&gt; [45] "sRNA"                               "macro_lncRNA"                      
#&gt; [47] "non_coding"                         "IG_pseudogene"                     
#&gt; [49] "LRG_gene"                          </div><div class='input'>
<span class='co'>## Update the EnsDb; this is in most instances not necessary at all.</span>
<span class='fu'>updateEnsDb</span>(<span class='no'>EnsDb.Hsapiens.v86</span>)</div><div class='output co'>#&gt; EnsDb for Ensembl:
#&gt; |Backend: SQLite
#&gt; |Db type: EnsDb
#&gt; |Type of Gene ID: Ensembl Gene ID
#&gt; |Supporting package: ensembldb
#&gt; |Db created by: ensembldb package from Bioconductor
#&gt; |script_version: 0.3.0
#&gt; |Creation time: Thu May 18 16:32:27 2017
#&gt; |ensembl_version: 86
#&gt; |ensembl_host: localhost
#&gt; |Organism: homo_sapiens
#&gt; |taxonomy_id: 9606
#&gt; |genome_build: GRCh38
#&gt; |DBSCHEMAVERSION: 2.0
#&gt; | No. of genes: 63970.
#&gt; | No. of transcripts: 216741.
#&gt; |Protein data available.</div><div class='input'>
<span class='co'>######    returnFilterColumns</span>
<span class='fu'>returnFilterColumns</span>(<span class='no'>EnsDb.Hsapiens.v86</span>)</div><div class='output co'>#&gt; [1] TRUE</div><div class='input'>
<span class='co'>## Get protein coding genes on chromosome X, specifying to return</span>
<span class='co'>## only columns gene_name as additional column.</span>
<span class='fu'><a href='EnsDb-exonsBy.html'>genes</a></span>(<span class='no'>EnsDb.Hsapiens.v86</span>, <span class='kw'>filter</span><span class='kw'>=</span><span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span>(<span class='fu'>SeqNameFilter</span>(<span class='st'>"X"</span>),
                                      <span class='fu'>GeneBiotypeFilter</span>(<span class='st'>"protein_coding"</span>)),
      <span class='kw'>columns</span><span class='kw'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"gene_name"</span>))</div><div class='output co'>#&gt; GRanges object with 841 ranges and 3 metadata columns:
#&gt;                   seqnames              ranges strand |   gene_name
#&gt;                      &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
#&gt;   ENSG00000182378        X       276322-303356      + |      PLCXD1
#&gt;   ENSG00000178605        X       304529-318819      - |      GTPBP6
#&gt;   ENSG00000167393        X       333963-386955      - |     PPP2R3B
#&gt;   ENSG00000185960        X       624344-659411      + |        SHOX
#&gt;   ENSG00000205755        X     1187549-1212750      - |       CRLF2
#&gt;               ...      ...                 ...    ... .         ...
#&gt;   ENSG00000277745        X 155459415-155460005      - |      H2AFB3
#&gt;   ENSG00000185973        X 155490115-155669944      - |       TMLHE
#&gt;   ENSG00000168939        X 155767812-155782459      + |       SPRY3
#&gt;   ENSG00000124333        X 155881293-155943769      + |       VAMP7
#&gt;   ENSG00000124334        X 155997581-156010817      + |        IL9R
#&gt;                           gene_id   gene_biotype
#&gt;                       &lt;character&gt;    &lt;character&gt;
#&gt;   ENSG00000182378 ENSG00000182378 protein_coding
#&gt;   ENSG00000178605 ENSG00000178605 protein_coding
#&gt;   ENSG00000167393 ENSG00000167393 protein_coding
#&gt;   ENSG00000185960 ENSG00000185960 protein_coding
#&gt;   ENSG00000205755 ENSG00000205755 protein_coding
#&gt;               ...             ...            ...
#&gt;   ENSG00000277745 ENSG00000277745 protein_coding
#&gt;   ENSG00000185973 ENSG00000185973 protein_coding
#&gt;   ENSG00000168939 ENSG00000168939 protein_coding
#&gt;   ENSG00000124333 ENSG00000124333 protein_coding
#&gt;   ENSG00000124334 ENSG00000124334 protein_coding
#&gt;   -------
#&gt;   seqinfo: 1 sequence from GRCh38 genome</div><div class='input'><span class='co'>## By default we get also the gene_biotype column as the data was filtered</span>
<span class='co'>## on this column.</span>

<span class='co'>## This can be changed using the returnFilterColumns option</span>
<span class='fu'>returnFilterColumns</span>(<span class='no'>EnsDb.Hsapiens.v86</span>) <span class='kw'>&lt;-</span> <span class='fl'>FALSE</span>
<span class='fu'><a href='EnsDb-exonsBy.html'>genes</a></span>(<span class='no'>EnsDb.Hsapiens.v86</span>, <span class='kw'>filter</span><span class='kw'>=</span><span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span>(<span class='fu'>SeqNameFilter</span>(<span class='st'>"X"</span>),
                                      <span class='fu'>GeneBiotypeFilter</span>(<span class='st'>"protein_coding"</span>)),
      <span class='kw'>columns</span><span class='kw'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"gene_name"</span>))</div><div class='output co'>#&gt; GRanges object with 841 ranges and 2 metadata columns:
#&gt;                   seqnames              ranges strand |   gene_name
#&gt;                      &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
#&gt;   ENSG00000182378        X       276322-303356      + |      PLCXD1
#&gt;   ENSG00000178605        X       304529-318819      - |      GTPBP6
#&gt;   ENSG00000167393        X       333963-386955      - |     PPP2R3B
#&gt;   ENSG00000185960        X       624344-659411      + |        SHOX
#&gt;   ENSG00000205755        X     1187549-1212750      - |       CRLF2
#&gt;               ...      ...                 ...    ... .         ...
#&gt;   ENSG00000277745        X 155459415-155460005      - |      H2AFB3
#&gt;   ENSG00000185973        X 155490115-155669944      - |       TMLHE
#&gt;   ENSG00000168939        X 155767812-155782459      + |       SPRY3
#&gt;   ENSG00000124333        X 155881293-155943769      + |       VAMP7
#&gt;   ENSG00000124334        X 155997581-156010817      + |        IL9R
#&gt;                           gene_id
#&gt;                       &lt;character&gt;
#&gt;   ENSG00000182378 ENSG00000182378
#&gt;   ENSG00000178605 ENSG00000178605
#&gt;   ENSG00000167393 ENSG00000167393
#&gt;   ENSG00000185960 ENSG00000185960
#&gt;   ENSG00000205755 ENSG00000205755
#&gt;               ...             ...
#&gt;   ENSG00000277745 ENSG00000277745
#&gt;   ENSG00000185973 ENSG00000185973
#&gt;   ENSG00000168939 ENSG00000168939
#&gt;   ENSG00000124333 ENSG00000124333
#&gt;   ENSG00000124334 ENSG00000124334
#&gt;   -------
#&gt;   seqinfo: 1 sequence from GRCh38 genome</div><div class='input'>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Johannes Rainer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


