<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Utility functions — getGeneRegionTrackForGviz • ensembldb</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Utility functions — getGeneRegionTrackForGviz"><meta property="og:description" content="Utility functions integrating EnsDb objects with other
  Bioconductor packages."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ensembldb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.21.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/ensembldb.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/MySQL-backend.html">Using a MySQL server backend</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping-use-cases.html">Use cases for coordinate mapping with ensembldb</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping.html">Mapping between genome, transcript and protein coordinates</a>
    </li>
    <li>
      <a href="../articles/proteins.html">Querying protein features</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/jorainer/ensembldb/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Utility functions</h1>
    
    <div class="hidden name"><code>EnsDb-utils.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Utility functions integrating <code>EnsDb</code> objects with other
  Bioconductor packages.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="co"># S4 method for EnsDb</span>
<span class="fu">getGeneRegionTrackForGviz</span><span class="op">(</span><span class="va">x</span>,
        filter <span class="op">=</span> <span class="fu">AnnotationFilterList</span><span class="op">(</span><span class="op">)</span>, chromosome <span class="op">=</span> <span class="cn">NULL</span>,
        start <span class="op">=</span> <span class="cn">NULL</span>, end <span class="op">=</span> <span class="cn">NULL</span>, featureIs <span class="op">=</span> <span class="st">"gene_biotype"</span><span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    


<p>(In alphabetic order)</p>
<dl><dt>chromosome</dt>
<dd><p>For <code>getGeneRegionTrackForGviz</code>: optional chromosome name to
    restrict the returned entry to a specific chromosome.</p></dd>


  <dt>end</dt>
<dd><p>For <code>getGeneRegionTrackForGviz</code>: optional chromosomal end
    coordinate specifying, together with <code>start</code>, the chromosomal
    region from which features should be retrieved.</p></dd>


  <dt>featureIs</dt>
<dd><p>For <code>getGeneRegionTrackForGviz</code>: whether the gene
    (<code>"gene_biotype"</code>) or the transcript biotype
    (<code>"tx_biotype"</code>) should be returned in column <code>"feature"</code>.</p></dd>


  <dt>filter</dt>
<dd><p>A filter describing which results to retrieve from the database. Can
    be a single object extending
    <code><a href="https://rdrr.io/pkg/AnnotationFilter/man/AnnotationFilter.html" class="external-link">AnnotationFilter</a></code>, an
    <code><a href="https://rdrr.io/pkg/AnnotationFilter/man/AnnotationFilterList.html" class="external-link">AnnotationFilterList</a></code> object
    combining several such objects or a <code>formula</code> representing a
    filter expression (see examples below or
    <code><a href="https://rdrr.io/pkg/AnnotationFilter/man/AnnotationFilter.html" class="external-link">AnnotationFilter</a></code> for more details).</p></dd>


  <dt>start</dt>
<dd><p>For <code>getGeneRegionTrackForGviz</code>: optional chromosomal start
    coordinate specifying, together with <code>end</code>, the chromosomal
    region from which features should be retrieved.</p></dd>


  <dt>x</dt>
<dd><p>For <code>toSAF</code> a <code>GRangesList</code> object. For all other
    methods an <code>EnsDb</code> instance.</p></dd>


</dl></div>
    <div id="methods-and-functions">
    <h2>Methods and Functions</h2>
    <p></p><dl><dt>getGeneRegionTrackForGviz</dt>
<dd><p>Retrieve a <code>GRanges</code> object with transcript features from the
      <code>EnsDb</code> that can be used directly in the <code>Gviz</code> package
      to create a <code>GeneRegionTrack</code>. Using the <code>filter</code>,
      <code>chromosome</code>, <code>start</code> and <code>end</code> arguments it is
      possible to fetch specific features (e.g. lincRNAs) from the
      database.</p>
<p>If <code>chromosome</code>, <code>start</code> and <code>end</code> is provided the
      function internally first retrieves all transcripts that have an
      exon or an intron in the specified chromosomal region and
      subsequently fetch all of these transcripts. This ensures that all
      transcripts of the region are returned, even those that have
      <em>only</em> an intron in the region.</p>
<p>The function returns a <code>GRanges</code> object with additional
      annotation columns <code>"feature"</code>, <code>"gene"</code>, <code>"exon"</code>,
      <code>"exon_rank"</code>, <code>"trancript"</code>, <code>"symbol"</code> specifying
      the feature type (either gene or transcript biotype), the
      (Ensembl) gene ID, the exon ID, the rank/index of the exon in the
      transcript, the transcript ID and the gene symbol/name.</p></dd>

  
</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>For <code>getGeneRegionTrackForGviz</code>: see method description above.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Johannes Rainer</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="EnsDb-exonsBy.html">transcripts</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"></span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span>
<span class="r-in"><span class="va">edb</span> <span class="op">&lt;-</span> <span class="va">EnsDb.Hsapiens.v86</span></span>
<span class="r-in"><span class="co">######   getGeneRegionTrackForGviz</span></span>
<span class="r-in"><span class="co">##</span></span>
<span class="r-in"><span class="co">## Get all genes encoded on chromosome Y in the specifyed region.</span></span>
<span class="r-in"><span class="va">AllY</span> <span class="op">&lt;-</span> <span class="fu">getGeneRegionTrackForGviz</span><span class="op">(</span><span class="va">edb</span>, chromosome <span class="op">=</span> <span class="st">"Y"</span>, start <span class="op">=</span> <span class="fl">5131959</span>,</span>
<span class="r-in">                                  end <span class="op">=</span> <span class="fl">7131959</span><span class="op">)</span></span>
<span class="r-in"><span class="co">## We could plot this now using plotTracks(GeneRegionTrack(AllY))</span></span>
<span class="r-in"></span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Johannes Rainer.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.4.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

