<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Integration into the AnnotationDbi framework — select • ensembldb</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Integration into the AnnotationDbi framework — select" />
<meta property="og:description" content="Several of the methods available for AnnotationDbi objects are
  also implemented for EnsDb objects. This enables to extract
  data from EnsDb objects in a similar fashion than from objects
  inheriting from the base annotation package class
  AnnotationDbi.
  In addition to the standard usage, the select and
  mapIds for EnsDb objects support also the filter
  framework of the ensembdb package and thus allow to perform more
  fine-grained queries to retrieve data." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ensembldb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.13.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/ensembldb.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/MySQL-backend.html">Using a MySQL server backend</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping-use-cases.html">Use cases for coordinate mapping with ensembldb</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping.html">Mapping between genome, transcript and protein coordinates</a>
    </li>
    <li>
      <a href="../articles/proteins.html">Querying protein features</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jorainer/ensembldb/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Integration into the AnnotationDbi framework</h1>
    
    <div class="hidden name"><code>EnsDb-AnnotationDbi.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Several of the methods available for <code>AnnotationDbi</code> objects are
  also implemented for <code>EnsDb</code> objects. This enables to extract
  data from <code>EnsDb</code> objects in a similar fashion than from objects
  inheriting from the base annotation package class
  <code>AnnotationDbi</code>.
  In addition to the <em>standard</em> usage, the <code>select</code> and
  <code>mapIds</code> for <code>EnsDb</code> objects support also the filter
  framework of the ensembdb package and thus allow to perform more
  fine-grained queries to retrieve data.</p>
    </div>

    <pre class="usage"><span class='co'># S4 method for EnsDb</span>
<span class='fu'>columns</span>(<span class='no'>x</span>)
<span class='co'># S4 method for EnsDb</span>
<span class='fu'>keys</span>(<span class='no'>x</span>, <span class='no'>keytype</span>, <span class='no'>filter</span>,<span class='no'>...</span>)
<span class='co'># S4 method for EnsDb</span>
<span class='fu'>keytypes</span>(<span class='no'>x</span>)
<span class='co'># S4 method for EnsDb</span>
<span class='fu'>mapIds</span>(<span class='no'>x</span>, <span class='no'>keys</span>, <span class='no'>column</span>, <span class='no'>keytype</span>, <span class='no'>...</span>, <span class='no'>multiVals</span>)
<span class='co'># S4 method for EnsDb</span>
<span class='fu'>select</span>(<span class='no'>x</span>, <span class='no'>keys</span>, <span class='no'>columns</span>, <span class='no'>keytype</span>, <span class='no'>...</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>column</th>
      <td><p>For <code>mapIds</code>: the column to search on, i.e. from which values
    should be retrieved.</p></td>
    </tr>
    <tr>
      <th>columns</th>
      <td><p>For <code>select</code>: the columns from which values should be
    retrieved. Use the <code>columns</code> method to list all possible
    columns.</p></td>
    </tr>
    <tr>
      <th>keys</th>
      <td><p>The keys/ids for which data should be retrieved from the
    database. This can be either a character vector of keys/IDs, a
    single filter object extending
    <code><a href='https://rdrr.io/pkg/AnnotationFilter/man/AnnotationFilter.html'>AnnotationFilter</a></code>, an combination of
    filters <code><a href='https://rdrr.io/pkg/AnnotationFilter/man/AnnotationFilterList.html'>AnnotationFilterList</a></code> or a
    <code>formula</code> representing a filter expression (see
    <code><a href='https://rdrr.io/pkg/AnnotationFilter/man/AnnotationFilter.html'>AnnotationFilter</a></code> for more details).</p></td>
    </tr>
    <tr>
      <th>keytype</th>
      <td><p>For <code>mapIds</code> and <code>select</code>: the type (column) that matches
    the provided keys. This argument does not have to be specified if
    argument <code>keys</code> is a filter object extending
    <code>AnnotationFilter</code> or a <code>list</code> of such objects.</p>
<p>For <code>keys</code>: which keys should be returned from the database.</p></td>
    </tr>
    <tr>
      <th>filter</th>
      <td><p>For <code>keys</code>: either a single object extending
    <code>AnnotationFilter</code> or a list of such object to
    retrieve only specific keys from the database.</p></td>
    </tr>
    <tr>
      <th>multiVals</th>
      <td><p>What should <code>mapIds</code> do when there are multiple values that
    could be returned? Options are: <code>"first"</code> (default), <code>"list"</code>,
    <code>"filter"</code>, <code>"asNA"</code>. See
    <code>mapIds</code> in the <code>AnnotationDbi</code> package  
    for a detailed description.</p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>The <code>EnsDb</code> object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Not used.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="methods-and-functions"><a class="anchor" href="#methods-and-functions"></a>Methods and Functions</h2>

    <p></p><dl>

    <dt>columns</dt><dd><p>List all the columns that can be retrieved by the <code>mapIds</code>
      and <code>select</code> methods. Note that these column names are
      different from the ones supported by the <code><a href='EnsDb-exonsBy.html'>genes</a></code>,
      <code><a href='EnsDb-exonsBy.html'>transcripts</a></code> etc. methods that can be listed by the
      <code><a href='EnsDb-class.html'>listColumns</a></code> method.      Returns a character vector of supported column names.</p></dd>

    <dt>keys</dt><dd><p>Retrieves all keys from the column name specified with
      <code>keytype</code>. By default (if <code>keytype</code> is not provided) it
      returns all gene IDs. Note that <code>keytype="TXNAME"</code> will
      return transcript ids, since no transcript names are available in
      the database.      Returns a character vector of IDs.</p></dd>

    <dt>keytypes</dt><dd><p>List all supported key types (column names).      Returns a character vector of key types.</p></dd>

    <dt>mapIds</dt><dd><p>Retrieve the mapped ids for a set of keys that are of a particular
      keytype. Argument <code>keys</code> can be either a character vector of
      keys/IDs, a single filter object extending
      <code>AnnotationFilter</code> or a list of such objects. For
      the latter, the argument <code>keytype</code> does not have to be
      specified. Importantly however, if the filtering system is used,
      the ordering of the results might not represent the ordering of
      the keys.      The method usually returns a named character vector or, depending
      on the argument <code>multiVals</code> a named list, with names
      corresponding to the keys (same ordering is only guaranteed if
      <code>keys</code> is a character vector).</p></dd>

    <dt>select</dt><dd><p>Retrieve the data as a <code>data.frame</code> based on parameters for
      selected <code>keys</code>, <code>columns</code> and <code>keytype</code>
      arguments. Multiple matches of the keys are returned in one row
      for each possible match. Argument <code>keys</code> can be either a
      character vector of keys/IDs, a single filter object extending
      <code>AnnotationFilter</code> or a list of such objects. For
      the latter, the argument <code>keytype</code> does not have to be
      specified.      Note that values from a column <code>"TXNAME"</code> will be the same
      than for a column <code>"TXID"</code>, since internally no database
      column <code>"tx_name"</code> is present and the column is thus mapped
      to <code>"tx_id"</code>.      Returns a <code>data.frame</code> with the column names corresponding to
      the argument <code>columns</code> and rows with all data matching the
      criteria specified with <code>keys</code>.      The use of <code>select</code> without filters or keys and without
      restricting to specicic columns is strongly discouraged, as the
      SQL query to join all of the tables, especially if protein
      annotation data is available is very expensive.</p></dd>

  
</dl>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>See method description above.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='EnsDb-class.html'>listColumns</a></code>
  <code><a href='EnsDb-exonsBy.html'>transcripts</a></code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='fu'><a href='https://rdrr.io/r/base/library.html'>library</a></span>(<span class='no'>EnsDb.Hsapiens.v86</span>)
<span class='no'>edb</span> <span class='kw'>&lt;-</span> <span class='no'>EnsDb.Hsapiens.v86</span>

<span class='co'>## List all supported keytypes.</span>
<span class='fu'>keytypes</span>(<span class='no'>edb</span>)</div><div class='output co'>#&gt;  [1] "ENTREZID"            "EXONID"              "GENEBIOTYPE"        
#&gt;  [4] "GENEID"              "GENENAME"            "PROTDOMID"          
#&gt;  [7] "PROTEINDOMAINID"     "PROTEINDOMAINSOURCE" "PROTEINID"          
#&gt; [10] "SEQNAME"             "SEQSTRAND"           "SYMBOL"             
#&gt; [13] "TXBIOTYPE"           "TXID"                "TXNAME"             
#&gt; [16] "UNIPROTID"          </div><div class='input'>
<span class='co'>## List all supported columns for the select and mapIds methods.</span>
<span class='fu'>columns</span>(<span class='no'>edb</span>)</div><div class='output co'>#&gt;  [1] "ENTREZID"            "EXONID"              "EXONIDX"            
#&gt;  [4] "EXONSEQEND"          "EXONSEQSTART"        "GENEBIOTYPE"        
#&gt;  [7] "GENEID"              "GENENAME"            "GENESEQEND"         
#&gt; [10] "GENESEQSTART"        "INTERPROACCESSION"   "ISCIRCULAR"         
#&gt; [13] "PROTDOMEND"          "PROTDOMSTART"        "PROTEINDOMAINID"    
#&gt; [16] "PROTEINDOMAINSOURCE" "PROTEINID"           "PROTEINSEQUENCE"    
#&gt; [19] "SEQCOORDSYSTEM"      "SEQLENGTH"           "SEQNAME"            
#&gt; [22] "SEQSTRAND"           "SYMBOL"              "TXBIOTYPE"          
#&gt; [25] "TXCDSSEQEND"         "TXCDSSEQSTART"       "TXID"               
#&gt; [28] "TXNAME"              "TXSEQEND"            "TXSEQSTART"         
#&gt; [31] "UNIPROTDB"           "UNIPROTID"           "UNIPROTMAPPINGTYPE" </div><div class='input'>
<span class='co'>## List /real/ database column names.</span>
<span class='fu'><a href='EnsDb-class.html'>listColumns</a></span>(<span class='no'>edb</span>)</div><div class='output co'>#&gt;  [1] "seq_name"              "seq_length"            "is_circular"          
#&gt;  [4] "gene_id"               "entrezid"              "exon_id"              
#&gt;  [7] "exon_seq_start"        "exon_seq_end"          "gene_name"            
#&gt; [10] "gene_biotype"          "gene_seq_start"        "gene_seq_end"         
#&gt; [13] "seq_strand"            "seq_coord_system"      "symbol"               
#&gt; [16] "name"                  "value"                 "tx_id"                
#&gt; [19] "protein_id"            "protein_sequence"      "protein_domain_id"    
#&gt; [22] "protein_domain_source" "interpro_accession"    "prot_dom_start"       
#&gt; [25] "prot_dom_end"          "tx_biotype"            "tx_seq_start"         
#&gt; [28] "tx_seq_end"            "tx_cds_seq_start"      "tx_cds_seq_end"       
#&gt; [31] "tx_name"               "exon_idx"              "uniprot_id"           
#&gt; [34] "uniprot_db"            "uniprot_mapping_type" </div><div class='input'>
<span class='co'>## Retrieve all keys corresponding to transcript ids.</span>
<span class='no'>txids</span> <span class='kw'>&lt;-</span> <span class='fu'>keys</span>(<span class='no'>edb</span>, <span class='kw'>keytype</span> <span class='kw'>=</span> <span class='st'>"TXID"</span>)
<span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span>(<span class='no'>txids</span>)</div><div class='output co'>#&gt; [1] 216741</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span>(<span class='no'>txids</span>)</div><div class='output co'>#&gt; [1] "ENST00000000233" "ENST00000000412" "ENST00000000442" "ENST00000001008"
#&gt; [5] "ENST00000001146" "ENST00000002125"</div><div class='input'>
<span class='co'>## Retrieve all keys corresponding to gene names of genes encoded on chromosome X</span>
<span class='no'>gids</span> <span class='kw'>&lt;-</span> <span class='fu'>keys</span>(<span class='no'>edb</span>, <span class='kw'>keytype</span> <span class='kw'>=</span> <span class='st'>"GENENAME"</span>, <span class='kw'>filter</span> <span class='kw'>=</span> <span class='fu'>SeqNameFilter</span>(<span class='st'>"X"</span>))
<span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span>(<span class='no'>gids</span>)</div><div class='output co'>#&gt; [1] 2300</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span>(<span class='no'>gids</span>)</div><div class='output co'>#&gt; [1] "TSPAN6" "TNMD"   "LAS1L"  "CD99"   "KLHL13" "ARX"   </div><div class='input'>
<span class='co'>## Get a mapping of the genes BCL2 and BCL2L11 to all of their</span>
<span class='co'>## transcript ids and return the result as list</span>
<span class='no'>maps</span> <span class='kw'>&lt;-</span> <span class='fu'>mapIds</span>(<span class='no'>edb</span>, <span class='kw'>keys</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"BCL2"</span>, <span class='st'>"BCL2L11"</span>), <span class='kw'>column</span> <span class='kw'>=</span> <span class='st'>"TXID"</span>,
               <span class='kw'>keytype</span> <span class='kw'>=</span> <span class='st'>"GENENAME"</span>, <span class='kw'>multiVals</span> <span class='kw'>=</span> <span class='st'>"list"</span>)
<span class='no'>maps</span></div><div class='output co'>#&gt; $BCL2
#&gt; [1] "ENST00000398117" "ENST00000333681" "ENST00000590515" "ENST00000589955"
#&gt; 
#&gt; $BCL2L11
#&gt;  [1] "ENST00000432179" "ENST00000308659" "ENST00000393256" "ENST00000393252"
#&gt;  [5] "ENST00000433098" "ENST00000405953" "ENST00000415458" "ENST00000436733"
#&gt;  [9] "ENST00000437029" "ENST00000452231" "ENST00000361493" "ENST00000431217"
#&gt; [13] "ENST00000439718" "ENST00000438054" "ENST00000337565" "ENST00000622509"
#&gt; [17] "ENST00000619294" "ENST00000610735" "ENST00000622612" "ENST00000357757"
#&gt; [21] "ENST00000615946" "ENST00000621302" "ENST00000620862" "LRG_620t1"      
#&gt; [25] "LRG_620t2"       "LRG_620t3"       "LRG_620t4"       "LRG_620t5"      
#&gt; </div><div class='input'>
<span class='co'>## Perform the same query using a combination of a GeneNameFilter and a</span>
<span class='co'>## TxBiotypeFilter to just retrieve protein coding transcripts for these</span>
<span class='co'>## two genes.</span>
<span class='fu'>mapIds</span>(<span class='no'>edb</span>, <span class='kw'>keys</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span>(<span class='fu'>GeneNameFilter</span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"BCL2"</span>, <span class='st'>"BCL2L11"</span>)),
                        <span class='fu'>TxBiotypeFilter</span>(<span class='st'>"protein_coding"</span>)), <span class='kw'>column</span> <span class='kw'>=</span> <span class='st'>"TXID"</span>,
       <span class='kw'>multiVals</span> <span class='kw'>=</span> <span class='st'>"list"</span>)</div><div class='output co'>#&gt; <span class='warning'>Warning: Got 2 filter objects. Will use the keys of the first for the mapping!</span></div><div class='output co'>#&gt; <span class='message'>Note: ordering of the results might not match ordering of keys!</span></div><div class='output co'>#&gt; $BCL2
#&gt; [1] "ENST00000398117" "ENST00000333681" "ENST00000589955"
#&gt; 
#&gt; $BCL2L11
#&gt;  [1] "ENST00000432179" "ENST00000308659" "ENST00000393256" "ENST00000393252"
#&gt;  [5] "ENST00000405953" "ENST00000438054" "ENST00000337565" "ENST00000622509"
#&gt;  [9] "ENST00000619294" "ENST00000610735" "ENST00000622612" "ENST00000357757"
#&gt; [13] "ENST00000615946" "ENST00000621302" "ENST00000620862"
#&gt; </div><div class='input'>
<span class='co'>## select:</span>
<span class='co'>## Retrieve all transcript and gene related information for the above example.</span>
<span class='fu'>select</span>(<span class='no'>edb</span>, <span class='kw'>keys</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span>(<span class='fu'>GeneNameFilter</span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"BCL2"</span>, <span class='st'>"BCL2L11"</span>)),
                        <span class='fu'>TxBiotypeFilter</span>(<span class='st'>"protein_coding"</span>)),
       <span class='kw'>columns</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"GENEID"</span>, <span class='st'>"GENENAME"</span>, <span class='st'>"TXID"</span>, <span class='st'>"TXBIOTYPE"</span>, <span class='st'>"TXSEQSTART"</span>,
                   <span class='st'>"TXSEQEND"</span>, <span class='st'>"SEQNAME"</span>, <span class='st'>"SEQSTRAND"</span>))</div><div class='output co'>#&gt; <span class='message'>Note: ordering of the results might not match ordering of keys!</span></div><div class='output co'>#&gt;             GENEID GENENAME            TXID      TXBIOTYPE TXSEQSTART  TXSEQEND
#&gt; 1  ENSG00000171791     BCL2 ENST00000398117 protein_coding   63123346  63320128
#&gt; 2  ENSG00000171791     BCL2 ENST00000333681 protein_coding   63127035  63319786
#&gt; 3  ENSG00000171791     BCL2 ENST00000589955 protein_coding   63313802  63318812
#&gt; 4  ENSG00000153094  BCL2L11 ENST00000432179 protein_coding  111119378 111124112
#&gt; 5  ENSG00000153094  BCL2L11 ENST00000308659 protein_coding  111120914 111165048
#&gt; 6  ENSG00000153094  BCL2L11 ENST00000393256 protein_coding  111120929 111168447
#&gt; 7  ENSG00000153094  BCL2L11 ENST00000393252 protein_coding  111122670 111123960
#&gt; 8  ENSG00000153094  BCL2L11 ENST00000405953 protein_coding  111123746 111128837
#&gt; 9  ENSG00000153094  BCL2L11 ENST00000438054 protein_coding  111123752 111146284
#&gt; 10 ENSG00000153094  BCL2L11 ENST00000337565 protein_coding  111120914 111128844
#&gt; 11 ENSG00000153094  BCL2L11 ENST00000622509 protein_coding  111120914 111168445
#&gt; 12 ENSG00000153094  BCL2L11 ENST00000619294 protein_coding  111120914 111168445
#&gt; 13 ENSG00000153094  BCL2L11 ENST00000610735 protein_coding  111120914 111168445
#&gt; 14 ENSG00000153094  BCL2L11 ENST00000622612 protein_coding  111120914 111168445
#&gt; 15 ENSG00000153094  BCL2L11 ENST00000357757 protein_coding  111120914 111146281
#&gt; 16 ENSG00000153094  BCL2L11 ENST00000615946 protein_coding  111120914 111168445
#&gt; 17 ENSG00000153094  BCL2L11 ENST00000621302 protein_coding  111120914 111168445
#&gt; 18 ENSG00000153094  BCL2L11 ENST00000620862 protein_coding  111120914 111168445
#&gt;    SEQNAME SEQSTRAND
#&gt; 1       18        -1
#&gt; 2       18        -1
#&gt; 3       18        -1
#&gt; 4        2         1
#&gt; 5        2         1
#&gt; 6        2         1
#&gt; 7        2         1
#&gt; 8        2         1
#&gt; 9        2         1
#&gt; 10       2         1
#&gt; 11       2         1
#&gt; 12       2         1
#&gt; 13       2         1
#&gt; 14       2         1
#&gt; 15       2         1
#&gt; 16       2         1
#&gt; 17       2         1
#&gt; 18       2         1</div><div class='input'>
<span class='co'>## Get all data for genes encoded on chromosome Y</span>
<span class='no'>Y</span> <span class='kw'>&lt;-</span> <span class='fu'>select</span>(<span class='no'>edb</span>, <span class='kw'>keys</span> <span class='kw'>=</span> <span class='st'>"Y"</span>, <span class='kw'>keytype</span> <span class='kw'>=</span> <span class='st'>"SEQNAME"</span>)
<span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span>(<span class='no'>Y</span>)</div><div class='output co'>#&gt;   ENTREZID          EXONID EXONIDX EXONSEQEND EXONSEQSTART    GENEBIOTYPE
#&gt; 1     8284 ENSE00001902471       1   19716695     19716596 protein_coding
#&gt; 2     8284 ENSE00003665408       2   19716473     19716279 protein_coding
#&gt; 3     8284 ENSE00003572891       3   19716004     19715823 protein_coding
#&gt; 4     8284 ENSE00003452486       4   19715739     19715584 protein_coding
#&gt; 5     8284 ENSE00003535737       5   19710472     19710376 protein_coding
#&gt; 6     8284 ENSE00003306433       6   19710260     19709451 protein_coding
#&gt;            GENEID GENENAME GENESEQEND GENESEQSTART INTERPROACCESSION ISCIRCULAR
#&gt; 1 ENSG00000012817    KDM5D   19744939     19703865              &lt;NA&gt;          0
#&gt; 2 ENSG00000012817    KDM5D   19744939     19703865              &lt;NA&gt;          0
#&gt; 3 ENSG00000012817    KDM5D   19744939     19703865              &lt;NA&gt;          0
#&gt; 4 ENSG00000012817    KDM5D   19744939     19703865              &lt;NA&gt;          0
#&gt; 5 ENSG00000012817    KDM5D   19744939     19703865              &lt;NA&gt;          0
#&gt; 6 ENSG00000012817    KDM5D   19744939     19703865              &lt;NA&gt;          0
#&gt;   PROTDOMEND PROTDOMSTART PROTEINDOMAINID PROTEINDOMAINSOURCE PROTEINID
#&gt; 1         NA           NA            &lt;NA&gt;                &lt;NA&gt;      &lt;NA&gt;
#&gt; 2         NA           NA            &lt;NA&gt;                &lt;NA&gt;      &lt;NA&gt;
#&gt; 3         NA           NA            &lt;NA&gt;                &lt;NA&gt;      &lt;NA&gt;
#&gt; 4         NA           NA            &lt;NA&gt;                &lt;NA&gt;      &lt;NA&gt;
#&gt; 5         NA           NA            &lt;NA&gt;                &lt;NA&gt;      &lt;NA&gt;
#&gt; 6         NA           NA            &lt;NA&gt;                &lt;NA&gt;      &lt;NA&gt;
#&gt;   PROTEINSEQUENCE SEQCOORDSYSTEM SEQLENGTH SEQNAME SEQSTRAND SYMBOL
#&gt; 1            &lt;NA&gt;     chromosome  57227415       Y        -1  KDM5D
#&gt; 2            &lt;NA&gt;     chromosome  57227415       Y        -1  KDM5D
#&gt; 3            &lt;NA&gt;     chromosome  57227415       Y        -1  KDM5D
#&gt; 4            &lt;NA&gt;     chromosome  57227415       Y        -1  KDM5D
#&gt; 5            &lt;NA&gt;     chromosome  57227415       Y        -1  KDM5D
#&gt; 6            &lt;NA&gt;     chromosome  57227415       Y        -1  KDM5D
#&gt;         TXBIOTYPE TXCDSSEQEND TXCDSSEQSTART            TXID          TXNAME
#&gt; 1 retained_intron          NA            NA ENST00000469599 ENST00000469599
#&gt; 2 retained_intron          NA            NA ENST00000469599 ENST00000469599
#&gt; 3 retained_intron          NA            NA ENST00000469599 ENST00000469599
#&gt; 4 retained_intron          NA            NA ENST00000469599 ENST00000469599
#&gt; 5 retained_intron          NA            NA ENST00000469599 ENST00000469599
#&gt; 6 retained_intron          NA            NA ENST00000469599 ENST00000469599
#&gt;   TXSEQEND TXSEQSTART UNIPROTDB UNIPROTID UNIPROTMAPPINGTYPE
#&gt; 1 19716695   19703865      &lt;NA&gt;      &lt;NA&gt;               &lt;NA&gt;
#&gt; 2 19716695   19703865      &lt;NA&gt;      &lt;NA&gt;               &lt;NA&gt;
#&gt; 3 19716695   19703865      &lt;NA&gt;      &lt;NA&gt;               &lt;NA&gt;
#&gt; 4 19716695   19703865      &lt;NA&gt;      &lt;NA&gt;               &lt;NA&gt;
#&gt; 5 19716695   19703865      &lt;NA&gt;      &lt;NA&gt;               &lt;NA&gt;
#&gt; 6 19716695   19703865      &lt;NA&gt;      &lt;NA&gt;               &lt;NA&gt;</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span>(<span class='no'>Y</span>)</div><div class='output co'>#&gt; [1] 28264</div><div class='input'>
<span class='co'>## Get selected columns for all lincRNAs encoded on chromosome Y. Here we use</span>
<span class='co'>## a filter expression to define what data to retrieve.</span>
<span class='no'>Y</span> <span class='kw'>&lt;-</span> <span class='fu'>select</span>(<span class='no'>edb</span>, <span class='kw'>keys</span> <span class='kw'>=</span> ~ <span class='no'>seq_name</span> <span class='kw'>==</span> <span class='st'>"Y"</span> <span class='kw'>&amp;</span> <span class='no'>gene_biotype</span> <span class='kw'>==</span> <span class='st'>"lincRNA"</span>,
            <span class='kw'>columns</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"GENEID"</span>, <span class='st'>"GENEBIOTYPE"</span>, <span class='st'>"TXID"</span>, <span class='st'>"GENENAME"</span>))</div><div class='output co'>#&gt; <span class='message'>Note: ordering of the results might not match ordering of keys!</span></div><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span>(<span class='no'>Y</span>)</div><div class='output co'>#&gt;            GENEID GENEBIOTYPE            TXID GENENAME SEQNAME
#&gt; 1 ENSG00000129816     lincRNA ENST00000250776   TTTY1B       Y
#&gt; 2 ENSG00000129845     lincRNA ENST00000250805    TTTY1       Y
#&gt; 3 ENSG00000131538     lincRNA ENST00000253838    TTTY6       Y
#&gt; 4 ENSG00000147753     lincRNA ENST00000457100    TTTY7       Y
#&gt; 5 ENSG00000147753     lincRNA ENST00000276770    TTTY7       Y
#&gt; 6 ENSG00000147753     lincRNA ENST00000449828    TTTY7       Y</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span>(<span class='no'>Y</span>)</div><div class='output co'>#&gt; [1] 73</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Johannes Rainer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


