<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Filters supported by ensembldb — Filter-classes • ensembldb</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Filters supported by ensembldb — Filter-classes" />
<meta property="og:description" content="ensembldb supports most of the filters from the AnnotationFilter
package to retrieve specific content from EnsDb databases. These filters
can be passed to the methods such as genes() with the filter parameter
or can be added as a global filter to an EnsDb object (see
addFilter() for more details). Use supportedFilters() to get an
overview of all filters supported by EnsDb object.
seqnames: accessor for the sequence names of the GRanges
object within a GRangesFilter.
seqnames: accessor for the seqlevels of the GRanges
object within a GRangesFilter.
supportedFilters returns a data.frame with the
names of all filters and the corresponding field supported by the
EnsDb object." />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



<!-- dependencies from examples -->


  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ensembldb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.15.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/ensembldb.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/MySQL-backend.html">Using a MySQL server backend</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping-use-cases.html">Use cases for coordinate mapping with ensembldb</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping.html">Mapping between genome, transcript and protein coordinates</a>
    </li>
    <li>
      <a href="../articles/proteins.html">Querying protein features</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jorainer/ensembldb/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Filters supported by ensembldb</h1>
    <small class="dont-index">Source: <a href='https://github.com/jorainer/ensembldb/blob/master/R/Classes.R'><code>R/Classes.R</code></a>, <a href='https://github.com/jorainer/ensembldb/blob/master/R/Methods-Filter.R'><code>R/Methods-Filter.R</code></a>, <a href='https://github.com/jorainer/ensembldb/blob/master/R/Methods.R'><code>R/Methods.R</code></a></small>
    <div class="hidden name"><code>Filter-classes.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>ensembldb</code> supports most of the filters from the AnnotationFilter
package to retrieve specific content from <a href='EnsDb.html'>EnsDb</a> databases. These filters
can be passed to the methods such as <code><a href='EnsDb-exonsBy.html'>genes()</a></code> with the <code>filter</code> parameter
or can be added as a <em>global</em> filter to an <code>EnsDb</code> object (see
<code><a href='global-filters.html'>addFilter()</a></code> for more details). Use <code>supportedFilters()</code> to get an
overview of all filters supported by <code>EnsDb</code> object.</p>
<p><code>seqnames</code>: accessor for the sequence names of the <code>GRanges</code>
object within a <code>GRangesFilter</code>.</p>
<p><code>seqnames</code>: accessor for the <code>seqlevels</code> of the <code>GRanges</code>
object within a <code>GRangesFilter</code>.</p>
<p><code>supportedFilters</code> returns a <code>data.frame</code> with the
names of all filters and the corresponding field supported by the
<code>EnsDb</code> object.</p>
    </div>

    <pre class="usage"><span class='fu'>OnlyCodingTxFilter</span><span class='op'>(</span><span class='op'>)</span>

<span class='fu'>ProtDomIdFilter</span><span class='op'>(</span><span class='va'>value</span>, condition <span class='op'>=</span> <span class='st'>"=="</span><span class='op'>)</span>

<span class='fu'>ProteinDomainIdFilter</span><span class='op'>(</span><span class='va'>value</span>, condition <span class='op'>=</span> <span class='st'>"=="</span><span class='op'>)</span>

<span class='fu'>ProteinDomainSourceFilter</span><span class='op'>(</span><span class='va'>value</span>, condition <span class='op'>=</span> <span class='st'>"=="</span><span class='op'>)</span>

<span class='fu'>UniprotDbFilter</span><span class='op'>(</span><span class='va'>value</span>, condition <span class='op'>=</span> <span class='st'>"=="</span><span class='op'>)</span>

<span class='fu'>UniprotMappingTypeFilter</span><span class='op'>(</span><span class='va'>value</span>, condition <span class='op'>=</span> <span class='st'>"=="</span><span class='op'>)</span>

<span class='fu'>TxSupportLevelFilter</span><span class='op'>(</span><span class='va'>value</span>, condition <span class='op'>=</span> <span class='st'>"=="</span><span class='op'>)</span>

<span class='co'># S4 method for GRangesFilter</span>
<span class='fu'>seqnames</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for GRangesFilter</span>
<span class='fu'><a href='EnsDb-class.html'>seqlevels</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>supportedFilters</span><span class='op'>(</span><span class='va'>object</span>, <span class='va'>...</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>value</th>
      <td><p>The value(s) for the filter. For <code>GRangesFilter</code> it has to be a
<code>GRanges</code> object.</p></td>
    </tr>
    <tr>
      <th>condition</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code> specifying the <em>condition</em> of the
filter. For <code>character</code>-based filters (such as
<code>GeneIdFilter</code>) <code>"=="</code>, <code>"!="</code>, <code>"startsWith"</code> and <code>"endsWith"</code> are
supported. Allowed values for <code>integer</code>-based filters (such as
<code>GeneStartFilter</code>) are <code>"=="</code>, <code>"!="</code>, <code>"&lt;"</code>. <code>"&lt;="</code>, <code>"&gt;"</code> and <code>"&gt;="</code>.</p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>For <code>seqnames</code>, <code>seqlevels</code>: a <code>GRangesFilter</code> object.</p></td>
    </tr>
    <tr>
      <th>object</th>
      <td><p>For <code>supportedFilters</code>: an <code>EnsDb</code> object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>For <code>supportedFilters</code>: currently not used.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>For <code>ProtDomIdFilter</code>: A <code>ProtDomIdFilter</code> object.</p>
<p>For <code>ProteinDomainIdFilter</code>: A <code>ProteinDomainIdFilter</code> object.</p>
<p>For <code>ProteinDomainSourceFilter</code>: A <code>ProteinDomainSourceFilter</code>
object.</p>
<p>For <code>UniprotDbFilter</code>: A <code>UniprotDbFilter</code> object.</p>
<p>For <code>UniprotMappingTypeFilter</code>: A <code>UniprotMappingTypeFilter</code> object.</p>
<p>For <code>TxSupportLevel</code>: A <code>TxSupportLevel</code> object.</p>
<p>For <code>supportedFilters</code>: a <code>data.frame</code> with the names and
the corresponding field of the supported filter classes.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><code>ensembldb</code> supports the following filters from the <code>AnnotationFilter</code>
package:</p><ul>
<li><p><code>GeneIdFilter</code>: filter based on the Ensembl gene ID.</p></li>
<li><p><code>GeneNameFilter</code>: filter based on the name of the gene as provided
Ensembl. In most cases this will correspond to the official gene symbol.</p></li>
<li><p><code>SymbolFilter</code> filter based on the gene names. <code>EnsDb</code> objects don't
have a dedicated <em>symbol</em> column, the filtering is hence based on the
gene names.</p></li>
<li><p><code>GeneBiotype</code>: filter based on the biotype of genes (e.g.
<code>"protein_coding"</code>).</p></li>
<li><p><code>GeneStartFilter</code>: filter based on the genomic start coordinate of genes.</p></li>
<li><p><code>GeneEndFilter</code>: filter based on the genomic end coordinate of genes.</p></li>
<li><p><code>EntrezidFilter</code>: filter based on the genes' NCBI Entrezgene ID.</p></li>
<li><p><code>TxIdFilter</code>: filter based on the Ensembld transcript ID.</p></li>
<li><p><code>TxNameFilter</code>: filter based on the Ensembld transcript ID; no transcript
names are provided in <code>EnsDb</code> databases.</p></li>
<li><p><code>TxBiotypeFilter</code>: filter based on the transcripts' biotype.</p></li>
<li><p><code>TxStartFilter</code>: filter based on the genomic start coordinate of the
transcripts.</p></li>
<li><p><code>TxEndFilter</code>: filter based on the genonic end coordinates of the
transcripts.</p></li>
<li><p><code>ExonIdFilter</code>: filter based on Ensembl exon IDs.</p></li>
<li><p><code>ExonRankFilter</code>: filter based on the index/rank of the exon within the
transcrips.</p></li>
<li><p><code>ExonStartFilter</code>: filter based on the genomic start coordinates of the
exons.</p></li>
<li><p><code>ExonEndFilter</code>: filter based on the genomic end coordinates of the exons.</p></li>
<li><p><code>GRangesFilter</code>: Allows to fetch features within or overlapping specified
genomic region(s)/range(s). This filter takes a <code>GRanges</code> object
as input and, if <code>type = "any"</code> (the default) will restrict results to
features (genes, transcripts or exons) that are partially overlapping the
region. Alternatively, by specifying <code>condition = "within"</code> it will
return features located within the range. In addition, the <code>GRangesFilter</code>
<code>condition = "start"</code>, <code>condition = "end"</code> and <code>condition = "equal"</code>
filtering for features with the same start or end coordinate or that are
equal to the <code>GRanges</code>.</p>
<p>Note that the type of feature on which the filter is applied depends on
the method that is called, i.e. <code><a href='EnsDb-exonsBy.html'>genes()</a></code> will filter on the
genomic coordinates of genes, <code><a href='EnsDb-exonsBy.html'>transcripts()</a></code> on those of
transcripts and <code><a href='EnsDb-exonsBy.html'>exons()</a></code> on exon coordinates.</p>
<p>Calls to the methods <code><a href='EnsDb-exonsBy.html'>exonsBy()</a></code>, <code><a href='EnsDb-exonsBy.html'>cdsBy()</a></code> and
<code><a href='EnsDb-exonsBy.html'>transcriptsBy()</a></code> use the start and end coordinates of the
feature type specified with argument <code>by</code> (i.e. <code>"gene"</code>,
<code>"transcript"</code> or <code>"exon"</code>) for the filtering.</p>
<p>If the specified <code>GRanges</code> object defines multiple regions, all
features within (or overlapping) any of these regions are returned.</p>
<p>Chromosome names/seqnames can be provided in UCSC format (e.g.
<code>"chrX"</code>) or Ensembl format (e.g. <code>"X"</code>); see <code><a href='EnsDb-seqlevels.html'>seqlevelsStyle()</a></code> for
more information.</p></li>
<li><p><code>SeqNameFilter</code>: filter based on chromosome names.</p></li>
<li><p><code>SeqStrandFilter</code>: filter based on the chromosome strand. The strand can
be specified with <code>value = "+"</code>, <code>value = "-"</code>, <code>value = -1</code> or
<code>value = 1</code>.</p></li>
<li><p><code>ProteinIdFilter</code>: filter based on Ensembl protein IDs. This filter is
only supported if the <code>EnsDb</code> provides protein annotations; use the
<code><a href='hasProteinData-EnsDb-method.html'>hasProteinData()</a></code> method to check.</p></li>
<li><p><code>UniprotFilter</code>: filter based on Uniprot IDs. This filter is only
supported if the <code>EnsDb</code> provides protein annotations; use the
<code><a href='hasProteinData-EnsDb-method.html'>hasProteinData()</a></code> method to check.</p></li>
</ul>

<p>In addition, the following filters are defined by <code>ensembldb</code>:</p><ul>
<li><p><code>TxSupportLevel</code>: allows to filter results using the provided transcript
support level. Support levels for transcripts are defined by Ensembl
based on the available evidences for a transcript with 1 being the
highest evidence grade and 5 the lowest level. This filter is only
supported on <code>EnsDb</code> databases with a db schema version higher 2.1.</p></li>
<li><p><code>UniprotDbFilter</code>: allows to filter results based on the specified Uniprot
database name(s).</p></li>
<li><p><code>UniprotMappingTypeFilter</code>: allows to filter results based on the mapping
method/type that was used to assign Uniprot IDs to Ensembl protein IDs.</p></li>
<li><p><code>ProtDomIdFilter</code>, <code>ProteinDomainIdFilter</code>: allows to retrieve entries
from the database matching the provided filter criteria based on their
protein domain ID (<em>protein_domain_id</em>).</p></li>
<li><p><code>ProteinDomainSourceFilter</code>: filter results based on the source
(database/method) defining the protein domain (e.g. <code>"pfam"</code>).</p></li>
<li><p><code>OnlyCodingTxFilter</code>: allows to retrieve entries only for protein coding
transcripts, i.e. transcripts with a CDS. This filter does not take any
input arguments.</p></li>
</ul>

    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>For users of <code>ensembldb</code> version &lt; 2.0: in the <code>GRangesFilter</code> from the
<code>AnnotationFilter</code> package the <code>condition</code> parameter was renamed to <code>type</code>
(to be consistent with the <code>IRanges</code> package). In addition,
<code>condition = "overlapping"</code> is no longer recognized. To retrieve all
features overlapping the range <code>type = "any"</code> has to be used.</p>
<p>Protein annotation based filters can only be used if the
<code>EnsDb</code> database contains protein annotations, i.e. if <code>hasProteinData</code>
is <code>TRUE</code>. Also, only protein coding transcripts will have protein
annotations available, thus, non-coding transcripts/genes will not be
returned by the queries using protein annotation filters.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code>supportedFilters()</code> to list all filters supported for <code>EnsDb</code> objects.</p>
<p><code><a href='ProteinFunctionality.html'>listUniprotDbs()</a></code> and <code><a href='ProteinFunctionality.html'>listUniprotMappingTypes()</a></code> to list all Uniprot
database names respectively mapping method types from the database.</p>
<p><code>GeneIdFilter()</code> in the <code>AnnotationFilter</code> package for more details on the
filter objects.</p>
<p><code><a href='EnsDb-exonsBy.html'>genes()</a></code>, <code><a href='EnsDb-exonsBy.html'>transcripts()</a></code>, <code><a href='EnsDb-exonsBy.html'>exons()</a></code>, <code><a href='EnsDb-class.html'>listGenebiotypes()</a></code>,
<code><a href='EnsDb-class.html'>listTxbiotypes()</a></code>.</p>
<p><code><a href='global-filters.html'>addFilter()</a></code> and <code><a href='global-filters.html'>filter()</a></code> for globally adding filters to an <code>EnsDb</code>.</p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Johannes Rainer</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='co'>## Create a filter that could be used to retrieve all informations for</span>
<span class='co'>## the respective gene.</span>
<span class='va'>gif</span> <span class='op'>&lt;-</span> <span class='fu'>GeneIdFilter</span><span class='op'>(</span><span class='st'>"ENSG00000012817"</span><span class='op'>)</span>
<span class='va'>gif</span>
</div><div class='output co'>#&gt; class: GeneIdFilter 
#&gt; condition: == 
#&gt; value: ENSG00000012817 </div><div class='input'>
<span class='co'>## Create a filter for a chromosomal end position of a gene</span>
<span class='va'>sef</span> <span class='op'>&lt;-</span> <span class='fu'>GeneEndFilter</span><span class='op'>(</span><span class='fl'>10000</span>, condition <span class='op'>=</span> <span class='st'>"&gt;"</span><span class='op'>)</span>
<span class='va'>sef</span>
</div><div class='output co'>#&gt; class: GeneEndFilter 
#&gt; condition: &gt; 
#&gt; value: 10000 </div><div class='input'>
<span class='co'>## For additional examples see the help page of "genes".</span>


<span class='co'>## Example for GRangesFilter:</span>
<span class='co'>## retrieve all genes overlapping the specified region</span>
<span class='va'>grf</span> <span class='op'>&lt;-</span> <span class='fu'>GRangesFilter</span><span class='op'>(</span><span class='fu'>GRanges</span><span class='op'>(</span><span class='st'>"11"</span>, ranges <span class='op'>=</span> <span class='fu'>IRanges</span><span class='op'>(</span><span class='fl'>114129278</span>, <span class='fl'>114129328</span><span class='op'>)</span>,
                             strand <span class='op'>=</span> <span class='st'>"+"</span><span class='op'>)</span>, type <span class='op'>=</span> <span class='st'>"any"</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>EnsDb.Hsapiens.v86</span><span class='op'>)</span>
<span class='va'>edb</span> <span class='op'>&lt;-</span> <span class='va'>EnsDb.Hsapiens.v86</span>
<span class='fu'><a href='EnsDb-exonsBy.html'>genes</a></span><span class='op'>(</span><span class='va'>edb</span>, filter <span class='op'>=</span> <span class='va'>grf</span><span class='op'>)</span>
</div><div class='output co'>#&gt; GRanges object with 1 range and 6 metadata columns:
#&gt;                   seqnames              ranges strand |         gene_id
#&gt;                      &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
#&gt;   ENSG00000109906       11 114059593-114250676      + | ENSG00000109906
#&gt;                     gene_name   gene_biotype seq_coord_system      symbol
#&gt;                   &lt;character&gt;    &lt;character&gt;      &lt;character&gt; &lt;character&gt;
#&gt;   ENSG00000109906      ZBTB16 protein_coding       chromosome      ZBTB16
#&gt;                   entrezid
#&gt;                     &lt;list&gt;
#&gt;   ENSG00000109906     7704
#&gt;   -------
#&gt;   seqinfo: 1 sequence from GRCh38 genome</div><div class='input'>
<span class='co'>## Get also all transcripts overlapping that region.</span>
<span class='fu'><a href='EnsDb-exonsBy.html'>transcripts</a></span><span class='op'>(</span><span class='va'>edb</span>, filter <span class='op'>=</span> <span class='va'>grf</span><span class='op'>)</span>
</div><div class='output co'>#&gt; GRanges object with 4 ranges and 6 metadata columns:
#&gt;                   seqnames              ranges strand |           tx_id
#&gt;                      &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
#&gt;   ENST00000335953       11 114059593-114250676      + | ENST00000335953
#&gt;   ENST00000541602       11 114059725-114189764      + | ENST00000541602
#&gt;   ENST00000392996       11 114060507-114250652      + | ENST00000392996
#&gt;   ENST00000539918       11 114064412-114247344      + | ENST00000539918
#&gt;                               tx_biotype tx_cds_seq_start tx_cds_seq_end
#&gt;                              &lt;character&gt;        &lt;integer&gt;      &lt;integer&gt;
#&gt;   ENST00000335953         protein_coding        114063301      114250555
#&gt;   ENST00000541602        retained_intron             &lt;NA&gt;           &lt;NA&gt;
#&gt;   ENST00000392996         protein_coding        114063301      114250555
#&gt;   ENST00000539918 nonsense_mediated_de..        114064412      114121827
#&gt;                           gene_id         tx_name
#&gt;                       &lt;character&gt;     &lt;character&gt;
#&gt;   ENST00000335953 ENSG00000109906 ENST00000335953
#&gt;   ENST00000541602 ENSG00000109906 ENST00000541602
#&gt;   ENST00000392996 ENSG00000109906 ENST00000392996
#&gt;   ENST00000539918 ENSG00000109906 ENST00000539918
#&gt;   -------
#&gt;   seqinfo: 1 sequence from GRCh38 genome</div><div class='input'>
<span class='co'>## Retrieve all transcripts for the above gene</span>
<span class='va'>gn</span> <span class='op'>&lt;-</span> <span class='fu'><a href='EnsDb-exonsBy.html'>genes</a></span><span class='op'>(</span><span class='va'>edb</span>, filter <span class='op'>=</span> <span class='va'>grf</span><span class='op'>)</span>
<span class='va'>txs</span> <span class='op'>&lt;-</span> <span class='fu'><a href='EnsDb-exonsBy.html'>transcripts</a></span><span class='op'>(</span><span class='va'>edb</span>, filter <span class='op'>=</span> <span class='fu'>GeneNameFilter</span><span class='op'>(</span><span class='va'>gn</span><span class='op'>$</span><span class='va'>gene_name</span><span class='op'>)</span><span class='op'>)</span>
<span class='co'>## Next we simply plot their start and end coordinates.</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='fl'>3</span>, <span class='fl'>3</span>, pch<span class='op'>=</span><span class='cn'>NA</span>, xlim<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/stats/start.html'>start</a></span><span class='op'>(</span><span class='va'>gn</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/stats/start.html'>end</a></span><span class='op'>(</span><span class='va'>gn</span><span class='op'>)</span><span class='op'>)</span>, ylim<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='va'>txs</span><span class='op'>)</span><span class='op'>)</span>,
yaxt<span class='op'>=</span><span class='st'>"n"</span>, ylab<span class='op'>=</span><span class='st'>""</span><span class='op'>)</span>
</div><div class='input'><span class='co'>## Highlight the GRangesFilter region</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/rect.html'>rect</a></span><span class='op'>(</span>xleft<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/stats/start.html'>start</a></span><span class='op'>(</span><span class='va'>grf</span><span class='op'>)</span>, xright<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/stats/start.html'>end</a></span><span class='op'>(</span><span class='va'>grf</span><span class='op'>)</span>, ybottom<span class='op'>=</span><span class='fl'>0</span>, ytop<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='va'>txs</span><span class='op'>)</span>,
col<span class='op'>=</span><span class='st'>"red"</span>, border<span class='op'>=</span><span class='st'>"red"</span><span class='op'>)</span>
</div><div class='input'><span class='kw'>for</span><span class='op'>(</span><span class='va'>i</span> <span class='kw'>in</span> <span class='fl'>1</span><span class='op'>:</span><span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='va'>txs</span><span class='op'>)</span><span class='op'>)</span><span class='op'>{</span>
    <span class='va'>current</span> <span class='op'>&lt;-</span> <span class='va'>txs</span><span class='op'>[</span><span class='va'>i</span><span class='op'>]</span>
    <span class='fu'><a href='https://rdrr.io/r/graphics/rect.html'>rect</a></span><span class='op'>(</span>xleft<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/stats/start.html'>start</a></span><span class='op'>(</span><span class='va'>current</span><span class='op'>)</span>, xright<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/stats/start.html'>end</a></span><span class='op'>(</span><span class='va'>current</span><span class='op'>)</span>, ybottom<span class='op'>=</span><span class='va'>i</span><span class='op'>-</span><span class='fl'>0.975</span>, ytop<span class='op'>=</span><span class='va'>i</span><span class='op'>-</span><span class='fl'>0.125</span>, border<span class='op'>=</span><span class='st'>"grey"</span><span class='op'>)</span>
    <span class='fu'><a href='https://rdrr.io/r/graphics/text.html'>text</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/stats/start.html'>start</a></span><span class='op'>(</span><span class='va'>current</span><span class='op'>)</span>, y<span class='op'>=</span><span class='va'>i</span><span class='op'>-</span><span class='fl'>0.5</span>,pos<span class='op'>=</span><span class='fl'>4</span>, cex<span class='op'>=</span><span class='fl'>0.75</span>, labels<span class='op'>=</span><span class='va'>current</span><span class='op'>$</span><span class='va'>tx_id</span><span class='op'>)</span>
<span class='op'>}</span>
</div><div class='img'><img src='Filter-classes-1.png' alt='' width='700' height='433' /></div><div class='input'><span class='co'>## Thus, we can see that only 4 transcripts of that gene are indeed</span>
<span class='co'>## overlapping the region.</span>


<span class='co'>## No exon is overlapping that region, thus we're not getting anything</span>
<span class='fu'><a href='EnsDb-exonsBy.html'>exons</a></span><span class='op'>(</span><span class='va'>edb</span>, filter <span class='op'>=</span> <span class='va'>grf</span><span class='op'>)</span>
</div><div class='output co'>#&gt; GRanges object with 0 ranges and 1 metadata column:
#&gt;    seqnames    ranges strand |     exon_id
#&gt;       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
#&gt;   -------
#&gt;   seqinfo: no sequences</div><div class='input'>

<span class='co'>## Example for ExonRankFilter</span>
<span class='co'>## Extract all exons 1 and (if present) 2 for all genes encoded on the</span>
<span class='co'>## Y chromosome</span>
<span class='fu'><a href='EnsDb-exonsBy.html'>exons</a></span><span class='op'>(</span><span class='va'>edb</span>, columns <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"tx_id"</span>, <span class='st'>"exon_idx"</span><span class='op'>)</span>,
      filter<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='fu'>SeqNameFilter</span><span class='op'>(</span><span class='st'>"Y"</span><span class='op'>)</span>,
                  <span class='fu'>ExonRankFilter</span><span class='op'>(</span><span class='fl'>3</span>, condition <span class='op'>=</span> <span class='st'>"&lt;"</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt; GRanges object with 1294 ranges and 3 metadata columns:
#&gt;                   seqnames            ranges strand |           tx_id  exon_idx
#&gt;                      &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt; &lt;integer&gt;
#&gt;   ENSE00002088309        Y   2784749-2784853      + | ENST00000516032         1
#&gt;   ENSE00001494622        Y   2786855-2787699      - | ENST00000383070         1
#&gt;   ENSE00001772499        Y   2789827-2790328      + | ENST00000454281         1
#&gt;   ENSE00001614266        Y   2827982-2828218      + | ENST00000430735         1
#&gt;   ENSE00002490412        Y   2841486-2841627      + | ENST00000250784         1
#&gt;               ...      ...               ...    ... .             ...       ...
#&gt;   ENSE00001632993        Y 26591548-26591601      - | ENST00000456738         1
#&gt;   ENSE00001616687        Y 26626520-26627159      - | ENST00000435741         1
#&gt;   ENSE00001638296        Y 26633345-26633431      - | ENST00000435945         2
#&gt;   ENSE00001797328        Y 26634523-26634652      - | ENST00000435945         1
#&gt;   ENSE00001794473        Y 56855244-56855488      + | ENST00000431853         1
#&gt;                           exon_id
#&gt;                       &lt;character&gt;
#&gt;   ENSE00002088309 ENSE00002088309
#&gt;   ENSE00001494622 ENSE00001494622
#&gt;   ENSE00001772499 ENSE00001772499
#&gt;   ENSE00001614266 ENSE00001614266
#&gt;   ENSE00002490412 ENSE00002490412
#&gt;               ...             ...
#&gt;   ENSE00001632993 ENSE00001632993
#&gt;   ENSE00001616687 ENSE00001616687
#&gt;   ENSE00001638296 ENSE00001638296
#&gt;   ENSE00001797328 ENSE00001797328
#&gt;   ENSE00001794473 ENSE00001794473
#&gt;   -------
#&gt;   seqinfo: 1 sequence from GRCh38 genome</div><div class='input'>

<span class='co'>## Get all transcripts for the gene SKA2</span>
<span class='fu'><a href='EnsDb-exonsBy.html'>transcripts</a></span><span class='op'>(</span><span class='va'>edb</span>, filter <span class='op'>=</span> <span class='fu'>GeneNameFilter</span><span class='op'>(</span><span class='st'>"SKA2"</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt; GRanges object with 10 ranges and 7 metadata columns:
#&gt;                   seqnames            ranges strand |           tx_id
#&gt;                      &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
#&gt;   ENST00000330137       17 59109951-59155269      - | ENST00000330137
#&gt;   ENST00000580541       17 59111886-59155184      - | ENST00000580541
#&gt;   ENST00000578105       17 59111888-59155184      - | ENST00000578105
#&gt;   ENST00000578519       17 59111890-59155182      - | ENST00000578519
#&gt;   ENST00000583927       17 59111943-59155260      - | ENST00000583927
#&gt;   ENST00000581068       17 59111956-59155248      - | ENST00000581068
#&gt;   ENST00000437036       17 59112047-59155194      - | ENST00000437036
#&gt;   ENST00000583380       17 59112091-59155174      - | ENST00000583380
#&gt;   ENST00000583976       17 59112262-59155177      - | ENST00000583976
#&gt;   ENST00000584089       17 59119387-59151801      - | ENST00000584089
#&gt;                               tx_biotype tx_cds_seq_start tx_cds_seq_end
#&gt;                              &lt;character&gt;        &lt;integer&gt;      &lt;integer&gt;
#&gt;   ENST00000330137         protein_coding         59112277       59155163
#&gt;   ENST00000580541         protein_coding         59112333       59155083
#&gt;   ENST00000578105         protein_coding         59112277       59155163
#&gt;   ENST00000578519 nonsense_mediated_de..         59124363       59155163
#&gt;   ENST00000583927   processed_transcript             &lt;NA&gt;           &lt;NA&gt;
#&gt;   ENST00000581068         protein_coding         59112277       59155163
#&gt;   ENST00000437036         protein_coding         59112333       59155083
#&gt;   ENST00000583380         protein_coding         59112333       59155163
#&gt;   ENST00000583976 nonsense_mediated_de..         59119474       59155083
#&gt;   ENST00000584089         protein_coding         59119387       59151771
#&gt;                           gene_id         tx_name   gene_name
#&gt;                       &lt;character&gt;     &lt;character&gt; &lt;character&gt;
#&gt;   ENST00000330137 ENSG00000182628 ENST00000330137        SKA2
#&gt;   ENST00000580541 ENSG00000182628 ENST00000580541        SKA2
#&gt;   ENST00000578105 ENSG00000182628 ENST00000578105        SKA2
#&gt;   ENST00000578519 ENSG00000182628 ENST00000578519        SKA2
#&gt;   ENST00000583927 ENSG00000182628 ENST00000583927        SKA2
#&gt;   ENST00000581068 ENSG00000182628 ENST00000581068        SKA2
#&gt;   ENST00000437036 ENSG00000182628 ENST00000437036        SKA2
#&gt;   ENST00000583380 ENSG00000182628 ENST00000583380        SKA2
#&gt;   ENST00000583976 ENSG00000182628 ENST00000583976        SKA2
#&gt;   ENST00000584089 ENSG00000182628 ENST00000584089        SKA2
#&gt;   -------
#&gt;   seqinfo: 1 sequence from GRCh38 genome</div><div class='input'>
<span class='co'>## Which is the same as using a SymbolFilter</span>
<span class='fu'><a href='EnsDb-exonsBy.html'>transcripts</a></span><span class='op'>(</span><span class='va'>edb</span>, filter <span class='op'>=</span> <span class='fu'>SymbolFilter</span><span class='op'>(</span><span class='st'>"SKA2"</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt; GRanges object with 10 ranges and 7 metadata columns:
#&gt;                   seqnames            ranges strand |           tx_id
#&gt;                      &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
#&gt;   ENST00000330137       17 59109951-59155269      - | ENST00000330137
#&gt;   ENST00000580541       17 59111886-59155184      - | ENST00000580541
#&gt;   ENST00000578105       17 59111888-59155184      - | ENST00000578105
#&gt;   ENST00000578519       17 59111890-59155182      - | ENST00000578519
#&gt;   ENST00000583927       17 59111943-59155260      - | ENST00000583927
#&gt;   ENST00000581068       17 59111956-59155248      - | ENST00000581068
#&gt;   ENST00000437036       17 59112047-59155194      - | ENST00000437036
#&gt;   ENST00000583380       17 59112091-59155174      - | ENST00000583380
#&gt;   ENST00000583976       17 59112262-59155177      - | ENST00000583976
#&gt;   ENST00000584089       17 59119387-59151801      - | ENST00000584089
#&gt;                               tx_biotype tx_cds_seq_start tx_cds_seq_end
#&gt;                              &lt;character&gt;        &lt;integer&gt;      &lt;integer&gt;
#&gt;   ENST00000330137         protein_coding         59112277       59155163
#&gt;   ENST00000580541         protein_coding         59112333       59155083
#&gt;   ENST00000578105         protein_coding         59112277       59155163
#&gt;   ENST00000578519 nonsense_mediated_de..         59124363       59155163
#&gt;   ENST00000583927   processed_transcript             &lt;NA&gt;           &lt;NA&gt;
#&gt;   ENST00000581068         protein_coding         59112277       59155163
#&gt;   ENST00000437036         protein_coding         59112333       59155083
#&gt;   ENST00000583380         protein_coding         59112333       59155163
#&gt;   ENST00000583976 nonsense_mediated_de..         59119474       59155083
#&gt;   ENST00000584089         protein_coding         59119387       59151771
#&gt;                           gene_id         tx_name      symbol
#&gt;                       &lt;character&gt;     &lt;character&gt; &lt;character&gt;
#&gt;   ENST00000330137 ENSG00000182628 ENST00000330137        SKA2
#&gt;   ENST00000580541 ENSG00000182628 ENST00000580541        SKA2
#&gt;   ENST00000578105 ENSG00000182628 ENST00000578105        SKA2
#&gt;   ENST00000578519 ENSG00000182628 ENST00000578519        SKA2
#&gt;   ENST00000583927 ENSG00000182628 ENST00000583927        SKA2
#&gt;   ENST00000581068 ENSG00000182628 ENST00000581068        SKA2
#&gt;   ENST00000437036 ENSG00000182628 ENST00000437036        SKA2
#&gt;   ENST00000583380 ENSG00000182628 ENST00000583380        SKA2
#&gt;   ENST00000583976 ENSG00000182628 ENST00000583976        SKA2
#&gt;   ENST00000584089 ENSG00000182628 ENST00000584089        SKA2
#&gt;   -------
#&gt;   seqinfo: 1 sequence from GRCh38 genome</div><div class='input'>

<span class='co'>## Create a ProteinIdFilter:</span>
<span class='va'>pf</span> <span class='op'>&lt;-</span> <span class='fu'>ProteinIdFilter</span><span class='op'>(</span><span class='st'>"ENSP00000362111"</span><span class='op'>)</span>
<span class='va'>pf</span>
</div><div class='output co'>#&gt; class: ProteinIdFilter 
#&gt; condition: == 
#&gt; value: ENSP00000362111 </div><div class='input'><span class='co'>## Using this filter would retrieve all database entries that are associated</span>
<span class='co'>## with a protein with the ID "ENSP00000362111"</span>
<span class='kw'>if</span> <span class='op'>(</span><span class='fu'><a href='hasProteinData-EnsDb-method.html'>hasProteinData</a></span><span class='op'>(</span><span class='va'>edb</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='va'>res</span> <span class='op'>&lt;-</span> <span class='fu'><a href='EnsDb-exonsBy.html'>genes</a></span><span class='op'>(</span><span class='va'>edb</span>, filter <span class='op'>=</span> <span class='va'>pf</span><span class='op'>)</span>
    <span class='va'>res</span>
<span class='op'>}</span>
</div><div class='output co'>#&gt; GRanges object with 1 range and 7 metadata columns:
#&gt;                   seqnames              ranges strand |         gene_id
#&gt;                      &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
#&gt;   ENSG00000000003        X 100627109-100639991      - | ENSG00000000003
#&gt;                     gene_name   gene_biotype seq_coord_system      symbol
#&gt;                   &lt;character&gt;    &lt;character&gt;      &lt;character&gt; &lt;character&gt;
#&gt;   ENSG00000000003      TSPAN6 protein_coding       chromosome      TSPAN6
#&gt;                   entrezid      protein_id
#&gt;                     &lt;list&gt;     &lt;character&gt;
#&gt;   ENSG00000000003     7105 ENSP00000362111
#&gt;   -------
#&gt;   seqinfo: 1 sequence from GRCh38 genome</div><div class='input'>
<span class='co'>## UniprotFilter:</span>
<span class='va'>uf</span> <span class='op'>&lt;-</span> <span class='fu'>UniprotFilter</span><span class='op'>(</span><span class='st'>"O60762"</span><span class='op'>)</span>
<span class='co'>## Get the transcripts encoding that protein:</span>
<span class='kw'>if</span> <span class='op'>(</span><span class='fu'><a href='hasProteinData-EnsDb-method.html'>hasProteinData</a></span><span class='op'>(</span><span class='va'>edb</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='fu'><a href='EnsDb-exonsBy.html'>transcripts</a></span><span class='op'>(</span><span class='va'>edb</span>, filter <span class='op'>=</span> <span class='va'>uf</span><span class='op'>)</span>
    <span class='co'>## The mapping Ensembl protein ID to Uniprot ID can however be 1:n:</span>
    <span class='fu'><a href='EnsDb-exonsBy.html'>transcripts</a></span><span class='op'>(</span><span class='va'>edb</span>, filter <span class='op'>=</span> <span class='fu'>TxIdFilter</span><span class='op'>(</span><span class='st'>"ENST00000371588"</span><span class='op'>)</span>,
        columns <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"protein_id"</span>, <span class='st'>"uniprot_id"</span><span class='op'>)</span><span class='op'>)</span>
<span class='op'>}</span>
</div><div class='output co'>#&gt; GRanges object with 2 ranges and 3 metadata columns:
#&gt;                   seqnames            ranges strand |      protein_id
#&gt;                      &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
#&gt;   ENST00000371588       20 50934867-50958550      - | ENSP00000360644
#&gt;   ENST00000371588       20 50934867-50958550      - | ENSP00000360644
#&gt;                    uniprot_id           tx_id
#&gt;                   &lt;character&gt;     &lt;character&gt;
#&gt;   ENST00000371588      O60762 ENST00000371588
#&gt;   ENST00000371588  A0A0S2Z4Y5 ENST00000371588
#&gt;   -------
#&gt;   seqinfo: 1 sequence from GRCh38 genome</div><div class='input'>
<span class='co'>## ProtDomIdFilter:</span>
<span class='va'>pdf</span> <span class='op'>&lt;-</span> <span class='fu'>ProtDomIdFilter</span><span class='op'>(</span><span class='st'>"PF00335"</span><span class='op'>)</span>
<span class='co'>## Also here we could get all transcripts related to that protein domain</span>
<span class='kw'>if</span> <span class='op'>(</span><span class='fu'><a href='hasProteinData-EnsDb-method.html'>hasProteinData</a></span><span class='op'>(</span><span class='va'>edb</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='fu'><a href='EnsDb-exonsBy.html'>transcripts</a></span><span class='op'>(</span><span class='va'>edb</span>, filter <span class='op'>=</span> <span class='va'>pdf</span>, columns <span class='op'>=</span> <span class='st'>"protein_id"</span><span class='op'>)</span>
<span class='op'>}</span>
</div><div class='output co'>#&gt; GRanges object with 165 ranges and 3 metadata columns:
#&gt;                   seqnames              ranges strand |      protein_id
#&gt;                      &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
#&gt;   ENST00000372003        1   46175087-46185958      + | ENSP00000361072
#&gt;   ENST00000271324        1 110873154-110899928      + | ENSP00000271324
#&gt;   ENST00000369516        1 115048011-115089464      - | ENSP00000358529
#&gt;   ENST00000369515        1 115050356-115089500      - | ENSP00000358528
#&gt;   ENST00000433172        1 115050383-115089414      - | ENSP00000415256
#&gt;               ...      ...                 ...    ... .             ...
#&gt;   ENST00000378482        X   38561370-38688913      + | ENSP00000367743
#&gt;   ENST00000286824        X   38561544-38688397      + | ENSP00000286824
#&gt;   ENST00000612152        X 100627109-100637104      - | ENSP00000482130
#&gt;   ENST00000373020        X 100628670-100636806      - | ENSP00000362111
#&gt;   ENST00000614008        X 100632063-100637104      - | ENSP00000482894
#&gt;                             tx_id protein_domain_id
#&gt;                       &lt;character&gt;       &lt;character&gt;
#&gt;   ENST00000372003 ENST00000372003           PF00335
#&gt;   ENST00000271324 ENST00000271324           PF00335
#&gt;   ENST00000369516 ENST00000369516           PF00335
#&gt;   ENST00000369515 ENST00000369515           PF00335
#&gt;   ENST00000433172 ENST00000433172           PF00335
#&gt;               ...             ...               ...
#&gt;   ENST00000378482 ENST00000378482           PF00335
#&gt;   ENST00000286824 ENST00000286824           PF00335
#&gt;   ENST00000612152 ENST00000612152           PF00335
#&gt;   ENST00000373020 ENST00000373020           PF00335
#&gt;   ENST00000614008 ENST00000614008           PF00335
#&gt;   -------
#&gt;   seqinfo: 13 sequences from GRCh38 genome</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by Johannes Rainer.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p></p>
</div>

      </footer>
   </div>

  


  

  </body>
</html>


