<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Map genomic coordinates to transcript coordinates — genomeToTranscript • ensembldb</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Map genomic coordinates to transcript coordinates — genomeToTranscript" />
<meta property="og:description" content="genomeToTranscript maps genomic coordinates to positions within the
transcript (if at the provided genomic position a transcript is encoded).
The function does only support mapping of genomic coordinates that are
completely within the genomic region at which an exon is encoded. If the
genomic region crosses the exon boundary an empty IRanges is returned.
See examples for details." />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ensembldb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.15.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/ensembldb.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/MySQL-backend.html">Using a MySQL server backend</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping-use-cases.html">Use cases for coordinate mapping with ensembldb</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping.html">Mapping between genome, transcript and protein coordinates</a>
    </li>
    <li>
      <a href="../articles/proteins.html">Querying protein features</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jorainer/ensembldb/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Map genomic coordinates to transcript coordinates</h1>
    <small class="dont-index">Source: <a href='https://github.com/jorainer/ensembldb/blob/master/R/genomeToX.R'><code>R/genomeToX.R</code></a></small>
    <div class="hidden name"><code>genomeToTranscript.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>genomeToTranscript</code> maps genomic coordinates to positions within the
transcript (if at the provided genomic position a transcript is encoded).
The function does only support mapping of genomic coordinates that are
completely within the genomic region at which an exon is encoded. If the
genomic region crosses the exon boundary an empty <code>IRanges</code> is returned.
See examples for details.</p>
    </div>

    <pre class="usage"><span class='fu'>genomeToTranscript</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>db</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p><code>GRanges</code> object with the genomic coordinates that should be
mapped.</p></td>
    </tr>
    <tr>
      <th>db</th>
      <td><p><code>EnsDb</code> object.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>An <code>IRangesList</code> with length equal to <code><a href='https://rdrr.io/r/base/length.html'>length(x)</a></code>. Each element providing
the mapping(s) to position within any encoded transcripts at the respective
genomic location as an <code>IRanges</code> object. An <code>IRanges</code> with negative start
coordinates is returned, if the provided genomic coordinates are not
completely within the genomic coordinates of an exon.</p>
<p>The ID of the exon and its rank (index of the exon in the transcript) are
provided in the result's <code>IRanges</code> metadata columns as well as the genomic
position of <code>x</code>.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The function first retrieves all exons overlapping the provided genomic
coordinates and identifies then exons that are fully containing the
coordinates in <code>x</code>. The transcript-relative coordinates are calculated based
on the relative position of the provided genomic coordinates in this exon.</p>
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>The function throws a warning and returns an empty <code>IRanges</code> object if the
genomic coordinates can not be mapped to a transcript.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other coordinate mapping functions: 
<code><a href='cdsToTranscript.html'>cdsToTranscript</a>()</code>,
<code><a href='genomeToProtein.html'>genomeToProtein</a>()</code>,
<code><a href='proteinToGenome.html'>proteinToGenome</a>()</code>,
<code><a href='proteinToTranscript.html'>proteinToTranscript</a>()</code>,
<code><a href='transcriptToCds.html'>transcriptToCds</a>()</code>,
<code><a href='transcriptToGenome.html'>transcriptToGenome</a>()</code>,
<code><a href='transcriptToProtein.html'>transcriptToProtein</a>()</code></p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Johannes Rainer</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>EnsDb.Hsapiens.v86</span><span class='op'>)</span>

<span class='co'>## Subsetting the EnsDb object to chromosome X only to speed up execution</span>
<span class='co'>## time of examples</span>
<span class='va'>edbx</span> <span class='op'>&lt;-</span> <span class='fu'><a href='addFilter,EnsDb-method.html'>filter</a></span><span class='op'>(</span><span class='va'>EnsDb.Hsapiens.v86</span>, filter <span class='op'>=</span> <span class='op'>~</span> <span class='va'>seq_name</span> <span class='op'>==</span> <span class='st'>"X"</span><span class='op'>)</span>

<span class='co'>## Define a genomic region and calculate within-transcript coordinates</span>
<span class='va'>gnm</span> <span class='op'>&lt;-</span> <span class='fu'>GRanges</span><span class='op'>(</span><span class='st'>"X:107716399-107716401"</span><span class='op'>)</span>

<span class='va'>res</span> <span class='op'>&lt;-</span> <span class='fu'>genomeToTranscript</span><span class='op'>(</span><span class='va'>gnm</span>, <span class='va'>edbx</span><span class='op'>)</span>
<span class='co'>## Result is an IRanges object with the start and end coordinates within</span>
<span class='co'>## each transcript that has an exon at the genomic range.</span>
<span class='va'>res</span>
</div><div class='output co'>#&gt; IRangesList object of length 1:
#&gt; [[1]]
#&gt; IRanges object with 2 ranges and 7 metadata columns:
#&gt;                       start       end     width |           tx_id
#&gt;                   &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |     &lt;character&gt;
#&gt;   ENST00000372390       145       147         3 | ENST00000372390
#&gt;   ENST00000486554         1         3         3 | ENST00000486554
#&gt;                           exon_id exon_rank seq_start   seq_end    seq_name
#&gt;                       &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;
#&gt;   ENST00000372390 ENSE00001457675         1 107716399 107716401           X
#&gt;   ENST00000486554 ENSE00001927337         1 107716399 107716401           X
#&gt;                    seq_strand
#&gt;                   &lt;character&gt;
#&gt;   ENST00000372390           *
#&gt;   ENST00000486554           *
#&gt; </div><div class='input'>
<span class='co'>## An IRanges with negative coordinates is returned if at the provided</span>
<span class='co'>## position no exon is present. Below we use the same coordinates but</span>
<span class='co'>## specify that the coordinates are on the forward (+) strand</span>
<span class='va'>gnm</span> <span class='op'>&lt;-</span> <span class='fu'>GRanges</span><span class='op'>(</span><span class='st'>"X:107716399-107716401:+"</span><span class='op'>)</span>
<span class='fu'>genomeToTranscript</span><span class='op'>(</span><span class='va'>gnm</span>, <span class='va'>edbx</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: 1 genomic region(s) could not be mapped to a transcript; hint: see ?seqlevelsStyle if you used UCSC chromosome names</span></div><div class='output co'>#&gt; IRangesList object of length 1:
#&gt; [[1]]
#&gt; IRanges object with 1 range and 7 metadata columns:
#&gt;           start       end     width |       tx_id     exon_id exon_rank
#&gt;       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; | &lt;character&gt; &lt;character&gt; &lt;integer&gt;
#&gt;   [1]        -1        -1         1 |        &lt;NA&gt;        &lt;NA&gt;      &lt;NA&gt;
#&gt;       seq_start   seq_end    seq_name  seq_strand
#&gt;       &lt;integer&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;
#&gt;   [1] 107716399 107716401           X           +
#&gt; </div><div class='input'>
<span class='co'>## Next we provide multiple genomic positions.</span>
<span class='va'>gnm</span> <span class='op'>&lt;-</span> <span class='fu'>GRanges</span><span class='op'>(</span><span class='st'>"X"</span>, <span class='fu'>IRanges</span><span class='op'>(</span>start <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>644635</span>, <span class='fl'>107716399</span>, <span class='fl'>107716399</span><span class='op'>)</span>,
    end <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>644639</span>, <span class='fl'>107716401</span>, <span class='fl'>107716401</span><span class='op'>)</span><span class='op'>)</span>, strand <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"*"</span>, <span class='st'>"*"</span>, <span class='st'>"+"</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'>## The result of the mapping is an IRangesList each element providing the</span>
<span class='co'>## within-transcript coordinates for each input region</span>
<span class='fu'>genomeToTranscript</span><span class='op'>(</span><span class='va'>gnm</span>, <span class='va'>edbx</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: 1 genomic region(s) could not be mapped to a transcript; hint: see ?seqlevelsStyle if you used UCSC chromosome names</span></div><div class='output co'>#&gt; IRangesList object of length 3:
#&gt; [[1]]
#&gt; IRanges object with 2 ranges and 7 metadata columns:
#&gt;                       start       end     width |           tx_id
#&gt;                   &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |     &lt;character&gt;
#&gt;   ENST00000381578      1569      1573         5 | ENST00000381578
#&gt;   ENST00000554971       969       973         5 | ENST00000554971
#&gt;                           exon_id exon_rank seq_start   seq_end    seq_name
#&gt;                       &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;
#&gt;   ENST00000381578 ENSE00001489174         6    644635    644639           X
#&gt;   ENST00000554971 ENSE00002438186         5    644635    644639           X
#&gt;                    seq_strand
#&gt;                   &lt;character&gt;
#&gt;   ENST00000381578           *
#&gt;   ENST00000554971           *
#&gt; 
#&gt; [[2]]
#&gt; IRanges object with 2 ranges and 7 metadata columns:
#&gt;                       start       end     width |           tx_id
#&gt;                   &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |     &lt;character&gt;
#&gt;   ENST00000372390       145       147         3 | ENST00000372390
#&gt;   ENST00000486554         1         3         3 | ENST00000486554
#&gt;                           exon_id exon_rank seq_start   seq_end    seq_name
#&gt;                       &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;
#&gt;   ENST00000372390 ENSE00001457675         1 107716399 107716401           X
#&gt;   ENST00000486554 ENSE00001927337         1 107716399 107716401           X
#&gt;                    seq_strand
#&gt;                   &lt;character&gt;
#&gt;   ENST00000372390           *
#&gt;   ENST00000486554           *
#&gt; 
#&gt; [[3]]
#&gt; IRanges object with 1 range and 7 metadata columns:
#&gt;        start       end     width |       tx_id     exon_id exon_rank seq_start
#&gt;    &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; | &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;integer&gt;
#&gt;           -1        -1         1 |        &lt;NA&gt;        &lt;NA&gt;      &lt;NA&gt; 107716399
#&gt;      seq_end    seq_name  seq_strand
#&gt;    &lt;integer&gt; &lt;character&gt; &lt;character&gt;
#&gt;    107716401           X           +
#&gt; </div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Johannes Rainer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


