<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Map genomic coordinates to protein coordinates — genomeToProtein • ensembldb</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Map genomic coordinates to protein coordinates — genomeToProtein" />
<meta property="og:description" content="Map positions along the genome to positions within the protein sequence if
a protein is encoded at the location. The provided coordinates have to be
completely within the genomic position of an exon of a protein coding
transcript (see genomeToTranscript() for details). Also, the provided
positions have to be within the genomic region encoding the CDS of a
transcript (excluding its stop codon; soo transcriptToProtein() for
details).
For genomic positions for which the mapping failed an IRanges with
negative coordinates (i.e. a start position of -1) is returned." />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ensembldb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.17.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/ensembldb.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/MySQL-backend.html">Using a MySQL server backend</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping-use-cases.html">Use cases for coordinate mapping with ensembldb</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping.html">Mapping between genome, transcript and protein coordinates</a>
    </li>
    <li>
      <a href="../articles/proteins.html">Querying protein features</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jorainer/ensembldb/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Map genomic coordinates to protein coordinates</h1>
    <small class="dont-index">Source: <a href='https://github.com/jorainer/ensembldb/blob/master/R/genomeToX.R'><code>R/genomeToX.R</code></a></small>
    <div class="hidden name"><code>genomeToProtein.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Map positions along the genome to positions within the protein sequence if
a protein is encoded at the location. The provided coordinates have to be
completely within the genomic position of an exon of a protein coding
transcript (see <code><a href='genomeToTranscript.html'>genomeToTranscript()</a></code> for details). Also, the provided
positions have to be within the genomic region encoding the CDS of a
transcript (excluding its stop codon; soo <code><a href='transcriptToProtein.html'>transcriptToProtein()</a></code> for
details).</p>
<p>For genomic positions for which the mapping failed an <code>IRanges</code> with
negative coordinates (i.e. a start position of -1) is returned.</p>
    </div>

    <div class="ref-usage sourceCode"><pre class='sourceCode r'><code><span class='fu'>genomeToProtein</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>db</span><span class='op'>)</span></code></pre></div>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p><code>GRanges</code> with the genomic coordinates that should be mapped to
within-protein coordinates.</p></td>
    </tr>
    <tr>
      <th>db</th>
      <td><p><code>EnsDb</code> object.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>An <code>IRangesList</code> with each element representing the mapping of one of the
<code>GRanges</code> in <code>x</code> (i.e. the length of the <code>IRangesList</code> is <code>length(x)</code>).
Each element in <code>IRanges</code> provides the coordinates within the protein
sequence, names being the (Ensembl) IDs of the protein. The ID of the
transcript encoding the protein, the ID of the exon within which the
genomic coordinates are located and its rank in the transcript are provided
in metadata columns <code>"tx_id"</code>, <code>"exon_id"</code> and <code>"exon_rank"</code>. Metadata
columns <code>"cds_ok"</code> indicates whether the length of the CDS matches the
length of the encoded protein. Coordinates for which <code>cds_ok = FALSE</code> should
be taken with caution, as they might not be correct. Metadata columns
<code>"seq_start"</code>, <code>"seq_end"</code>, <code>"seq_name"</code> and <code>"seq_strand"</code> provide the
provided genomic coordinates.</p>
<p>For genomic coordinates that can not be mapped to within-protein sequences
an <code>IRanges</code> with a start coordinate of -1 is returned.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><code>genomeToProtein</code> combines calls to <code><a href='genomeToTranscript.html'>genomeToTranscript()</a></code> and
<code><a href='transcriptToProtein.html'>transcriptToProtein()</a></code>.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other coordinate mapping functions: 
<code><a href='cdsToTranscript.html'>cdsToTranscript</a>()</code>,
<code><a href='genomeToTranscript.html'>genomeToTranscript</a>()</code>,
<code><a href='proteinToGenome.html'>proteinToGenome</a>()</code>,
<code><a href='proteinToTranscript.html'>proteinToTranscript</a>()</code>,
<code><a href='transcriptToCds.html'>transcriptToCds</a>()</code>,
<code><a href='transcriptToGenome.html'>transcriptToGenome</a>()</code>,
<code><a href='transcriptToProtein.html'>transcriptToProtein</a>()</code></p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Johannes Rainer</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <div class="ref-examples sourceCode"><pre class='sourceCode r'><code><span class='r-in'></span>
<span class='r-in'><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>EnsDb.Hsapiens.v86</span><span class='op'>)</span></span>
<span class='r-in'><span class='co'>## Restrict all further queries to chromosome x to speed up the examples</span></span>
<span class='r-in'><span class='va'>edbx</span> <span class='op'>&lt;-</span> <span class='fu'><a href='global-filters.html'>filter</a></span><span class='op'>(</span><span class='va'>EnsDb.Hsapiens.v86</span>, filter <span class='op'>=</span> <span class='op'>~</span> <span class='va'>seq_name</span> <span class='op'>==</span> <span class='st'>"X"</span><span class='op'>)</span></span>
<span class='r-in'></span>
<span class='r-in'><span class='co'>## In the example below we define 4 genomic regions:</span></span>
<span class='r-in'><span class='co'>## 630898: corresponds to the first nt of the CDS of ENST00000381578</span></span>
<span class='r-in'><span class='co'>## 644636: last nt of the CDS of ENST00000381578</span></span>
<span class='r-in'><span class='co'>## 644633: last nt before the stop codon in ENST00000381578</span></span>
<span class='r-in'><span class='co'>## 634829: position within an intron.</span></span>
<span class='r-in'><span class='va'>gnm</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges</a></span><span class='op'>(</span><span class='st'>"X"</span>, <span class='fu'>IRanges</span><span class='op'>(</span>start <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>630898</span>, <span class='fl'>644636</span>, <span class='fl'>644633</span>, <span class='fl'>634829</span><span class='op'>)</span>,</span>
<span class='r-in'>    width <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>5</span>, <span class='fl'>1</span>, <span class='fl'>1</span>, <span class='fl'>3</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>res</span> <span class='op'>&lt;-</span> <span class='fu'>genomeToProtein</span><span class='op'>(</span><span class='va'>gnm</span>, <span class='va'>edbx</span><span class='op'>)</span></span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>1 genomic region(s) could not be mapped to a transcript; hint: see ?seqlevelsStyle if you used UCSC chromosome names</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>Transcript(s) '' could not be found</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>Provided coordinates for 'ENST00000381578', 'ENST00000554971' are not within the coding region</span>
<span class='r-in'></span>
<span class='r-in'><span class='co'>## The result is an IRangesList with the same length as gnm</span></span>
<span class='r-in'><span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='va'>res</span><span class='op'>)</span></span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> [1] 4</span>
<span class='r-in'><span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='va'>gnm</span><span class='op'>)</span></span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> [1] 4</span>
<span class='r-in'></span>
<span class='r-in'><span class='co'>## The first element represents the mapping for the first GRanges:</span></span>
<span class='r-in'><span class='co'>## the coordinate is mapped to the first amino acid of the protein(s).</span></span>
<span class='r-in'><span class='co'>## The genomic coordinates can be mapped to several transcripts (and hence</span></span>
<span class='r-in'><span class='co'>## proteins).</span></span>
<span class='r-in'><span class='va'>res</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span></span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> IRanges object with 4 ranges and 9 metadata columns:</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                       start       end     width |           tx_id    cds_ok</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                   &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |     &lt;character&gt; &lt;logical&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENSP00000335505         1         2         2 | ENST00000334060      TRUE</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENSP00000370987         1         2         2 | ENST00000381575      TRUE</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENSP00000370990         1         2         2 | ENST00000381578      TRUE</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENSP00000452016         1         2         2 | ENST00000554971      TRUE</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                             tx_id         exon_id exon_rank seq_start   seq_end</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                       &lt;character&gt;     &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENSP00000335505 ENST00000334060 ENSE00001489177         2    630898    630902</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENSP00000370987 ENST00000381575 ENSE00001489169         1    630898    630902</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENSP00000370990 ENST00000381578 ENSE00001489177         2    630898    630902</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENSP00000452016 ENST00000554971 ENSE00001489169         1    630898    630902</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                      seq_name  seq_strand</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                   &lt;character&gt; &lt;character&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENSP00000335505           X           *</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENSP00000370987           X           *</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENSP00000370990           X           *</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENSP00000452016           X           *</span>
<span class='r-in'></span>
<span class='r-in'><span class='co'>## The stop codon is not translated, thus the mapping for the second</span></span>
<span class='r-in'><span class='co'>## GRanges fails</span></span>
<span class='r-in'><span class='va'>res</span><span class='op'>[[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>]</span></span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> IRanges object with 2 ranges and 9 metadata columns:</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>        start       end     width |           tx_id    cds_ok           tx_id</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>    &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |     &lt;character&gt; &lt;logical&gt;     &lt;character&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>           -1        -1         1 | ENST00000381578      &lt;NA&gt; ENST00000381578</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>           -1        -1         1 | ENST00000554971      &lt;NA&gt; ENST00000554971</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>            exon_id exon_rank seq_start   seq_end    seq_name  seq_strand</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>        &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>    ENSE00001489174         6    644636    644636           X           *</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>    ENSE00002438186         5    644636    644636           X           *</span>
<span class='r-in'></span>
<span class='r-in'><span class='co'>## The 3rd GRanges is mapped to the last amino acid.</span></span>
<span class='r-in'><span class='va'>res</span><span class='op'>[[</span><span class='fl'>3</span><span class='op'>]</span><span class='op'>]</span></span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> IRanges object with 2 ranges and 9 metadata columns:</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                       start       end     width |           tx_id    cds_ok</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                   &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |     &lt;character&gt; &lt;logical&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENSP00000370990       292       292         1 | ENST00000381578      TRUE</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENSP00000452016       292       292         1 | ENST00000554971      TRUE</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                             tx_id         exon_id exon_rank seq_start   seq_end</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                       &lt;character&gt;     &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENSP00000370990 ENST00000381578 ENSE00001489174         6    644633    644633</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENSP00000452016 ENST00000554971 ENSE00002438186         5    644633    644633</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                      seq_name  seq_strand</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                   &lt;character&gt; &lt;character&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENSP00000370990           X           *</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENSP00000452016           X           *</span>
<span class='r-in'></span>
<span class='r-in'><span class='co'>## Mapping of intronic positions fail</span></span>
<span class='r-in'><span class='va'>res</span><span class='op'>[[</span><span class='fl'>4</span><span class='op'>]</span><span class='op'>]</span></span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> IRanges object with 1 range and 9 metadata columns:</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>        start       end     width |       tx_id    cds_ok       tx_id</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>    &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; | &lt;character&gt; &lt;logical&gt; &lt;character&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>           -1        -1         1 |                  &lt;NA&gt;        &lt;NA&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>        exon_id exon_rank seq_start   seq_end    seq_name  seq_strand</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>    &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>           &lt;NA&gt;      &lt;NA&gt;    634829    634831           X           *</span>
</code></pre></div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by Johannes Rainer.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p></p>
</div>

      </footer>
   </div>

  


  

  </body>
</html>


