<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ ensembldb</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ensembldb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.19.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/ensembldb.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/MySQL-backend.html">Using a MySQL server backend</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping-use-cases.html">Use cases for coordinate mapping with ensembldb</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping.html">Mapping between genome, transcript and protein coordinates</a>
    </li>
    <li>
      <a href="../articles/proteins.html">Querying protein features</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/jorainer/ensembldb/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="Deprecated.html">GeneidFilter()</a></code> <code><a href="Deprecated.html">GenebiotypeFilter()</a></code> <code><a href="Deprecated.html">EntrezidFilter()</a></code> <code><a href="Deprecated.html">TxidFilter()</a></code> <code><a href="Deprecated.html">TxbiotypeFilter()</a></code> <code><a href="Deprecated.html">ExonidFilter()</a></code> <code><a href="Deprecated.html">ExonrankFilter()</a></code> <code><a href="Deprecated.html">SeqnameFilter()</a></code> <code><a href="Deprecated.html">SeqstrandFilter()</a></code> <code><a href="Deprecated.html">SeqstartFilter()</a></code> <code><a href="Deprecated.html">SeqendFilter()</a></code> </p>
        </td>
        <td><p>Deprecated functionality</p></td>
      </tr><tr><td>
          <p><code><a href="EnsDb-AnnotationDbi.html">columns(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-AnnotationDbi.html">keys(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-AnnotationDbi.html">keytypes(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-AnnotationDbi.html">mapIds(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-AnnotationDbi.html">select(<i>&lt;EnsDb&gt;</i>)</a></code> </p>
        </td>
        <td><p>Integration into the AnnotationDbi framework</p></td>
      </tr><tr><td>
          <p><code><a href="EnsDb-class.html">dbconn(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-class.html">ensemblVersion(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-class.html">listColumns(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-class.html">listGenebiotypes(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-class.html">listTxbiotypes(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-class.html">listTables(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-class.html">metadata(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-class.html">organism(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-class.html">returnFilterColumns(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-class.html">returnFilterColumns(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-class.html">`returnFilterColumns&lt;-`(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-class.html">seqinfo(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-class.html">seqlevels(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-class.html">updateEnsDb(<i>&lt;EnsDb&gt;</i>)</a></code> </p>
        </td>
        <td><p>Basic usage of an Ensembl based annotation database</p></td>
      </tr><tr><td>
          <p><code><a href="EnsDb-exonsBy.html">exons(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-exonsBy.html">exonsBy(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-exonsBy.html">intronsByTranscript(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-exonsBy.html">exonsByOverlaps(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-exonsBy.html">transcripts(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-exonsBy.html">transcriptsBy(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-exonsBy.html">transcriptsByOverlaps(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-exonsBy.html">promoters(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-exonsBy.html">genes(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-exonsBy.html">disjointExons(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-exonsBy.html">cdsBy(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-exonsBy.html">fiveUTRsByTranscript(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-exonsBy.html">threeUTRsByTranscript(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-exonsBy.html">toSAF(<i>&lt;GRangesList&gt;</i>)</a></code> </p>
        </td>
        <td><p>Retrieve annotation data from an Ensembl based package</p></td>
      </tr><tr><td>
          <p><code><a href="EnsDb-lengths.html">lengthOf(<i>&lt;EnsDb&gt;</i>)</a></code> </p>
        </td>
        <td><p>Calculating lengths of features</p></td>
      </tr><tr><td>
          <p><code><a href="EnsDb-seqlevels.html">seqlevelsStyle(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-seqlevels.html">`seqlevelsStyle&lt;-`(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-seqlevels.html">supportedSeqlevelsStyles(<i>&lt;EnsDb&gt;</i>)</a></code> </p>
        </td>
        <td><p>Support for other than Ensembl seqlevel style</p></td>
      </tr><tr><td>
          <p><code><a href="EnsDb-sequences.html">getGenomeFaFile(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="EnsDb-sequences.html">getGenomeTwoBitFile(<i>&lt;EnsDb&gt;</i>)</a></code> </p>
        </td>
        <td><p>Functionality related to DNA/RNA sequences</p></td>
      </tr><tr><td>
          <p><code><a href="EnsDb-utils.html">getGeneRegionTrackForGviz(<i>&lt;EnsDb&gt;</i>)</a></code> </p>
        </td>
        <td><p>Utility functions</p></td>
      </tr><tr><td>
          <p><code><a href="EnsDb.html">EnsDb()</a></code> </p>
        </td>
        <td><p>Connect to an EnsDb object</p></td>
      </tr><tr><td>
          <p><code><a href="Filter-classes.html">OnlyCodingTxFilter()</a></code> <code><a href="Filter-classes.html">ProtDomIdFilter()</a></code> <code><a href="Filter-classes.html">ProteinDomainIdFilter()</a></code> <code><a href="Filter-classes.html">ProteinDomainSourceFilter()</a></code> <code><a href="Filter-classes.html">UniprotDbFilter()</a></code> <code><a href="Filter-classes.html">UniprotMappingTypeFilter()</a></code> <code><a href="Filter-classes.html">TxSupportLevelFilter()</a></code> <code><a href="Filter-classes.html">seqnames(<i>&lt;GRangesFilter&gt;</i>)</a></code> <code><a href="Filter-classes.html">seqlevels(<i>&lt;GRangesFilter&gt;</i>)</a></code> <code><a href="Filter-classes.html">supportedFilters(<i>&lt;EnsDb&gt;</i>)</a></code> </p>
        </td>
        <td><p>Filters supported by ensembldb</p></td>
      </tr><tr><td>
          <p><code><a href="ProteinFunctionality.html">proteins(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="ProteinFunctionality.html">listUniprotDbs(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="ProteinFunctionality.html">listUniprotMappingTypes(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="ProteinFunctionality.html">listProteinColumns()</a></code> </p>
        </td>
        <td><p>Protein related functionality</p></td>
      </tr><tr><td>
          <p><code><a href="cdsToTranscript.html">cdsToTranscript()</a></code> </p>
        </td>
        <td><p>Map positions within the CDS to coordinates relative to the start of
the transcript</p></td>
      </tr><tr><td>
          <p><code><a href="convertFilter.html">convertFilter(<i>&lt;AnnotationFilter&gt;</i>,<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="convertFilter.html">convertFilter(<i>&lt;AnnotationFilterList&gt;</i>,<i>&lt;EnsDb&gt;</i>)</a></code> </p>
        </td>
        <td><p>Convert an AnnotationFilter to a SQL WHERE condition for EnsDb</p></td>
      </tr><tr><td>
          <p><code><a href="genomeToProtein.html">genomeToProtein()</a></code> </p>
        </td>
        <td><p>Map genomic coordinates to protein coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="genomeToTranscript.html">genomeToTranscript()</a></code> </p>
        </td>
        <td><p>Map genomic coordinates to transcript coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="global-filters.html">addFilter(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="global-filters.html">dropFilter(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="global-filters.html">activeFilter(<i>&lt;EnsDb&gt;</i>)</a></code> <code><a href="global-filters.html">filter()</a></code> </p>
        </td>
        <td><p>Globally add filters to an EnsDb database</p></td>
      </tr><tr><td>
          <p><code><a href="hasProteinData-EnsDb-method.html">hasProteinData(<i>&lt;EnsDb&gt;</i>)</a></code> </p>
        </td>
        <td><p>Determine whether protein data is available in the database</p></td>
      </tr><tr><td>
          <p><code><a href="listEnsDbs.html">listEnsDbs()</a></code> </p>
        </td>
        <td><p>List EnsDb databases in a MariaDB/MySQL server</p></td>
      </tr><tr><td>
          <p><code><a href="makeEnsemblDbPackage.html">ensDbFromAH()</a></code> <code><a href="makeEnsemblDbPackage.html">ensDbFromGRanges()</a></code> <code><a href="makeEnsemblDbPackage.html">ensDbFromGff()</a></code> <code><a href="makeEnsemblDbPackage.html">ensDbFromGtf()</a></code> <code><a href="makeEnsemblDbPackage.html">fetchTablesFromEnsembl()</a></code> <code><a href="makeEnsemblDbPackage.html">makeEnsemblSQLiteFromTables()</a></code> <code><a href="makeEnsemblDbPackage.html">makeEnsembldbPackage()</a></code> </p>
        </td>
        <td><p>Generating a Ensembl annotation package from Ensembl</p></td>
      </tr><tr><td>
          <p><code><a href="proteinToGenome.html">proteinToGenome()</a></code> </p>
        </td>
        <td><p>Map within-protein coordinates to genomic coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="proteinToTranscript.html">proteinToTranscript()</a></code> </p>
        </td>
        <td><p>Map protein-relative coordinates to positions within the transcript</p></td>
      </tr><tr><td>
          <p><code><a href="runEnsDbApp.html">runEnsDbApp()</a></code> </p>
        </td>
        <td><p>Search annotations interactively</p></td>
      </tr><tr><td>
          <p><code><a href="transcriptToCds.html">transcriptToCds()</a></code> </p>
        </td>
        <td><p>Map transcript-relative coordinates to positions within the CDS</p></td>
      </tr><tr><td>
          <p><code><a href="transcriptToGenome.html">transcriptToGenome()</a></code> </p>
        </td>
        <td><p>Map transcript-relative coordinates to genomic coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="transcriptToProtein.html">transcriptToProtein()</a></code> </p>
        </td>
        <td><p>Map transcript-relative coordinates to amino acid residues of the
encoded protein</p></td>
      </tr><tr><td>
          <p><code><a href="useMySQL-EnsDb-method.html">useMySQL(<i>&lt;EnsDb&gt;</i>)</a></code> </p>
        </td>
        <td><p>Use a MariaDB/MySQL backend</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Johannes Rainer.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.9000.9000.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

