<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Generating a Ensembl annotation package from Ensembl — makeEnsembldbPackage • ensembldb</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Generating a Ensembl annotation package from Ensembl — makeEnsembldbPackage" />
<meta property="og:description" content="The functions described on this page allow to build EnsDb
  annotation objects/databases from Ensembl annotations. The most
  complete set of annotations, which include also the NCBI Entrezgene
  identifiers for each gene, can be retrieved by the functions using
  the Ensembl Perl API (i.e. functions fetchTablesFromEnsembl,
  makeEnsemblSQLiteFromTables). Alternatively the functions
  ensDbFromAH, ensDbFromGRanges, ensDbFromGff and
  ensDbFromGtf can be used to build EnsDb objects using
  GFF or GTF files from Ensembl, which can be either manually downloaded
  from the Ensembl ftp server, or directly form within R using
  AnnotationHub.
  The generated SQLite database can be packaged into an R package using
  the makeEnsembldbPackage." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ensembldb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.15.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/ensembldb.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/MySQL-backend.html">Using a MySQL server backend</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping-use-cases.html">Use cases for coordinate mapping with ensembldb</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping.html">Mapping between genome, transcript and protein coordinates</a>
    </li>
    <li>
      <a href="../articles/proteins.html">Querying protein features</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jorainer/ensembldb/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Generating a Ensembl annotation package from Ensembl</h1>
    
    <div class="hidden name"><code>makeEnsemblDbPackage.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The functions described on this page allow to build <code>EnsDb</code>
  annotation objects/databases from Ensembl annotations. The most
  complete set of annotations, which include also the NCBI Entrezgene
  identifiers for each gene, can be retrieved by the functions using
  the Ensembl Perl API (i.e. functions <code>fetchTablesFromEnsembl</code>,
  <code>makeEnsemblSQLiteFromTables</code>). Alternatively the functions
  <code>ensDbFromAH</code>, <code>ensDbFromGRanges</code>, <code>ensDbFromGff</code> and
  <code>ensDbFromGtf</code> can be used to build <code>EnsDb</code> objects using
  GFF or GTF files from Ensembl, which can be either manually downloaded
  from the Ensembl ftp server, or directly form within R using
  <code>AnnotationHub</code>.
  The generated SQLite database can be packaged into an R package using
  the <code>makeEnsembldbPackage</code>.</p>
    </div>

    <pre class="usage"><span class='fu'>ensDbFromAH</span><span class='op'>(</span><span class='va'>ah</span>, <span class='va'>outfile</span>, <span class='va'>path</span>, <span class='va'>organism</span>, <span class='va'>genomeVersion</span>, <span class='va'>version</span><span class='op'>)</span>

<span class='fu'>ensDbFromGRanges</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>outfile</span>, <span class='va'>path</span>, <span class='va'>organism</span>, <span class='va'>genomeVersion</span>,
                 <span class='va'>version</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='fu'>ensDbFromGff</span><span class='op'>(</span><span class='va'>gff</span>, <span class='va'>outfile</span>, <span class='va'>path</span>, <span class='va'>organism</span>, <span class='va'>genomeVersion</span>,
             <span class='va'>version</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='fu'>ensDbFromGtf</span><span class='op'>(</span><span class='va'>gtf</span>, <span class='va'>outfile</span>, <span class='va'>path</span>, <span class='va'>organism</span>, <span class='va'>genomeVersion</span>,
             <span class='va'>version</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='fu'>fetchTablesFromEnsembl</span><span class='op'>(</span><span class='va'>version</span>, <span class='va'>ensemblapi</span>, user<span class='op'>=</span><span class='st'>"anonymous"</span>,
                       host<span class='op'>=</span><span class='st'>"ensembldb.ensembl.org"</span>, pass<span class='op'>=</span><span class='st'>""</span>,
                       port<span class='op'>=</span><span class='fl'>5306</span>, species<span class='op'>=</span><span class='st'>"human"</span><span class='op'>)</span>

<span class='fu'>makeEnsemblSQLiteFromTables</span><span class='op'>(</span>path<span class='op'>=</span><span class='st'>"."</span>, <span class='va'>dbname</span><span class='op'>)</span>

<span class='fu'>makeEnsembldbPackage</span><span class='op'>(</span><span class='va'>ensdb</span>, <span class='va'>version</span>, <span class='va'>maintainer</span>, <span class='va'>author</span>,
                     destDir<span class='op'>=</span><span class='st'>"."</span>, license<span class='op'>=</span><span class='st'>"Artistic-2.0"</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>ah</th>
      <td><p>For <code>ensDbFromAH</code>: an <code>AnnotationHub</code> object representing
    a single resource (i.e. GTF file from Ensembl) from
    <code>AnnotationHub</code>.</p></td>
    </tr>
    <tr>
      <th>author</th>
      <td><p>The author of the package.</p></td>
    </tr>
    <tr>
      <th>dbname</th>
      <td><p>The name for the database (optional). By default a name based on the
    species and Ensembl version will be automatically generated (and
    returned by the function).</p></td>
    </tr>
    <tr>
      <th>destDir</th>
      <td><p>Where the package should be saved to.</p></td>
    </tr>
    <tr>
      <th>ensdb</th>
      <td><p>The file name of the SQLite database generated by <code>makeEnsemblSQLiteFromTables</code>.</p></td>
    </tr>
    <tr>
      <th> ensemblapi </th>
      <td><p>The path to the Ensembl perl API installed locally on the
  system. The Ensembl perl API version has to fit the version.</p></td>
    </tr>
    <tr>
      <th>genomeVersion</th>
      <td><p>For <code>ensDbFromAH</code>, <code>ensDbFromGtf</code> and <code>ensDbFromGff</code>:
    the version of the genome (e.g. <code>"GRCh37"</code>). If not provided
    the function will try to guess it from the file name (assuming file
    name convention of Ensembl GTF files).</p></td>
    </tr>
    <tr>
      <th>gff</th>
      <td><p>The GFF file to import.</p></td>
    </tr>
    <tr>
      <th>gtf</th>
      <td><p>The GTF file name.</p></td>
    </tr>
    <tr>
      <th>host</th>
      <td><p>The hostname to access the Ensembl database.</p></td>
    </tr>
    <tr>
      <th>license</th>
      <td><p>The license of the package.</p></td>
    </tr>
    <tr>
      <th>maintainer</th>
      <td><p>The maintainer of the package.</p></td>
    </tr>
    <tr>
      <th>organism</th>
      <td><p>For <code>ensDbFromAH</code>, <code>ensDbFromGff</code> and <code>ensDbFromGtf</code>:
    the organism name (e.g. <code>"Homo_sapiens"</code>). If not provided the
    function will try to guess it from the file name (assuming file name
    convention of Ensembl GTF files).</p></td>
    </tr>
    <tr>
      <th>outfile</th>
      <td><p>The desired file name of the SQLite file. If not provided the name
    of the GTF file will be used.</p></td>
    </tr>
    <tr>
      <th>pass</th>
      <td><p>The password for the Ensembl database.</p></td>
    </tr>
    <tr>
      <th>path</th>
      <td><p>The directory in which the tables retrieved by
    <code>fetchTablesFromEnsembl</code> or the SQLite database file generated
    by <code>ensDbFromGtf</code> are stored.</p></td>
    </tr>
    <tr>
      <th>port</th>
      <td><p>The port to be used to connect to the Ensembl database.</p></td>
    </tr>
    <tr>
      <th>species</th>
      <td><p>The species for which the annotations should be retrieved.</p></td>
    </tr>
    <tr>
      <th>user</th>
      <td><p>The username for the Ensembl database.</p></td>
    </tr>
    <tr>
      <th>version</th>
      <td><p>For <code>fetchTablesFromEnsembl</code>, <code>ensDbFromGRanges</code> and <code>ensDbFromGtf</code>: the
    Ensembl version for which the annotation should be retrieved
    (e.g. 75). The <code>ensDbFromGtf</code> function will try to guess the
    Ensembl version from the GTF file name if not provided.</p>
<p>For <code>makeEnsemblDbPackage</code>: the version for the package.</p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>For <code>ensDbFromGRanges</code>: the <code>GRanges</code> object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Currently not used.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="functions"><a class="anchor" href="#functions"></a>Functions</h2>

    <p></p><dl>
    <dt>ensDbFromAH</dt><dd><p>Create an <code>EnsDb</code> (SQLite) database from a GTF file provided
      by <code>AnnotationHub</code>. The function returns the file name of the
      generated database file. For usage see the examples below.</p></dd>

    <dt>ensDbFromGff</dt><dd><p>Create an <code>EnsDb</code> (SQLite) database from a GFF file from
      Ensembl. The function returns the file name of the
      generated database file. For usage see the examples below.</p></dd>

    <dt>ensDbFromGtf</dt><dd><p>Create an <code>EnsDb</code> (SQLite) database from a GTF file from
      Ensembl. The function returns the file name of the generated
      database file. For usage see the examplesbelow.</p></dd>

    <dt>ensDbFromGRanges</dt><dd><p>Create an <code>EnsDb</code> (SQLite) database from a GRanges object
      (e.g. from <code>AnnotationHub</code>). The function returns the file
      name of the generated database file. For usage see the examples
      below.</p></dd>

    <dt>fetchTablesFromEnsembl</dt><dd><p>Uses the Ensembl Perl API to fetch all required data from an
      Ensembl database server and stores them locally to text files
      (that can be used as input for the
      <code>makeEnsembldbSQLiteFromTables</code> function).</p></dd>

    <dt>makeEnsemblSQLiteFromTables</dt><dd><p>Creates the SQLite <code>EnsDb</code> database from the tables generated
      by the <code>fetchTablesFromEnsembl</code>.</p></dd>

    <dt>makeEnsembldbPackage</dt><dd><p>Creates an R package containing the <code>EnsDb</code> database from a
      <code>EnsDb</code> SQLite database created by any of the above
      functions <code>ensDbFromAH</code>, <code>ensDbFromGff</code>,
      <code>ensDbFromGtf</code> or <code>makeEnsemblSQLiteFromTables</code>.</p></dd>
  
</dl>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The <code>fetchTablesFromEnsembl</code> function internally calls the perl
  script <code>get_gene_transcript_exon_tables.pl</code> to retrieve all
  required information from the Ensembl database using the Ensembl perl
  API.</p>
<p>As an alternative way, a EnsDb database file can be generated by the
  <code>ensDbFromGtf</code> or <code>ensDbFromGff</code> from a GTF or GFF file
  downloaded from the Ensembl ftp server or using the <code>ensDbFromAH</code>
  to build a database directly from corresponding resources from the
  AnnotationHub. The returned database file name can then
  be used as an input to the <code>makeEnsembldbPackage</code> or it can be
  directly loaded and used by the <code>EnsDb</code> constructor.</p>
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>A local installation of the Ensembl perl API is required for the
  <code>fetchTablesFromEnsembl</code>. See
  <a href='http://www.ensembl.org/info/docs/api/api_installation.html'>http://www.ensembl.org/info/docs/api/api_installation.html</a> for
  installation inscructions.</p>
<p>A database generated from a GTF/GFF files lacks some features as they are
  not available in the GTF files from Ensembl. These are: NCBI
  Entrezgene IDs.</p>
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>makeEnsemblSQLiteFromTables</code>, <code>ensDbFromAH</code>,
  <code>ensDbFromGRanges</code> and <code>ensDbFromGtf</code>: the name of the
  SQLite file.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='EnsDb.html'>EnsDb</a></code>, <code><a href='exonsBy.html'>genes</a></code></p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Johannes Rainer</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>

    <span class='co'>## get all human gene/transcript/exon annotations from Ensembl (75)</span>
    <span class='co'>## the resulting tables will be stored by default to the current working</span>
    <span class='co'>## directory; if the correct Ensembl api (version 75) is defined in the</span>
    <span class='co'>## PERL5LIB environment variable, the ensemblapi parameter can also be omitted.</span>
    <span class='fu'>fetchTablesFromEnsembl</span><span class='op'>(</span><span class='fl'>75</span>,
                           ensemblapi<span class='op'>=</span><span class='st'>"/home/bioinfo/ensembl/75/API/ensembl/modules"</span>,
                           species<span class='op'>=</span><span class='st'>"human"</span><span class='op'>)</span>

    <span class='co'>## These tables can then be processed to generate a SQLite database</span>
    <span class='co'>## containing the annotations</span>
    <span class='va'>DBFile</span> <span class='op'>&lt;-</span> <span class='fu'>makeEnsemblSQLiteFromTables</span><span class='op'>(</span><span class='op'>)</span>

    <span class='co'>## and finally we can generate the package</span>
    <span class='fu'>makeEnsembldbPackage</span><span class='op'>(</span>ensdb<span class='op'>=</span><span class='va'>DBFile</span>, version<span class='op'>=</span><span class='st'>"0.0.1"</span>,
                         maintainer<span class='op'>=</span><span class='st'>"Johannes Rainer &lt;johannes.rainer@eurac.edu&gt;"</span>,
                         author<span class='op'>=</span><span class='st'>"J Rainer"</span><span class='op'>)</span>

    <span class='co'>## Build an annotation database form a GFF file from Ensembl.</span>
    <span class='co'>## ftp://ftp.ensembl.org/pub/release-83/gff3/rattus_norvegicus</span>
    <span class='va'>gff</span> <span class='op'>&lt;-</span> <span class='st'>"Rattus_norvegicus.Rnor_6.0.83.gff3.gz"</span>
    <span class='va'>DB</span> <span class='op'>&lt;-</span> <span class='fu'>ensDbFromGff</span><span class='op'>(</span>gff<span class='op'>=</span><span class='va'>gff</span><span class='op'>)</span>
    <span class='va'>edb</span> <span class='op'>&lt;-</span> <span class='fu'><a href='EnsDb.html'>EnsDb</a></span><span class='op'>(</span><span class='va'>DB</span><span class='op'>)</span>
    <span class='va'>edb</span>

    <span class='co'>## Build an annotation file from a GTF file.</span>
    <span class='co'>## the GTF file can be downloaded from</span>
    <span class='co'>## ftp://ftp.ensembl.org/pub/release-75/gtf/homo_sapiens/</span>
    <span class='va'>gtffile</span> <span class='op'>&lt;-</span> <span class='st'>"Homo_sapiens.GRCh37.75.gtf.gz"</span>
    <span class='co'>## generate the SQLite database file</span>
    <span class='va'>DB</span> <span class='op'>&lt;-</span> <span class='fu'>ensDbFromGtf</span><span class='op'>(</span>gtf<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='va'>ensemblhost</span>, <span class='va'>gtffile</span><span class='op'>)</span><span class='op'>)</span>

    <span class='co'>## load the DB file directly</span>
    <span class='va'>EDB</span> <span class='op'>&lt;-</span> <span class='fu'><a href='EnsDb.html'>EnsDb</a></span><span class='op'>(</span><span class='va'>DB</span><span class='op'>)</span>

    <span class='co'>## Alternatively, we could fetch a GTF file directly from AnnotationHub</span>
    <span class='co'>## and build the database from that:</span>
    <span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>AnnotationHub</span><span class='op'>)</span>
    <span class='va'>ah</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html'>AnnotationHub</a></span><span class='op'>(</span><span class='op'>)</span>
    <span class='co'>## Query for all GTF files from Ensembl for Ensembl version 81</span>
    <span class='fu'><a href='https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html'>query</a></span><span class='op'>(</span><span class='va'>ah</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Ensembl"</span>, <span class='st'>"release-81"</span>, <span class='st'>"GTF"</span><span class='op'>)</span><span class='op'>)</span>
    <span class='co'>## We could get the one from e.g. Bos taurus:</span>
    <span class='va'>DB</span> <span class='op'>&lt;-</span> <span class='fu'>ensDbFromAH</span><span class='op'>(</span><span class='va'>ah</span><span class='op'>[</span><span class='st'>"AH47941"</span><span class='op'>]</span><span class='op'>)</span>
    <span class='va'>edb</span> <span class='op'>&lt;-</span> <span class='fu'><a href='EnsDb.html'>EnsDb</a></span><span class='op'>(</span><span class='va'>DB</span><span class='op'>)</span>
    <span class='va'>edb</span>
<span class='op'>}</span>

<span class='co'>## Generate a sqlite database for genes encoded on chromosome Y</span>
<span class='va'>chrY</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"chrY"</span>, package<span class='op'>=</span><span class='st'>"ensembldb"</span><span class='op'>)</span>
<span class='va'>DBFile</span> <span class='op'>&lt;-</span> <span class='fu'>makeEnsemblSQLiteFromTables</span><span class='op'>(</span>path<span class='op'>=</span><span class='va'>chrY</span> ,dbname<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempfile</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Processing 'chromosome' table ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Processing 'gene' table ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Processing 'trancript' table ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Processing 'exon' table ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Processing 'tx2exon' table ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Creating indices ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Checking validity of the database ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='input'><span class='co'>## load this database:</span>
<span class='va'>edb</span> <span class='op'>&lt;-</span> <span class='fu'><a href='EnsDb.html'>EnsDb</a></span><span class='op'>(</span><span class='va'>DBFile</span><span class='op'>)</span>

<span class='va'>edb</span>
</div><div class='output co'>#&gt; EnsDb for Ensembl:
#&gt; |Backend: SQLite
#&gt; |Db type: EnsDb
#&gt; |Type of Gene ID: Ensembl Gene ID
#&gt; |Supporting package: ensembldb
#&gt; |Db created by: ensembldb package from Bioconductor
#&gt; |script_version: 0.1.2
#&gt; |Creation time: Wed Mar 18 09:30:54 2015
#&gt; |ensembl_version: 75
#&gt; |ensembl_host: manny.i-med.ac.at
#&gt; |Organism: homo_sapiens
#&gt; |genome_build: GRCh37
#&gt; |DBSCHEMAVERSION: 1.0
#&gt; | No. of genes: 495.
#&gt; | No. of transcripts: 731.</div><div class='input'>
<span class='co'>## Generate a sqlite database from a GRanges object specifying</span>
<span class='co'>## genes encoded on chromosome Y</span>
<span class='fu'><a href='https://rdrr.io/r/base/load.html'>load</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"YGRanges.RData"</span>, package<span class='op'>=</span><span class='st'>"ensembldb"</span><span class='op'>)</span><span class='op'>)</span>

<span class='va'>Y</span>
</div><div class='output co'>#&gt; GRanges object with 7155 ranges and 16 metadata columns:
#&gt;          seqnames            ranges strand |               source       type
#&gt;             &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |             &lt;factor&gt;   &lt;factor&gt;
#&gt;      [1]        Y   2652790-2652894      + |       snRNA          gene      
#&gt;      [2]        Y   2652790-2652894      + |       snRNA          transcript
#&gt;      [3]        Y   2652790-2652894      + |       snRNA          exon      
#&gt;      [4]        Y   2654896-2655740      - |       protein_coding gene      
#&gt;      [5]        Y   2654896-2655740      - |       protein_coding transcript
#&gt;      ...      ...               ...    ... .                  ...        ...
#&gt;   [7151]        Y 28772667-28773306      - | processed_pseudogene transcript
#&gt;   [7152]        Y 28772667-28773306      - | processed_pseudogene exon      
#&gt;   [7153]        Y 59001391-59001635      + | pseudogene           gene      
#&gt;   [7154]        Y 59001391-59001635      + | processed_pseudogene transcript
#&gt;   [7155]        Y 59001391-59001635      + | processed_pseudogene exon      
#&gt;              score     phase         gene_id   gene_name    gene_source
#&gt;          &lt;numeric&gt; &lt;integer&gt;     &lt;character&gt; &lt;character&gt;    &lt;character&gt;
#&gt;      [1]        NA      &lt;NA&gt; ENSG00000251841  RNU6-1334P        ensembl
#&gt;      [2]        NA      &lt;NA&gt; ENSG00000251841  RNU6-1334P        ensembl
#&gt;      [3]        NA      &lt;NA&gt; ENSG00000251841  RNU6-1334P        ensembl
#&gt;      [4]        NA      &lt;NA&gt; ENSG00000184895         SRY ensembl_havana
#&gt;      [5]        NA      &lt;NA&gt; ENSG00000184895         SRY ensembl_havana
#&gt;      ...       ...       ...             ...         ...            ...
#&gt;   [7151]        NA      &lt;NA&gt; ENSG00000231514     FAM58CP         havana
#&gt;   [7152]        NA      &lt;NA&gt; ENSG00000231514     FAM58CP         havana
#&gt;   [7153]        NA      &lt;NA&gt; ENSG00000235857     CTBP2P1         havana
#&gt;   [7154]        NA      &lt;NA&gt; ENSG00000235857     CTBP2P1         havana
#&gt;   [7155]        NA      &lt;NA&gt; ENSG00000235857     CTBP2P1         havana
#&gt;            gene_biotype   transcript_id transcript_name transcript_source
#&gt;             &lt;character&gt;     &lt;character&gt;     &lt;character&gt;       &lt;character&gt;
#&gt;      [1]          snRNA            &lt;NA&gt;            &lt;NA&gt;              &lt;NA&gt;
#&gt;      [2]          snRNA ENST00000516032  RNU6-1334P-201           ensembl
#&gt;      [3]          snRNA ENST00000516032  RNU6-1334P-201           ensembl
#&gt;      [4] protein_coding            &lt;NA&gt;            &lt;NA&gt;              &lt;NA&gt;
#&gt;      [5] protein_coding ENST00000383070         SRY-001    ensembl_havana
#&gt;      ...            ...             ...             ...               ...
#&gt;   [7151]     pseudogene ENST00000435741     FAM58CP-001            havana
#&gt;   [7152]     pseudogene ENST00000435741     FAM58CP-001            havana
#&gt;   [7153]     pseudogene            &lt;NA&gt;            &lt;NA&gt;              &lt;NA&gt;
#&gt;   [7154]     pseudogene ENST00000431853     CTBP2P1-001            havana
#&gt;   [7155]     pseudogene ENST00000431853     CTBP2P1-001            havana
#&gt;          exon_number         exon_id         tag     ccds_id  protein_id
#&gt;            &lt;numeric&gt;     &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;
#&gt;      [1]          NA            &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
#&gt;      [2]          NA            &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
#&gt;      [3]           1 ENSE00002088309        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
#&gt;      [4]          NA            &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
#&gt;      [5]          NA            &lt;NA&gt;        CCDS   CCDS14772        &lt;NA&gt;
#&gt;      ...         ...             ...         ...         ...         ...
#&gt;   [7151]          NA            &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
#&gt;   [7152]           1 ENSE00001616687        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
#&gt;   [7153]          NA            &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
#&gt;   [7154]          NA            &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
#&gt;   [7155]           1 ENSE00001794473        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
#&gt;   -------
#&gt;   seqinfo: 1 sequence from GRCh37 genome</div><div class='input'>
<span class='va'>DB</span> <span class='op'>&lt;-</span> <span class='fu'>ensDbFromGRanges</span><span class='op'>(</span><span class='va'>Y</span>, path<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempdir</a></span><span class='op'>(</span><span class='op'>)</span>, version<span class='op'>=</span><span class='fl'>75</span>,
                       organism<span class='op'>=</span><span class='st'>"Homo_sapiens"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Processing metadata ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Processing genes ... </span></div><div class='output co'>#&gt; <span class='warning'>Warning:  I'm missing column(s): 'entrezid'. The corresponding database column(s) will be empty!</span></div><div class='output co'>#&gt; <span class='message'> Attribute availability:</span></div><div class='output co'>#&gt; <span class='message'>  o gene_id ... OK</span></div><div class='output co'>#&gt; <span class='message'>  o gene_name ... OK</span></div><div class='output co'>#&gt; <span class='message'>  o entrezid ... Nope</span></div><div class='output co'>#&gt; <span class='message'>  o gene_biotype ... OK</span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Processing transcripts ... </span></div><div class='output co'>#&gt; <span class='message'> Attribute availability:</span></div><div class='output co'>#&gt; <span class='message'>  o transcript_id ... OK</span></div><div class='output co'>#&gt; <span class='message'>  o gene_id ... OK</span></div><div class='output co'>#&gt; <span class='message'>  o source ... OK</span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Processing exons ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Processing chromosomes ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Generating index ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>  -------------</span></div><div class='output co'>#&gt; <span class='message'>Verifying validity of the information in the database:</span></div><div class='output co'>#&gt; <span class='message'>Checking transcripts ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span>
#&gt; <span class='message'>Checking exons ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='input'><span class='va'>edb</span> <span class='op'>&lt;-</span> <span class='fu'><a href='EnsDb.html'>EnsDb</a></span><span class='op'>(</span><span class='va'>DB</span><span class='op'>)</span>


</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Johannes Rainer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


