<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Map within-protein coordinates to genomic coordinates — proteinToGenome • ensembldb</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Map within-protein coordinates to genomic coordinates — proteinToGenome" />
<meta property="og:description" content="proteinToGenome maps protein-relative coordinates to genomic coordinates
based on the genomic coordinates of the CDS of the encoding transcript. The
encoding transcript is identified using protein-to-transcript annotations
(and eventually Uniprot to Ensembl protein identifier mappings) from the
submitted EnsDb object (and thus based on annotations from Ensembl).
Not all coding regions for protein coding transcripts are complete, and the
function thus checks also if the length of the coding region matches the
length of the protein sequence and throws a warning if that is not the case.
The genomic coordinates for the within-protein coordinates, the Ensembl
protein ID, the ID of the encoding transcript and the within protein start
and end coordinates are reported for each input range." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ensembldb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.15.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/ensembldb.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/MySQL-backend.html">Using a MySQL server backend</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping-use-cases.html">Use cases for coordinate mapping with ensembldb</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping.html">Mapping between genome, transcript and protein coordinates</a>
    </li>
    <li>
      <a href="../articles/proteins.html">Querying protein features</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jorainer/ensembldb/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Map within-protein coordinates to genomic coordinates</h1>
    <small class="dont-index">Source: <a href='https://github.com/jorainer/ensembldb/blob/master/R/proteinToX.R'><code>R/proteinToX.R</code></a></small>
    <div class="hidden name"><code>proteinToGenome.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>proteinToGenome</code> maps protein-relative coordinates to genomic coordinates
based on the genomic coordinates of the CDS of the encoding transcript. The
encoding transcript is identified using protein-to-transcript annotations
(and eventually Uniprot to Ensembl protein identifier mappings) from the
submitted <code>EnsDb</code> object (and thus based on annotations from Ensembl).</p>
<p>Not all coding regions for protein coding transcripts are complete, and the
function thus checks also if the length of the coding region matches the
length of the protein sequence and throws a warning if that is not the case.</p>
<p>The genomic coordinates for the within-protein coordinates, the Ensembl
protein ID, the ID of the encoding transcript and the within protein start
and end coordinates are reported for each input range.</p>
    </div>

    <pre class="usage"><span class='fu'>proteinToGenome</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>db</span>, id <span class='op'>=</span> <span class='st'>"name"</span>, idType <span class='op'>=</span> <span class='st'>"protein_id"</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p><code>IRanges</code> with the coordinates within the protein(s). The
object has also to provide some means to identify the protein (see
details).</p></td>
    </tr>
    <tr>
      <th>db</th>
      <td><p><code>EnsDb</code> object to be used to retrieve genomic coordinates of
encoding transcripts.</p></td>
    </tr>
    <tr>
      <th>id</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code> specifying where the protein identifier can be
found. Has to be either <code>"name"</code> or one of <code><a href='https://rdrr.io/r/base/colnames.html'>colnames(mcols(prng))</a></code>.</p></td>
    </tr>
    <tr>
      <th>idType</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code> defining what type of IDs are provided. Has to
be one of <code>"protein_id"</code> (default), <code>"uniprot_id"</code> or <code>"tx_id"</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>list</code>, each element being the mapping results for one of the input
ranges in <code>x</code> and names being the IDs used for the mapping. Each
element can be either a:</p><ul>
<li><p><code>GRanges</code> object with the genomic coordinates calculated on the
protein-relative coordinates for the respective Ensembl protein (stored in
the <code>"protein_id"</code> metadata column.</p></li>
<li><p><code>GRangesList</code> object, if the provided protein identifier in <code>x</code> was
mapped to several Ensembl protein IDs (e.g. if Uniprot identifiers were
used). Each element in this <code>GRangesList</code> is a <code>GRanges</code> with the genomic
coordinates calculated for the protein-relative coordinates from the
respective Ensembl protein ID.</p></li>
</ul>

<p>The following metadata columns are available in each <code>GRanges</code> in the result:</p><ul>
<li><p><code>"protein_id"</code>: the ID of the Ensembl protein for which the within-protein
coordinates were mapped to the genome.</p></li>
<li><p><code>"tx_id"</code>: the Ensembl transcript ID of the encoding transcript.</p></li>
<li><p><code>"exon_id"</code>: ID of the exons that have overlapping genomic coordinates.</p></li>
<li><p><code>"exon_rank"</code>: the rank/index of the exon within the encoding transcript.</p></li>
<li><p><code>"cds_ok"</code>: contains <code>TRUE</code> if the length of the CDS matches the length
of the amino acid sequence and <code>FALSE</code> otherwise.</p></li>
<li><p><code>"protein_start"</code>: the within-protein sequence start coordinate of the
mapping.</p></li>
<li><p><code>"protein_end"</code>: the within-protein sequence end coordinate of the mapping.</p></li>
</ul>

<p>Genomic coordinates are returned ordered by the exon index within the
transcript.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Protein identifiers (supported are Ensembl protein IDs or Uniprot IDs) can
be passed to the function as <code>names</code> of the <code>x</code> <code>IRanges</code> object, or
alternatively in any one of the metadata columns (<code>mcols</code>) of <code>x</code>.</p>
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>While the mapping for Ensembl protein IDs to encoding transcripts (and
thus CDS) is 1:1, the mapping between Uniprot identifiers and encoding
transcripts (which is based on Ensembl annotations) can be one to many. In
such cases <code>proteinToGenome</code> calculates genomic coordinates for
within-protein coordinates for all of the annotated Ensembl proteins and
returns all of them. See below for examples.</p>
<p>Mapping using Uniprot identifiers needs also additional internal checks that
have a significant impact on the performance of the function. It is thus
strongly suggested to first identify the Ensembl protein identifiers for the
list of input Uniprot identifiers (e.g. using the <code><a href='proteins,EnsDb-method.html'>proteins()</a></code> function and
use these as input for the mapping function.</p>
<p>A warning is thrown for proteins which sequence does not match the coding
sequence length of any encoding transcripts. For such proteins/transcripts
a <code>FALSE</code> is reported in the respective <code>"cds_ok"</code> metadata column.
The most common reason for such discrepancies are incomplete 3' or 5' ends
of the CDS. The positions within the protein might not be correclty
mapped to the genome in such cases and it might be required to check
the mapping manually in the Ensembl genome browser.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other coordinate mapping functions: 
<code><a href='cdsToTranscript.html'>cdsToTranscript</a>()</code>,
<code><a href='genomeToProtein.html'>genomeToProtein</a>()</code>,
<code><a href='genomeToTranscript.html'>genomeToTranscript</a>()</code>,
<code><a href='proteinToTranscript.html'>proteinToTranscript</a>()</code>,
<code><a href='transcriptToCds.html'>transcriptToCds</a>()</code>,
<code><a href='transcriptToGenome.html'>transcriptToGenome</a>()</code>,
<code><a href='transcriptToProtein.html'>transcriptToProtein</a>()</code></p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Johannes Rainer based on initial code from Laurent Gatto and
Sebastian Gibb</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>EnsDb.Hsapiens.v86</span><span class='op'>)</span>
<span class='co'>## Restrict all further queries to chromosome x to speed up the examples</span>
<span class='va'>edbx</span> <span class='op'>&lt;-</span> <span class='fu'><a href='addFilter,EnsDb-method.html'>filter</a></span><span class='op'>(</span><span class='va'>EnsDb.Hsapiens.v86</span>, filter <span class='op'>=</span> <span class='op'>~</span> <span class='va'>seq_name</span> <span class='op'>==</span> <span class='st'>"X"</span><span class='op'>)</span>

<span class='co'>## Define an IRange with protein-relative coordinates within a protein for</span>
<span class='co'>## the gene SYP</span>
<span class='va'>syp</span> <span class='op'>&lt;-</span> <span class='fu'>IRanges</span><span class='op'>(</span>start <span class='op'>=</span> <span class='fl'>4</span>, end <span class='op'>=</span> <span class='fl'>17</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>syp</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='st'>"ENSP00000418169"</span>
<span class='va'>res</span> <span class='op'>&lt;-</span> <span class='fu'>proteinToGenome</span><span class='op'>(</span><span class='va'>syp</span>, <span class='va'>edbx</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Fetching CDS for 1 proteins ... </span></div><div class='output co'>#&gt; <span class='message'>1 found</span></div><div class='output co'>#&gt; <span class='message'>Checking CDS and protein sequence lengths ... </span></div><div class='output co'>#&gt; <span class='message'>1/1 OK</span></div><div class='input'><span class='va'>res</span>
</div><div class='output co'>#&gt; $ENSP00000418169
#&gt; GRanges object with 2 ranges and 7 metadata columns:
#&gt;       seqnames            ranges strand |      protein_id           tx_id
#&gt;          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;     &lt;character&gt;
#&gt;   [1]        X 49200151-49200177      - | ENSP00000418169 ENST00000479808
#&gt;   [2]        X 49199019-49199033      - | ENSP00000418169 ENST00000479808
#&gt;               exon_id exon_rank    cds_ok protein_start protein_end
#&gt;           &lt;character&gt; &lt;integer&gt; &lt;logical&gt;     &lt;integer&gt;   &lt;integer&gt;
#&gt;   [1] ENSE00001902363         1      TRUE             4          17
#&gt;   [2] ENSE00003520347         2      TRUE             4          17
#&gt;   -------
#&gt;   seqinfo: 1 sequence from GRCh38 genome
#&gt; </div><div class='input'><span class='co'>## Positions 4 to 17 within the protein span two exons of the encoding</span>
<span class='co'>## transcript.</span>

<span class='co'>## Perform the mapping for multiple proteins identified by their Uniprot</span>
<span class='co'>## IDs.</span>
<span class='va'>ids</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"O15266"</span>, <span class='st'>"Q9HBJ8"</span>, <span class='st'>"unexistant"</span><span class='op'>)</span>
<span class='va'>prngs</span> <span class='op'>&lt;-</span> <span class='fu'>IRanges</span><span class='op'>(</span>start <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>13</span>, <span class='fl'>43</span>, <span class='fl'>100</span><span class='op'>)</span>, end <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>21</span>, <span class='fl'>80</span>, <span class='fl'>100</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>prngs</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='va'>ids</span>

<span class='va'>res</span> <span class='op'>&lt;-</span> <span class='fu'>proteinToGenome</span><span class='op'>(</span><span class='va'>prngs</span>, <span class='va'>edbx</span>, idType <span class='op'>=</span> <span class='st'>"uniprot_id"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Fetching CDS for 3 proteins ... </span></div><div class='output co'>#&gt; <span class='warning'>Warning: No CDS found for: unexistant</span></div><div class='output co'>#&gt; <span class='message'>2 found</span></div><div class='output co'>#&gt; <span class='message'>Checking CDS and protein sequence lengths ... </span></div><div class='output co'>#&gt; <span class='message'>2/2 OK</span></div><div class='input'>
<span class='co'>## The result is a list, same length as the input object</span>
<span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='va'>res</span><span class='op'>)</span>
</div><div class='output co'>#&gt; [1] 3</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>res</span><span class='op'>)</span>
</div><div class='output co'>#&gt; [1] "O15266"     "Q9HBJ8"     "unexistant"</div><div class='input'>
<span class='co'>## No protein/encoding transcript could be found for the last one</span>
<span class='va'>res</span><span class='op'>[[</span><span class='fl'>3</span><span class='op'>]</span><span class='op'>]</span>
</div><div class='output co'>#&gt; GRanges object with 0 ranges and 0 metadata columns:
#&gt;    seqnames    ranges strand
#&gt;       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
#&gt;   -------
#&gt;   seqinfo: no sequences</div><div class='input'>
<span class='co'>## The first protein could be mapped to multiple Ensembl proteins. The</span>
<span class='co'>## mapping result using all of their encoding transcripts are returned</span>
<span class='va'>res</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>
</div><div class='output co'>#&gt; GRangesList object of length 4:
#&gt; $ENSP00000335505
#&gt; GRanges object with 1 range and 8 metadata columns:
#&gt;       seqnames        ranges strand |  uniprot_id           tx_id
#&gt;          &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;     &lt;character&gt;
#&gt;   [1]        X 630934-630960      + |      O15266 ENST00000334060
#&gt;            protein_id         exon_id exon_rank    cds_ok protein_start
#&gt;           &lt;character&gt;     &lt;character&gt; &lt;integer&gt; &lt;logical&gt;     &lt;integer&gt;
#&gt;   [1] ENSP00000335505 ENSE00001489177         2      TRUE            13
#&gt;       protein_end
#&gt;         &lt;integer&gt;
#&gt;   [1]          21
#&gt;   -------
#&gt;   seqinfo: 1 sequence from GRCh38 genome
#&gt; 
#&gt; $ENSP00000370987
#&gt; GRanges object with 1 range and 8 metadata columns:
#&gt;       seqnames        ranges strand |  uniprot_id           tx_id
#&gt;          &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;     &lt;character&gt;
#&gt;   [1]        X 630934-630960      + |      O15266 ENST00000381575
#&gt;            protein_id         exon_id exon_rank    cds_ok protein_start
#&gt;           &lt;character&gt;     &lt;character&gt; &lt;integer&gt; &lt;logical&gt;     &lt;integer&gt;
#&gt;   [1] ENSP00000370987 ENSE00001489169         1      TRUE            13
#&gt;       protein_end
#&gt;         &lt;integer&gt;
#&gt;   [1]          21
#&gt;   -------
#&gt;   seqinfo: 1 sequence from GRCh38 genome
#&gt; 
#&gt; $ENSP00000370990
#&gt; GRanges object with 1 range and 8 metadata columns:
#&gt;       seqnames        ranges strand |  uniprot_id           tx_id
#&gt;          &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;     &lt;character&gt;
#&gt;   [1]        X 630934-630960      + |      O15266 ENST00000381578
#&gt;            protein_id         exon_id exon_rank    cds_ok protein_start
#&gt;           &lt;character&gt;     &lt;character&gt; &lt;integer&gt; &lt;logical&gt;     &lt;integer&gt;
#&gt;   [1] ENSP00000370990 ENSE00001489177         2      TRUE            13
#&gt;       protein_end
#&gt;         &lt;integer&gt;
#&gt;   [1]          21
#&gt;   -------
#&gt;   seqinfo: 1 sequence from GRCh38 genome
#&gt; 
#&gt; $ENSP00000452016
#&gt; GRanges object with 1 range and 8 metadata columns:
#&gt;       seqnames        ranges strand |  uniprot_id           tx_id
#&gt;          &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;     &lt;character&gt;
#&gt;   [1]        X 630934-630960      + |      O15266 ENST00000554971
#&gt;            protein_id         exon_id exon_rank    cds_ok protein_start
#&gt;           &lt;character&gt;     &lt;character&gt; &lt;integer&gt; &lt;logical&gt;     &lt;integer&gt;
#&gt;   [1] ENSP00000452016 ENSE00001489169         1      TRUE            13
#&gt;       protein_end
#&gt;         &lt;integer&gt;
#&gt;   [1]          21
#&gt;   -------
#&gt;   seqinfo: 1 sequence from GRCh38 genome
#&gt; </div><div class='input'>
<span class='co'>## The coordinates within the second protein span two exons</span>
<span class='va'>res</span><span class='op'>[[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>]</span>
</div><div class='output co'>#&gt; GRanges object with 2 ranges and 8 metadata columns:
#&gt;       seqnames            ranges strand |  uniprot_id           tx_id
#&gt;          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;     &lt;character&gt;
#&gt;   [1]        X 15659016-15659092      - |      Q9HBJ8 ENST00000380342
#&gt;   [2]        X 15644993-15645029      - |      Q9HBJ8 ENST00000380342
#&gt;            protein_id         exon_id exon_rank    cds_ok protein_start
#&gt;           &lt;character&gt;     &lt;character&gt; &lt;integer&gt; &lt;logical&gt;     &lt;integer&gt;
#&gt;   [1] ENSP00000369699 ENSE00001202097         3      TRUE            43
#&gt;   [2] ENSP00000369699 ENSE00000978331         4      TRUE            43
#&gt;       protein_end
#&gt;         &lt;integer&gt;
#&gt;   [1]          80
#&gt;   [2]          80
#&gt;   -------
#&gt;   seqinfo: 1 sequence from GRCh38 genome</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Johannes Rainer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


