<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Map protein-relative coordinates to positions within the transcript — proteinToTranscript • ensembldb</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Map protein-relative coordinates to positions within the transcript — proteinToTranscript" />
<meta property="og:description" content="proteinToTranscript maps protein-relative coordinates to positions within
the encoding transcript. Note that the returned positions are relative to
the complete transcript length, which includes the 5' UTR.
Similar to the proteinToGenome() function, proteinToTranscript compares
for each protein whether the length of its sequence matches the length of
the encoding CDS and throws a warning if that is not the case. Incomplete
3' or 5' CDS of the encoding transcript are the most common reasons for a
mismatch between protein and transcript sequences." />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ensembldb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.17.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/ensembldb.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/MySQL-backend.html">Using a MySQL server backend</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping-use-cases.html">Use cases for coordinate mapping with ensembldb</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping.html">Mapping between genome, transcript and protein coordinates</a>
    </li>
    <li>
      <a href="../articles/proteins.html">Querying protein features</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jorainer/ensembldb/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Map protein-relative coordinates to positions within the transcript</h1>
    <small class="dont-index">Source: <a href='https://github.com/jorainer/ensembldb/blob/master/R/proteinToX.R'><code>R/proteinToX.R</code></a></small>
    <div class="hidden name"><code>proteinToTranscript.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>proteinToTranscript</code> maps protein-relative coordinates to positions within
the encoding transcript. Note that the returned positions are relative to
the complete transcript length, which includes the 5' UTR.</p>
<p>Similar to the <code><a href='proteinToGenome.html'>proteinToGenome()</a></code> function, <code>proteinToTranscript</code> compares
for each protein whether the length of its sequence matches the length of
the encoding CDS and throws a warning if that is not the case. Incomplete
3' or 5' CDS of the encoding transcript are the most common reasons for a
mismatch between protein and transcript sequences.</p>
    </div>

    <div class="ref-usage sourceCode"><pre class='sourceCode r'><code><span class='fu'>proteinToTranscript</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>db</span>, id <span class='op'>=</span> <span class='st'>"name"</span>, idType <span class='op'>=</span> <span class='st'>"protein_id"</span><span class='op'>)</span></code></pre></div>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p><code>IRanges</code> with the coordinates within the protein(s). The
object has also to provide some means to identify the protein (see
details).</p></td>
    </tr>
    <tr>
      <th>db</th>
      <td><p><code>EnsDb</code> object to be used to retrieve genomic coordinates of
encoding transcripts.</p></td>
    </tr>
    <tr>
      <th>id</th>
      <td><p><code>character(1)</code> specifying where the protein identifier can be
found. Has to be either <code>"name"</code> or one of <code>colnames(mcols(prng))</code>.</p></td>
    </tr>
    <tr>
      <th>idType</th>
      <td><p><code>character(1)</code> defining what type of IDs are provided. Has to
be one of <code>"protein_id"</code> (default), <code>"uniprot_id"</code> or <code>"tx_id"</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>IRangesList</code>, each element being the mapping results for one of the input
ranges in <code>x</code>. Each element is a <code>IRanges</code> object with the positions within
the encoding transcript (relative to the start of the transcript, which
includes the 5' UTR). The transcript ID is reported as the name of each
<code>IRanges</code>. The <code>IRanges</code> can be of length &gt; 1 if the provided
protein identifier is annotated to more than one Ensembl protein ID (which
can be the case if Uniprot IDs are provided). If the coordinates can not be
mapped (because the protein identifier is unknown to the database) an
<code>IRanges</code> with negative coordinates is returned.</p>
<p>The following metadata columns are available in each <code>IRanges</code> in the result:</p><ul>
<li><p><code>"protein_id"</code>: the ID of the Ensembl protein for which the within-protein
coordinates were mapped to the genome.</p></li>
<li><p><code>"tx_id"</code>: the Ensembl transcript ID of the encoding transcript.</p></li>
<li><p><code>"cds_ok"</code>: contains <code>TRUE</code> if the length of the CDS matches the length
of the amino acid sequence and <code>FALSE</code> otherwise.</p></li>
<li><p><code>"protein_start"</code>: the within-protein sequence start coordinate of the
mapping.</p></li>
<li><p><code>"protein_end"</code>: the within-protein sequence end coordinate of the mapping.</p></li>
</ul>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Protein identifiers (supported are Ensembl protein IDs or Uniprot IDs) can
be passed to the function as <code>names</code> of the <code>x</code> <code>IRanges</code> object, or
alternatively in any one of the metadata columns (<code>mcols</code>) of <code>x</code>.</p>
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>While mapping of Ensembl protein IDs to Ensembl transcript IDs is 1:1, a
single Uniprot identifier can be annotated to several Ensembl protein IDs.
<code>proteinToTranscript</code> calculates in such cases transcript-relative
coordinates for each annotated Ensembl protein.</p>
<p>Mapping using Uniprot identifiers needs also additional internal checks that
can have a significant impact on the performance of the function. It is thus
strongly suggested to first identify the Ensembl protein identifiers for the
list of input Uniprot identifiers (e.g. using the <code><a href='ProteinFunctionality.html'>proteins()</a></code> function and
use these as input for the mapping function.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other coordinate mapping functions: 
<code><a href='cdsToTranscript.html'>cdsToTranscript</a>()</code>,
<code><a href='genomeToProtein.html'>genomeToProtein</a>()</code>,
<code><a href='genomeToTranscript.html'>genomeToTranscript</a>()</code>,
<code><a href='proteinToGenome.html'>proteinToGenome</a>()</code>,
<code><a href='transcriptToCds.html'>transcriptToCds</a>()</code>,
<code><a href='transcriptToGenome.html'>transcriptToGenome</a>()</code>,
<code><a href='transcriptToProtein.html'>transcriptToProtein</a>()</code></p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Johannes Rainer</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <div class="ref-examples sourceCode"><pre class='sourceCode r'><code><span class='r-in'></span>
<span class='r-in'><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>EnsDb.Hsapiens.v86</span><span class='op'>)</span></span>
<span class='r-in'><span class='co'>## Restrict all further queries to chromosome x to speed up the examples</span></span>
<span class='r-in'><span class='va'>edbx</span> <span class='op'>&lt;-</span> <span class='fu'><a href='global-filters.html'>filter</a></span><span class='op'>(</span><span class='va'>EnsDb.Hsapiens.v86</span>, filter <span class='op'>=</span> <span class='op'>~</span> <span class='va'>seq_name</span> <span class='op'>==</span> <span class='st'>"X"</span><span class='op'>)</span></span>
<span class='r-in'></span>
<span class='r-in'><span class='co'>## Define an IRange with protein-relative coordinates within a protein for</span></span>
<span class='r-in'><span class='co'>## the gene SYP</span></span>
<span class='r-in'><span class='va'>syp</span> <span class='op'>&lt;-</span> <span class='fu'>IRanges</span><span class='op'>(</span>start <span class='op'>=</span> <span class='fl'>4</span>, end <span class='op'>=</span> <span class='fl'>17</span><span class='op'>)</span></span>
<span class='r-in'><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>syp</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='st'>"ENSP00000418169"</span></span>
<span class='r-in'><span class='va'>res</span> <span class='op'>&lt;-</span> <span class='fu'>proteinToTranscript</span><span class='op'>(</span><span class='va'>syp</span>, <span class='va'>edbx</span><span class='op'>)</span></span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Fetching CDS for 1 proteins ... </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> 1 found</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Checking CDS and protein sequence lengths ... </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> 1/1 OK</span>
<span class='r-in'><span class='va'>res</span></span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> IRangesList object of length 1:</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> $ENSP00000418169</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> IRanges object with 1 range and 5 metadata columns:</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                       start       end     width |      protein_id</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                   &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |     &lt;character&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENST00000479808        23        64        42 | ENSP00000418169</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                             tx_id    cds_ok protein_start protein_end</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                       &lt;character&gt; &lt;logical&gt;     &lt;integer&gt;   &lt;integer&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENST00000479808 ENST00000479808      TRUE             4          17</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-in'><span class='co'>## Positions 4 to 17 within the protein span are encoded by the region</span></span>
<span class='r-in'><span class='co'>## from nt 23 to 64.</span></span>
<span class='r-in'></span>
<span class='r-in'><span class='co'>## Perform the mapping for multiple proteins identified by their Uniprot</span></span>
<span class='r-in'><span class='co'>## IDs.</span></span>
<span class='r-in'><span class='va'>ids</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"O15266"</span>, <span class='st'>"Q9HBJ8"</span>, <span class='st'>"unexistant"</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>prngs</span> <span class='op'>&lt;-</span> <span class='fu'>IRanges</span><span class='op'>(</span>start <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>13</span>, <span class='fl'>43</span>, <span class='fl'>100</span><span class='op'>)</span>, end <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>21</span>, <span class='fl'>80</span>, <span class='fl'>100</span><span class='op'>)</span><span class='op'>)</span></span>
<span class='r-in'><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>prngs</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='va'>ids</span></span>
<span class='r-in'></span>
<span class='r-in'><span class='va'>res</span> <span class='op'>&lt;-</span> <span class='fu'>proteinToTranscript</span><span class='op'>(</span><span class='va'>prngs</span>, <span class='va'>edbx</span>, idType <span class='op'>=</span> <span class='st'>"uniprot_id"</span><span class='op'>)</span></span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Fetching CDS for 3 proteins ... </span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>No CDS found for: unexistant</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> 2 found</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Checking CDS and protein sequence lengths ... </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> 2/2 OK</span>
<span class='r-in'></span>
<span class='r-in'><span class='co'>## The result is a list, same length as the input object</span></span>
<span class='r-in'><span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='va'>res</span><span class='op'>)</span></span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> [1] 3</span>
<span class='r-in'><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>res</span><span class='op'>)</span></span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> [1] "O15266"     "Q9HBJ8"     "unexistant"</span>
<span class='r-in'></span>
<span class='r-in'><span class='co'>## No protein/encoding transcript could be found for the last one</span></span>
<span class='r-in'><span class='va'>res</span><span class='op'>[[</span><span class='fl'>3</span><span class='op'>]</span><span class='op'>]</span></span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> IRanges object with 1 range and 6 metadata columns:</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>           start       end     width |  protein_id       tx_id    cds_ok</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; | &lt;character&gt; &lt;character&gt; &lt;logical&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   [1]        -1        -1         1 |        &lt;NA&gt;        &lt;NA&gt;      &lt;NA&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>       protein_start protein_end  uniprot_id</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>           &lt;integer&gt;   &lt;integer&gt; &lt;character&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   [1]           100         100  unexistant</span>
<span class='r-in'></span>
<span class='r-in'><span class='co'>## The first protein could be mapped to multiple Ensembl proteins. The</span></span>
<span class='r-in'><span class='co'>## region within all transcripts encoding the region in the protein are</span></span>
<span class='r-in'><span class='co'>## returned</span></span>
<span class='r-in'><span class='va'>res</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span></span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> IRanges object with 4 ranges and 6 metadata columns:</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                       start       end     width |      protein_id</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                   &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |     &lt;character&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENST00000334060       728       754        27 | ENSP00000335505</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENST00000381575       128       154        27 | ENSP00000370987</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENST00000381578       728       754        27 | ENSP00000370990</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENST00000554971       128       154        27 | ENSP00000452016</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                             tx_id    cds_ok protein_start protein_end</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                       &lt;character&gt; &lt;logical&gt;     &lt;integer&gt;   &lt;integer&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENST00000334060 ENST00000334060      TRUE            13          21</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENST00000381575 ENST00000381575      TRUE            13          21</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENST00000381578 ENST00000381578      TRUE            13          21</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENST00000554971 ENST00000554971      TRUE            13          21</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                    uniprot_id</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                   &lt;character&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENST00000334060      O15266</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENST00000381575      O15266</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENST00000381578      O15266</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENST00000554971      O15266</span>
<span class='r-in'></span>
<span class='r-in'><span class='co'>## The result for the region within the second protein</span></span>
<span class='r-in'><span class='va'>res</span><span class='op'>[[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>]</span></span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> IRanges object with 1 range and 6 metadata columns:</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                       start       end     width |      protein_id</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                   &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |     &lt;character&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENST00000380342       383       496       114 | ENSP00000369699</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                             tx_id    cds_ok protein_start protein_end</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                       &lt;character&gt; &lt;logical&gt;     &lt;integer&gt;   &lt;integer&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENST00000380342 ENST00000380342      TRUE            43          80</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                    uniprot_id</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                   &lt;character&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   ENST00000380342      Q9HBJ8</span>
</code></pre></div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by Johannes Rainer.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p></p>
</div>

      </footer>
   </div>

  


  

  </body>
</html>


