<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Protein related functionality — proteins,EnsDb-method • ensembldb</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Protein related functionality — proteins,EnsDb-method" />
<meta property="og:description" content="This help page provides information about most of the
    functionality related to protein annotations in ensembldb.
The proteins method retrieves protein related annotations from
    an EnsDb database.
The listUniprotDbs method lists all Uniprot database
    names in the EnsDb.
The listUniprotMappingTypes method lists all methods
    that were used for the mapping of Uniprot IDs to Ensembl protein IDs.
The listProteinColumns function allows to conveniently
    extract all database columns containing protein annotations from
    an EnsDb database." />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ensembldb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.17.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/ensembldb.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/MySQL-backend.html">Using a MySQL server backend</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping-use-cases.html">Use cases for coordinate mapping with ensembldb</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping.html">Mapping between genome, transcript and protein coordinates</a>
    </li>
    <li>
      <a href="../articles/proteins.html">Querying protein features</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jorainer/ensembldb/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Protein related functionality</h1>
    <small class="dont-index">Source: <a href='https://github.com/jorainer/ensembldb/blob/master/R/Methods.R'><code>R/Methods.R</code></a>, <a href='https://github.com/jorainer/ensembldb/blob/master/R/functions-utils.R'><code>R/functions-utils.R</code></a></small>
    <div class="hidden name"><code>ProteinFunctionality.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This help page provides information about most of the
    functionality related to protein annotations in <code>ensembldb</code>.</p>
<p>The <code>proteins</code> method retrieves protein related annotations from
    an <code><a href='EnsDb-class.html'>EnsDb</a></code> database.</p>
<p>The <code>listUniprotDbs</code> method lists all Uniprot database
    names in the <code>EnsDb</code>.</p>
<p>The <code>listUniprotMappingTypes</code> method lists all methods
    that were used for the mapping of Uniprot IDs to Ensembl protein IDs.</p>
<p>The <code>listProteinColumns</code> function allows to conveniently
    extract all database columns containing protein annotations from
    an <code><a href='EnsDb-class.html'>EnsDb</a></code> database.</p>
    </div>

    <div class="ref-usage sourceCode"><pre class='sourceCode r'><code><span class='co'># S4 method for EnsDb</span>
<span class='fu'>proteins</span><span class='op'>(</span>
  <span class='va'>object</span>,
  columns <span class='op'>=</span> <span class='fu'><a href='EnsDb-class.html'>listColumns</a></span><span class='op'>(</span><span class='va'>object</span>, <span class='st'>"protein"</span><span class='op'>)</span>,
  filter <span class='op'>=</span> <span class='fu'>AnnotationFilterList</span><span class='op'>(</span><span class='op'>)</span>,
  order.by <span class='op'>=</span> <span class='st'>""</span>,
  order.type <span class='op'>=</span> <span class='st'>"asc"</span>,
  return.type <span class='op'>=</span> <span class='st'>"DataFrame"</span>
<span class='op'>)</span>

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>listUniprotDbs</span><span class='op'>(</span><span class='va'>object</span><span class='op'>)</span>

<span class='co'># S4 method for EnsDb</span>
<span class='fu'>listUniprotMappingTypes</span><span class='op'>(</span><span class='va'>object</span><span class='op'>)</span>

<span class='fu'>listProteinColumns</span><span class='op'>(</span><span class='va'>object</span><span class='op'>)</span></code></pre></div>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>The <code><a href='EnsDb-class.html'>EnsDb</a></code> object.</p></td>
    </tr>
    <tr>
      <th>columns</th>
      <td><p>For <code>proteins</code>: character vector defining the columns to
be extracted from the database. Can be any column(s) listed by the
<code><a href='EnsDb-class.html'>listColumns</a></code> method.</p></td>
    </tr>
    <tr>
      <th>filter</th>
      <td><p>For <code>proteins</code>: A filter object extending
<code>AnnotationFilter</code> or a list of such objects to select
specific entries from the database. See <code><a href='Filter-classes.html'>Filter-classes</a></code> for
a documentation of available filters and use
<code>supportedFilters</code> to get the full list of supported filters.</p></td>
    </tr>
    <tr>
      <th>order.by</th>
      <td><p>For <code>proteins</code>: a character vector specifying the
column(s) by which the result should be ordered.</p></td>
    </tr>
    <tr>
      <th>order.type</th>
      <td><p>For <code>proteins</code>: if the results should be ordered
ascending (<code>order.type = "asc"</code>) or descending
(<code>order.type = "desc"</code>)</p></td>
    </tr>
    <tr>
      <th>return.type</th>
      <td><p>For <code>proteins</code>: character of lenght one specifying
the type of the returned object. Can be either <code>"DataFrame"</code>,
<code>"data.frame"</code> or <code>"AAStringSet"</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>The <code>proteins</code> method returns protein related annotations from
    an <code><a href='EnsDb-class.html'>EnsDb</a></code> object with its <code>return.type</code> argument
    allowing to define the type of the returned object. Note that if
    <code>return.type = "AAStringSet"</code> additional annotation columns are
    stored in a <code>DataFrame</code> that can be accessed with the <code>mcols</code>
    method on the returned object.</p>
<p>The <code>listProteinColumns</code> function returns a character vector
    with the column names containing protein annotations or throws an error
    if no such annotations are available.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The <code>proteins</code> method performs the query starting from the
    <code>protein</code> tables and can hence return all annotations from the
    database that are related to proteins and transcripts encoding these
    proteins from the database. Since <code>proteins</code> does thus only query
    annotations for protein coding transcripts, the <code><a href='EnsDb-exonsBy.html'>genes</a></code> or
    <code><a href='EnsDb-exonsBy.html'>transcripts</a></code> methods have to be used to retrieve annotations
    for non-coding transcripts.</p>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Johannes Rainer</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <div class="ref-examples sourceCode"><pre class='sourceCode r'><code><span class='r-in'><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/jorainer/ensembldb'>ensembldb</a></span><span class='op'>)</span></span>
<span class='r-in'><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>EnsDb.Hsapiens.v86</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>edb</span> <span class='op'>&lt;-</span> <span class='va'>EnsDb.Hsapiens.v86</span></span>
<span class='r-in'><span class='co'>## Get all proteins from tha database for the gene ZBTB16, if protein</span></span>
<span class='r-in'><span class='co'>## annotations are available</span></span>
<span class='r-in'><span class='kw'>if</span> <span class='op'>(</span><span class='fu'><a href='hasProteinData-EnsDb-method.html'>hasProteinData</a></span><span class='op'>(</span><span class='va'>edb</span><span class='op'>)</span><span class='op'>)</span></span>
<span class='r-in'>    <span class='fu'>proteins</span><span class='op'>(</span><span class='va'>edb</span>, filter <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/AnnotationFilter/man/AnnotationFilter.html'>GeneNameFilter</a></span><span class='op'>(</span><span class='st'>"ZBTB16"</span><span class='op'>)</span><span class='op'>)</span></span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> DataFrame with 5 rows and 4 columns</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>             tx_id      protein_id       protein_sequence   gene_name</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>       &lt;character&gt;     &lt;character&gt;            &lt;character&gt; &lt;character&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> 1 ENST00000335953 ENSP00000338157 MDLTKMGMIQLQNPSHPTGL..      ZBTB16</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> 2 ENST00000544220 ENSP00000437716 MDLTKMGMIQLQNPSHPTGL..      ZBTB16</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> 3 ENST00000535700 ENSP00000443013 MDLTKMGMIQLQNPSHPTGL..      ZBTB16</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> 4 ENST00000392996 ENSP00000376721 MDLTKMGMIQLQNPSHPTGL..      ZBTB16</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> 5 ENST00000539918 ENSP00000445047 XGGLLPQGFIQRELFSKLGE..      ZBTB16</span>
<span class='r-in'></span>
<span class='r-in'><span class='co'>## List the names of all Uniprot databases from which Uniprot IDs are</span></span>
<span class='r-in'><span class='co'>## available in the EnsDb</span></span>
<span class='r-in'><span class='kw'>if</span> <span class='op'>(</span><span class='fu'><a href='hasProteinData-EnsDb-method.html'>hasProteinData</a></span><span class='op'>(</span><span class='va'>edb</span><span class='op'>)</span><span class='op'>)</span></span>
<span class='r-in'>    <span class='fu'>listUniprotDbs</span><span class='op'>(</span><span class='va'>edb</span><span class='op'>)</span></span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> [1] "SWISSPROT" "SPTREMBL" </span>
<span class='r-in'></span>
<span class='r-in'></span>
<span class='r-in'><span class='co'>## List the type of all methods that were used to map Uniprot IDs to Ensembl</span></span>
<span class='r-in'><span class='co'>## protein IDs</span></span>
<span class='r-in'><span class='kw'>if</span> <span class='op'>(</span><span class='fu'><a href='hasProteinData-EnsDb-method.html'>hasProteinData</a></span><span class='op'>(</span><span class='va'>edb</span><span class='op'>)</span><span class='op'>)</span></span>
<span class='r-in'>    <span class='fu'>listUniprotMappingTypes</span><span class='op'>(</span><span class='va'>edb</span><span class='op'>)</span></span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> [1] "DIRECT"         "SEQUENCE_MATCH"</span>
<span class='r-in'></span>
<span class='r-in'></span>
<span class='r-in'><span class='co'>## List all columns containing protein annotations</span></span>
<span class='r-in'><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>EnsDb.Hsapiens.v86</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>edb</span> <span class='op'>&lt;-</span> <span class='va'>EnsDb.Hsapiens.v86</span></span>
<span class='r-in'><span class='kw'>if</span> <span class='op'>(</span><span class='fu'><a href='hasProteinData-EnsDb-method.html'>hasProteinData</a></span><span class='op'>(</span><span class='va'>edb</span><span class='op'>)</span><span class='op'>)</span></span>
<span class='r-in'>    <span class='fu'>listProteinColumns</span><span class='op'>(</span><span class='va'>edb</span><span class='op'>)</span></span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>  [1] "tx_id"                 "protein_id"            "protein_sequence"     </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>  [4] "protein_domain_id"     "protein_domain_source" "interpro_accession"   </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>  [7] "prot_dom_start"        "prot_dom_end"          "uniprot_id"           </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> [10] "uniprot_db"            "uniprot_mapping_type" </span>
</code></pre></div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by Johannes Rainer.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p></p>
</div>

      </footer>
   </div>

  


  

  </body>
</html>


