<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Map transcript-relative coordinates to genomic coordinates — transcriptToGenome • ensembldb</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Map transcript-relative coordinates to genomic coordinates — transcriptToGenome"><meta property="og:description" content="transcriptToGenome maps transcript-relative coordinates to genomic
coordinates. Provided coordinates are expected to be relative to the first
nucleotide of the transcript, not the CDS. CDS-relative coordinates
have to be converted to transcript-relative positions first with the
cdsToTranscript() function."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ensembldb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.23.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/ensembldb.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/MySQL-backend.html">Using a MySQL server backend</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping-use-cases.html">Use cases for coordinate mapping with ensembldb</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping.html">Mapping between genome, transcript and protein coordinates</a>
    </li>
    <li>
      <a href="../articles/proteins.html">Querying protein features</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/jorainer/ensembldb/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Map transcript-relative coordinates to genomic coordinates</h1>
    <small class="dont-index">Source: <a href="https://github.com/jorainer/ensembldb/blob/HEAD/R/transcriptToX.R" class="external-link"><code>R/transcriptToX.R</code></a></small>
    <div class="hidden name"><code>transcriptToGenome.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>transcriptToGenome</code> maps transcript-relative coordinates to genomic
coordinates. Provided coordinates are expected to be relative to the first
nucleotide of the <strong>transcript</strong>, not the <strong>CDS</strong>. CDS-relative coordinates
have to be converted to transcript-relative positions first with the
<code><a href="cdsToTranscript.html">cdsToTranscript()</a></code> function.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">transcriptToGenome</span><span class="op">(</span><span class="va">x</span>, <span class="va">db</span>, id <span class="op">=</span> <span class="st">"name"</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>x</dt>
<dd><p><code>IRanges</code> with the coordinates within the transcript. Coordinates
are counted from the start of the transcript (including the 5' UTR). The
Ensembl IDs of the corresponding transcripts have to be provided either
as <code>names</code> of the <code>IRanges</code>, or in one of its metadata columns.</p></dd>


<dt>db</dt>
<dd><p><code>EnsDb</code> object.</p></dd>


<dt>id</dt>
<dd><p><code>character(1)</code> specifying where the transcript identifier can be
found. Has to be either <code>"name"</code> or one of <code>colnames(mcols(prng))</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p><code>GRangesList</code> with the same length (and order) than the input <code>IRanges</code></p>


<p><code>x</code>. Each <code>GRanges</code> in the <code>GRangesList</code> provides the genomic coordinates
corresponding to the provided within-transcript coordinates. The
original transcript ID and the transcript-relative coordinates are provided
as metadata columns as well as the ID of the individual exon(s). An empty
<code>GRanges</code> is returned for transcripts that can not be found in the database.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="cdsToTranscript.html">cdsToTranscript()</a></code> and <code><a href="transcriptToCds.html">transcriptToCds()</a></code> for the mapping between
CDS- and transcript-relative coordinates.</p>
<p>Other coordinate mapping functions: 
<code><a href="cdsToTranscript.html">cdsToTranscript</a>()</code>,
<code><a href="genomeToProtein.html">genomeToProtein</a>()</code>,
<code><a href="genomeToTranscript.html">genomeToTranscript</a>()</code>,
<code><a href="proteinToGenome.html">proteinToGenome</a>()</code>,
<code><a href="proteinToTranscript.html">proteinToTranscript</a>()</code>,
<code><a href="transcriptToCds.html">transcriptToCds</a>()</code>,
<code><a href="transcriptToProtein.html">transcriptToProtein</a>()</code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Johannes Rainer</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co">## Restrict all further queries to chromosome x to speed up the examples</span></span></span>
<span class="r-in"><span><span class="va">edbx</span> <span class="op">&lt;-</span> <span class="fu"><a href="global-filters.html">filter</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span>, filter <span class="op">=</span> <span class="op">~</span> <span class="va">seq_name</span> <span class="op">==</span> <span class="st">"X"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Below we map positions 1 to 5 within the transcript ENST00000381578 to</span></span></span>
<span class="r-in"><span><span class="co">## the genome. The ID of the transcript has to be provided either as names</span></span></span>
<span class="r-in"><span><span class="co">## or in one of the IRanges' metadata columns</span></span></span>
<span class="r-in"><span><span class="va">txpos</span> <span class="op">&lt;-</span> <span class="fu">IRanges</span><span class="op">(</span>start <span class="op">=</span> <span class="fl">1</span>, end <span class="op">=</span> <span class="fl">5</span>, names <span class="op">=</span> <span class="st">"ENST00000381578"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">transcriptToGenome</span><span class="op">(</span><span class="va">txpos</span>, <span class="va">edbx</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRangesList object of length 1:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $ENST00000381578</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 1 range and 5 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       seqnames        ranges strand |         exon_id           tx_id exon_rank</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;     &lt;character&gt; &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]        X 624344-624348      + | ENSE00001489178 ENST00000381578         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        tx_start    tx_end</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       &lt;integer&gt; &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]         1         5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from GRCh38 genome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="co">## The object returns a GRangesList with the genomic coordinates, in this</span></span></span>
<span class="r-in"><span><span class="co">## example the coordinates are within the same exon and map to a single</span></span></span>
<span class="r-in"><span><span class="co">## genomic region.</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Next we map nucleotides 501 to 505 of ENST00000486554 to the genome</span></span></span>
<span class="r-in"><span><span class="va">txpos</span> <span class="op">&lt;-</span> <span class="fu">IRanges</span><span class="op">(</span>start <span class="op">=</span> <span class="fl">501</span>, end <span class="op">=</span> <span class="fl">505</span>, names <span class="op">=</span> <span class="st">"ENST00000486554"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">transcriptToGenome</span><span class="op">(</span><span class="va">txpos</span>, <span class="va">edbx</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRangesList object of length 1:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $ENST00000486554</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 2 ranges and 5 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       seqnames              ranges strand |         exon_id           tx_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;     &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]        X 107715899-107715901      - | ENSE00001927337 ENST00000486554</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2]        X 107714748-107714749      - | ENSE00001837666 ENST00000486554</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       exon_rank  tx_start    tx_end</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]         1       501       505</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2]         2       501       505</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from GRCh38 genome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="co">## The positions within the transcript are located within two of the</span></span></span>
<span class="r-in"><span><span class="co">## transcripts exons and thus a `GRanges` of length 2 is returned.</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Next we map multiple regions, two within the same transcript and one</span></span></span>
<span class="r-in"><span><span class="co">## in a transcript that does not exist.</span></span></span>
<span class="r-in"><span><span class="va">txpos</span> <span class="op">&lt;-</span> <span class="fu">IRanges</span><span class="op">(</span>start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">501</span>, <span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span>, end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">505</span>, <span class="fl">10</span>, <span class="fl">6</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENST00000486554"</span>, <span class="st">"ENST00000486554"</span>, <span class="st">"some"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">transcriptToGenome</span><span class="op">(</span><span class="va">txpos</span>, <span class="va">edbx</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>1 transcript(s) could either not be found in the database or the specified range is outside the transcript's sequence</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## The length of the result GRangesList has the same length than the</span></span></span>
<span class="r-in"><span><span class="co">## input IRanges</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 3</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## The result for the last region is an empty GRanges, because the</span></span></span>
<span class="r-in"><span><span class="co">## transcript could not be found in the database</span></span></span>
<span class="r-in"><span><span class="va">res</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 0 ranges and 0 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    seqnames    ranges strand</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: no sequences</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">res</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRangesList object of length 3:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $ENST00000486554</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 2 ranges and 5 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       seqnames              ranges strand |         exon_id           tx_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;     &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]        X 107715899-107715901      - | ENSE00001927337 ENST00000486554</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2]        X 107714748-107714749      - | ENSE00001837666 ENST00000486554</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       exon_rank  tx_start    tx_end</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]         1       501       505</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2]         2       501       505</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from GRCh38 genome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $ENST00000486554</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 1 range and 5 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       seqnames              ranges strand |         exon_id           tx_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;     &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]        X 107716392-107716401      - | ENSE00001927337 ENST00000486554</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       exon_rank  tx_start    tx_end</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]         1         1        10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from GRCh38 genome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $some</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 0 ranges and 0 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    seqnames    ranges strand</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: no sequences</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Johannes Rainer.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

