<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Map transcript-relative coordinates to amino acid residues of the
encoded protein — transcriptToProtein • ensembldb</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Map transcript-relative coordinates to amino acid residues of the
encoded protein — transcriptToProtein" />
<meta property="og:description" content="transcriptToProtein maps within-transcript coordinates to the corresponding
coordinates within the encoded protein sequence. The provided coordinates
have to be within the coding region of the transcript (excluding the stop
codon) but are supposed to be relative to the first nucleotide of the
transcript (which includes the 5' UTR). Positions relative to the CDS of a
transcript (e.g. /PKP2 c.1643delg/) have to be first converted to
transcript-relative coordinates. This can be done with the
cdsToTranscript() function." />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ensembldb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.15.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/ensembldb.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/MySQL-backend.html">Using a MySQL server backend</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping-use-cases.html">Use cases for coordinate mapping with ensembldb</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping.html">Mapping between genome, transcript and protein coordinates</a>
    </li>
    <li>
      <a href="../articles/proteins.html">Querying protein features</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jorainer/ensembldb/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Map transcript-relative coordinates to amino acid residues of the
encoded protein</h1>
    <small class="dont-index">Source: <a href='https://github.com/jorainer/ensembldb/blob/master/R/transcriptToX.R'><code>R/transcriptToX.R</code></a></small>
    <div class="hidden name"><code>transcriptToProtein.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>transcriptToProtein</code> maps within-transcript coordinates to the corresponding
coordinates within the encoded protein sequence. The provided coordinates
have to be within the coding region of the transcript (excluding the stop
codon) but are supposed to be relative to the first nucleotide of the
transcript (which includes the 5' UTR). Positions relative to the CDS of a
transcript (e.g. /PKP2 c.1643delg/) have to be first converted to
transcript-relative coordinates. This can be done with the
<code><a href='cdsToTranscript.html'>cdsToTranscript()</a></code> function.</p>
    </div>

    <pre class="usage"><span class='fu'>transcriptToProtein</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>db</span>, id <span class='op'>=</span> <span class='st'>"name"</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p><code>IRanges</code> with the coordinates within the transcript. Coordinates
are counted from the start of the transcript (including the 5' UTR). The
Ensembl IDs of the corresponding transcripts have to be provided either
as <code>names</code> of the <code>IRanges</code>, or in one of its metadata columns.</p></td>
    </tr>
    <tr>
      <th>db</th>
      <td><p><code>EnsDb</code> object.</p></td>
    </tr>
    <tr>
      <th>id</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code> specifying where the transcript identifier can be
found. Has to be either <code>"name"</code> or one of <code><a href='https://rdrr.io/r/base/colnames.html'>colnames(mcols(prng))</a></code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>IRanges</code> with the same length (and order) than the input <code>IRanges</code>
<code>x</code>. Each element in <code>IRanges</code> provides the coordinates within the
protein sequence, names being the (Ensembl) IDs of the protein. The
original transcript ID and the transcript-relative coordinates are provided
as metadata columns. Metadata columns <code>"cds_ok"</code> indicates whether the
length of the transcript's CDS matches the length of the encoded protein.
<code>IRanges</code> with a start coordinate of <code>-1</code> is returned for transcript
coordinates that can not be mapped to protein-relative coordinates
(either no transcript was found for the provided ID, the transcript
does not encode a protein or the provided coordinates are not within
the coding region of the transcript).</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Transcript-relative coordinates are mapped to the amino acid residues they
encode. As an example, positions within the transcript that correspond to
nucleotides 1 to 3 in the CDS are mapped to the first position in the
protein sequence (see examples for more details).</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='cdsToTranscript.html'>cdsToTranscript()</a></code> and <code><a href='transcriptToCds.html'>transcriptToCds()</a></code> for conversion between
CDS- and transcript-relative coordinates.</p>
<p>Other coordinate mapping functions: 
<code><a href='cdsToTranscript.html'>cdsToTranscript</a>()</code>,
<code><a href='genomeToProtein.html'>genomeToProtein</a>()</code>,
<code><a href='genomeToTranscript.html'>genomeToTranscript</a>()</code>,
<code><a href='proteinToGenome.html'>proteinToGenome</a>()</code>,
<code><a href='proteinToTranscript.html'>proteinToTranscript</a>()</code>,
<code><a href='transcriptToCds.html'>transcriptToCds</a>()</code>,
<code><a href='transcriptToGenome.html'>transcriptToGenome</a>()</code></p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Johannes Rainer</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>EnsDb.Hsapiens.v86</span><span class='op'>)</span>
<span class='co'>## Restrict all further queries to chromosome x to speed up the examples</span>
<span class='va'>edbx</span> <span class='op'>&lt;-</span> <span class='fu'><a href='addFilter,EnsDb-method.html'>filter</a></span><span class='op'>(</span><span class='va'>EnsDb.Hsapiens.v86</span>, filter <span class='op'>=</span> <span class='op'>~</span> <span class='va'>seq_name</span> <span class='op'>==</span> <span class='st'>"X"</span><span class='op'>)</span>

<span class='co'>## Define an IRanges with the positions of the first 2 nucleotides of the</span>
<span class='co'>## coding region for the transcript ENST00000381578</span>
<span class='va'>txpos</span> <span class='op'>&lt;-</span> <span class='fu'>IRanges</span><span class='op'>(</span>start <span class='op'>=</span> <span class='fl'>692</span>, width <span class='op'>=</span> <span class='fl'>2</span>, names <span class='op'>=</span> <span class='st'>"ENST00000381578"</span><span class='op'>)</span>

<span class='co'>## Map these to the corresponding residues in the protein sequence</span>
<span class='co'>## The protein-relative coordinates are returned as an `IRanges` object,</span>
<span class='co'>## with the original, transcript-relative coordinates provided in metadata</span>
<span class='co'>## columns tx_start and tx_end</span>
<span class='fu'>transcriptToProtein</span><span class='op'>(</span><span class='va'>txpos</span>, <span class='va'>edbx</span><span class='op'>)</span>
</div><div class='output co'>#&gt; IRanges object with 1 range and 4 metadata columns:
#&gt;                       start       end     width |           tx_id  tx_start
#&gt;                   &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |     &lt;character&gt; &lt;integer&gt;
#&gt;   ENSP00000370990         1         1         1 | ENST00000381578       692
#&gt;                      tx_end    cds_ok
#&gt;                   &lt;integer&gt; &lt;logical&gt;
#&gt;   ENSP00000370990       693      TRUE</div><div class='input'>
<span class='co'>## We can also map multiple ranges. Note that for any of the 3 nucleotides</span>
<span class='co'>## encoding the same amino acid the position of this residue in the</span>
<span class='co'>## protein sequence is returned. To illustrate this we map below each of the</span>
<span class='co'>## first 4 nucleotides of the CDS to the corresponding position within the</span>
<span class='co'>## protein.</span>
<span class='va'>txpos</span> <span class='op'>&lt;-</span> <span class='fu'>IRanges</span><span class='op'>(</span>start <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>692</span>, <span class='fl'>693</span>, <span class='fl'>694</span>, <span class='fl'>695</span><span class='op'>)</span>,
    width <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>4</span><span class='op'>)</span>, names <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='st'>"ENST00000381578"</span>, <span class='fl'>4</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'>transcriptToProtein</span><span class='op'>(</span><span class='va'>txpos</span>, <span class='va'>edbx</span><span class='op'>)</span>
</div><div class='output co'>#&gt; IRanges object with 4 ranges and 4 metadata columns:
#&gt;                       start       end     width |           tx_id  tx_start
#&gt;                   &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |     &lt;character&gt; &lt;integer&gt;
#&gt;   ENSP00000370990         1         1         1 | ENST00000381578       692
#&gt;   ENSP00000370990         1         1         1 | ENST00000381578       693
#&gt;   ENSP00000370990         1         1         1 | ENST00000381578       694
#&gt;   ENSP00000370990         2         2         1 | ENST00000381578       695
#&gt;                      tx_end    cds_ok
#&gt;                   &lt;integer&gt; &lt;logical&gt;
#&gt;   ENSP00000370990       692      TRUE
#&gt;   ENSP00000370990       693      TRUE
#&gt;   ENSP00000370990       694      TRUE
#&gt;   ENSP00000370990       695      TRUE</div><div class='input'>
<span class='co'>## If the mapping fails, an IRanges with negative start position is returned.</span>
<span class='co'>## Mapping can fail (as below) because the ID is not known.</span>
<span class='fu'>transcriptToProtein</span><span class='op'>(</span><span class='fu'>IRanges</span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>1</span>, names <span class='op'>=</span> <span class='st'>"unknown"</span><span class='op'>)</span>, <span class='va'>edbx</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: Transcript(s) 'unknown' could not be found</span></div><div class='output co'>#&gt; IRanges object with 1 range and 4 metadata columns:
#&gt;           start       end     width |       tx_id  tx_start    tx_end    cds_ok
#&gt;       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; | &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;logical&gt;
#&gt;   [1]        -1        -1         1 |     unknown         1         1      &lt;NA&gt;</div><div class='input'>
<span class='co'>## Or because the provided coordinates are not within the CDS</span>
<span class='fu'>transcriptToProtein</span><span class='op'>(</span><span class='fu'>IRanges</span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>1</span>, names <span class='op'>=</span> <span class='st'>"ENST00000381578"</span><span class='op'>)</span>, <span class='va'>edbx</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: Provided coordinates for 'ENST00000381578' are not within the coding region</span></div><div class='output co'>#&gt; IRanges object with 1 range and 4 metadata columns:
#&gt;           start       end     width |           tx_id  tx_start    tx_end
#&gt;       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |     &lt;character&gt; &lt;integer&gt; &lt;integer&gt;
#&gt;   [1]        -1        -1         1 | ENST00000381578         1         1
#&gt;          cds_ok
#&gt;       &lt;logical&gt;
#&gt;   [1]      &lt;NA&gt;</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Johannes Rainer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


