<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Generating an using Ensembl based annotation packages • ensembldb</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Generating an using Ensembl based annotation packages">
<meta property="og:description" content="ensembldb">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ensembldb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.13.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/ensembldb.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/MySQL-backend.html">Using a MySQL server backend</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping-use-cases.html">Use cases for coordinate mapping with ensembldb</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping.html">Mapping between genome, transcript and protein coordinates</a>
    </li>
    <li>
      <a href="../articles/proteins.html">Querying protein features</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jorainer/ensembldb/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="ensembldb_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Generating an using Ensembl based annotation packages</h1>
                        <h4 class="author">Johannes Rainer</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jorainer/ensembldb/blob/master/vignettes/ensembldb.Rmd"><code>vignettes/ensembldb.Rmd</code></a></small>
      <div class="hidden name"><code>ensembldb.Rmd</code></div>

    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>The <code>ensembldb</code> package provides functions to create and use transcript centric annotation databases/packages. The annotation for the databases are directly fetched from Ensembl <sup><a id="fnr.1" class="footref" href="#fn.1">1</a></sup> using their Perl API. The functionality and data is similar to that of the <code>TxDb</code> packages from the <code>GenomicFeatures</code> package, but, in addition to retrieve all gene/transcript models and annotations from the database, the <code>ensembldb</code> package provides also a filter framework allowing to retrieve annotations for specific entries like genes encoded on a chromosome region or transcript models of lincRNA genes. From version 1.7 on, <code>EnsDb</code> databases created by the <code>ensembldb</code> package contain also protein annotation data (see Section <a href="#org4707794">11</a> for the database layout and an overview of available attributes/columns). For more information on the use of the protein annotations refer to the <em>proteins</em> vignette.</p>
<p>Another main goal of this package is to generate <em>versioned</em> annotation packages, i.e. annotation packages that are build for a specific Ensembl release, and are also named according to that (e.g. <code>EnsDb.Hsapiens.v86</code> for human gene definitions of the Ensembl code database version 86). This ensures reproducibility, as it allows to load annotations from a specific Ensembl release also if newer versions of annotation packages/releases are available. It also allows to load multiple annotation packages at the same time in order to e.g. compare gene models between Ensembl releases.</p>
<p>In the example below we load an Ensembl based annotation package for Homo sapiens, Ensembl version 86. The <code>EnsDb</code> object providing access to the underlying SQLite database is bound to the variable name <code>EnsDb.Hsapiens.v86</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">EnsDb.Hsapiens.v86</span>)

<span class="co">## Making a "short cut"</span>
<span class="kw">edb</span> <span class="op">&lt;-</span> <span class="kw">EnsDb.Hsapiens.v86</span>
<span class="co">## print some informations for this package</span>
<span class="kw">edb</span>
</pre></div>
<pre><code>## EnsDb for Ensembl:
## |Backend: SQLite
## |Db type: EnsDb
## |Type of Gene ID: Ensembl Gene ID
## |Supporting package: ensembldb
## |Db created by: ensembldb package from Bioconductor
## |script_version: 0.3.0
## |Creation time: Thu May 18 16:32:27 2017
## |ensembl_version: 86
## |ensembl_host: localhost
## |Organism: homo_sapiens
## |taxonomy_id: 9606
## |genome_build: GRCh38
## |DBSCHEMAVERSION: 2.0
## | No. of genes: 63970.
## | No. of transcripts: 216741.
## |Protein data available.</code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="co">## For what organism was the database generated?</span>
<span class="fu"><a href="../reference/EnsDb-class.html">organism</a></span>(<span class="kw">edb</span>)
</pre></div>
<pre><code>## [1] "Homo sapiens"</code></pre>
</div>
<div id="using-ensembldb-annotation-packages-to-retrieve-specific-annotations" class="section level1">
<h1 class="hasAnchor">
<a href="#using-ensembldb-annotation-packages-to-retrieve-specific-annotations" class="anchor"></a>Using <code>ensembldb</code> annotation packages to retrieve specific annotations</h1>
<p>One of the strengths of the <code>ensembldb</code> package and the related <code>EnsDb</code> databases is its implementation of a filter framework that enables to efficiently extract data sub-sets from the databases. The <code>ensembldb</code> package supports most of the filters defined in the <code>AnnotationFilter</code> Bioconductor package and defines some additional filters specific to the data stored in <code>EnsDb</code> databases. Filters can be passed directly to all methods extracting data from an <code>EnsDb</code> (such as <code>genes</code>, <code>transcripts</code> or <code>exons</code>). Alternatively it is possible with the <code>addFilter</code> or <code>filter</code> functions to add a filter directly to an <code>EnsDb</code> which will then be used in all queries on that object.</p>
<p>The <code>supportedFilters</code> method can be used to get an overview over all supported filter classes, each of them (except the <code>GRangesFilter</code>) working on a single column/field in the database.</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="fu">supportedFilters</span>(<span class="kw">edb</span>)
</pre></div>
<pre><code>##                       filter                 field
## 1               EntrezFilter                entrez
## 2              ExonEndFilter              exon_end
## 3               ExonIdFilter               exon_id
## 4             ExonRankFilter             exon_rank
## 5            ExonStartFilter            exon_start
## 6          GeneBiotypeFilter          gene_biotype
## 7              GeneEndFilter              gene_end
## 8               GeneIdFilter               gene_id
## 9             GenenameFilter              genename
## 10            GeneNameFilter             gene_name
## 11           GeneStartFilter            gene_start
## 12             GRangesFilter                  &lt;NA&gt;
## 13           ProtDomIdFilter           prot_dom_id
## 14     ProteinDomainIdFilter     protein_domain_id
## 15 ProteinDomainSourceFilter protein_domain_source
## 16           ProteinIdFilter            protein_id
## 17             SeqNameFilter              seq_name
## 18           SeqStrandFilter            seq_strand
## 19              SymbolFilter                symbol
## 20           TxBiotypeFilter            tx_biotype
## 21               TxEndFilter                tx_end
## 22                TxIdFilter                 tx_id
## 23              TxNameFilter               tx_name
## 24             TxStartFilter              tx_start
## 25           UniprotDbFilter            uniprot_db
## 26             UniprotFilter               uniprot
## 27  UniprotMappingTypeFilter  uniprot_mapping_type</code></pre>
<p>These filters can be divided into 3 main filter types:</p>
<ul>
<li>
<code>IntegerFilter</code>: filter classes extending this basic object can take a single numeric value as input and support the conditions <code><a href="https://rdrr.io/r/base/Comparison.html">==</a></code>, <code><a href="https://rdrr.io/r/base/Comparison.html">!=</a></code>, <code><a href="https://rdrr.io/r/base/Comparison.html">&gt;</a></code>, <code><a href="https://rdrr.io/r/base/Comparison.html">&lt;</a></code>, <code><a href="https://rdrr.io/r/base/Comparison.html">&gt;=</a></code> and <code><a href="https://rdrr.io/r/base/Comparison.html">&lt;=</a></code>. All filters that work on chromosomal coordinates, such as the <code>GeneEndFilter</code> extend <code>IntegerFilter</code>.</li>
<li>
<code>CharacterFilter</code>: filter classes extending this object can take a single or multiple character values as input and allow conditions: <code><a href="https://rdrr.io/r/base/Comparison.html">==</a></code>, <code><a href="https://rdrr.io/r/base/Comparison.html">!=</a></code>, <code>"startsWith"</code> , <code>"endsWith"</code> and <code>"contains"</code>. All filters working on IDs extend this class.</li>
<li>
<code>GRangesFilter</code>: takes a <code>GRanges</code> object as input and supports all conditions that <code>findOverlaps</code> from the <code>IRanges</code> package supports (<code>"any"</code>, <code>"start"</code>, <code>"end"</code>, <code>"within"</code>, <code>"equal"</code>). Note that these have to be passed using the parameter <code>type</code> to the constructor function.</li>
</ul>
<p>The supported filters are:</p>
<ul>
<li>
<code>EntrezFilter</code>: allows to filter results based on NCBI Entrezgene identifiers of the genes.</li>
<li>
<code>ExonEndFilter</code>: filter using the chromosomal end coordinate of exons.</li>
<li>
<code>ExonIdFilter</code>: filter based on the (Ensembl) exon identifiers.</li>
<li>
<code>ExonRankFilter</code>: filter based on the rank (index) of an exon within the transcript model. Exons are always numbered from 5’ to 3’ end of the transcript, thus, also on the reverse strand, the exon 1 is the most 5’ exon of the transcript.</li>
<li>
<code>ExonStartFilter</code>: filter using the chromosomal start coordinate of exons.</li>
<li>
<code>GeneBiotypeFilter</code>: filter using the gene biotypes defined in the Ensembl database; use the <code>listGenebiotypes</code> method to list all available biotypes.</li>
<li>
<code>GeneEndFilter</code>: filter using the chromosomal end coordinate of gene.</li>
<li>
<code>GeneIdFilter</code>: filter based on the Ensembl gene IDs.</li>
<li>
<code>GeneNameFilter</code>: filter based on the names (symbols) of the genes.</li>
<li>
<code>GeneStartFilter</code>: filter using the chromosomal start coordinate of gene.</li>
<li>
<code>GRangesFilter</code>: allows to retrieve all features (genes, transcripts or exons) that are either within (setting parameter <code>type</code> to “within”) or partially overlapping (setting <code>type</code> to “any”) the defined genomic region/range. Note that, depending on the called method (<code>genes</code>, <code>transcripts</code> or <code>exons</code>) the start and end coordinates of either the genes, transcripts or exons are used for the filter. For methods <code>exonsBy</code>, <code>cdsBy</code> and <code>txBy</code> the coordinates of <code>by</code> are used.</li>
<li>
<code>SeqNameFilter</code>: filter by the name of the chromosomes the genes are encoded on.</li>
<li>
<code>SeqStrandFilter</code>: filter for the chromosome strand on which the genes are encoded.</li>
<li>
<code>SymbolFilter</code>: filter on gene symbols; note that no database columns <em>symbol</em> is available in an <code>EnsDb</code> database and hence the gene name is used for filtering.</li>
<li>
<code>TxBiotypeFilter</code>: filter on the transcript biotype defined in Ensembl; use the <code>listTxbiotypes</code> method to list all available biotypes.</li>
<li>
<code>TxEndFilter</code>: filter using the chromosomal end coordinate of transcripts.</li>
<li>
<code>TxIdFilter</code>: filter on the Ensembl transcript identifiers.</li>
<li>
<code>TxNameFilter</code>: filter on the Ensembl transcript names (currently identical to the transcript IDs).</li>
<li>
<code>TxStartFilter</code>: filter using the chromosomal start coordinate of transcripts.</li>
</ul>
<p>In addition to the above listed <em>DNA-RNA-based</em> filters, <em>protein-specific</em> filters are also available:</p>
<ul>
<li>
<code>ProtDomIdFilter</code>, <code>ProteinDomainIdFilter</code>: filter by the protein domain ID.</li>
<li>
<code>ProteinDomainSourceFilter</code>: filter by the source of the protein domain (database or method, e.g. <em>pfam</em>).</li>
<li>
<code>ProteinIdFilter</code>: filter by Ensembl protein ID filters.</li>
<li>
<code>UniprotDbFilter</code>: filter by the name of the Uniprot database.</li>
<li>
<code>UniprotFilter</code>: filter by the Uniprot ID.</li>
<li>
<code>UniprotMappingTypeFilter</code>: filter by the mapping type of Ensembl protein IDs to Uniprot IDs.</li>
</ul>
<p>These can however only be used on <code>EnsDb</code> databases that provide protein annotations, i.e. for which a call to <code>hasProteinData</code> returns <code>TRUE</code>.</p>
<p><code>EnsDb</code> databases for more recent Ensembl versions (starting from Ensembl 87) provide also evidence levels for individual transcripts in the <code>tx_support_level</code> database column. Such databases support also a <code>TxSupportLevelFilter</code> filter to use this columns for filtering.</p>
<p>A simple use case for the filter framework would be to get all transcripts for the gene <em>BCL2L11</em>. To this end we specify a <code>GeneNameFilter</code> with the value <em>BCL2L11</em>. As a result we get a <code>GRanges</code> object with <code>start</code>, <code>end</code>, <code>strand</code> and <code>seqname</code> being the start coordinate, end coordinate, chromosome name and strand for the respective transcripts. All additional annotations are available as metadata columns. Alternatively, by setting <code>return.type</code> to “DataFrame”, or “data.frame” the method would return a <code>DataFrame</code> or <code>data.frame</code> object instead of the default <code>GRanges</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="kw">Tx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/exonsBy.html">transcripts</a></span>(<span class="kw">edb</span>, filter = <span class="fu">GeneNameFilter</span>(<span class="st">"BCL2L11"</span>))

<span class="kw">Tx</span>
</pre></div>
<pre><code>## GRanges object with 28 ranges and 7 metadata columns:
##                   seqnames              ranges strand |           tx_id
##                      &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
##   ENST00000432179        2 111119378-111124112      + | ENST00000432179
##   ENST00000308659        2 111120914-111165048      + | ENST00000308659
##   ENST00000337565        2 111120914-111128844      + | ENST00000337565
##   ENST00000622509        2 111120914-111168445      + | ENST00000622509
##   ENST00000619294        2 111120914-111168445      + | ENST00000619294
##               ...      ...                 ...    ... .             ...
##   ENST00000452231        2 111123746-111164231      + | ENST00000452231
##   ENST00000361493        2 111123746-111164231      + | ENST00000361493
##   ENST00000431217        2 111123746-111164352      + | ENST00000431217
##   ENST00000439718        2 111123746-111164643      + | ENST00000439718
##   ENST00000438054        2 111123752-111146284      + | ENST00000438054
##                               tx_biotype tx_cds_seq_start tx_cds_seq_end
##                              &lt;character&gt;        &lt;integer&gt;      &lt;integer&gt;
##   ENST00000432179         protein_coding        111123746      111124112
##   ENST00000308659         protein_coding        111123746      111164231
##   ENST00000337565         protein_coding        111123746      111128751
##   ENST00000622509         protein_coding        111123746      111161439
##   ENST00000619294         protein_coding        111123746      111144501
##               ...                    ...              ...            ...
##   ENST00000452231 nonsense_mediated_de..        111123746      111161439
##   ENST00000361493 nonsense_mediated_de..        111123746      111130235
##   ENST00000431217 nonsense_mediated_de..        111123746      111144501
##   ENST00000439718 nonsense_mediated_de..        111123746      111151851
##   ENST00000438054         protein_coding        111123752      111144491
##                           gene_id         tx_name   gene_name
##                       &lt;character&gt;     &lt;character&gt; &lt;character&gt;
##   ENST00000432179 ENSG00000153094 ENST00000432179     BCL2L11
##   ENST00000308659 ENSG00000153094 ENST00000308659     BCL2L11
##   ENST00000337565 ENSG00000153094 ENST00000337565     BCL2L11
##   ENST00000622509 ENSG00000153094 ENST00000622509     BCL2L11
##   ENST00000619294 ENSG00000153094 ENST00000619294     BCL2L11
##               ...             ...             ...         ...
##   ENST00000452231 ENSG00000153094 ENST00000452231     BCL2L11
##   ENST00000361493 ENSG00000153094 ENST00000361493     BCL2L11
##   ENST00000431217 ENSG00000153094 ENST00000431217     BCL2L11
##   ENST00000439718 ENSG00000153094 ENST00000439718     BCL2L11
##   ENST00000438054 ENSG00000153094 ENST00000438054     BCL2L11
##   -------
##   seqinfo: 1 sequence from GRCh38 genome</code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="co">## as this is a GRanges object we can access e.g. the start coordinates with</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/start.html">start</a></span>(<span class="kw">Tx</span>))
</pre></div>
<pre><code>## [1] 111119378 111120914 111120914 111120914 111120914 111120914</code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="co">## or extract the biotype with</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">Tx</span><span class="op">$</span><span class="kw">tx_biotype</span>)
</pre></div>
<pre><code>## [1] "protein_coding" "protein_coding" "protein_coding" "protein_coding"
## [5] "protein_coding" "protein_coding"</code></pre>
<p>The parameter <code>columns</code> of the extractor methods (such as <code>exons</code>, <code>genes</code> or <code>transcripts)</code> allows to specify which database attributes (columns) should be retrieved. The <code>exons</code> method returns by default all exon-related columns, the <code>transcripts</code> all columns from the transcript database table and the <code>genes</code> all from the gene table. Note however that in the example above we got also a column <code>gene_name</code> although this column is not present in the transcript database table. By default the methods return also all columns that are used by any of the filters submitted with the <code>filter</code> argument (thus, because a <code>GeneNameFilter</code> was used, the column <code>gene_name</code> is also returned). Setting <code>returnFilterColumns(edb) &lt;- FALSE</code> disables this option and only the columns specified by the <code>columns</code> parameter are retrieved.</p>
<p>Instead of passing a filter <em>object</em> to the method it is also possible to provide a filter <em>expression</em> written as a <code>formula</code>. The <code>formula</code> has to be written in the form <code>~ &lt;field&gt; &lt;condition&gt; &lt;value&gt;</code> with <code>&lt;field&gt;</code> being the field (database column) in the database, <code>&lt;condition&gt;</code> the condition for the filter object and <code>&lt;value&gt;</code> its value. Use the <code>supportedFilter</code> method to get the field names corresponding to each filter class.</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="co">## Use a filter expression to perform the filtering.</span>
<span class="fu"><a href="../reference/exonsBy.html">transcripts</a></span>(<span class="kw">edb</span>, filter = <span class="op">~</span> <span class="kw">gene_name</span> <span class="op">==</span> <span class="st">"ZBTB16"</span>)
</pre></div>
<pre><code>## GRanges object with 9 ranges and 7 metadata columns:
##                   seqnames              ranges strand |           tx_id
##                      &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
##   ENST00000335953       11 114059593-114250676      + | ENST00000335953
##   ENST00000541602       11 114059725-114189764      + | ENST00000541602
##   ENST00000544220       11 114059737-114063646      + | ENST00000544220
##   ENST00000535700       11 114060257-114063744      + | ENST00000535700
##   ENST00000392996       11 114060507-114250652      + | ENST00000392996
##   ENST00000539918       11 114064412-114247344      + | ENST00000539918
##   ENST00000545851       11 114180766-114247296      + | ENST00000545851
##   ENST00000535379       11 114237207-114250557      + | ENST00000535379
##   ENST00000535509       11 114246790-114250476      + | ENST00000535509
##                               tx_biotype tx_cds_seq_start tx_cds_seq_end
##                              &lt;character&gt;        &lt;integer&gt;      &lt;integer&gt;
##   ENST00000335953         protein_coding        114063301      114250555
##   ENST00000541602        retained_intron             &lt;NA&gt;           &lt;NA&gt;
##   ENST00000544220         protein_coding        114063301      114063646
##   ENST00000535700         protein_coding        114063301      114063744
##   ENST00000392996         protein_coding        114063301      114250555
##   ENST00000539918 nonsense_mediated_de..        114064412      114121827
##   ENST00000545851   processed_transcript             &lt;NA&gt;           &lt;NA&gt;
##   ENST00000535379   processed_transcript             &lt;NA&gt;           &lt;NA&gt;
##   ENST00000535509        retained_intron             &lt;NA&gt;           &lt;NA&gt;
##                           gene_id         tx_name   gene_name
##                       &lt;character&gt;     &lt;character&gt; &lt;character&gt;
##   ENST00000335953 ENSG00000109906 ENST00000335953      ZBTB16
##   ENST00000541602 ENSG00000109906 ENST00000541602      ZBTB16
##   ENST00000544220 ENSG00000109906 ENST00000544220      ZBTB16
##   ENST00000535700 ENSG00000109906 ENST00000535700      ZBTB16
##   ENST00000392996 ENSG00000109906 ENST00000392996      ZBTB16
##   ENST00000539918 ENSG00000109906 ENST00000539918      ZBTB16
##   ENST00000545851 ENSG00000109906 ENST00000545851      ZBTB16
##   ENST00000535379 ENSG00000109906 ENST00000535379      ZBTB16
##   ENST00000535509 ENSG00000109906 ENST00000535509      ZBTB16
##   -------
##   seqinfo: 1 sequence from GRCh38 genome</code></pre>
<p>Filter expression have to be written as a formula (i.e. starting with a <code><a href="https://rdrr.io/r/base/tilde.html">~</a></code>) in the form <em>column name</em> followed by the logical condition.</p>
<p>Alternatively, <code>EnsDb</code> objects can be filtered directly using the <code>filter</code> function. In the example below we use the <code>filter</code> function to filter the <code>EnsDb</code> object and pass that filtered database to the <code>transcripts</code> method using the <code><a href="https://rdrr.io/pkg/magrittr/man/pipe.html">%&gt;%</a></code> from the <code>magrittr</code> package.</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">magrittr</span>)

<span class="kw">edb</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/addFilter,EnsDb-method.html">filter</a></span>(<span class="op">~</span> <span class="kw">symbol</span> <span class="op">==</span> <span class="st">"BCL2"</span> <span class="op">&amp;</span> <span class="kw">tx_biotype</span> != <span class="st">"protein_coding"</span>) <span class="op">%&gt;%</span>
    <span class="kw">transcripts</span>
</pre></div>
<pre><code>## GRanges object with 1 range and 6 metadata columns:
##                   seqnames            ranges strand |           tx_id
##                      &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
##   ENST00000590515       18 63128212-63161869      - | ENST00000590515
##                             tx_biotype tx_cds_seq_start tx_cds_seq_end
##                            &lt;character&gt;        &lt;integer&gt;      &lt;integer&gt;
##   ENST00000590515 processed_transcript             &lt;NA&gt;           &lt;NA&gt;
##                           gene_id         tx_name
##                       &lt;character&gt;     &lt;character&gt;
##   ENST00000590515 ENSG00000171791 ENST00000590515
##   -------
##   seqinfo: 1 sequence from GRCh38 genome</code></pre>
<p>Adding a filter to an <code>EnsDb</code> enables this filter (globally) on all subsequent queries on that object. We could thus filter an <code>EnsDb</code> to (virtually) contain only features encoded on chromosome Y.</p>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="kw">edb_y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addFilter,EnsDb-method.html">addFilter</a></span>(<span class="kw">edb</span>, <span class="fu">SeqNameFilter</span>(<span class="st">"Y"</span>))

<span class="co">## All subsequent filters on that EnsDb will only work on features encoded on</span>
<span class="co">## chromosome Y</span>
<span class="fu"><a href="../reference/exonsBy.html">genes</a></span>(<span class="kw">edb_y</span>)
</pre></div>
<pre><code>## GRanges object with 523 ranges and 6 metadata columns:
##                   seqnames            ranges strand |         gene_id
##                      &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
##   ENSG00000251841        Y   2784749-2784853      + | ENSG00000251841
##   ENSG00000184895        Y   2786855-2787699      - | ENSG00000184895
##   ENSG00000237659        Y   2789827-2790328      + | ENSG00000237659
##   ENSG00000232195        Y   2827982-2828218      + | ENSG00000232195
##   ENSG00000129824        Y   2841486-2932000      + | ENSG00000129824
##               ...      ...               ...    ... .             ...
##   ENSG00000224240        Y 26549425-26549743      + | ENSG00000224240
##   ENSG00000227629        Y 26586642-26591601      - | ENSG00000227629
##   ENSG00000237917        Y 26594851-26634652      - | ENSG00000237917
##   ENSG00000231514        Y 26626520-26627159      - | ENSG00000231514
##   ENSG00000235857        Y 56855244-56855488      + | ENSG00000235857
##                     gene_name           gene_biotype seq_coord_system
##                   &lt;character&gt;            &lt;character&gt;      &lt;character&gt;
##   ENSG00000251841  RNU6-1334P                  snRNA       chromosome
##   ENSG00000184895         SRY         protein_coding       chromosome
##   ENSG00000237659  RNASEH2CP1   processed_pseudogene       chromosome
##   ENSG00000232195    TOMM22P2   processed_pseudogene       chromosome
##   ENSG00000129824      RPS4Y1         protein_coding       chromosome
##               ...         ...                    ...              ...
##   ENSG00000224240     CYCSP49   processed_pseudogene       chromosome
##   ENSG00000227629  SLC25A15P1 unprocessed_pseudogene       chromosome
##   ENSG00000237917     PARP4P1 unprocessed_pseudogene       chromosome
##   ENSG00000231514     FAM58CP   processed_pseudogene       chromosome
##   ENSG00000235857     CTBP2P1   processed_pseudogene       chromosome
##                        symbol entrezid
##                   &lt;character&gt;   &lt;list&gt;
##   ENSG00000251841  RNU6-1334P       NA
##   ENSG00000184895         SRY     6736
##   ENSG00000237659  RNASEH2CP1       NA
##   ENSG00000232195    TOMM22P2       NA
##   ENSG00000129824      RPS4Y1     6192
##               ...         ...      ...
##   ENSG00000224240     CYCSP49       NA
##   ENSG00000227629  SLC25A15P1       NA
##   ENSG00000237917     PARP4P1       NA
##   ENSG00000231514     FAM58CP       NA
##   ENSG00000235857     CTBP2P1       NA
##   -------
##   seqinfo: 1 sequence from GRCh38 genome</code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit">
<span class="co">## Get all lincRNAs on chromosome Y</span>
<span class="fu"><a href="../reference/exonsBy.html">genes</a></span>(<span class="kw">edb_y</span>, filter = <span class="op">~</span> <span class="kw">gene_biotype</span> <span class="op">==</span> <span class="st">"lincRNA"</span>)
</pre></div>
<pre><code>## GRanges object with 52 ranges and 6 metadata columns:
##                   seqnames            ranges strand |         gene_id
##                      &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
##   ENSG00000278847        Y   2934406-2934771      - | ENSG00000278847
##   ENSG00000231535        Y   3002912-3102272      + | ENSG00000231535
##   ENSG00000229308        Y   4036497-4100320      + | ENSG00000229308
##   ENSG00000277930        Y   4993858-4999650      - | ENSG00000277930
##   ENSG00000237069        Y   6242446-6243610      - | ENSG00000237069
##               ...      ...               ...    ... .             ...
##   ENSG00000228296        Y 25063083-25099892      - | ENSG00000228296
##   ENSG00000223641        Y 25183643-25184773      - | ENSG00000223641
##   ENSG00000228786        Y 25378300-25394719      - | ENSG00000228786
##   ENSG00000240450        Y 25482908-25486705      + | ENSG00000240450
##   ENSG00000231141        Y 25728490-25733388      + | ENSG00000231141
##                       gene_name gene_biotype seq_coord_system        symbol
##                     &lt;character&gt;  &lt;character&gt;      &lt;character&gt;   &lt;character&gt;
##   ENSG00000278847 RP11-414C23.1      lincRNA       chromosome RP11-414C23.1
##   ENSG00000231535     LINC00278      lincRNA       chromosome     LINC00278
##   ENSG00000229308    AC010084.1      lincRNA       chromosome    AC010084.1
##   ENSG00000277930  RP11-122L9.1      lincRNA       chromosome  RP11-122L9.1
##   ENSG00000237069       TTTY23B      lincRNA       chromosome       TTTY23B
##               ...           ...          ...              ...           ...
##   ENSG00000228296        TTTY4C      lincRNA       chromosome        TTTY4C
##   ENSG00000223641       TTTY17C      lincRNA       chromosome       TTTY17C
##   ENSG00000228786  LINC00266-4P      lincRNA       chromosome  LINC00266-4P
##   ENSG00000240450      CSPG4P1Y      lincRNA       chromosome      CSPG4P1Y
##   ENSG00000231141         TTTY3      lincRNA       chromosome         TTTY3
##                               entrezid
##                                 &lt;list&gt;
##   ENSG00000278847                   NA
##   ENSG00000231535            100873962
##   ENSG00000229308                   NA
##   ENSG00000277930                   NA
##   ENSG00000237069     100101121,252955
##               ...                  ...
##   ENSG00000228296 474150,474149,114761
##   ENSG00000223641 474152,474151,252949
##   ENSG00000228786                   NA
##   ENSG00000240450               114758
##   ENSG00000231141        474148,114760
##   -------
##   seqinfo: 1 sequence from GRCh38 genome</code></pre>
<p>To get an overview of database tables and available columns the function <code>listTables</code> can be used. The method <code>listColumns</code> on the other hand lists columns for the specified database table.</p>
<div class="sourceCode" id="cb21"><pre class="downlit">
<span class="co">## list all database tables along with their columns</span>
<span class="fu"><a href="../reference/EnsDb-class.html">listTables</a></span>(<span class="kw">edb</span>)
</pre></div>
<pre><code>## $gene
## [1] "gene_id"          "gene_name"        "gene_biotype"     "gene_seq_start"  
## [5] "gene_seq_end"     "seq_name"         "seq_strand"       "seq_coord_system"
## [9] "symbol"          
## 
## $tx
## [1] "tx_id"            "tx_biotype"       "tx_seq_start"     "tx_seq_end"      
## [5] "tx_cds_seq_start" "tx_cds_seq_end"   "gene_id"          "tx_name"         
## 
## $tx2exon
## [1] "tx_id"    "exon_id"  "exon_idx"
## 
## $exon
## [1] "exon_id"        "exon_seq_start" "exon_seq_end"  
## 
## $chromosome
## [1] "seq_name"    "seq_length"  "is_circular"
## 
## $protein
## [1] "tx_id"            "protein_id"       "protein_sequence"
## 
## $uniprot
## [1] "protein_id"           "uniprot_id"           "uniprot_db"          
## [4] "uniprot_mapping_type"
## 
## $protein_domain
## [1] "protein_id"            "protein_domain_id"     "protein_domain_source"
## [4] "interpro_accession"    "prot_dom_start"        "prot_dom_end"         
## 
## $entrezgene
## [1] "gene_id"  "entrezid"
## 
## $metadata
## [1] "name"  "value"</code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit">
<span class="co">## list columns from a specific table</span>
<span class="fu"><a href="../reference/EnsDb-class.html">listColumns</a></span>(<span class="kw">edb</span>, <span class="st">"tx"</span>)
</pre></div>
<pre><code>## [1] "tx_id"            "tx_biotype"       "tx_seq_start"     "tx_seq_end"      
## [5] "tx_cds_seq_start" "tx_cds_seq_end"   "gene_id"          "tx_name"</code></pre>
<p>Thus, we could retrieve all transcripts of the biotype <em>nonsense_mediated_decay</em> (which, according to the definitions by Ensembl are transcribed, but most likely not translated in a protein, but rather degraded after transcription) along with the name of the gene for each transcript. Note that we are changing here the <code>return.type</code> to <code>DataFrame</code>, so the method will return a <code>DataFrame</code> with the results instead of the default <code>GRanges</code>.</p>
<div class="sourceCode" id="cb25"><pre class="downlit">
<span class="kw">Tx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/exonsBy.html">transcripts</a></span>(<span class="kw">edb</span>,
          columns = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fu"><a href="../reference/EnsDb-class.html">listColumns</a></span>(<span class="kw">edb</span> , <span class="st">"tx"</span>), <span class="st">"gene_name"</span>),
          filter = <span class="fu">TxBiotypeFilter</span>(<span class="st">"nonsense_mediated_decay"</span>),
          return.type = <span class="st">"DataFrame"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="kw">Tx</span>)
</pre></div>
<pre><code>## [1] 14423</code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit">
<span class="kw">Tx</span>
</pre></div>
<pre><code>## DataFrame with 14423 rows and 9 columns
##                 tx_id             tx_biotype tx_seq_start tx_seq_end
##           &lt;character&gt;            &lt;character&gt;    &lt;integer&gt;  &lt;integer&gt;
## 1     ENST00000567466 nonsense_mediated_de..        47578      49521
## 2     ENST00000397876 nonsense_mediated_de..        53887      57372
## 3     ENST00000428730 nonsense_mediated_de..        58062      65039
## 4     ENST00000417043 nonsense_mediated_de..        62973      65037
## 5     ENST00000622194 nonsense_mediated_de..        85386     138349
## ...               ...                    ...          ...        ...
## 14419 ENST00000496411 nonsense_mediated_de..    248855728  248859018
## 14420 ENST00000483223 nonsense_mediated_de..    248856515  248858529
## 14421 ENST00000533647 nonsense_mediated_de..    248857273  248858324
## 14422 ENST00000528141 nonsense_mediated_de..    248857391  248859085
## 14423 ENST00000530986 nonsense_mediated_de..    248857469  248859085
##       tx_cds_seq_start tx_cds_seq_end         gene_id         tx_name
##              &lt;integer&gt;      &lt;integer&gt;     &lt;character&gt;     &lt;character&gt;
## 1                48546          48893 ENSG00000261456 ENST00000567466
## 2                54017          56360 ENSG00000161981 ENST00000397876
## 3                62884          65015 ENSG00000007384 ENST00000428730
## 4                63904          65015 ENSG00000007384 ENST00000417043
## 5               117330         138267 ENSG00000103148 ENST00000622194
## ...                ...            ...             ...             ...
## 14419        248857954      248858309 ENSG00000171163 ENST00000496411
## 14420        248857954      248858309 ENSG00000171163 ENST00000483223
## 14421        248857954      248858309 ENSG00000171163 ENST00000533647
## 14422        248858004      248858309 ENSG00000171163 ENST00000528141
## 14423        248858004      248858309 ENSG00000171163 ENST00000530986
##         gene_name
##       &lt;character&gt;
## 1           TUBB8
## 2         SNRNP25
## 3          RHBDF1
## 4          RHBDF1
## 5           NPRL3
## ...           ...
## 14419      ZNF692
## 14420      ZNF692
## 14421      ZNF692
## 14422      ZNF692
## 14423      ZNF692</code></pre>
<p>For protein coding transcripts, we can also specifically extract their coding region. In the example below we extract the CDS for all transcripts encoded on chromosome Y.</p>
<div class="sourceCode" id="cb29"><pre class="downlit">
<span class="kw">yCds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/exonsBy.html">cdsBy</a></span>(<span class="kw">edb</span>, filter = <span class="fu">SeqNameFilter</span>(<span class="st">"Y"</span>))
<span class="kw">yCds</span>
</pre></div>
<pre><code>## GRangesList object of length 151:
## $ENST00000155093
## GRanges object with 7 ranges and 3 metadata columns:
##       seqnames          ranges strand |    seq_name         exon_id exon_rank
##          &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;     &lt;character&gt; &lt;integer&gt;
##   [1]        Y 2953937-2953997      + |           Y ENSE00002223884         2
##   [2]        Y 2961074-2961646      + |           Y ENSE00003645989         3
##   [3]        Y 2975095-2975244      + |           Y ENSE00003764421         4
##   [4]        Y 2975511-2975654      + |           Y ENSE00003768468         5
##   [5]        Y 2976670-2976822      + |           Y ENSE00003766362         6
##   [6]        Y 2977940-2978080      + |           Y ENSE00003766086         7
##   [7]        Y 2978810-2979993      + |           Y ENSE00001368923         8
##   -------
##   seqinfo: 1 sequence from GRCh38 genome
## 
## $ENST00000215473
## GRanges object with 2 ranges and 3 metadata columns:
##       seqnames          ranges strand |    seq_name         exon_id exon_rank
##          &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;     &lt;character&gt; &lt;integer&gt;
##   [1]        Y 5056824-5057459      + |           Y ENSE00001436852         1
##   [2]        Y 5098215-5100740      + |           Y ENSE00003741448         2
##   -------
##   seqinfo: 1 sequence from GRCh38 genome
## 
## $ENST00000215479
## GRanges object with 5 ranges and 3 metadata columns:
##       seqnames          ranges strand |    seq_name         exon_id exon_rank
##          &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;     &lt;character&gt; &lt;integer&gt;
##   [1]        Y 6872555-6872608      - |           Y ENSE00001671586         2
##   [2]        Y 6870006-6870053      - |           Y ENSE00001645681         3
##   [3]        Y 6868732-6868776      - |           Y ENSE00000652250         4
##   [4]        Y 6868037-6868462      - |           Y ENSE00001667251         5
##   [5]        Y 6866073-6866078      - |           Y ENSE00001494454         6
##   -------
##   seqinfo: 1 sequence from GRCh38 genome
## 
## ...
## &lt;148 more elements&gt;</code></pre>
<p>Using a <code>GRangesFilter</code> we can retrieve all features from the database that are either within or overlapping the specified genomic region. In the example below we query all genes that are partially overlapping with a small region on chromosome 11. The filter restricts to all genes for which either an exon or an intron is partially overlapping with the region.</p>
<div class="sourceCode" id="cb31"><pre class="downlit">
<span class="co">## Define the filter</span>
<span class="kw">grf</span> <span class="op">&lt;-</span> <span class="fu">GRangesFilter</span>(<span class="fu">GRanges</span>(<span class="st">"11"</span>, ranges = <span class="fu">IRanges</span>(<span class="fl">114129278</span>, <span class="fl">114129328</span>),
                 strand = <span class="st">"+"</span>), type = <span class="st">"any"</span>)

<span class="co">## Query genes:</span>
<span class="kw">gn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/exonsBy.html">genes</a></span>(<span class="kw">edb</span>, filter = <span class="kw">grf</span>)
<span class="kw">gn</span>
</pre></div>
<pre><code>## GRanges object with 1 range and 6 metadata columns:
##                   seqnames              ranges strand |         gene_id
##                      &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
##   ENSG00000109906       11 114059593-114250676      + | ENSG00000109906
##                     gene_name   gene_biotype seq_coord_system      symbol
##                   &lt;character&gt;    &lt;character&gt;      &lt;character&gt; &lt;character&gt;
##   ENSG00000109906      ZBTB16 protein_coding       chromosome      ZBTB16
##                   entrezid
##                     &lt;list&gt;
##   ENSG00000109906     7704
##   -------
##   seqinfo: 1 sequence from GRCh38 genome</code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit">
<span class="co">## Next we retrieve all transcripts for that gene so that we can plot them.</span>
<span class="kw">txs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/exonsBy.html">transcripts</a></span>(<span class="kw">edb</span>, filter = <span class="fu">GeneNameFilter</span>(<span class="kw">gn</span><span class="op">$</span><span class="kw">gene_name</span>))
</pre></div>
<p><img src="ensembldb_files/figure-html/granges-zbtb16-1.png" width="700"></p>
<p>As we can see, 4 transcripts of the gene ZBTB16 are also overlapping the region. Below we fetch these 4 transcripts. Note, that a call to <code>exons</code> will not return any features from the database, as no exon is overlapping with the region.</p>
<div class="sourceCode" id="cb34"><pre class="downlit">
<span class="fu"><a href="../reference/exonsBy.html">transcripts</a></span>(<span class="kw">edb</span>, filter = <span class="kw">grf</span>)
</pre></div>
<pre><code>## GRanges object with 4 ranges and 6 metadata columns:
##                   seqnames              ranges strand |           tx_id
##                      &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
##   ENST00000335953       11 114059593-114250676      + | ENST00000335953
##   ENST00000541602       11 114059725-114189764      + | ENST00000541602
##   ENST00000392996       11 114060507-114250652      + | ENST00000392996
##   ENST00000539918       11 114064412-114247344      + | ENST00000539918
##                               tx_biotype tx_cds_seq_start tx_cds_seq_end
##                              &lt;character&gt;        &lt;integer&gt;      &lt;integer&gt;
##   ENST00000335953         protein_coding        114063301      114250555
##   ENST00000541602        retained_intron             &lt;NA&gt;           &lt;NA&gt;
##   ENST00000392996         protein_coding        114063301      114250555
##   ENST00000539918 nonsense_mediated_de..        114064412      114121827
##                           gene_id         tx_name
##                       &lt;character&gt;     &lt;character&gt;
##   ENST00000335953 ENSG00000109906 ENST00000335953
##   ENST00000541602 ENSG00000109906 ENST00000541602
##   ENST00000392996 ENSG00000109906 ENST00000392996
##   ENST00000539918 ENSG00000109906 ENST00000539918
##   -------
##   seqinfo: 1 sequence from GRCh38 genome</code></pre>
<p>The <code>GRangesFilter</code> supports also <code>GRanges</code> defining multiple regions and a query will return all features overlapping any of these regions. Besides using the <code>GRangesFilter</code> it is also possible to search for transcripts or exons overlapping genomic regions using the <code>exonsByOverlaps</code> or <code>transcriptsByOverlaps</code> known from the <code>GenomicFeatures</code> package. Note that the implementation of these methods for <code>EnsDb</code> objects supports also to use filters to further fine-tune the query.</p>
<p>The functions <code>listGenebiotypes</code> and <code>listTxbiotypes</code> can be used to get an overview of allowed/available gene and transcript biotype</p>
<div class="sourceCode" id="cb36"><pre class="downlit">
<span class="co">## Get all gene biotypes from the database. The GeneBiotypeFilter</span>
<span class="co">## allows to filter on these values.</span>
<span class="fu"><a href="../reference/EnsDb-class.html">listGenebiotypes</a></span>(<span class="kw">edb</span>)
</pre></div>
<pre><code>##  [1] "protein_coding"                     "unitary_pseudogene"                
##  [3] "unprocessed_pseudogene"             "processed_pseudogene"              
##  [5] "processed_transcript"               "transcribed_unprocessed_pseudogene"
##  [7] "antisense"                          "transcribed_unitary_pseudogene"    
##  [9] "polymorphic_pseudogene"             "lincRNA"                           
## [11] "sense_intronic"                     "transcribed_processed_pseudogene"  
## [13] "sense_overlapping"                  "IG_V_pseudogene"                   
## [15] "pseudogene"                         "TR_V_gene"                         
## [17] "3prime_overlapping_ncRNA"           "IG_V_gene"                         
## [19] "bidirectional_promoter_lncRNA"      "snRNA"                             
## [21] "miRNA"                              "misc_RNA"                          
## [23] "snoRNA"                             "rRNA"                              
## [25] "Mt_tRNA"                            "Mt_rRNA"                           
## [27] "IG_C_gene"                          "IG_J_gene"                         
## [29] "TR_J_gene"                          "TR_C_gene"                         
## [31] "TR_V_pseudogene"                    "TR_J_pseudogene"                   
## [33] "IG_D_gene"                          "ribozyme"                          
## [35] "IG_C_pseudogene"                    "TR_D_gene"                         
## [37] "TEC"                                "IG_J_pseudogene"                   
## [39] "scRNA"                              "scaRNA"                            
## [41] "vaultRNA"                           "sRNA"                              
## [43] "macro_lncRNA"                       "non_coding"                        
## [45] "IG_pseudogene"                      "LRG_gene"</code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit">
<span class="co">## Get all transcript biotypes from the database.</span>
<span class="fu"><a href="../reference/EnsDb-class.html">listTxbiotypes</a></span>(<span class="kw">edb</span>)
</pre></div>
<pre><code>##  [1] "protein_coding"                     "processed_transcript"              
##  [3] "nonsense_mediated_decay"            "retained_intron"                   
##  [5] "unitary_pseudogene"                 "TEC"                               
##  [7] "miRNA"                              "misc_RNA"                          
##  [9] "non_stop_decay"                     "unprocessed_pseudogene"            
## [11] "processed_pseudogene"               "transcribed_unprocessed_pseudogene"
## [13] "lincRNA"                            "antisense"                         
## [15] "transcribed_unitary_pseudogene"     "polymorphic_pseudogene"            
## [17] "sense_intronic"                     "transcribed_processed_pseudogene"  
## [19] "sense_overlapping"                  "IG_V_pseudogene"                   
## [21] "pseudogene"                         "TR_V_gene"                         
## [23] "3prime_overlapping_ncRNA"           "IG_V_gene"                         
## [25] "bidirectional_promoter_lncRNA"      "snRNA"                             
## [27] "snoRNA"                             "rRNA"                              
## [29] "Mt_tRNA"                            "Mt_rRNA"                           
## [31] "IG_C_gene"                          "IG_J_gene"                         
## [33] "TR_J_gene"                          "TR_C_gene"                         
## [35] "TR_V_pseudogene"                    "TR_J_pseudogene"                   
## [37] "IG_D_gene"                          "ribozyme"                          
## [39] "IG_C_pseudogene"                    "TR_D_gene"                         
## [41] "IG_J_pseudogene"                    "scRNA"                             
## [43] "scaRNA"                             "vaultRNA"                          
## [45] "sRNA"                               "macro_lncRNA"                      
## [47] "non_coding"                         "IG_pseudogene"                     
## [49] "LRG_gene"</code></pre>
<p>Data can be fetched in an analogous way using the <code>exons</code> and <code>genes</code> methods. In the example below we retrieve <code>gene_name</code>, <code>entrezid</code> and the <code>gene_biotype</code> of all genes in the database which names start with “BCL2”.</p>
<div class="sourceCode" id="cb40"><pre class="downlit">
<span class="co">## We're going to fetch all genes which names start with BCL.</span>
<span class="kw">BCLs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/exonsBy.html">genes</a></span>(<span class="kw">edb</span>,
          columns = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"gene_name"</span>, <span class="st">"entrezid"</span>, <span class="st">"gene_biotype"</span>),
          filter = <span class="fu">GeneNameFilter</span>(<span class="st">"BCL"</span>, condition = <span class="st">"startsWith"</span>),
          return.type = <span class="st">"DataFrame"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="kw">BCLs</span>)
</pre></div>
<pre><code>## [1] 30</code></pre>
<div class="sourceCode" id="cb42"><pre class="downlit">
<span class="kw">BCLs</span>
</pre></div>
<pre><code>## DataFrame with 30 rows and 4 columns
##       gene_name entrezid         gene_biotype         gene_id
##     &lt;character&gt;   &lt;list&gt;          &lt;character&gt;     &lt;character&gt;
## 1         BCL10     8915       protein_coding ENSG00000142867
## 2        BCL11A    53335       protein_coding ENSG00000119866
## 3        BCL11B    64919       protein_coding ENSG00000127152
## 4          BCL2      596       protein_coding ENSG00000171791
## 5        BCL2A1      597       protein_coding ENSG00000140379
## ...         ...      ...                  ...             ...
## 26        BCL9L   283149       protein_coding ENSG00000186174
## 27       BCL9P1       NA processed_pseudogene ENSG00000249238
## 28       BCLAF1     9774       protein_coding ENSG00000029363
## 29     BCLAF1P1       NA processed_pseudogene ENSG00000248966
## 30     BCLAF1P2       NA processed_pseudogene ENSG00000279800</code></pre>
<p>Sometimes it might be useful to know the length of genes or transcripts (i.e. the total sum of nucleotides covered by their exons). Below we calculate the mean length of transcripts from protein coding genes on chromosomes X and Y as well as the average length of snoRNA, snRNA and rRNA transcripts encoded on these chromosomes. For the first query we combine two <code>AnnotationFilter</code> objects using an <code>AnnotationFilterList</code> object, in the second we define the query using a filter expression.</p>
<div class="sourceCode" id="cb44"><pre class="downlit">
<span class="co">## determine the average length of snRNA, snoRNA and rRNA genes encoded on</span>
<span class="co">## chromosomes X and Y.</span>
<span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(<span class="fu"><a href="../reference/lengthOf.html">lengthOf</a></span>(<span class="kw">edb</span>, of = <span class="st">"tx"</span>,
          filter = <span class="fu">AnnotationFilterList</span>(
          <span class="fu">GeneBiotypeFilter</span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"snRNA"</span>, <span class="st">"snoRNA"</span>, <span class="st">"rRNA"</span>)),
          <span class="fu">SeqNameFilter</span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"X"</span>, <span class="st">"Y"</span>)))))
</pre></div>
<pre><code>## [1] 118.2458</code></pre>
<div class="sourceCode" id="cb46"><pre class="downlit">
<span class="co">## determine the average length of protein coding genes encoded on the same</span>
<span class="co">## chromosomes.</span>
<span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(<span class="fu"><a href="../reference/lengthOf.html">lengthOf</a></span>(<span class="kw">edb</span>, of = <span class="st">"tx"</span>,
          filter = <span class="op">~</span> <span class="kw">gene_biotype</span> <span class="op">==</span> <span class="st">"protein_coding"</span> <span class="op">&amp;</span>
          <span class="kw">seq_name</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"X"</span>, <span class="st">"Y"</span>)))
</pre></div>
<pre><code>## [1] 1943.554</code></pre>
<p>Not unexpectedly, transcripts of protein coding genes are longer than those of snRNA, snoRNA or rRNA genes.</p>
<p>At last we extract the first two exons of each transcript model from the database.</p>
<div class="sourceCode" id="cb48"><pre class="downlit">
<span class="co">## Extract all exons 1 and (if present) 2 for all genes encoded on the</span>
<span class="co">## Y chromosome</span>
<span class="fu"><a href="../reference/exonsBy.html">exons</a></span>(<span class="kw">edb</span>, columns = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"tx_id"</span>, <span class="st">"exon_idx"</span>),
      filter = <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="fu">SeqNameFilter</span>(<span class="st">"Y"</span>),
                    <span class="fu">ExonRankFilter</span>(<span class="fl">3</span>, condition = <span class="st">"&lt;"</span>)))
</pre></div>
<pre><code>## GRanges object with 1294 ranges and 3 metadata columns:
##                   seqnames            ranges strand |           tx_id  exon_idx
##                      &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt; &lt;integer&gt;
##   ENSE00002088309        Y   2784749-2784853      + | ENST00000516032         1
##   ENSE00001494622        Y   2786855-2787699      - | ENST00000383070         1
##   ENSE00001772499        Y   2789827-2790328      + | ENST00000454281         1
##   ENSE00001614266        Y   2827982-2828218      + | ENST00000430735         1
##   ENSE00002490412        Y   2841486-2841627      + | ENST00000250784         1
##               ...      ...               ...    ... .             ...       ...
##   ENSE00001632993        Y 26591548-26591601      - | ENST00000456738         1
##   ENSE00001616687        Y 26626520-26627159      - | ENST00000435741         1
##   ENSE00001638296        Y 26633345-26633431      - | ENST00000435945         2
##   ENSE00001797328        Y 26634523-26634652      - | ENST00000435945         1
##   ENSE00001794473        Y 56855244-56855488      + | ENST00000431853         1
##                           exon_id
##                       &lt;character&gt;
##   ENSE00002088309 ENSE00002088309
##   ENSE00001494622 ENSE00001494622
##   ENSE00001772499 ENSE00001772499
##   ENSE00001614266 ENSE00001614266
##   ENSE00002490412 ENSE00002490412
##               ...             ...
##   ENSE00001632993 ENSE00001632993
##   ENSE00001616687 ENSE00001616687
##   ENSE00001638296 ENSE00001638296
##   ENSE00001797328 ENSE00001797328
##   ENSE00001794473 ENSE00001794473
##   -------
##   seqinfo: 1 sequence from GRCh38 genome</code></pre>
</div>
<div id="extracting-genetranscriptexon-models-for-rnaseq-feature-counting" class="section level1">
<h1 class="hasAnchor">
<a href="#extracting-genetranscriptexon-models-for-rnaseq-feature-counting" class="anchor"></a>Extracting gene/transcript/exon models for RNASeq feature counting</h1>
<p>For the feature counting step of an RNAseq experiment, the gene or transcript models (defined by the chromosomal start and end positions of their exons) have to be known. To extract these from an Ensembl based annotation package, the <code>exonsBy</code>, <code>genesBy</code> and <code>transcriptsBy</code> methods can be used in an analogous way as in <code>TxDb</code> packages generated by the <code>GenomicFeatures</code> package. However, the <code>transcriptsBy</code> method does not, in contrast to the method in the <code>GenomicFeatures</code> package, allow to return transcripts by “cds”. While the annotation packages built by the <code>ensembldb</code> contain the chromosomal start and end coordinates of the coding region (for protein coding genes) they do not assign an ID to each CDS.</p>
<p>A simple use case is to retrieve all genes encoded on chromosomes X and Y from the database.</p>
<div class="sourceCode" id="cb50"><pre class="downlit">
<span class="kw">TxByGns</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/exonsBy.html">transcriptsBy</a></span>(<span class="kw">edb</span>, by = <span class="st">"gene"</span>, filter = <span class="fu">SeqNameFilter</span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"X"</span>, <span class="st">"Y"</span>)))
<span class="kw">TxByGns</span>
</pre></div>
<pre><code>## GRangesList object of length 2922:
## $ENSG00000000003
## GRanges object with 5 ranges and 6 metadata columns:
##       seqnames              ranges strand |           tx_id
##          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
##   [1]        X 100633442-100639991      - | ENST00000494424
##   [2]        X 100627109-100637104      - | ENST00000612152
##   [3]        X 100632063-100637104      - | ENST00000614008
##   [4]        X 100628670-100636806      - | ENST00000373020
##   [5]        X 100632541-100636689      - | ENST00000496771
##                 tx_biotype tx_cds_seq_start tx_cds_seq_end         gene_id
##                &lt;character&gt;        &lt;integer&gt;      &lt;integer&gt;     &lt;character&gt;
##   [1] processed_transcript             &lt;NA&gt;           &lt;NA&gt; ENSG00000000003
##   [2]       protein_coding        100630798      100635569 ENSG00000000003
##   [3]       protein_coding        100632063      100635569 ENSG00000000003
##   [4]       protein_coding        100630798      100636694 ENSG00000000003
##   [5] processed_transcript             &lt;NA&gt;           &lt;NA&gt; ENSG00000000003
##               tx_name
##           &lt;character&gt;
##   [1] ENST00000494424
##   [2] ENST00000612152
##   [3] ENST00000614008
##   [4] ENST00000373020
##   [5] ENST00000496771
##   -------
##   seqinfo: 2 sequences from GRCh38 genome
## 
## $ENSG00000000005
## GRanges object with 2 ranges and 6 metadata columns:
##       seqnames              ranges strand |           tx_id
##          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
##   [1]        X 100584802-100599885      + | ENST00000373031
##   [2]        X 100593624-100597531      + | ENST00000485971
##                 tx_biotype tx_cds_seq_start tx_cds_seq_end         gene_id
##                &lt;character&gt;        &lt;integer&gt;      &lt;integer&gt;     &lt;character&gt;
##   [1]       protein_coding        100585019      100599717 ENSG00000000005
##   [2] processed_transcript             &lt;NA&gt;           &lt;NA&gt; ENSG00000000005
##               tx_name
##           &lt;character&gt;
##   [1] ENST00000373031
##   [2] ENST00000485971
##   -------
##   seqinfo: 2 sequences from GRCh38 genome
## 
## $ENSG00000001497
## GRanges object with 5 ranges and 6 metadata columns:
##       seqnames            ranges strand |           tx_id
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
##   [1]        X 65512583-65534775      - | ENST00000484069
##   [2]        X 65512582-65534756      - | ENST00000374811
##   [3]        X 65512583-65534756      - | ENST00000374804
##   [4]        X 65512582-65534754      - | ENST00000374807
##   [5]        X 65520429-65523617      - | ENST00000469091
##                   tx_biotype tx_cds_seq_start tx_cds_seq_end         gene_id
##                  &lt;character&gt;        &lt;integer&gt;      &lt;integer&gt;     &lt;character&gt;
##   [1] nonsense_mediated_de..         65525021       65534715 ENSG00000001497
##   [2]         protein_coding         65512775       65534715 ENSG00000001497
##   [3]         protein_coding         65512775       65534715 ENSG00000001497
##   [4]         protein_coding         65512775       65534715 ENSG00000001497
##   [5]         protein_coding         65520655       65523617 ENSG00000001497
##               tx_name
##           &lt;character&gt;
##   [1] ENST00000484069
##   [2] ENST00000374811
##   [3] ENST00000374804
##   [4] ENST00000374807
##   [5] ENST00000469091
##   -------
##   seqinfo: 2 sequences from GRCh38 genome
## 
## ...
## &lt;2919 more elements&gt;</code></pre>
<p>Since Ensembl contains also definitions of genes that are on chromosome variants (supercontigs), it is advisable to specify the chromosome names for which the gene models should be returned.</p>
<p>In a real use case, we might thus want to retrieve all genes encoded on the <em>standard</em> chromosomes. In addition it is advisable to use a <code>GeneIdFilter</code> to restrict to Ensembl genes only, as also <em>LRG</em> (Locus Reference Genomic) genes<sup><a id="fnr.2" class="footref" href="#fn.2">2</a></sup> are defined in the database, which are partially redundant with Ensembl genes.</p>
<div class="sourceCode" id="cb52"><pre class="downlit">
<span class="co">## will just get exons for all genes on chromosomes 1 to 22, X and Y.</span>
<span class="co">## Note: want to get rid of the "LRG" genes!!!</span>
<span class="kw">EnsGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/exonsBy.html">exonsBy</a></span>(<span class="kw">edb</span>, by = <span class="st">"gene"</span>, filter = <span class="fu">AnnotationFilterList</span>(
                      <span class="fu">SeqNameFilter</span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span><span class="op">:</span><span class="fl">22</span>, <span class="st">"X"</span>, <span class="st">"Y"</span>)),
                      <span class="fu">GeneIdFilter</span>(<span class="st">"ENSG"</span>, <span class="st">"startsWith"</span>)))
</pre></div>
<p>The code above returns a <code>GRangesList</code> that can be used directly as an input for the <code>summarizeOverlaps</code> function from the <code>GenomicAlignments</code> package <sup><a id="fnr.3" class="footref" href="#fn.3">3</a></sup>.</p>
<p>Alternatively, the above <code>GRangesList</code> can be transformed to a <code>data.frame</code> in <em>SAF</em> format that can be used as an input to the <code>featureCounts</code> function of the <code>Rsubread</code> package <sup><a id="fnr.4" class="footref" href="#fn.4">4</a></sup>.</p>
<div class="sourceCode" id="cb53"><pre class="downlit">
<span class="co">## Transforming the GRangesList into a data.frame in SAF format</span>
<span class="kw">EnsGenes.SAF</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/exonsBy.html">toSAF</a></span>(<span class="kw">EnsGenes</span>)
</pre></div>
<p>Note that the ID by which the <code>GRangesList</code> is split is used in the SAF formatted <code>data.frame</code> as the <code>GeneID</code>. In the example below this would be the Ensembl gene IDs, while the start, end coordinates (along with the strand and chromosomes) are those of the the exons.</p>
<p>In addition, the <code>disjointExons</code> function (similar to the one defined in <code>GenomicFeatures</code>) can be used to generate a <code>GRanges</code> of non-overlapping exon parts which can be used in the <code>DEXSeq</code> package.</p>
<div class="sourceCode" id="cb54"><pre class="downlit">
<span class="co">## Create a GRanges of non-overlapping exon parts.</span>
<span class="kw">DJE</span> <span class="op">&lt;-</span> <span class="fu">disjointExons</span>(<span class="kw">edb</span>, filter = <span class="fu">AnnotationFilterList</span>(
                  <span class="fu">SeqNameFilter</span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span><span class="op">:</span><span class="fl">22</span>, <span class="st">"X"</span>, <span class="st">"Y"</span>)),
                  <span class="fu">GeneIdFilter</span>(<span class="st">"ENSG%"</span>, <span class="st">"startsWith"</span>)))
</pre></div>
</div>
<div id="retrieving-sequences-for-genetranscriptexon-models" class="section level1">
<h1 class="hasAnchor">
<a href="#retrieving-sequences-for-genetranscriptexon-models" class="anchor"></a>Retrieving sequences for gene/transcript/exon models</h1>
<p>The methods to retrieve exons, transcripts and genes (i.e. <code>exons</code>, <code>transcripts</code> and <code>genes</code>) return by default <code>GRanges</code> objects that can be used to retrieve sequences using the <code>getSeq</code> method e.g. from BSgenome packages. The basic workflow is thus identical to the one for <code>TxDb</code> packages, however, it is not straight forward to identify the BSgenome package with the matching genomic sequence. Most BSgenome packages are named according to the genome build identifier used in UCSC which does not (always) match the genome build name used by Ensembl. Using the Ensembl version provided by the <code>EnsDb</code>, the correct genomic sequence can however be retrieved easily from the <code>AnnotationHub</code> using the <code>getGenomeTeoBitFile</code>. If no 2bit file matching the Ensembl version is available, the function tries to identify a file with the correct genome build from the <em>closest</em> Ensembl release and returns that instead.</p>
<p>In the code block below we retrieve first the <code>TwoBitFile</code> with the genomic DNA sequence, extract the genomic start and end coordinates for all genes defined in the package, subset to genes encoded on sequences available in the <code>TwoBitFile</code> and extract all of their sequences. Note: these sequences represent the sequence between the chromosomal start and end coordinates of the gene.</p>
<div class="sourceCode" id="cb55"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">EnsDb.Hsapiens.v86</span>)
<span class="kw">edb</span> <span class="op">&lt;-</span> <span class="kw">EnsDb.Hsapiens.v86</span>

<span class="co">## Get the TwoBit with the genomic sequence matching the Ensembl version</span>
<span class="co">## using the AnnotationHub package.</span>
<span class="kw">dna</span> <span class="op">&lt;-</span> <span class="kw">ensembldb</span>:::<span class="fu"><a href="../reference/getGenomeFaFile.html">getGenomeTwoBitFile</a></span>(<span class="kw">edb</span>)

<span class="co">## Get start/end coordinates of all genes.</span>
<span class="kw">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/exonsBy.html">genes</a></span>(<span class="kw">edb</span>)
<span class="co">## Subset to all genes that are encoded on chromosomes for which</span>
<span class="co">## we do have DNA sequence available.</span>
<span class="kw">genes</span> <span class="op">&lt;-</span> <span class="kw">genes</span>[<span class="fu">seqnames</span>(<span class="kw">genes</span>) <span class="op">%in%</span> <span class="fu">seqnames</span>(<span class="fu"><a href="../reference/EnsDb-class.html">seqinfo</a></span>(<span class="kw">dna</span>))]

<span class="co">## Get the gene sequences, i.e. the sequence including the sequence of</span>
<span class="co">## all of the gene's exons and introns.</span>
<span class="kw">geneSeqs</span> <span class="op">&lt;-</span> <span class="fu">getSeq</span>(<span class="kw">dna</span>, <span class="kw">genes</span>)
</pre></div>
<p>To retrieve the (exonic) sequence of transcripts (i.e. without introns) we can use directly the <code>extractTranscriptSeqs</code> method defined in the <code>GenomicFeatures</code> on the <code>EnsDb</code> object, eventually using a filter to restrict the query.</p>
<div class="sourceCode" id="cb56"><pre class="downlit">
<span class="co">## get all exons of all transcripts encoded on chromosome Y</span>
<span class="kw">yTx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/exonsBy.html">exonsBy</a></span>(<span class="kw">edb</span>, filter = <span class="fu">SeqNameFilter</span>(<span class="st">"Y"</span>))

<span class="co">## Retrieve the sequences for these transcripts from the TwoBitile.</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">GenomicFeatures</span>)
<span class="kw">yTxSeqs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/extractTranscriptSeqs.html">extractTranscriptSeqs</a></span>(<span class="kw">dna</span>, <span class="kw">yTx</span>)
<span class="kw">yTxSeqs</span>

<span class="co">## Extract the sequences of all transcripts encoded on chromosome Y.</span>
<span class="kw">yTx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/extractTranscriptSeqs.html">extractTranscriptSeqs</a></span>(<span class="kw">dna</span>, <span class="kw">edb</span>, filter = <span class="fu">SeqNameFilter</span>(<span class="st">"Y"</span>))

<span class="co">## Along these lines, we could use the method also to retrieve the coding sequence</span>
<span class="co">## of all transcripts on the Y chromosome.</span>
<span class="kw">cdsY</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/exonsBy.html">cdsBy</a></span>(<span class="kw">edb</span>, filter = <span class="fu">SeqNameFilter</span>(<span class="st">"Y"</span>))
<span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/extractTranscriptSeqs.html">extractTranscriptSeqs</a></span>(<span class="kw">dna</span>, <span class="kw">cdsY</span>)
</pre></div>
<p>Next we retrieve transcript sequences from genes encoded on chromosome Y using the <code>BSGenome</code> package for the human genome. Ensembl version 86 based on the <code>GRCh38</code> genome build and we thus load the corresponding <code>BSGenome</code> package.</p>
<div class="sourceCode" id="cb57"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">BSgenome.Hsapiens.NCBI.GRCh38</span>)
<span class="kw">bsg</span> <span class="op">&lt;-</span> <span class="kw">BSgenome.Hsapiens.NCBI.GRCh38</span>

<span class="co">## Get the genome version</span>
<span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="fu">genome</span>(<span class="kw">bsg</span>))
</pre></div>
<pre><code>## [1] "GRCh38"</code></pre>
<div class="sourceCode" id="cb59"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="fu">genome</span>(<span class="kw">edb</span>))
</pre></div>
<pre><code>## [1] "GRCh38"</code></pre>
<div class="sourceCode" id="cb61"><pre class="downlit">
<span class="co">## Extract the full transcript sequences.</span>
<span class="kw">yTxSeqs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/extractTranscriptSeqs.html">extractTranscriptSeqs</a></span>(
  <span class="kw">bsg</span>, <span class="fu"><a href="../reference/exonsBy.html">exonsBy</a></span>(<span class="kw">edb</span>, <span class="st">"tx"</span>, filter = <span class="fu">SeqNameFilter</span>(<span class="st">"Y"</span>)))

<span class="kw">yTxSeqs</span>
</pre></div>
<pre><code>## DNAStringSet object of length 740:
##       width seq                                             names               
##   [1]  5239 GCCTAGTGCGCGCGCAGTAACC...AATAAATGTTTACTTGTATATG ENST00000155093
##   [2]  4595 CTGGTGGTCCAGTACCTCCAAA...TGAGCCCTTCAGAAGACATTCT ENST00000215473
##   [3]   802 AGAGGACCAAGCCTCCCTGTGT...CAATAAAATGTTTTAAAAATCA ENST00000215479
##   [4]   910 TGTCTGTCAGAGCTGTCAGCCT...TAAACACTGGTATATTTCTGTT ENST00000250776
##   [5]  1305 TTCCAGGATATGAACTCTACAG...TAAATCCTGTGGCTGTAGGAAA ENST00000250784
##   ...   ... ...
## [736]   792 ATGGCCCGGGGCCCCAAGAAGC...TGCCAAACAGAGCAGTGGCTAA ENST00000629237
## [737]   344 GGTTGCCACTTCAAGGGACTAC...CTGGCTCTTCTGGCAGTTTTTT ENST00000631331
## [738]   933 CTCTCCCAGCTTCTACCCACAG...GCATACTATAAAAATGCTTTAA ENST00000634531
## [739]  1832 ATGTCTGCTGCAAATCCTGAGA...AGTATTTAAATCTGTTGGATCC ENST00000634662
## [740]   890 CTCTCCCAGCTTCTACCCACAG...GCATACTATAAAAATGCTTTAA ENST00000635343</code></pre>
<div class="sourceCode" id="cb63"><pre class="downlit">
<span class="co">## Extract just the CDS</span>
<span class="kw">Test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/exonsBy.html">cdsBy</a></span>(<span class="kw">edb</span>, <span class="st">"tx"</span>, filter = <span class="fu">SeqNameFilter</span>(<span class="st">"Y"</span>))
<span class="kw">yTxCds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/extractTranscriptSeqs.html">extractTranscriptSeqs</a></span>(
  <span class="kw">bsg</span>, <span class="fu"><a href="../reference/exonsBy.html">cdsBy</a></span>(<span class="kw">edb</span>, <span class="st">"tx"</span>, filter = <span class="fu">SeqNameFilter</span>(<span class="st">"Y"</span>)))
<span class="kw">yTxCds</span>
</pre></div>
<pre><code>## DNAStringSet object of length 151:
##       width seq                                             names               
##   [1]  2406 ATGGATGAAGATGAATTTGAAT...TAAAGAAGTTGGTCTGCCCTAA ENST00000155093
##   [2]  3162 ATGTTTAGGGTTGGCTTCTTAA...AGTTTCTAACACAACTTTCTAA ENST00000215473
##   [3]   579 ATGGGGACCTGGATTTTGTTTG...CAAGCAGGAGGAAGTGGATTAA ENST00000215479
##   [4]   792 ATGGCCCGGGGCCCCAAGAAGC...CACCAAACAGAGCAGTGGCTAA ENST00000250784
##   [5]   378 ATGAGTCCAAAGCCGAGAGCCT...ATCTACTCCCCTATCTCCCTGA ENST00000250823
##   ...   ... ...
## [147]   387 ATGCAAAGCCAGAGAGGTCTCC...CACACTCTGTGTCCCAAAATGA ENST00000624507
## [148]    78 ATGAGAGCCAAGTGGAGGAAGA...GATGAGGCAGAAGTCCAAGTAA ENST00000624575
## [149]  1833 ATGGATGAAGATGAATTTGAAT...TAAAGAAGTTGGTCTGCCCTAA ENST00000625061
## [150]   792 ATGGCCCGGGGCCCCAAGAAGC...TGCCAAACAGAGCAGTGGCTAA ENST00000629237
## [151]  1740 ATGTCTGCTGCAAATCCTGAGA...TTTAATCCAGAGAAGAGACTGA ENST00000634662</code></pre>
</div>
<div id="integrating-annotations-from-ensembl-based-ensdb-packages-with-ucsc-based-annotations" class="section level1">
<h1 class="hasAnchor">
<a href="#integrating-annotations-from-ensembl-based-ensdb-packages-with-ucsc-based-annotations" class="anchor"></a>Integrating annotations from Ensembl based <code>EnsDb</code> packages with UCSC based annotations</h1>
<p>Sometimes it might be useful to combine (Ensembl based) annotations from <code>EnsDb</code> packages/objects with annotations from other Bioconductor packages, that might base on UCSC annotations. To support such an integration of annotations, the <code>ensembldb</code> packages implements the <code>seqlevelsStyle</code> and <code>seqlevelsStyle&lt;-</code> from the <code>GenomeInfoDb</code> package that allow to change the style of chromosome naming. Thus, sequence/chromosome names other than those used by Ensembl can be used in, and are returned by, the queries to <code>EnsDb</code> objects as long as a mapping for them is provided by the <code>GenomeInfoDb</code> package (which provides a mapping mostly between UCSC, NCBI and Ensembl chromosome names for the <em>main</em> chromosomes).</p>
<p>In the example below we change the seqnames style to UCSC.</p>
<div class="sourceCode" id="cb65"><pre class="downlit">
<span class="co">## Change the seqlevels style form Ensembl (default) to UCSC:</span>
<span class="fu"><a href="../reference/seqlevelsStyle.html">seqlevelsStyle</a></span>(<span class="kw">edb</span>) <span class="op">&lt;-</span> <span class="st">"UCSC"</span>

<span class="co">## Now we can use UCSC style seqnames in SeqNameFilters or GRangesFilter:</span>
<span class="kw">genesY</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/exonsBy.html">genes</a></span>(<span class="kw">edb</span>, filter = <span class="op">~</span> <span class="kw">seq_name</span> <span class="op">==</span> <span class="st">"chrY"</span>)
<span class="co">## The seqlevels of the returned GRanges are also in UCSC style</span>
<span class="fu"><a href="../reference/EnsDb-class.html">seqlevels</a></span>(<span class="kw">genesY</span>)
</pre></div>
<pre><code>## [1] "chrY"</code></pre>
<p>Note that in most instances no mapping is available for sequences not corresponding to the main chromosomes (i.e. contigs, patched chromosomes etc). What is returned in cases in which no mapping is available can be specified with the global <code>ensembldb.seqnameNotFound</code> option. By default (with <code>ensembldb.seqnameNotFound</code> set to “ORIGINAL”), the original seqnames (i.e. the ones from Ensembl) are returned. With <code>ensembldb.seqnameNotFound</code> “MISSING” each time a seqname can not be found an error is thrown. For all other cases (e.g. <code>ensembldb.seqnameNotFound = NA</code>) the value of the option is returned.</p>
<div class="sourceCode" id="cb67"><pre class="downlit">
<span class="fu"><a href="../reference/seqlevelsStyle.html">seqlevelsStyle</a></span>(<span class="kw">edb</span>) <span class="op">&lt;-</span> <span class="st">"UCSC"</span>

<span class="co">## Getting the default option:</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">getOption</a></span>(<span class="st">"ensembldb.seqnameNotFound"</span>)
</pre></div>
<pre><code>## [1] "ORIGINAL"</code></pre>
<div class="sourceCode" id="cb69"><pre class="downlit">
<span class="co">## Listing all seqlevels in the database.</span>
<span class="fu"><a href="../reference/EnsDb-class.html">seqlevels</a></span>(<span class="kw">edb</span>)[<span class="fl">1</span><span class="op">:</span><span class="fl">30</span>]
</pre></div>
<pre><code>## Warning in .formatSeqnameByStyleFromQuery(x, sn, ifNotFound): More than 5
## seqnames with seqlevels style of the database (Ensembl) could not be mapped to
## the seqlevels style: UCSC!) Returning the orginal seqnames for these.</code></pre>
<pre><code>##  [1] "chr1"                       "chr10"                     
##  [3] "chr11"                      "chr12"                     
##  [5] "chr13"                      "chr14"                     
##  [7] "chr15"                      "chr16"                     
##  [9] "chr17"                      "chr18"                     
## [11] "chr19"                      "chr2"                      
## [13] "chr20"                      "chr21"                     
## [15] "chr22"                      "chr3"                      
## [17] "chr4"                       "chr5"                      
## [19] "chr6"                       "chr7"                      
## [21] "chr8"                       "chr9"                      
## [23] "CHR_HG107_PATCH"            "CHR_HG126_PATCH"           
## [25] "CHR_HG1311_PATCH"           "CHR_HG1342_HG2282_PATCH"   
## [27] "CHR_HG1362_PATCH"           "CHR_HG142_HG150_NOVEL_TEST"
## [29] "CHR_HG151_NOVEL_TEST"       "CHR_HG1651_PATCH"</code></pre>
<div class="sourceCode" id="cb72"><pre class="downlit">
<span class="co">## Setting the option to NA, thus, for each seqname for which no mapping is available,</span>
<span class="co">## NA is returned.</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(ensembldb.seqnameNotFound=<span class="fl">NA</span>)
<span class="fu"><a href="../reference/EnsDb-class.html">seqlevels</a></span>(<span class="kw">edb</span>)[<span class="fl">1</span><span class="op">:</span><span class="fl">30</span>]
</pre></div>
<pre><code>## Warning in .formatSeqnameByStyleFromQuery(x, sn, ifNotFound): More than 5
## seqnames with seqlevels style of the database (Ensembl) could not be mapped to
## the seqlevels style: UCSC!) Returning NA for these.</code></pre>
<pre><code>##  [1] "chr1"  "chr10" "chr11" "chr12" "chr13" "chr14" "chr15" "chr16" "chr17"
## [10] "chr18" "chr19" "chr2"  "chr20" "chr21" "chr22" "chr3"  "chr4"  "chr5" 
## [19] "chr6"  "chr7"  "chr8"  "chr9"  NA      NA      NA      NA      NA     
## [28] NA      NA      NA</code></pre>
<div class="sourceCode" id="cb75"><pre class="downlit">
<span class="co">## Resetting the option.</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(ensembldb.seqnameNotFound = <span class="st">"ORIGINAL"</span>)
</pre></div>
<p>At last changing the seqname style to the default value <code>"Ensembl"</code>.</p>
<div class="sourceCode" id="cb76"><pre class="downlit">
<span class="fu"><a href="../reference/seqlevelsStyle.html">seqlevelsStyle</a></span>(<span class="kw">edb</span>) <span class="op">&lt;-</span> <span class="st">"Ensembl"</span>
</pre></div>
</div>
<div id="interactive-annotation-lookup-using-the-shiny-web-app" class="section level1">
<h1 class="hasAnchor">
<a href="#interactive-annotation-lookup-using-the-shiny-web-app" class="anchor"></a>Interactive annotation lookup using the <code>shiny</code> web app</h1>
<p>In addition to the <code>genes</code>, <code>transcripts</code> and <code>exons</code> methods it is possibly to search interactively for gene/transcript/exon annotations using the internal, <code>shiny</code> based, web application. The application can be started with the <code><a href="../reference/runEnsDbApp.html">runEnsDbApp()</a></code> function. The search results from this app can also be returned to the R workspace either as a <code>data.frame</code> or <code>GRanges</code> object.</p>
</div>
<div id="plotting-genetranscript-features-using-ensembldb-and-gviz-and-ggbio" class="section level1">
<h1 class="hasAnchor">
<a href="#plotting-genetranscript-features-using-ensembldb-and-gviz-and-ggbio" class="anchor"></a>Plotting gene/transcript features using <code>ensembldb</code> and <code>Gviz</code> and <code>ggbio</code>
</h1>
<p>The <code>Gviz</code> package provides functions to plot genes and transcripts along with other data on a genomic scale. Gene models can be provided either as a <code>data.frame</code>, <code>GRanges</code>, <code>TxDB</code> database, can be fetched from biomart and can also be retrieved from <code>ensembldb</code>.</p>
<p>Below we generate a <code>GeneRegionTrack</code> fetching all transcripts from a certain region on chromosome Y.</p>
<p>Note that if we want in addition to work also with BAM files that were aligned against DNA sequences retrieved from Ensembl or FASTA files representing genomic DNA sequences from Ensembl we should change the <code>ucscChromosomeNames</code> option from <code>Gviz</code> to <code>FALSE</code> (i.e. by calling <code><a href="https://rdrr.io/r/base/options.html">options(ucscChromosomeNames = FALSE)</a></code>). This is not necessary if we just want to retrieve gene models from an <code>EnsDb</code> object, as the <code>ensembldb</code> package internally checks the <code>ucscChromosomeNames</code> option and, depending on that, maps Ensembl chromosome names to UCSC chromosome names.</p>
<div class="sourceCode" id="cb77"><pre class="downlit">
<span class="co">## Loading the Gviz library</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://github.com/ivanek/Gviz">Gviz</a></span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">EnsDb.Hsapiens.v86</span>)
<span class="kw">edb</span> <span class="op">&lt;-</span> <span class="kw">EnsDb.Hsapiens.v86</span>

<span class="co">## Retrieving a Gviz compatible GRanges object with all genes</span>
<span class="co">## encoded on chromosome Y.</span>
<span class="kw">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGeneRegionTrackForGviz.html">getGeneRegionTrackForGviz</a></span>(<span class="kw">edb</span>, chromosome = <span class="st">"Y"</span>,
                start = <span class="fl">20400000</span>, end = <span class="fl">21400000</span>)
<span class="co">## Define a genome axis track</span>
<span class="kw">gat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/GenomeAxisTrack-class.html">GenomeAxisTrack</a></span>()

<span class="co">## We have to change the ucscChromosomeNames option to FALSE to enable Gviz usage</span>
<span class="co">## with non-UCSC chromosome names.</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(ucscChromosomeNames = <span class="fl">FALSE</span>)

<span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/plotTracks.html">plotTracks</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">gat</span>, <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/GeneRegionTrack-class.html">GeneRegionTrack</a></span>(<span class="kw">gr</span>)))
</pre></div>
<p><img src="ensembldb_files/figure-html/gviz-plot-1.png" width="700"></p>
<div class="sourceCode" id="cb78"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(ucscChromosomeNames = <span class="fl">TRUE</span>)
</pre></div>
<p>Above we had to change the option <code>ucscChromosomeNames</code> to <code>FALSE</code> in order to use it with non-UCSC chromosome names. Alternatively, we could however also change the <code>seqnamesStyle</code> of the <code>EnsDb</code> object to <code>UCSC</code>. Note that we have to use now also chromosome names in the <em>UCSC style</em> in the <code>SeqNameFilter</code> (i.e. “chrY” instead of “Y”).</p>
<div class="sourceCode" id="cb79"><pre class="downlit">
<span class="fu"><a href="../reference/seqlevelsStyle.html">seqlevelsStyle</a></span>(<span class="kw">edb</span>) <span class="op">&lt;-</span> <span class="st">"UCSC"</span>
<span class="co">## Retrieving the GRanges objects with seqnames corresponding to UCSC chromosome names.</span>
<span class="kw">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGeneRegionTrackForGviz.html">getGeneRegionTrackForGviz</a></span>(<span class="kw">edb</span>, chromosome = <span class="st">"chrY"</span>,
                                start = <span class="fl">20400000</span>, end = <span class="fl">21400000</span>)
</pre></div>
<pre><code>## Warning in .formatSeqnameByStyleForQuery(x, sn, ifNotFound): Seqnames: Y could
## not be mapped to the seqlevels style of the database (Ensembl)!Returning the
## orginal seqnames for these.</code></pre>
<div class="sourceCode" id="cb81"><pre class="downlit">
<span class="fu">seqnames</span>(<span class="kw">gr</span>)
</pre></div>
<pre><code>## factor-Rle of length 89 with 1 run
##   Lengths:   89
##   Values : chrY
## Levels(1): chrY</code></pre>
<div class="sourceCode" id="cb83"><pre class="downlit">
<span class="co">## Define a genome axis track</span>
<span class="kw">gat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/GenomeAxisTrack-class.html">GenomeAxisTrack</a></span>()
<span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/plotTracks.html">plotTracks</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">gat</span>, <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/GeneRegionTrack-class.html">GeneRegionTrack</a></span>(<span class="kw">gr</span>)))
</pre></div>
<p><img src="ensembldb_files/figure-html/unnamed-chunk-1-1.png" width="700"></p>
<p>We can also use the filters from the <code>ensembldb</code> package to further refine what transcripts are fetched, like in the example below, in which we create two different gene region tracks, one for protein coding genes and one for lincRNAs.</p>
<div class="sourceCode" id="cb84"><pre class="downlit">
<span class="kw">protCod</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGeneRegionTrackForGviz.html">getGeneRegionTrackForGviz</a></span>(<span class="kw">edb</span>, chromosome = <span class="st">"chrY"</span>,
                     start = <span class="fl">20400000</span>, end = <span class="fl">21400000</span>,
                     filter = <span class="fu">GeneBiotypeFilter</span>(<span class="st">"protein_coding"</span>))
<span class="kw">lincs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGeneRegionTrackForGviz.html">getGeneRegionTrackForGviz</a></span>(<span class="kw">edb</span>, chromosome = <span class="st">"chrY"</span>,
                   start = <span class="fl">20400000</span>, end = <span class="fl">21400000</span>,
                   filter = <span class="fu">GeneBiotypeFilter</span>(<span class="st">"lincRNA"</span>))

<span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/plotTracks.html">plotTracks</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">gat</span>, <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/GeneRegionTrack-class.html">GeneRegionTrack</a></span>(<span class="kw">protCod</span>, name = <span class="st">"protein coding"</span>),
        <span class="fu"><a href="https://rdrr.io/pkg/Gviz/man/GeneRegionTrack-class.html">GeneRegionTrack</a></span>(<span class="kw">lincs</span>, name = <span class="st">"lincRNAs"</span>)), transcriptAnnotation = <span class="st">"symbol"</span>)
</pre></div>
<p><img src="ensembldb_files/figure-html/gviz-separate-tracks-1.png" width="700"></p>
<div class="sourceCode" id="cb85"><pre class="downlit">
<span class="co">## At last we change the seqlevels style again to Ensembl</span>
<span class="kw">seqlevelsStyle</span> <span class="op">&lt;-</span> <span class="st">"Ensembl"</span>
</pre></div>
<p>Alternatively, we can also use <code>ggbio</code> for plotting. For <code>ggbio</code> we can directly pass the <code>EnsDb</code> object along with optional filters (or as in the example below a filter expression as a <code>formula</code>).</p>
<div class="sourceCode" id="cb86"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://tengfei.github.com/ggbio">ggbio</a></span>)

<span class="co">## Create a plot for all transcripts of the gene SKA2</span>
<span class="fu"><a href="https://rdrr.io/pkg/ggbio/man/autoplot-method.html">autoplot</a></span>(<span class="kw">edb</span>, <span class="op">~</span> <span class="kw">gene_name</span> <span class="op">==</span> <span class="st">"SKA2"</span>)
</pre></div>
<p>To plot the genomic region and plot genes from both strands we can use a <code>GRangesFilter</code>.</p>
<div class="sourceCode" id="cb87"><pre class="downlit">
<span class="co">## Get the chromosomal region in which the gene is encoded</span>
<span class="kw">ska2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/exonsBy.html">genes</a></span>(<span class="kw">edb</span>, filter = <span class="op">~</span> <span class="kw">gene_name</span> <span class="op">==</span> <span class="st">"SKA2"</span>)
<span class="fu">strand</span>(<span class="kw">ska2</span>) <span class="op">&lt;-</span> <span class="st">"*"</span>
<span class="fu"><a href="https://rdrr.io/pkg/ggbio/man/autoplot-method.html">autoplot</a></span>(<span class="kw">edb</span>, <span class="fu">GRangesFilter</span>(<span class="kw">ska2</span>), names.expr = <span class="st">"gene_name"</span>)
</pre></div>
</div>
<div id="using-ensdb-objects-in-the-annotationdbi-framework" class="section level1">
<h1 class="hasAnchor">
<a href="#using-ensdb-objects-in-the-annotationdbi-framework" class="anchor"></a>Using <code>EnsDb</code> objects in the <code>AnnotationDbi</code> framework</h1>
<p>Most of the methods defined for objects extending the basic annotation package class <code>AnnotationDbi</code> are also defined for <code>EnsDb</code> objects (i.e. methods <code>columns</code>, <code>keytypes</code>, <code>keys</code>, <code>mapIds</code> and <code>select</code>). While these methods can be used analogously to basic annotation packages, the implementation for <code>EnsDb</code> objects also support the filtering framework of the <code>ensembldb</code> package.</p>
<p>In the example below we first evaluate all the available columns and keytypes in the database and extract then the gene names for all genes encoded on chromosome X.</p>
<div class="sourceCode" id="cb88"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">EnsDb.Hsapiens.v86</span>)
<span class="kw">edb</span> <span class="op">&lt;-</span> <span class="kw">EnsDb.Hsapiens.v86</span>

<span class="co">## List all available columns in the database.</span>
<span class="fu">columns</span>(<span class="kw">edb</span>)
</pre></div>
<pre><code>##  [1] "ENTREZID"            "EXONID"              "EXONIDX"            
##  [4] "EXONSEQEND"          "EXONSEQSTART"        "GENEBIOTYPE"        
##  [7] "GENEID"              "GENENAME"            "GENESEQEND"         
## [10] "GENESEQSTART"        "INTERPROACCESSION"   "ISCIRCULAR"         
## [13] "PROTDOMEND"          "PROTDOMSTART"        "PROTEINDOMAINID"    
## [16] "PROTEINDOMAINSOURCE" "PROTEINID"           "PROTEINSEQUENCE"    
## [19] "SEQCOORDSYSTEM"      "SEQLENGTH"           "SEQNAME"            
## [22] "SEQSTRAND"           "SYMBOL"              "TXBIOTYPE"          
## [25] "TXCDSSEQEND"         "TXCDSSEQSTART"       "TXID"               
## [28] "TXNAME"              "TXSEQEND"            "TXSEQSTART"         
## [31] "UNIPROTDB"           "UNIPROTID"           "UNIPROTMAPPINGTYPE"</code></pre>
<div class="sourceCode" id="cb90"><pre class="downlit">
<span class="co">## Note that these do *not* correspond to the actual column names</span>
<span class="co">## of the database that can be passed to methods like exons, genes,</span>
<span class="co">## transcripts etc. These column names can be listed with the listColumns</span>
<span class="co">## method.</span>
<span class="fu"><a href="../reference/EnsDb-class.html">listColumns</a></span>(<span class="kw">edb</span>)
</pre></div>
<pre><code>##  [1] "seq_name"              "seq_length"            "is_circular"          
##  [4] "gene_id"               "entrezid"              "exon_id"              
##  [7] "exon_seq_start"        "exon_seq_end"          "gene_name"            
## [10] "gene_biotype"          "gene_seq_start"        "gene_seq_end"         
## [13] "seq_strand"            "seq_coord_system"      "symbol"               
## [16] "name"                  "value"                 "tx_id"                
## [19] "protein_id"            "protein_sequence"      "protein_domain_id"    
## [22] "protein_domain_source" "interpro_accession"    "prot_dom_start"       
## [25] "prot_dom_end"          "tx_biotype"            "tx_seq_start"         
## [28] "tx_seq_end"            "tx_cds_seq_start"      "tx_cds_seq_end"       
## [31] "tx_name"               "exon_idx"              "uniprot_id"           
## [34] "uniprot_db"            "uniprot_mapping_type"</code></pre>
<div class="sourceCode" id="cb92"><pre class="downlit">
<span class="co">## List all of the supported key types.</span>
<span class="fu">keytypes</span>(<span class="kw">edb</span>)
</pre></div>
<pre><code>##  [1] "ENTREZID"            "EXONID"              "GENEBIOTYPE"        
##  [4] "GENEID"              "GENENAME"            "PROTDOMID"          
##  [7] "PROTEINDOMAINID"     "PROTEINDOMAINSOURCE" "PROTEINID"          
## [10] "SEQNAME"             "SEQSTRAND"           "SYMBOL"             
## [13] "TXBIOTYPE"           "TXID"                "TXNAME"             
## [16] "UNIPROTID"</code></pre>
<div class="sourceCode" id="cb94"><pre class="downlit">
<span class="co">## Get all gene ids from the database.</span>
<span class="kw">gids</span> <span class="op">&lt;-</span> <span class="fu">keys</span>(<span class="kw">edb</span>, keytype = <span class="st">"GENEID"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="kw">gids</span>)
</pre></div>
<pre><code>## [1] 63970</code></pre>
<div class="sourceCode" id="cb96"><pre class="downlit">
<span class="co">## Get all gene names for genes encoded on chromosome Y.</span>
<span class="kw">gnames</span> <span class="op">&lt;-</span> <span class="fu">keys</span>(<span class="kw">edb</span>, keytype = <span class="st">"GENENAME"</span>, filter = <span class="fu">SeqNameFilter</span>(<span class="st">"Y"</span>))
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">gnames</span>)
</pre></div>
<pre><code>## [1] "KDM5D"   "DDX3Y"   "ZFY"     "TBL1Y"   "PCDH11Y" "AMELY"</code></pre>
<p>In the next example we retrieve specific information from the database using the <code>select</code> method. First we fetch all transcripts for the genes <em>BCL2</em> and <em>BCL2L11</em>. In the first call we provide the gene names, while in the second call we employ the filtering system to perform a more fine-grained query to fetch only the protein coding transcripts for these genes.</p>
<div class="sourceCode" id="cb98"><pre class="downlit">
<span class="co">## Use the /standard/ way to fetch data.</span>
<span class="fu"><a href="../reference/select.html">select</a></span>(<span class="kw">edb</span>, keys = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"BCL2"</span>, <span class="st">"BCL2L11"</span>), keytype = <span class="st">"GENENAME"</span>,
       columns = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"GENEID"</span>, <span class="st">"GENENAME"</span>, <span class="st">"TXID"</span>, <span class="st">"TXBIOTYPE"</span>))
</pre></div>
<pre><code>##             GENEID GENENAME            TXID               TXBIOTYPE
## 1  ENSG00000171791     BCL2 ENST00000398117          protein_coding
## 2  ENSG00000171791     BCL2 ENST00000333681          protein_coding
## 3  ENSG00000171791     BCL2 ENST00000590515    processed_transcript
## 4  ENSG00000171791     BCL2 ENST00000589955          protein_coding
## 5  ENSG00000153094  BCL2L11 ENST00000432179          protein_coding
## 6  ENSG00000153094  BCL2L11 ENST00000308659          protein_coding
## 7  ENSG00000153094  BCL2L11 ENST00000393256          protein_coding
## 8  ENSG00000153094  BCL2L11 ENST00000393252          protein_coding
## 9  ENSG00000153094  BCL2L11 ENST00000433098 nonsense_mediated_decay
## 10 ENSG00000153094  BCL2L11 ENST00000405953          protein_coding
## 11 ENSG00000153094  BCL2L11 ENST00000415458 nonsense_mediated_decay
## 12 ENSG00000153094  BCL2L11 ENST00000436733 nonsense_mediated_decay
## 13 ENSG00000153094  BCL2L11 ENST00000437029 nonsense_mediated_decay
## 14 ENSG00000153094  BCL2L11 ENST00000452231 nonsense_mediated_decay
## 15 ENSG00000153094  BCL2L11 ENST00000361493 nonsense_mediated_decay
## 16 ENSG00000153094  BCL2L11 ENST00000431217 nonsense_mediated_decay
## 17 ENSG00000153094  BCL2L11 ENST00000439718 nonsense_mediated_decay
## 18 ENSG00000153094  BCL2L11 ENST00000438054          protein_coding
## 19 ENSG00000153094  BCL2L11 ENST00000337565          protein_coding
## 20 ENSG00000153094  BCL2L11 ENST00000622509          protein_coding
## 21 ENSG00000153094  BCL2L11 ENST00000619294          protein_coding
## 22 ENSG00000153094  BCL2L11 ENST00000610735          protein_coding
## 23 ENSG00000153094  BCL2L11 ENST00000622612          protein_coding
## 24 ENSG00000153094  BCL2L11 ENST00000357757          protein_coding
## 25 ENSG00000153094  BCL2L11 ENST00000615946          protein_coding
## 26 ENSG00000153094  BCL2L11 ENST00000621302          protein_coding
## 27 ENSG00000153094  BCL2L11 ENST00000620862          protein_coding
## 28         LRG_620  BCL2L11       LRG_620t1                LRG_gene
## 29         LRG_620  BCL2L11       LRG_620t2                LRG_gene
## 30         LRG_620  BCL2L11       LRG_620t3                LRG_gene
## 31         LRG_620  BCL2L11       LRG_620t4                LRG_gene
## 32         LRG_620  BCL2L11       LRG_620t5                LRG_gene</code></pre>
<div class="sourceCode" id="cb100"><pre class="downlit">
<span class="co">## Use the filtering system of ensembldb</span>
<span class="fu"><a href="../reference/select.html">select</a></span>(<span class="kw">edb</span>, keys = <span class="op">~</span> <span class="kw">gene_name</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"BCL2"</span>, <span class="st">"BCL2L11"</span>) <span class="op">&amp;</span>
        <span class="kw">tx_biotype</span> <span class="op">==</span> <span class="st">"protein_coding"</span>,
       columns = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"GENEID"</span>, <span class="st">"GENENAME"</span>, <span class="st">"TXID"</span>, <span class="st">"TXBIOTYPE"</span>))
</pre></div>
<pre><code>##             GENEID GENENAME            TXID      TXBIOTYPE
## 1  ENSG00000171791     BCL2 ENST00000398117 protein_coding
## 2  ENSG00000171791     BCL2 ENST00000333681 protein_coding
## 3  ENSG00000171791     BCL2 ENST00000589955 protein_coding
## 4  ENSG00000153094  BCL2L11 ENST00000432179 protein_coding
## 5  ENSG00000153094  BCL2L11 ENST00000308659 protein_coding
## 6  ENSG00000153094  BCL2L11 ENST00000393256 protein_coding
## 7  ENSG00000153094  BCL2L11 ENST00000393252 protein_coding
## 8  ENSG00000153094  BCL2L11 ENST00000405953 protein_coding
## 9  ENSG00000153094  BCL2L11 ENST00000438054 protein_coding
## 10 ENSG00000153094  BCL2L11 ENST00000337565 protein_coding
## 11 ENSG00000153094  BCL2L11 ENST00000622509 protein_coding
## 12 ENSG00000153094  BCL2L11 ENST00000619294 protein_coding
## 13 ENSG00000153094  BCL2L11 ENST00000610735 protein_coding
## 14 ENSG00000153094  BCL2L11 ENST00000622612 protein_coding
## 15 ENSG00000153094  BCL2L11 ENST00000357757 protein_coding
## 16 ENSG00000153094  BCL2L11 ENST00000615946 protein_coding
## 17 ENSG00000153094  BCL2L11 ENST00000621302 protein_coding
## 18 ENSG00000153094  BCL2L11 ENST00000620862 protein_coding</code></pre>
<p>Finally, we use the <code>mapIds</code> method to establish a mapping between ids and values. In the example below we fetch transcript ids for the two genes from the example above.</p>
<div class="sourceCode" id="cb102"><pre class="downlit">
<span class="co">## Use the default method, which just returns the first value for multi mappings.</span>
<span class="fu">mapIds</span>(<span class="kw">edb</span>, keys = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"BCL2"</span>, <span class="st">"BCL2L11"</span>), column = <span class="st">"TXID"</span>, keytype = <span class="st">"GENENAME"</span>)
</pre></div>
<pre><code>##              BCL2           BCL2L11 
## "ENST00000398117" "ENST00000432179"</code></pre>
<div class="sourceCode" id="cb104"><pre class="downlit">
<span class="co">## Alternatively, specify multiVals="list" to return all mappings.</span>
<span class="fu">mapIds</span>(<span class="kw">edb</span>, keys = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"BCL2"</span>, <span class="st">"BCL2L11"</span>), column = <span class="st">"TXID"</span>, keytype = <span class="st">"GENENAME"</span>,
       multiVals = <span class="st">"list"</span>)
</pre></div>
<pre><code>## $BCL2
## [1] "ENST00000398117" "ENST00000333681" "ENST00000590515" "ENST00000589955"
## 
## $BCL2L11
##  [1] "ENST00000432179" "ENST00000308659" "ENST00000393256" "ENST00000393252"
##  [5] "ENST00000433098" "ENST00000405953" "ENST00000415458" "ENST00000436733"
##  [9] "ENST00000437029" "ENST00000452231" "ENST00000361493" "ENST00000431217"
## [13] "ENST00000439718" "ENST00000438054" "ENST00000337565" "ENST00000622509"
## [17] "ENST00000619294" "ENST00000610735" "ENST00000622612" "ENST00000357757"
## [21] "ENST00000615946" "ENST00000621302" "ENST00000620862" "LRG_620t1"      
## [25] "LRG_620t2"       "LRG_620t3"       "LRG_620t4"       "LRG_620t5"</code></pre>
<div class="sourceCode" id="cb106"><pre class="downlit">
<span class="co">## And, just like before, we can use filters to map only to protein coding transcripts.</span>
<span class="fu">mapIds</span>(<span class="kw">edb</span>, keys = <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="fu">GeneNameFilter</span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"BCL2"</span>, <span class="st">"BCL2L11"</span>)),
                        <span class="fu">TxBiotypeFilter</span>(<span class="st">"protein_coding"</span>)), column = <span class="st">"TXID"</span>,
       multiVals = <span class="st">"list"</span>)
</pre></div>
<pre><code>## Warning in .mapIds(x = x, keys = keys, column = column, keytype = keytype, : Got
## 2 filter objects. Will use the keys of the first for the mapping!</code></pre>
<pre><code>## $BCL2
## [1] "ENST00000398117" "ENST00000333681" "ENST00000589955"
## 
## $BCL2L11
##  [1] "ENST00000432179" "ENST00000308659" "ENST00000393256" "ENST00000393252"
##  [5] "ENST00000405953" "ENST00000438054" "ENST00000337565" "ENST00000622509"
##  [9] "ENST00000619294" "ENST00000610735" "ENST00000622612" "ENST00000357757"
## [13] "ENST00000615946" "ENST00000621302" "ENST00000620862"</code></pre>
<p>Note that, if the filters are used, the ordering of the result does no longer match the ordering of the genes.</p>
</div>
<div id="important-notes" class="section level1">
<h1 class="hasAnchor">
<a href="#important-notes" class="anchor"></a>Important notes</h1>
<p>These notes might explain eventually unexpected results (and, more importantly, help avoiding them):</p>
<ul>
<li><p>The ordering of the results returned by the <code>genes</code>, <code>exons</code>, <code>transcripts</code> methods can be specified with the <code>order.by</code> parameter. The ordering of the results does however <strong>not</strong> correspond to the ordering of values in submitted filter objects. The exception is the <code>select</code> method. If a character vector of values or a single filter is passed with argument <code>keys</code> the ordering of results of this method matches the ordering of the key values or the values of the filter.</p></li>
<li><p>Results of <code>exonsBy</code>, <code>transcriptsBy</code> are always ordered by the <code>by</code> argument.</p></li>
<li><p>The CDS provided by <code>EnsDb</code> objects <strong>always</strong> includes both, the start and the stop codon.</p></li>
<li><p>Transcripts with multiple CDS are at present not supported by <code>EnsDb</code>.</p></li>
<li><p>At present, <code>EnsDb</code> support only genes/transcripts for which all of their exons are encoded on the same chromosome and the same strand.</p></li>
<li><p>Since a single Ensembl gene ID might be mapped to multiple NCBI Entrezgene IDs methods such as <code>genes</code>, <code>transcripts</code> etc return a <code>list</code> in the <code>"entrezid"</code> column of the resulting result object.</p></li>
</ul>
</div>
<div id="getting-or-building-ensdb-databasespackages" class="section level1">
<h1 class="hasAnchor">
<a href="#getting-or-building-ensdb-databasespackages" class="anchor"></a>Getting or building <code>EnsDb</code> databases/packages</h1>
<p>Some of the code in this section is not supposed to be automatically executed when the vignette is built, as this would require a working installation of the Ensembl Perl API, which is not expected to be available on each system. Also, building <code>EnsDb</code> from alternative sources, like GFF or GTF files takes some time and thus also these examples are not directly executed when the vignette is build.</p>
<div id="getting-ensdb-databases" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-ensdb-databases" class="anchor"></a>Getting <code>EnsDb</code> databases</h2>
<p>Some <code>EnsDb</code> databases are available as <code>R</code> packages from Bioconductor and can be simply installed with the <code>install</code> function from the <code>BiocManager</code> package. The name of such annotation packages starts with <em>EnsDb</em> followed by the abbreviation of the organism and the Ensembl version on which the annotation bases. <code>EnsDb.Hsapiens.v86</code> provides thus an <code>EnsDb</code> database for homo sapiens with annotations from Ensembl version 86.</p>
<p>Since Bioconductor version 3.5 <code>EnsDb</code> databases can also be retrieved directly from <code>AnnotationHub</code>.</p>
<div class="sourceCode" id="cb109"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">AnnotationHub</span>)
<span class="co">## Load the annotation resource.</span>
<span class="kw">ah</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html">AnnotationHub</a></span>()

<span class="co">## Query for all available EnsDb databases</span>
<span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html">query</a></span>(<span class="kw">ah</span>, <span class="st">"EnsDb"</span>)
</pre></div>
<pre><code>## AnnotationHub with 1747 records
## # snapshotDate(): 2020-06-18
## # $dataprovider: Ensembl
## # $species: Homo sapiens, Xiphophorus maculatus, Xenopus tropicalis, Vicugna...
## # $rdataclass: EnsDb
## # additional mcols(): taxonomyid, genome, description,
## #   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,
## #   rdatapath, sourceurl, sourcetype 
## # retrieve records with, e.g., 'object[["AH53185"]]' 
## 
##             title                                       
##   AH53185 | Ensembl 87 EnsDb for Anolis Carolinensis    
##   AH53186 | Ensembl 87 EnsDb for Ailuropoda Melanoleuca 
##   AH53187 | Ensembl 87 EnsDb for Astyanax Mexicanus     
##   AH53188 | Ensembl 87 EnsDb for Anas Platyrhynchos     
##   AH53189 | Ensembl 87 EnsDb for Bos Taurus             
##   ...       ...                                         
##   AH79812 | Ensembl 100 EnsDb for Vulpes vulpes         
##   AH79813 | Ensembl 100 EnsDb for Xiphophorus couchianus
##   AH79814 | Ensembl 100 EnsDb for Xiphophorus maculatus 
##   AH79815 | Ensembl 100 EnsDb for Xenopus tropicalis    
##   AH79816 | Ensembl 100 EnsDb for Zonotrichia albicollis</code></pre>
<p>We can simply fetch one of the databases.</p>
<div class="sourceCode" id="cb111"><pre class="downlit">
<span class="kw">ahDb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html">query</a></span>(<span class="kw">ah</span>, pattern = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Xiphophorus Maculatus"</span>, <span class="st">"EnsDb"</span>, <span class="fl">87</span>))
<span class="co">## What have we got</span>
<span class="kw">ahDb</span>
</pre></div>
<pre><code>## AnnotationHub with 1 record
## # snapshotDate(): 2020-06-18
## # names(): AH53251
## # $dataprovider: Ensembl
## # $species: Xiphophorus maculatus
## # $rdataclass: EnsDb
## # $rdatadateadded: 2017-02-07
## # $title: Ensembl 87 EnsDb for Xiphophorus Maculatus
## # $description: Gene and protein annotations for Xiphophorus Maculatus based...
## # $taxonomyid: 8083
## # $genome: Xipmac4.4.2
## # $sourcetype: ensembl
## # $sourceurl: http://www.ensembl.org
## # $sourcesize: NA
## # $tags: c("EnsDb", "Ensembl", "Gene", "Transcript", "Protein",
## #   "Annotation", "87", "AHEnsDbs") 
## # retrieve record with 'object[["AH53251"]]'</code></pre>
<p>Fetch the <code>EnsDb</code> and use it.</p>
<div class="sourceCode" id="cb113"><pre class="downlit">
<span class="kw">ahEdb</span> <span class="op">&lt;-</span> <span class="kw">ahDb</span>[[<span class="fl">1</span>]]

<span class="co">## retriebe all genes</span>
<span class="kw">gns</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/exonsBy.html">genes</a></span>(<span class="kw">ahEdb</span>)
</pre></div>
<p>We could even make an annotation package from this <code>EnsDb</code> object using the <code>makeEnsembldbPackage</code> and passing <code>dbfile(dbconn(ahEdb))</code> as <code>ensdb</code> argument.</p>
</div>
<div id="building-annotation-packages" class="section level2">
<h2 class="hasAnchor">
<a href="#building-annotation-packages" class="anchor"></a>Building annotation packages</h2>
<div id="directly-from-ensembl-databases" class="section level3">
<h3 class="hasAnchor">
<a href="#directly-from-ensembl-databases" class="anchor"></a>Directly from Ensembl databases</h3>
<p>The <code>fetchTablesFromEnsembl</code> function uses the Ensembl Perl API to retrieve the required annotations from an Ensembl database (e.g. from the main site <em>ensembldb.ensembl.org</em>). Thus, to use this functionality to build databases, the Ensembl Perl API needs to be installed (see <sup><a id="fnr.5" class="footref" href="#fn.5">5</a></sup> for details).</p>
<p>Below we create an <code>EnsDb</code> database by fetching the required data directly from the Ensembl core databases. The <code>makeEnsembldbPackage</code> function is then used to create an annotation package from this <code>EnsDb</code> containing all human genes for Ensembl version 75.</p>
<div class="sourceCode" id="cb114"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://github.com/jorainer/ensembldb">ensembldb</a></span>)

<span class="co">## get all human gene/transcript/exon annotations from Ensembl (75)</span>
<span class="co">## the resulting tables will be stored by default to the current working</span>
<span class="co">## directory</span>
<span class="fu"><a href="../reference/makeEnsembldbPackage.html">fetchTablesFromEnsembl</a></span>(<span class="fl">75</span>, species = <span class="st">"human"</span>)

<span class="co">## These tables can then be processed to generate a SQLite database</span>
<span class="co">## containing the annotations (again, the function assumes the required</span>
<span class="co">## txt files to be present in the current working directory)</span>
<span class="kw">DBFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeEnsembldbPackage.html">makeEnsemblSQLiteFromTables</a></span>()

<span class="co">## and finally we can generate the package</span>
<span class="fu"><a href="../reference/makeEnsembldbPackage.html">makeEnsembldbPackage</a></span>(ensdb = <span class="kw">DBFile</span>, version = <span class="st">"0.99.12"</span>,
                     maintainer = <span class="st">"Johannes Rainer &lt;johannes.rainer@eurac.edu&gt;"</span>,
                     author = <span class="st">"J Rainer"</span>)
</pre></div>
<p>The generated package can then be build using <code>R CMD build EnsDb.Hsapiens.v75</code> and installed with <code>R CMD INSTALL EnsDb.Hsapiens.v75*</code>. Note that we could directly generate an <code>EnsDb</code> instance by loading the database file, i.e. by calling <code>edb &lt;- EnsDb(DBFile)</code> and work with that annotation object.</p>
<p>To fetch and build annotation packages for plant genomes (e.g. arabidopsis thaliana), the <em>Ensembl genomes</em> should be specified as a host, i.e. setting <code>host</code> to “mysql-eg-publicsql.ebi.ac.uk”, <code>port</code> to <code>4157</code> and <code>species</code> to e.g. “arabidopsis thaliana”.</p>
</div>
<div id="from-a-gtf-or-gff-file" class="section level3">
<h3 class="hasAnchor">
<a href="#from-a-gtf-or-gff-file" class="anchor"></a>From a GTF or GFF file</h3>
<p>Alternatively, the <code>ensDbFromAH</code>, <code>ensDbFromGff</code>, <code>ensDbFromGRanges</code> and <code>ensDbFromGtf</code> functions allow to build EnsDb SQLite files from a <code>GRanges</code> object or GFF/GTF files from Ensembl (either provided as files or <em>via</em> <code>AnnotationHub</code>). These functions do not depend on the Ensembl Perl API, but require a working internet connection to fetch the chromosome lengths from Ensembl as these are not provided within GTF or GFF files. Also note that protein annotations are usually not available in GTF or GFF files, thus, such annotations will not be included in the generated <code>EnsDb</code> database - protein annotations are only available in <code>EnsDb</code> databases created with the Ensembl Perl API (such as the ones provided through <code>AnnotationHub</code> or as Bioconductor packages).</p>
<p>In the next example we create an <code>EnsDb</code> database using the <code>AnnotationHub</code> package and load also the corresponding genomic DNA sequence matching the Ensembl version. We thus first query the <code>AnnotationHub</code> package for all resources available for <code>Mus musculus</code> and the Ensembl release 77. Next we create the <code>EnsDb</code> object from the appropriate <code>AnnotationHub</code> resource. We then use the <code>getGenomeTwoBitFile</code> method on the <code>EnsDb</code> to directly look up and retrieve the correct or best matching <code>TwoBitFile</code> with the genomic DNA sequence. At last we retrieve the sequences of all exons using the <code>getSeq</code> method.</p>
<div class="sourceCode" id="cb115"><pre class="downlit">
<span class="co">## Load the AnnotationHub data.</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">AnnotationHub</span>)
<span class="kw">ah</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html">AnnotationHub</a></span>()

<span class="co">## Query all available files for Ensembl release 77 for</span>
<span class="co">## Mus musculus.</span>
<span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html">query</a></span>(<span class="kw">ah</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Mus musculus"</span>, <span class="st">"release-77"</span>))

<span class="co">## Get the resource for the gtf file with the gene/transcript definitions.</span>
<span class="kw">Gtf</span> <span class="op">&lt;-</span> <span class="kw">ah</span>[<span class="st">"AH28822"</span>]
<span class="co">## Create a EnsDb database file from this.</span>
<span class="kw">DbFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeEnsembldbPackage.html">ensDbFromAH</a></span>(<span class="kw">Gtf</span>)
<span class="co">## We can either generate a database package, or directly load the data</span>
<span class="kw">edb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EnsDb.html">EnsDb</a></span>(<span class="kw">DbFile</span>)


<span class="co">## Identify and get the TwoBit object with the genomic DNA sequence matching</span>
<span class="co">## the EnsDb annotation.</span>
<span class="kw">Dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGenomeFaFile.html">getGenomeTwoBitFile</a></span>(<span class="kw">edb</span>)
<span class="co">## We next retrieve the sequence of all exons on chromosome Y.</span>
<span class="kw">exons</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/exonsBy.html">exons</a></span>(<span class="kw">edb</span>, filter = <span class="fu">SeqNameFilter</span>(<span class="st">"Y"</span>))
<span class="kw">exonSeq</span> <span class="op">&lt;-</span> <span class="fu">getSeq</span>(<span class="kw">Dna</span>, <span class="kw">exons</span>)
</pre></div>
<p>In the example below we load a <code>GRanges</code> containing gene definitions for genes encoded on chromosome Y and generate a <code>EnsDb</code> SQLite database from that information.</p>
<div class="sourceCode" id="cb116"><pre class="downlit">
<span class="co">## Generate a sqlite database from a GRanges object specifying</span>
<span class="co">## genes encoded on chromosome Y</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"YGRanges.RData"</span>, package = <span class="st">"ensembldb"</span>))
<span class="kw">Y</span>
</pre></div>
<pre><code>## GRanges object with 7155 ranges and 16 metadata columns:
##          seqnames            ranges strand |               source       type
##             &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |             &lt;factor&gt;   &lt;factor&gt;
##      [1]        Y   2652790-2652894      + |       snRNA          gene      
##      [2]        Y   2652790-2652894      + |       snRNA          transcript
##      [3]        Y   2652790-2652894      + |       snRNA          exon      
##      [4]        Y   2654896-2655740      - |       protein_coding gene      
##      [5]        Y   2654896-2655740      - |       protein_coding transcript
##      ...      ...               ...    ... .                  ...        ...
##   [7151]        Y 28772667-28773306      - | processed_pseudogene transcript
##   [7152]        Y 28772667-28773306      - | processed_pseudogene exon      
##   [7153]        Y 59001391-59001635      + | pseudogene           gene      
##   [7154]        Y 59001391-59001635      + | processed_pseudogene transcript
##   [7155]        Y 59001391-59001635      + | processed_pseudogene exon      
##              score     phase         gene_id   gene_name    gene_source
##          &lt;numeric&gt; &lt;integer&gt;     &lt;character&gt; &lt;character&gt;    &lt;character&gt;
##      [1]        NA      &lt;NA&gt; ENSG00000251841  RNU6-1334P        ensembl
##      [2]        NA      &lt;NA&gt; ENSG00000251841  RNU6-1334P        ensembl
##      [3]        NA      &lt;NA&gt; ENSG00000251841  RNU6-1334P        ensembl
##      [4]        NA      &lt;NA&gt; ENSG00000184895         SRY ensembl_havana
##      [5]        NA      &lt;NA&gt; ENSG00000184895         SRY ensembl_havana
##      ...       ...       ...             ...         ...            ...
##   [7151]        NA      &lt;NA&gt; ENSG00000231514     FAM58CP         havana
##   [7152]        NA      &lt;NA&gt; ENSG00000231514     FAM58CP         havana
##   [7153]        NA      &lt;NA&gt; ENSG00000235857     CTBP2P1         havana
##   [7154]        NA      &lt;NA&gt; ENSG00000235857     CTBP2P1         havana
##   [7155]        NA      &lt;NA&gt; ENSG00000235857     CTBP2P1         havana
##            gene_biotype   transcript_id transcript_name transcript_source
##             &lt;character&gt;     &lt;character&gt;     &lt;character&gt;       &lt;character&gt;
##      [1]          snRNA            &lt;NA&gt;            &lt;NA&gt;              &lt;NA&gt;
##      [2]          snRNA ENST00000516032  RNU6-1334P-201           ensembl
##      [3]          snRNA ENST00000516032  RNU6-1334P-201           ensembl
##      [4] protein_coding            &lt;NA&gt;            &lt;NA&gt;              &lt;NA&gt;
##      [5] protein_coding ENST00000383070         SRY-001    ensembl_havana
##      ...            ...             ...             ...               ...
##   [7151]     pseudogene ENST00000435741     FAM58CP-001            havana
##   [7152]     pseudogene ENST00000435741     FAM58CP-001            havana
##   [7153]     pseudogene            &lt;NA&gt;            &lt;NA&gt;              &lt;NA&gt;
##   [7154]     pseudogene ENST00000431853     CTBP2P1-001            havana
##   [7155]     pseudogene ENST00000431853     CTBP2P1-001            havana
##          exon_number         exon_id         tag     ccds_id  protein_id
##            &lt;numeric&gt;     &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;
##      [1]          NA            &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
##      [2]          NA            &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
##      [3]           1 ENSE00002088309        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
##      [4]          NA            &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
##      [5]          NA            &lt;NA&gt;        CCDS   CCDS14772        &lt;NA&gt;
##      ...         ...             ...         ...         ...         ...
##   [7151]          NA            &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
##   [7152]           1 ENSE00001616687        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
##   [7153]          NA            &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
##   [7154]          NA            &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
##   [7155]           1 ENSE00001794473        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
##   -------
##   seqinfo: 1 sequence from GRCh37 genome</code></pre>
<div class="sourceCode" id="cb118"><pre class="downlit">
<span class="co">## Create the EnsDb database file</span>
<span class="kw">DB</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeEnsembldbPackage.html">ensDbFromGRanges</a></span>(<span class="kw">Y</span>, path = <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span>(), version = <span class="fl">75</span>,
               organism = <span class="st">"Homo_sapiens"</span>)
</pre></div>
<pre><code>## Warning in ensDbFromGRanges(Y, path = tempdir(), version = 75, organism =
## "Homo_sapiens"): I'm missing column(s): 'entrezid'. The corresponding database
## column(s) will be empty!</code></pre>
<div class="sourceCode" id="cb120"><pre class="downlit">
<span class="co">## Load the database</span>
<span class="kw">edb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EnsDb.html">EnsDb</a></span>(<span class="kw">DB</span>)
<span class="kw">edb</span>
</pre></div>
<pre><code>## EnsDb for Ensembl:
## |Backend: SQLite
## |Db type: EnsDb
## |Type of Gene ID: Ensembl Gene ID
## |Supporting package: ensembldb
## |Db created by: ensembldb package from Bioconductor
## |script_version: 0.0.1
## |Creation time: Thu Jun 25 08:26:44 2020
## |ensembl_version: 75
## |ensembl_host: unknown
## |Organism: Homo_sapiens
## |genome_build: GRCh37
## |DBSCHEMAVERSION: 1.0
## |source_file: GRanges object
## | No. of genes: 495.
## | No. of transcripts: 731.</code></pre>
<p>Alternatively we can build the annotation database using the <code>ensDbFromGtf</code> <code>ensDbFromGff</code> functions, that extract most of the required data from a GTF respectively GFF (version 3) file which can be downloaded from Ensembl (e.g. from <a href="ftp://ftp.ensembl.org/pub/release-75/gtf/homo_sapiens" class="uri">ftp://ftp.ensembl.org/pub/release-75/gtf/homo_sapiens</a> for human gene definitions from Ensembl version 75; for plant genomes etc, files can be retrieved from <a href="ftp://ftp.ensemblgenomes.org" class="uri">ftp://ftp.ensemblgenomes.org</a>). All information except the chromosome lengths, the NCBI Entrezgene IDs and protein annotations can be extracted from these GTF files. The function also tries to retrieve chromosome length information automatically from Ensembl.</p>
<p>Below we create the annotation from a gtf file that we fetch directly from Ensembl.</p>
<div class="sourceCode" id="cb122"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://github.com/jorainer/ensembldb">ensembldb</a></span>)

<span class="co">## the GTF file can be downloaded from</span>
<span class="co">## ftp://ftp.ensembl.org/pub/release-75/gtf/homo_sapiens/</span>
<span class="kw">gtffile</span> <span class="op">&lt;-</span> <span class="st">"Homo_sapiens.GRCh37.75.gtf.gz"</span>
<span class="co">## generate the SQLite database file</span>
<span class="kw">DB</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeEnsembldbPackage.html">ensDbFromGtf</a></span>(gtf = <span class="kw">gtffile</span>)

<span class="co">## load the DB file directly</span>
<span class="kw">EDB</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EnsDb.html">EnsDb</a></span>(<span class="kw">DB</span>)

<span class="co">## alternatively, build the annotation package</span>
<span class="co">## and finally we can generate the package</span>
<span class="fu"><a href="../reference/makeEnsembldbPackage.html">makeEnsembldbPackage</a></span>(ensdb = <span class="kw">DB</span>, version = <span class="st">"0.99.12"</span>,
                     maintainer = <span class="st">"Johannes Rainer &lt;johannes.rainer@eurac.edu&gt;"</span>,
                     author = <span class="st">"J Rainer"</span>)
</pre></div>
</div>
</div>
</div>
<div id="database-layout" class="section level1">
<h1 class="hasAnchor">
<a href="#database-layout" class="anchor"></a>Database layout<a id="org4707794"></a>
</h1>
<p>The database consists of the following tables and attributes (the layout is also shown in Figure <a href="#orgfe7e95a">165</a>). Note that the protein-specific annotations might not be available in all <code>EnsDB</code> databases (e.g. such ones created with <code>ensembldb</code> version &lt; 1.7 or created from GTF or GFF files).</p>
<ul>
<li>
<p><strong>gene</strong>: all gene specific annotations.</p>
<ul>
<li>
<code>gene_id</code>: the Ensembl ID of the gene.</li>
<li>
<code>gene_name</code>: the name (symbol) of the gene.</li>
<li>
<code>gene_biotype</code>: the biotype of the gene.</li>
<li>
<code>gene_seq_start</code>: the start coordinate of the gene on the sequence (usually a chromosome).</li>
<li>
<code>gene_seq_end</code>: the end coordinate of the gene on the sequence.</li>
<li>
<code>seq_name</code>: the name of the sequence (usually the chromosome name).</li>
<li>
<code>seq_strand</code>: the strand on which the gene is encoded.</li>
<li>
<code>seq_coord_system</code>: the coordinate system of the sequence.</li>
<li>
<code>description</code>: the description of the gene.</li>
</ul>
</li>
<li>
<p><strong>entrezgene</strong>: mapping of Ensembl genes to NCBI Entrezgene identifiers. Note that this mapping can be a one-to-many mapping.</p>
<ul>
<li>
<code>gene_id</code>: the Ensembl gene ID.</li>
<li>
<code>entrezid</code>: the NCBI Entrezgene ID.</li>
</ul>
</li>
<li>
<p><strong>tx</strong>: all transcript related annotations. Note that while no <code>tx_name</code> column is available in this database column, all methods to retrieve data from the database support also this column. The returned values are however the ID of the transcripts.</p>
<ul>
<li>
<code>tx_id</code>: the Ensembl transcript ID.</li>
<li>
<code>tx_biotype</code>: the biotype of the transcript.</li>
<li>
<code>tx_seq_start</code>: the start coordinate of the transcript.</li>
<li>
<code>tx_seq_end</code>: the end coordinate of the transcript.</li>
<li>
<code>tx_cds_seq_start</code>: the start coordinate of the coding region of the transcript (NULL for non-coding transcripts).</li>
<li>
<code>tx_cds_seq_end</code>: the end coordinate of the coding region of the transcript.</li>
<li>
<code>gc_count</code>: from Ensembl release 98 on, the <strong>tx</strong> table contains also a column <code>gc_count</code> providing the transcript’s G-C content expressed as a percentage.</li>
<li>
<code>gene_id</code>: the gene to which the transcript belongs.</li>
</ul>
<p><code>EnsDb</code> databases for more recent Ensembl releases have also a column <code>tx_support_level</code> providing the evidence level for a transcript (1 high evidence, 5 low evidence, NA no evidence calculated).</p>
</li>
<li>
<p><strong>exon</strong>: all exon related annotation.</p>
<ul>
<li>
<code>exon_id</code>: the Ensembl exon ID.</li>
<li>
<code>exon_seq_start</code>: the start coordinate of the exon.</li>
<li>
<code>exon_seq_end</code>: the end coordinate of the exon.</li>
</ul>
</li>
<li>
<p><strong>tx2exon</strong>: provides the n:m mapping between transcripts and exons.</p>
<ul>
<li>
<code>tx_id</code>: the Ensembl transcript ID.</li>
<li>
<code>exon_id</code>: the Ensembl exon ID.</li>
<li>
<code>exon_idx</code>: the index of the exon in the corresponding transcript, always from 5’ to 3’ of the transcript.</li>
</ul>
</li>
<li>
<p><strong>chromosome</strong>: provides some information about the chromosomes.</p>
<ul>
<li>
<code>seq_name</code>: the name of the sequence/chromosome.</li>
<li>
<code>seq_length</code>: the length of the sequence.</li>
<li>
<code>is_circular</code>: whether the sequence in circular.</li>
</ul>
</li>
<li>
<p><strong>protein</strong>: provides protein annotation for a (coding) transcript.</p>
<ul>
<li>
<code>protein_id</code>: the Ensembl protein ID.</li>
<li>
<code>tx_id</code>: the transcript ID which CDS encodes the protein.</li>
<li>
<code>protein_sequence</code>: the peptide sequence of the protein (translated from the transcript’s coding sequence after applying eventual RNA editing).</li>
</ul>
</li>
<li>
<p><strong>uniprot</strong>: provides the mapping from Ensembl protein ID(s) to Uniprot ID(s). Not all Ensembl proteins are annotated to Uniprot IDs, also, each Ensembl protein might be mapped to multiple Uniprot IDs.</p>
<ul>
<li>
<code>protein_id</code>: the Ensembl protein ID.</li>
<li>
<code>uniprot_id</code>: the Uniprot ID.</li>
<li>
<code>uniprot_db</code>: the Uniprot database in which the ID is defined.</li>
<li>
<code>uniprot_mapping_type</code>: the type of the mapping method that was used to assign the Uniprot ID to an Ensembl protein ID.</li>
</ul>
</li>
<li>
<p><strong>protein_domain</strong>: provides protein domain annotations and mapping to proteins.</p>
<ul>
<li>
<code>protein_id</code>: the Ensembl protein ID on which the protein domain is present.</li>
<li>
<code>protein_domain_id</code>: the ID of the protein domain (from the protein domain source).</li>
<li>
<code>protein_domain_source</code>: the source/analysis method in/by which the protein domain was defined (such as pfam etc).</li>
<li>
<code>interpro_accession</code>: the Interpro accession ID of the protein domain.</li>
<li>
<code>prot_dom_start</code>: the start position of the protein domain within the protein’s sequence.</li>
<li>
<code>prot_dom_end</code>: the end position of the protein domain within the protein’s sequence.</li>
</ul>
</li>
<li>
<p><strong>metadata</strong>: some additional, internal, informations (Genome build, Ensembl version etc).</p>
<ul>
<li><code>name</code></li>
<li><code>value</code></li>
</ul>
</li>
<li>
<p><em>virtual</em> columns:</p>
<ul>
<li>
<code>symbol</code>: the database does not have such a database column, but it is still possible to use it in the <code>columns</code> parameter. This column is <em>symlinked</em> to the <code>gene_name</code> column.</li>
<li>
<code>tx_name</code>: similar to the <code>symbol</code> column, this column is <em>symlinked</em> to the <code>tx_id</code> column.</li>
</ul>
</li>
</ul>
<p>The database layout: as already described above, protein related annotations (green) might not be available in each <code>EnsDb</code> database.</p>
<div class="figure">
<img src="images/dblayout.png" title="Database layout." alt=""><p class="caption">img</p>
</div>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session information</h1>
<div class="sourceCode" id="cb123"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()
</pre></div>
<pre><code>## R version 4.0.0 (2020-04-24)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.4 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
##  [1] grid      stats4    parallel  stats     graphics  grDevices utils    
##  [8] datasets  methods   base     
## 
## other attached packages:
##  [1] AnnotationHub_2.21.1                  
##  [2] BiocFileCache_1.13.0                  
##  [3] dbplyr_1.4.4                          
##  [4] Gviz_1.33.1                           
##  [5] BSgenome.Hsapiens.NCBI.GRCh38_1.3.1000
##  [6] BSgenome_1.57.0                       
##  [7] rtracklayer_1.49.3                    
##  [8] Biostrings_2.57.2                     
##  [9] XVector_0.29.2                        
## [10] magrittr_1.5                          
## [11] EnsDb.Hsapiens.v86_2.99.0             
## [12] ensembldb_2.13.1                      
## [13] AnnotationFilter_1.13.0               
## [14] GenomicFeatures_1.41.0                
## [15] AnnotationDbi_1.51.0                  
## [16] Biobase_2.49.0                        
## [17] GenomicRanges_1.41.5                  
## [18] GenomeInfoDb_1.25.3                   
## [19] IRanges_2.23.10                       
## [20] S4Vectors_0.27.12                     
## [21] BiocGenerics_0.35.4                   
## [22] BiocStyle_2.17.0                      
## 
## loaded via a namespace (and not attached):
##   [1] colorspace_1.4-1              ellipsis_0.3.1               
##   [3] rprojroot_1.3-2               biovizBase_1.37.0            
##   [5] htmlTable_2.0.0               base64enc_0.1-3              
##   [7] fs_1.4.1                      dichromat_2.0-0              
##   [9] rstudioapi_0.11               bit64_0.9-7                  
##  [11] interactiveDisplayBase_1.27.5 splines_4.0.0                
##  [13] knitr_1.28                    Formula_1.2-3                
##  [15] Rsamtools_2.5.2               cluster_2.1.0                
##  [17] png_0.1-7                     shiny_1.4.0.2                
##  [19] BiocManager_1.30.10           compiler_4.0.0               
##  [21] httr_1.4.1                    backports_1.1.8              
##  [23] fastmap_1.0.1                 assertthat_0.2.1             
##  [25] Matrix_1.2-18                 lazyeval_0.2.2               
##  [27] later_1.1.0.1                 acepack_1.4.1                
##  [29] htmltools_0.5.0               prettyunits_1.1.1            
##  [31] tools_4.0.0                   gtable_0.3.0                 
##  [33] glue_1.4.1                    GenomeInfoDbData_1.2.3       
##  [35] dplyr_1.0.0                   rappdirs_0.3.1               
##  [37] Rcpp_1.0.4.6                  pkgdown_1.5.1.9000           
##  [39] vctrs_0.3.1                   xfun_0.15                    
##  [41] stringr_1.4.0                 mime_0.9                     
##  [43] lifecycle_0.2.0               XML_3.99-0.3                 
##  [45] zlibbioc_1.35.0               scales_1.1.1                 
##  [47] VariantAnnotation_1.35.2      promises_1.1.1               
##  [49] hms_0.5.3                     ProtGenerics_1.21.0          
##  [51] SummarizedExperiment_1.19.5   RColorBrewer_1.1-2           
##  [53] yaml_2.2.1                    curl_4.3                     
##  [55] memoise_1.1.0                 gridExtra_2.3                
##  [57] ggplot2_3.3.2                 biomaRt_2.45.1               
##  [59] rpart_4.1-15                  latticeExtra_0.6-29          
##  [61] stringi_1.4.6                 RSQLite_2.2.0                
##  [63] BiocVersion_3.12.0            desc_1.2.0                   
##  [65] checkmate_2.0.0               BiocParallel_1.23.0          
##  [67] rlang_0.4.6                   pkgconfig_2.0.3              
##  [69] matrixStats_0.56.0            bitops_1.0-6                 
##  [71] evaluate_0.14                 lattice_0.20-41              
##  [73] purrr_0.3.4                   GenomicAlignments_1.25.3     
##  [75] htmlwidgets_1.5.1             bit_1.1-15.2                 
##  [77] tidyselect_1.1.0              bookdown_0.19                
##  [79] R6_2.4.1                      generics_0.0.2               
##  [81] Hmisc_4.4-0                   DelayedArray_0.15.5          
##  [83] DBI_1.1.0                     pillar_1.4.4                 
##  [85] foreign_0.8-78                survival_3.1-12              
##  [87] RCurl_1.98-1.2                nnet_7.3-13                  
##  [89] tibble_3.0.1                  crayon_1.3.4                 
##  [91] rmarkdown_2.3.1               jpeg_0.1-8.1                 
##  [93] progress_1.2.2                data.table_1.12.8            
##  [95] blob_1.2.1                    digest_0.6.25                
##  [97] xtable_1.8-4                  httpuv_1.5.4                 
##  [99] openssl_1.4.1                 munsell_0.5.0                
## [101] askpass_1.1</code></pre>
</div>
<div id="footnotes" class="section level1">
<h1 class="hasAnchor">
<a href="#footnotes" class="anchor"></a>Footnotes</h1>
<p><sup><a id="fn.1" href="#fnr.1">1</a></sup><a href="http://www.ensembl.org" class="uri">http://www.ensembl.org</a></p>
<p><sup><a id="fn.2" href="#fnr.2">2</a></sup><a href="http://www.lrg-sequence.org" class="uri">http://www.lrg-sequence.org</a></p>
<p><sup><a id="fn.3" href="#fnr.3">3</a></sup><a href="http://www.ncbi.nlm.nih.gov/pubmed/23950696" class="uri">http://www.ncbi.nlm.nih.gov/pubmed/23950696</a></p>
<p><sup><a id="fn.4" href="#fnr.4">4</a></sup><a href="http://www.ncbi.nlm.nih.gov/pubmed/24227677" class="uri">http://www.ncbi.nlm.nih.gov/pubmed/24227677</a></p>
<p><sup><a id="fn.5" href="#fnr.5">5</a></sup><a href="http://www.ensembl.org/info/docs/api/api_installation.html" class="uri">http://www.ensembl.org/info/docs/api/api_installation.html</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Johannes Rainer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
