<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Querying protein features • ensembldb</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Querying protein features">
<meta property="og:description" content="ensembldb">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ensembldb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.17.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/ensembldb.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/MySQL-backend.html">Using a MySQL server backend</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping-use-cases.html">Use cases for coordinate mapping with ensembldb</a>
    </li>
    <li>
      <a href="../articles/coordinate-mapping.html">Mapping between genome, transcript and protein coordinates</a>
    </li>
    <li>
      <a href="../articles/proteins.html">Querying protein features</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jorainer/ensembldb/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="proteins_files/header-attrs-2.9/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Querying protein features</h1>
                        <h4 data-toc-skip class="author">Johannes Rainer</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jorainer/ensembldb/blob/master/vignettes/proteins.Rmd" class="external-link"><code>vignettes/proteins.Rmd</code></a></small>
      <div class="hidden name"><code>proteins.Rmd</code></div>

    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor" aria-hidden="true"></a>Introduction</h1>
<p>From Bioconductor release 3.5 on, <code>EnsDb</code> databases/packages created by the <code>ensembldb</code> package contain also, for transcripts with a coding regions, mappings between transcripts and proteins. Thus, in addition to the RNA/DNA-based features also the following protein related information is available:</p>
<ul>
<li>
<code>protein_id</code>: the Ensembl protein ID. This is the primary ID for the proteins defined in Ensembl and each (protein coding) Ensembl transcript has one protein ID assigned to it.</li>
<li>
<code>protein_sequence</code>: the amino acid sequence of a protein.</li>
<li>
<code>uniprot_id</code>: the Uniprot ID for a protein. Note that not every Ensembl <code>protein_id</code> has an Uniprot ID, and each <code>protein_id</code> might be mapped to several <code>uniprot_id</code>. Also, the same Uniprot ID might be mapped to different <code>protein_id</code>.</li>
<li>
<code>uniprot_db</code>: the name of the Uniprot database in which the feature is annotated. Can be either <em>SPTREMBL</em> or <em>SWISSPROT</em>.</li>
<li>
<code>uniprot_mapping_type</code>: the type of the mapping method that was used to assign the Uniprot ID to the Ensembl protein ID.</li>
<li>
<code>protein_domain_id</code>: the ID of the protein domain according to the source/analysis in/by which is was defined.</li>
<li>
<code>protein_domain_source</code>: the source of the protein domain information, one of <em>pfscan</em>, <em>scanprosite</em>, <em>superfamily</em>, <em>pfam</em>, <em>prints</em>, <em>smart</em>, <em>pirsf</em> or <em>tigrfam</em>.</li>
<li>
<code>interpro_accession</code>: the Interpro accession ID of the protein domain (if available).</li>
<li>
<code>prot_dom_start</code>: the start of the protein domain within the sequence of the protein.</li>
<li>
<code>prot_dom_start</code>: the end position of the protein domain within the sequence of the protein.</li>
</ul>
<p>Thus, for protein coding transcripts, these annotations can be fetched from the database too, given that protein annotations are available. Note that only <code>EnsDb</code> databases created through the Ensembl Perl API contain protein annotation, while databases created using <code>ensDbFromAH</code>, <code>ensDbFromGff</code>, <code>ensDbFromGRanges</code> and <code>ensDbFromGtf</code> don’t.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jorainer/ensembldb" class="external-link">ensembldb</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span>
<span class="va">edb</span> <span class="op">&lt;-</span> <span class="va">EnsDb.Hsapiens.v86</span>
<span class="co">## Evaluate whether we have protein annotation available</span>
<span class="fu"><a href="../reference/hasProteinData-EnsDb-method.html">hasProteinData</a></span><span class="op">(</span><span class="va">edb</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>If protein annotation is available, the additional tables and columns are also listed by the <code>listTables</code> and <code>listColumns</code> methods.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/EnsDb-class.html">listTables</a></span><span class="op">(</span><span class="va">edb</span><span class="op">)</span></code></pre></div>
<pre><code>## $gene
## [1] "gene_id"          "gene_name"        "gene_biotype"     "gene_seq_start"  
## [5] "gene_seq_end"     "seq_name"         "seq_strand"       "seq_coord_system"
## [9] "symbol"          
## 
## $tx
## [1] "tx_id"            "tx_biotype"       "tx_seq_start"     "tx_seq_end"      
## [5] "tx_cds_seq_start" "tx_cds_seq_end"   "gene_id"          "tx_name"         
## 
## $tx2exon
## [1] "tx_id"    "exon_id"  "exon_idx"
## 
## $exon
## [1] "exon_id"        "exon_seq_start" "exon_seq_end"  
## 
## $chromosome
## [1] "seq_name"    "seq_length"  "is_circular"
## 
## $protein
## [1] "tx_id"            "protein_id"       "protein_sequence"
## 
## $uniprot
## [1] "protein_id"           "uniprot_id"           "uniprot_db"          
## [4] "uniprot_mapping_type"
## 
## $protein_domain
## [1] "protein_id"            "protein_domain_id"     "protein_domain_source"
## [4] "interpro_accession"    "prot_dom_start"        "prot_dom_end"         
## 
## $entrezgene
## [1] "gene_id"  "entrezid"
## 
## $metadata
## [1] "name"  "value"</code></pre>
<p>In the following sections we show examples how to 1) fetch protein annotations as additional columns to gene/transcript annotations, 2) fetch protein annotation data and 3) map proteins to the genome.</p>
</div>
<div id="fetch-protein-annotation-for-genes-and-transcripts" class="section level1">
<h1 class="hasAnchor">
<a href="#fetch-protein-annotation-for-genes-and-transcripts" class="anchor" aria-hidden="true"></a>Fetch protein annotation for genes and transcripts</h1>
<p>Protein annotations for (protein coding) transcripts can be retrieved by simply adding the desired annotation columns to the <code>columns</code> parameter of the e.g. <code>genes</code> or <code>transcripts</code> methods.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Get also protein information for ZBTB16 transcripts</span>
<span class="va">txs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EnsDb-exonsBy.html">transcripts</a></span><span class="op">(</span><span class="va">edb</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationFilter/man/AnnotationFilter.html" class="external-link">GeneNameFilter</a></span><span class="op">(</span><span class="st">"ZBTB16"</span><span class="op">)</span>,
           columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"protein_id"</span>, <span class="st">"uniprot_id"</span>, <span class="st">"tx_biotype"</span><span class="op">)</span><span class="op">)</span>
<span class="va">txs</span></code></pre></div>
<pre><code>## GRanges object with 11 ranges and 5 metadata columns:
##                   seqnames              ranges strand |      protein_id
##                      &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
##   ENST00000335953       11 114059593-114250676      + | ENSP00000338157
##   ENST00000335953       11 114059593-114250676      + | ENSP00000338157
##   ENST00000541602       11 114059725-114189764      + |            &lt;NA&gt;
##   ENST00000544220       11 114059737-114063646      + | ENSP00000437716
##   ENST00000535700       11 114060257-114063744      + | ENSP00000443013
##   ENST00000392996       11 114060507-114250652      + | ENSP00000376721
##   ENST00000392996       11 114060507-114250652      + | ENSP00000376721
##   ENST00000539918       11 114064412-114247344      + | ENSP00000445047
##   ENST00000545851       11 114180766-114247296      + |            &lt;NA&gt;
##   ENST00000535379       11 114237207-114250557      + |            &lt;NA&gt;
##   ENST00000535509       11 114246790-114250476      + |            &lt;NA&gt;
##                    uniprot_id             tx_biotype           tx_id
##                   &lt;character&gt;            &lt;character&gt;     &lt;character&gt;
##   ENST00000335953      Q05516         protein_coding ENST00000335953
##   ENST00000335953  A0A024R3C6         protein_coding ENST00000335953
##   ENST00000541602        &lt;NA&gt;        retained_intron ENST00000541602
##   ENST00000544220      F5H6C3         protein_coding ENST00000544220
##   ENST00000535700      F5H5Y7         protein_coding ENST00000535700
##   ENST00000392996      Q05516         protein_coding ENST00000392996
##   ENST00000392996  A0A024R3C6         protein_coding ENST00000392996
##   ENST00000539918      H0YGW2 nonsense_mediated_de.. ENST00000539918
##   ENST00000545851        &lt;NA&gt;   processed_transcript ENST00000545851
##   ENST00000535379        &lt;NA&gt;   processed_transcript ENST00000535379
##   ENST00000535509        &lt;NA&gt;        retained_intron ENST00000535509
##                     gene_name
##                   &lt;character&gt;
##   ENST00000335953      ZBTB16
##   ENST00000335953      ZBTB16
##   ENST00000541602      ZBTB16
##   ENST00000544220      ZBTB16
##   ENST00000535700      ZBTB16
##   ENST00000392996      ZBTB16
##   ENST00000392996      ZBTB16
##   ENST00000539918      ZBTB16
##   ENST00000545851      ZBTB16
##   ENST00000535379      ZBTB16
##   ENST00000535509      ZBTB16
##   -------
##   seqinfo: 1 sequence from GRCh38 genome</code></pre>
<p>The gene ZBTB16 has protein coding and non-coding transcripts, thus, we get the protein ID for the coding- and <code>NA</code> for the non-coding transcripts. Note also that we have a transcript targeted for nonsense mediated mRNA-decay with a protein ID associated with it, but no Uniprot ID.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Subset to transcripts with tx_biotype other than protein_coding.</span>
<span class="va">txs</span><span class="op">[</span><span class="va">txs</span><span class="op">$</span><span class="va">tx_biotype</span> <span class="op">!=</span> <span class="st">"protein_coding"</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"uniprot_id"</span>, <span class="st">"tx_biotype"</span>,
                      <span class="st">"protein_id"</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<pre><code>## GRanges object with 5 ranges and 3 metadata columns:
##                   seqnames              ranges strand |  uniprot_id
##                      &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##   ENST00000541602       11 114059725-114189764      + |        &lt;NA&gt;
##   ENST00000539918       11 114064412-114247344      + |      H0YGW2
##   ENST00000545851       11 114180766-114247296      + |        &lt;NA&gt;
##   ENST00000535379       11 114237207-114250557      + |        &lt;NA&gt;
##   ENST00000535509       11 114246790-114250476      + |        &lt;NA&gt;
##                               tx_biotype      protein_id
##                              &lt;character&gt;     &lt;character&gt;
##   ENST00000541602        retained_intron            &lt;NA&gt;
##   ENST00000539918 nonsense_mediated_de.. ENSP00000445047
##   ENST00000545851   processed_transcript            &lt;NA&gt;
##   ENST00000535379   processed_transcript            &lt;NA&gt;
##   ENST00000535509        retained_intron            &lt;NA&gt;
##   -------
##   seqinfo: 1 sequence from GRCh38 genome</code></pre>
<p>While the mapping from a protein coding transcript to a Ensembl protein ID (column <code>protein_id</code>) is 1:1, the mapping between <code>protein_id</code> and <code>uniprot_id</code> can be n:m, i.e. each Ensembl protein ID can be mapped to 1 or more Uniprot IDs and each Uniprot ID can be mapped to more than one <code>protein_id</code> (and hence <code>tx_id</code>). This should be kept in mind if querying transcripts from the database fetching Uniprot related additional columns or even protein ID features, as in such cases a redundant list of transcripts is returned.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## List the protein IDs and uniprot IDs for the coding transcripts</span>
<span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">txs</span><span class="op">[</span><span class="va">txs</span><span class="op">$</span><span class="va">tx_biotype</span> <span class="op">==</span> <span class="st">"protein_coding"</span>,
      <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"tx_id"</span>, <span class="st">"protein_id"</span>, <span class="st">"uniprot_id"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 3 columns
##                           tx_id      protein_id  uniprot_id
##                     &lt;character&gt;     &lt;character&gt; &lt;character&gt;
## ENST00000335953 ENST00000335953 ENSP00000338157      Q05516
## ENST00000335953 ENST00000335953 ENSP00000338157  A0A024R3C6
## ENST00000544220 ENST00000544220 ENSP00000437716      F5H6C3
## ENST00000535700 ENST00000535700 ENSP00000443013      F5H5Y7
## ENST00000392996 ENST00000392996 ENSP00000376721      Q05516
## ENST00000392996 ENST00000392996 ENSP00000376721  A0A024R3C6</code></pre>
<p>Some of the n:m mappings for Uniprot IDs can be resolved by restricting either to entries from one Uniprot database (<em>SPTREMBL</em> or <em>SWISSPROT</em>) or to mappings of a certain type of mapping method. The corresponding filters are the <code>UniprotDbFilter</code> and the <code>UniprotMappingTypeFilter</code> (using the <code>uniprot_db</code> and <code>uniprot_mapping_type</code> columns of the <code>uniprot</code> database table). In the example below we restrict the result to Uniprot IDs with the mapping type <em>DIRECT</em>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## List all uniprot mapping types in the database.</span>
<span class="fu"><a href="../reference/ProteinFunctionality.html">listUniprotMappingTypes</a></span><span class="op">(</span><span class="va">edb</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "DIRECT"         "SEQUENCE_MATCH"</code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Get all protein_coding transcripts of ZBTB16 along with their protein_id</span>
<span class="co">## and Uniprot IDs, restricting to protein_id to uniprot_id mappings based</span>
<span class="co">## on "DIRECT" mapping methods.</span>
<span class="va">txs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EnsDb-exonsBy.html">transcripts</a></span><span class="op">(</span><span class="va">edb</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationFilter/man/AnnotationFilter.html" class="external-link">GeneNameFilter</a></span><span class="op">(</span><span class="st">"ZBTB16"</span><span class="op">)</span>,
                      <span class="fu"><a href="../reference/Filter-classes.html">UniprotMappingTypeFilter</a></span><span class="op">(</span><span class="st">"DIRECT"</span><span class="op">)</span><span class="op">)</span>,
           columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"protein_id"</span>, <span class="st">"uniprot_id"</span>, <span class="st">"uniprot_db"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">txs</span><span class="op">)</span></code></pre></div>
<pre><code>## DataFrame with 5 rows and 6 columns
##                      protein_id  uniprot_id  uniprot_db           tx_id
##                     &lt;character&gt; &lt;character&gt; &lt;character&gt;     &lt;character&gt;
## ENST00000335953 ENSP00000338157      Q05516   SWISSPROT ENST00000335953
## ENST00000544220 ENSP00000437716      F5H6C3    SPTREMBL ENST00000544220
## ENST00000535700 ENSP00000443013      F5H5Y7    SPTREMBL ENST00000535700
## ENST00000392996 ENSP00000376721      Q05516   SWISSPROT ENST00000392996
## ENST00000539918 ENSP00000445047      H0YGW2    SPTREMBL ENST00000539918
##                   gene_name uniprot_mapping_type
##                 &lt;character&gt;          &lt;character&gt;
## ENST00000335953      ZBTB16               DIRECT
## ENST00000544220      ZBTB16               DIRECT
## ENST00000535700      ZBTB16               DIRECT
## ENST00000392996      ZBTB16               DIRECT
## ENST00000539918      ZBTB16               DIRECT</code></pre>
<p>For this example the use of the <code>UniprotMappingTypeFilter</code> resolved the multiple mapping of Uniprot IDs to Ensembl protein IDs, but the Uniprot ID <em>Q05516</em> is still assigned to the two Ensembl protein IDs <em>ENSP00000338157</em> and <em>ENSP00000376721</em>.</p>
<p>All protein annotations can also be added as <em>metadata columns</em> to the results of the <code>genes</code>, <code>exons</code>, <code>exonsBy</code>, <code>transcriptsBy</code>, <code>cdsBy</code>, <code>fiveUTRsByTranscript</code> and <code>threeUTRsByTranscript</code> methods by specifying the desired column names with the <code>columns</code> parameter. For non coding transcripts <code>NA</code> will be reported in the protein annotation columns.</p>
<p>In addition to retrieve protein annotations from the database, we can also use protein data to filter the results. In the example below we fetch for example all genes from the database that have a certain protein domain in the protein encoded by any of its transcripts.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Get all genes encoded on chromosome 11 which protein contains </span>
<span class="co">## a certain protein domain.</span>
<span class="va">gns</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EnsDb-exonsBy.html">genes</a></span><span class="op">(</span><span class="va">edb</span>, filter <span class="op">=</span> <span class="op">~</span> <span class="va">prot_dom_id</span> <span class="op">==</span> <span class="st">"PS50097"</span> <span class="op">&amp;</span> <span class="va">seq_name</span> <span class="op">==</span> <span class="st">"11"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">gns</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 9</code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">gns</span><span class="op">$</span><span class="va">gene_name</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "ABTB2"  "BTBD18" "KBTBD3" "KBTBD4" "KCTD21" "KLHL35" "ZBTB16" "ZBTB3" 
## [9] "ZBTB44"</code></pre>
<p>So, in total we got 152 genes with that protein domain. In addition to the <code>ProtDomIdFilter</code>, also the <code>ProteinidFilter</code> and the <code>UniprotidFilter</code> can be used to query the database for entries matching conditions on their protein ID or Uniprot ID.</p>
</div>
<div id="use-methods-from-the-annotationdbi-package-to-query-protein-annotation" class="section level1">
<h1 class="hasAnchor">
<a href="#use-methods-from-the-annotationdbi-package-to-query-protein-annotation" class="anchor" aria-hidden="true"></a>Use methods from the <code>AnnotationDbi</code> package to query protein annotation</h1>
<p>The <code>select</code>, <code>keys</code> and <code>mapIds</code> methods from the <code>AnnotationDbi</code> package can also be used to query <code>EnsDb</code> objects for protein annotations. Supported columns and key types are returned by the <code>columns</code> and <code>keytypes</code> methods.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Show all columns that are provided by the database</span>
<span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">columns</a></span><span class="op">(</span><span class="va">edb</span><span class="op">)</span></code></pre></div>
<pre><code>##  [1] "ENTREZID"            "EXONID"              "EXONIDX"            
##  [4] "EXONSEQEND"          "EXONSEQSTART"        "GENEBIOTYPE"        
##  [7] "GENEID"              "GENENAME"            "GENESEQEND"         
## [10] "GENESEQSTART"        "INTERPROACCESSION"   "ISCIRCULAR"         
## [13] "PROTDOMEND"          "PROTDOMSTART"        "PROTEINDOMAINID"    
## [16] "PROTEINDOMAINSOURCE" "PROTEINID"           "PROTEINSEQUENCE"    
## [19] "SEQCOORDSYSTEM"      "SEQLENGTH"           "SEQNAME"            
## [22] "SEQSTRAND"           "SYMBOL"              "TXBIOTYPE"          
## [25] "TXCDSSEQEND"         "TXCDSSEQSTART"       "TXID"               
## [28] "TXNAME"              "TXSEQEND"            "TXSEQSTART"         
## [31] "UNIPROTDB"           "UNIPROTID"           "UNIPROTMAPPINGTYPE"</code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Show all key types/filters that are supported</span>
<span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">keytypes</a></span><span class="op">(</span><span class="va">edb</span><span class="op">)</span></code></pre></div>
<pre><code>##  [1] "ENTREZID"            "EXONID"              "GENEBIOTYPE"        
##  [4] "GENEID"              "GENENAME"            "PROTDOMID"          
##  [7] "PROTEINDOMAINID"     "PROTEINDOMAINSOURCE" "PROTEINID"          
## [10] "SEQNAME"             "SEQSTRAND"           "SYMBOL"             
## [13] "TXBIOTYPE"           "TXID"                "TXNAME"             
## [16] "UNIPROTID"</code></pre>
<p>Below we fetch all Uniprot IDs annotated to the gene <em>ZBTB16</em>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/EnsDb-AnnotationDbi.html">select</a></span><span class="op">(</span><span class="va">edb</span>, keys <span class="op">=</span> <span class="st">"ZBTB16"</span>, keytype <span class="op">=</span> <span class="st">"GENENAME"</span>,
       columns <span class="op">=</span> <span class="st">"UNIPROTID"</span><span class="op">)</span></code></pre></div>
<pre><code>##   GENENAME  UNIPROTID
## 1   ZBTB16     Q05516
## 2   ZBTB16 A0A024R3C6
## 3   ZBTB16       &lt;NA&gt;
## 4   ZBTB16     F5H6C3
## 5   ZBTB16     F5H5Y7
## 6   ZBTB16     H0YGW2</code></pre>
<p>This returns us all Uniprot IDs of all proteins encoded by the gene’s transcripts. One of the transcripts from ZBTB16, while having a CDS and being annotated to a protein, does not have an Uniprot ID assigned (thus <code>NA</code> is returned by the above call). As we see below, this transcript is targeted for non sense mediated mRNA decay.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Call select, this time providing a GeneNameFilter.</span>
<span class="fu"><a href="../reference/EnsDb-AnnotationDbi.html">select</a></span><span class="op">(</span><span class="va">edb</span>, keys <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationFilter/man/AnnotationFilter.html" class="external-link">GeneNameFilter</a></span><span class="op">(</span><span class="st">"ZBTB16"</span><span class="op">)</span>,
       columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TXBIOTYPE"</span>, <span class="st">"UNIPROTID"</span>, <span class="st">"PROTEINID"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>##                 TXBIOTYPE  UNIPROTID       PROTEINID GENENAME
## 1          protein_coding     Q05516 ENSP00000338157   ZBTB16
## 2          protein_coding A0A024R3C6 ENSP00000338157   ZBTB16
## 3         retained_intron       &lt;NA&gt;            &lt;NA&gt;   ZBTB16
## 4          protein_coding     F5H6C3 ENSP00000437716   ZBTB16
## 5          protein_coding     F5H5Y7 ENSP00000443013   ZBTB16
## 6          protein_coding     Q05516 ENSP00000376721   ZBTB16
## 7          protein_coding A0A024R3C6 ENSP00000376721   ZBTB16
## 8 nonsense_mediated_decay     H0YGW2 ENSP00000445047   ZBTB16
## 9    processed_transcript       &lt;NA&gt;            &lt;NA&gt;   ZBTB16</code></pre>
<p>Note also that we passed this time a <code>GeneMameFilter</code> with the <code>keys</code> parameter.</p>
</div>
<div id="retrieve-proteins-from-the-database" class="section level1">
<h1 class="hasAnchor">
<a href="#retrieve-proteins-from-the-database" class="anchor" aria-hidden="true"></a>Retrieve proteins from the database</h1>
<p>Proteins can be fetched using the dedicated <code>proteins</code> method that returns, unlike DNA/RNA-based methods like <code>genes</code> or <code>transcripts</code>, not a <code>GRanges</code> object by default, but a <code>DataFrame</code> object. Alternatively, results can be returned as a <code>data.frame</code> or as an <code>AAStringSet</code> object from the <code>Biobase</code> package. Note that this might change in future releases if a more appropriate object to represent protein annotations becomes available.</p>
<p>In the code chunk below we fetch all protein annotations for the gene <em>ZBTB16</em>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Get all proteins and return them as an AAStringSet</span>
<span class="va">prts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ProteinFunctionality.html">proteins</a></span><span class="op">(</span><span class="va">edb</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationFilter/man/AnnotationFilter.html" class="external-link">GeneNameFilter</a></span><span class="op">(</span><span class="st">"ZBTB16"</span><span class="op">)</span>,
         return.type <span class="op">=</span> <span class="st">"AAStringSet"</span><span class="op">)</span>
<span class="va">prts</span></code></pre></div>
<pre><code>## AAStringSet object of length 5:
##     width seq                                               names               
## [1]   673 MDLTKMGMIQLQNPSHPTGLLCK...GHKPEEIPPDWRIEKTYLYLCYV ENSP00000338157
## [2]   115 MDLTKMGMIQLQNPSHPTGLLCK...QAKAEDLDDLLYAAEILEIEYLE ENSP00000437716
## [3]   148 MDLTKMGMIQLQNPSHPTGLLCK...QASDDNDTEATMADGGAEEEEDR ENSP00000443013
## [4]   673 MDLTKMGMIQLQNPSHPTGLLCK...GHKPEEIPPDWRIEKTYLYLCYV ENSP00000376721
## [5]    55 XGGLLPQGFIQRELFSKLGELAV...GEQCSVCGVELPDNEAVEQHRVF ENSP00000445047</code></pre>
<p>Besides the amino acid sequence, the <code>prts</code> contains also additional annotations that can be accessed with the <code>mcols</code> method (metadata columns). All additional columns provided with the parameter <code>columns</code> are also added to the <code>mcols</code> <code>DataFrame</code>.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">prts</span><span class="op">)</span></code></pre></div>
<pre><code>## DataFrame with 5 rows and 3 columns
##                           tx_id      protein_id   gene_name
##                     &lt;character&gt;     &lt;character&gt; &lt;character&gt;
## ENSP00000338157 ENST00000335953 ENSP00000338157      ZBTB16
## ENSP00000437716 ENST00000544220 ENSP00000437716      ZBTB16
## ENSP00000443013 ENST00000535700 ENSP00000443013      ZBTB16
## ENSP00000376721 ENST00000392996 ENSP00000376721      ZBTB16
## ENSP00000445047 ENST00000539918 ENSP00000445047      ZBTB16</code></pre>
<p>Note that the <code>proteins</code> method will retrieve only gene/transcript annotations of transcripts encoding a protein. Thus annotations for the non-coding transcripts of the gene <em>ZBTB16</em>, that were returned by calls to <code>genes</code> or <code>transcripts</code> in the previous section are not fetched.</p>
<p>Querying in addition Uniprot identifiers or protein domain data will result at present in a redundant list of proteins as shown in the code block below.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Get also protein domain annotations in addition to the protein annotations.</span>
<span class="va">pd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ProteinFunctionality.html">proteins</a></span><span class="op">(</span><span class="va">edb</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationFilter/man/AnnotationFilter.html" class="external-link">GeneNameFilter</a></span><span class="op">(</span><span class="st">"ZBTB16"</span><span class="op">)</span>,
           columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"tx_id"</span>, <span class="fu"><a href="../reference/EnsDb-class.html">listColumns</a></span><span class="op">(</span><span class="va">edb</span>, <span class="st">"protein_domain"</span><span class="op">)</span><span class="op">)</span>,
           return.type <span class="op">=</span> <span class="st">"AAStringSet"</span><span class="op">)</span>
<span class="va">pd</span></code></pre></div>
<pre><code>## AAStringSet object of length 81:
##      width seq                                              names               
##  [1]   673 MDLTKMGMIQLQNPSHPTGLLCK...HKPEEIPPDWRIEKTYLYLCYV ENSP00000338157
##  [2]   673 MDLTKMGMIQLQNPSHPTGLLCK...HKPEEIPPDWRIEKTYLYLCYV ENSP00000338157
##  [3]   673 MDLTKMGMIQLQNPSHPTGLLCK...HKPEEIPPDWRIEKTYLYLCYV ENSP00000338157
##  [4]   673 MDLTKMGMIQLQNPSHPTGLLCK...HKPEEIPPDWRIEKTYLYLCYV ENSP00000338157
##  [5]   673 MDLTKMGMIQLQNPSHPTGLLCK...HKPEEIPPDWRIEKTYLYLCYV ENSP00000338157
##  ...   ... ...
## [77]   673 MDLTKMGMIQLQNPSHPTGLLCK...HKPEEIPPDWRIEKTYLYLCYV ENSP00000376721
## [78]   673 MDLTKMGMIQLQNPSHPTGLLCK...HKPEEIPPDWRIEKTYLYLCYV ENSP00000376721
## [79]   673 MDLTKMGMIQLQNPSHPTGLLCK...HKPEEIPPDWRIEKTYLYLCYV ENSP00000376721
## [80]   673 MDLTKMGMIQLQNPSHPTGLLCK...HKPEEIPPDWRIEKTYLYLCYV ENSP00000376721
## [81]    55 XGGLLPQGFIQRELFSKLGELAV...EQCSVCGVELPDNEAVEQHRVF ENSP00000445047</code></pre>
<p>The result contains one row/element for each protein domain in each of the proteins. The number of protein domains per protein and the <code>mcols</code> are shown below.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## The number of protein domains per protein:</span>
<span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">pd</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## 
## ENSP00000338157 ENSP00000376721 ENSP00000437716 ENSP00000443013 ENSP00000445047 
##              36              36               4               4               1</code></pre>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## The mcols</span>
<span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">pd</span><span class="op">)</span></code></pre></div>
<pre><code>## DataFrame with 81 rows and 8 columns
##                           tx_id      protein_id protein_domain_id
##                     &lt;character&gt;     &lt;character&gt;       &lt;character&gt;
## ENSP00000338157 ENST00000335953 ENSP00000338157           PS50157
## ENSP00000338157 ENST00000335953 ENSP00000338157           PS50157
## ENSP00000338157 ENST00000335953 ENSP00000338157           PS50157
## ENSP00000338157 ENST00000335953 ENSP00000338157           PS50157
## ENSP00000338157 ENST00000335953 ENSP00000338157           PS50157
## ...                         ...             ...               ...
## ENSP00000376721 ENST00000392996 ENSP00000376721           SM00355
## ENSP00000376721 ENST00000392996 ENSP00000376721           SM00355
## ENSP00000376721 ENST00000392996 ENSP00000376721           SM00355
## ENSP00000376721 ENST00000392996 ENSP00000376721           SM00355
## ENSP00000445047 ENST00000539918 ENSP00000445047                NA
##                 protein_domain_source interpro_accession prot_dom_start
##                           &lt;character&gt;        &lt;character&gt;      &lt;integer&gt;
## ENSP00000338157                pfscan          IPR007087            602
## ENSP00000338157                pfscan          IPR007087            490
## ENSP00000338157                pfscan          IPR007087            630
## ENSP00000338157                pfscan          IPR007087            432
## ENSP00000338157                pfscan          IPR007087            546
## ...                               ...                ...            ...
## ENSP00000376721                 smart          IPR015880            546
## ENSP00000376721                 smart          IPR015880            574
## ENSP00000376721                 smart          IPR015880            602
## ENSP00000376721                 smart          IPR015880            630
## ENSP00000445047                    NA                 NA             NA
##                 prot_dom_end   gene_name
##                    &lt;integer&gt; &lt;character&gt;
## ENSP00000338157          629      ZBTB16
## ENSP00000338157          517      ZBTB16
## ENSP00000338157          657      ZBTB16
## ENSP00000338157          459      ZBTB16
## ENSP00000338157          573      ZBTB16
## ...                      ...         ...
## ENSP00000376721          568      ZBTB16
## ENSP00000376721          596      ZBTB16
## ENSP00000376721          624      ZBTB16
## ENSP00000376721          652      ZBTB16
## ENSP00000445047           NA      ZBTB16</code></pre>
<p>As we can see each protein can have several protein domains with the start and end coordinates within the amino acid sequence being reported in columns <code>prot_dom_start</code> and <code>prot_dom_end</code>. Also, not all Ensembl protein IDs, like <code>protein_id</code> <em>ENSP00000445047</em> are mapped to an Uniprot ID or have protein domains.</p>
</div>
<div id="map-peptide-features-within-proteins-to-the-genome" class="section level1">
<h1 class="hasAnchor">
<a href="#map-peptide-features-within-proteins-to-the-genome" class="anchor" aria-hidden="true"></a>Map peptide features within proteins to the genome</h1>
<p>The <em>coordinate-mapping.Rmd</em> vignette provides a detailed description of all functions that allow to map between genomic, transcript and protein coordinates.</p>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor" aria-hidden="true"></a>Session information</h1>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span> </code></pre></div>
<pre><code>## R version 4.1.0 (2021-05-18)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.2 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] EnsDb.Hsapiens.v86_2.99.0 ensembldb_2.17.2         
##  [3] AnnotationFilter_1.17.0   GenomicFeatures_1.45.0   
##  [5] AnnotationDbi_1.55.1      Biobase_2.53.0           
##  [7] GenomicRanges_1.45.0      GenomeInfoDb_1.29.2      
##  [9] IRanges_2.27.0            S4Vectors_0.31.0         
## [11] BiocGenerics_0.39.1       BiocStyle_2.21.3         
## 
## loaded via a namespace (and not attached):
##  [1] ProtGenerics_1.25.0         bitops_1.0-7               
##  [3] matrixStats_0.59.0          fs_1.5.0                   
##  [5] bit64_4.0.5                 filelock_1.0.2             
##  [7] progress_1.2.2              httr_1.4.2                 
##  [9] rprojroot_2.0.2             tools_4.1.0                
## [11] bslib_0.2.5.1               utf8_1.2.1                 
## [13] R6_2.5.0                    DBI_1.1.1                  
## [15] lazyeval_0.2.2              tidyselect_1.1.1           
## [17] prettyunits_1.1.1           bit_4.0.4                  
## [19] curl_4.3.2                  compiler_4.1.0             
## [21] textshaping_0.3.5           xml2_1.3.2                 
## [23] desc_1.3.0                  DelayedArray_0.19.1        
## [25] rtracklayer_1.53.0          bookdown_0.22              
## [27] sass_0.4.0                  rappdirs_0.3.3             
## [29] pkgdown_1.6.1.9001          systemfonts_1.0.2          
## [31] stringr_1.4.0               digest_0.6.27              
## [33] Rsamtools_2.9.1             rmarkdown_2.9              
## [35] XVector_0.33.0              pkgconfig_2.0.3            
## [37] htmltools_0.5.1.1           MatrixGenerics_1.5.0       
## [39] dbplyr_2.1.1                fastmap_1.1.0              
## [41] rlang_0.4.11                RSQLite_2.2.7              
## [43] jquerylib_0.1.4             BiocIO_1.3.0               
## [45] generics_0.1.0              jsonlite_1.7.2             
## [47] BiocParallel_1.27.0         dplyr_1.0.7                
## [49] RCurl_1.98-1.3              magrittr_2.0.1             
## [51] GenomeInfoDbData_1.2.6      Matrix_1.3-4               
## [53] Rcpp_1.0.6                  fansi_0.5.0                
## [55] lifecycle_1.0.0             stringi_1.6.2              
## [57] yaml_2.2.1                  SummarizedExperiment_1.23.1
## [59] zlibbioc_1.39.0             BiocFileCache_2.1.1        
## [61] grid_4.1.0                  blob_1.2.1                 
## [63] parallel_4.1.0              crayon_1.4.1               
## [65] lattice_0.20-44             Biostrings_2.61.1          
## [67] hms_1.1.0                   KEGGREST_1.33.0            
## [69] knitr_1.33                  pillar_1.6.1               
## [71] rjson_0.2.20                biomaRt_2.49.2             
## [73] XML_3.99-0.6                glue_1.4.2                 
## [75] evaluate_0.14               BiocManager_1.30.16        
## [77] png_0.1-7                   vctrs_0.3.8                
## [79] purrr_0.3.4                 assertthat_0.2.1           
## [81] cachem_1.0.5                xfun_0.24                  
## [83] restfulr_0.0.13             ragg_1.1.3                 
## [85] tibble_3.1.2                GenomicAlignments_1.29.0   
## [87] memoise_2.0.0               ellipsis_0.3.2</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Johannes Rainer.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link external-link">pkgdown</a> 1.6.1.9001.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
